---
title: "R Notebook"
---

# IMPORTING REQUIRED PACKAGES
```{r}
# LIST OF PACKAGES
pkgs = c(
  "tidyverse",
  "here",
  "svglite",
  "ggplot2",
  "ggrepel",
  "stringr",
  "BiocManager",
  "RColorBrewer",
  "viridis",
  "magick",
  "nlme",
  "ashr",
  "factoextra",
  "plotly",
  "remotes",
  "cowplot", "tibble", "dplyr",
  "gridExtra",
  "eulerr", "data.table", "clusterProfiler", "zoo"
)
bio_pkgs = c("biomaRt",
             "DESeq2",
             "ComplexHeatmap",
             "apeglm",
             "vsn", "caTools",
             "sva",
             "limma",
             "mixOmics", "plotgardener", "org.Mm.eg.db")

# INSTALL PACKAGES
#install.packages(pkgs)
#BiocManager::install(bio_pkgs)

# LOAD PACKAGES
lapply(pkgs, require, character.only = TRUE)
lapply(bio_pkgs, require, character.only = TRUE)

#remotes::install_github("EvaYiwenWang/PLSDAbatch")
require(PLSDAbatch)

# CLEANING
rm(pkgs, bio_pkgs)

options(scipen = 999)
```
# Make v4c
(1) Extract read pairs which a readmate maps from the 10kb  bin that the read mate maps around the vitual read point
(2) Make successive over-lapping windows for each chromosome at a 10 kb resolution (1 kb step)
(3) Counted all secondmapped read mate in all overlapping bins
-- up until this part is done by step1.sh
(4) Normalize with CPM for total Micro_C reads
(5) Visualize
## Visualization
```{r}
library(Gviz)
library(plotgardener)
library(colorspace)
mmToLinePlotgarden <- 1/0.75

strong_red <- "#CB333A"
strong_blue <- "#4851A0"
weak_red <- lighten(strong_red, amount = 0.4)   # FF7D81
weak_blue <- lighten(strong_blue, amount = 0.4) # 8A91DD
no_grey <- "#A8A8A8"

strong_teel <- "#0892A5"
strong_green <- "#23CE6B" # A485
strong_darkgreen <- "#054A29"
strong_yellow <- "#FFBA49"
strong_orange <- "#F18F01" # dTAG
strong_lightpurple <- "#BD93D8"
strong_purple <- "#9E33CB" # Epi

color_esc_light <- "#f28e2c"
color_esc_dark <- colorspace::darken(color_esc_light, amount = 0.3)
"#AA5F00"
color_epi_light <- "#009b77"
color_epi_dark <- colorspace::darken(color_epi_light, amount = 0.3)
"#006B51"
################################################################################
# IMPORTING PARAMETERS & SAMPLE LIST
################################################################################


bgDir <- here("..", "data", "v4C", "bedgraph_parallel_chrLocalNorm")
allInfoDir <- here("..", "data", "v4C", "allInfo_parallel_chr")
loopDir <- here("..", "data", "loop_analysis")
refDir <- here("..", "reference")
figDir <- here("..", "figure", "v4C")
dir.create(figDir, recursive = TRUE, showWarnings = FALSE)

fileList <- as_tibble(list.files(allInfoDir))

fileList <- fileList %>% dplyr::rowwise() %>% dplyr::mutate(ensembl = unlist(strsplit(value, "\\."))[[2]],
                                                            note = unlist(strsplit(value, "\\."))[[3]],
                                                            gene = unlist(strsplit(note, "__"))[[1]],
                                                            vp = paste(ensembl, gene, sep = "."))
vpList <- unique(fileList$vp)
expList <- c("G1DMSO_pooled", "G1dTAG_pooled", "G1A485_pooled", "EpiG1DMSO_pooled", "EpiG1dTAG_pooled", "GSE178982_AsyncUT_pooled", "GSE178982_AsyncAID_pooled")
resList <- c(5, 10)
################################################################################
################################################################################
# TEMPORARY
vp <- "ENSMUSG00000022346.Myc"
res <- 10
expName <- expList[1]
cutoff <- 75
cutoff <- 200
```

### All VPs
```{r}
for(vp in vpList){
  for(res in resList){
    for(cutoff in c(200, 74)){
      
      
      ################################################################################
      # IMPORTING TRACKS
      ################################################################################
      # Import bedgraph
      tb_ESC_DMSO <- fread(here(bgDir, paste0("v4C_", vp, "__", "G1DMSO_pooled", "__", res, "kb_localNorm.bedGraph" )))
      colnames(tb_ESC_DMSO) <- c("chrom", "start", "end", "score")
      bg_ESC_DMSO <- makeGRangesFromDataFrame(tb_ESC_DMSO, keep.extra.columns = TRUE)
      keep <- countOverlaps(bg_ESC_DMSO) == 1
      bg_ESC_DMSO <- bg_ESC_DMSO[keep]
      # Import bedgraph
      tb_EpiLC_DMSO <- fread(here(bgDir, paste0("v4C_", vp, "__", "EpiG1DMSO_pooled", "__", res, "kb_localNorm.bedGraph" )))
      colnames(tb_EpiLC_DMSO) <- c("chrom", "start", "end", "score")
      bg_EpiLC_DMSO <- makeGRangesFromDataFrame(tb_EpiLC_DMSO, keep.extra.columns = TRUE)
      keep <- countOverlaps(bg_EpiLC_DMSO) == 1
      bg_EpiLC_DMSO <- bg_EpiLC_DMSO[keep]
      
      # Import bedgraph
      tb_ESC_dTAG <- fread(here(bgDir, paste0("v4C_", vp, "__", "G1dTAG_pooled", "__", res, "kb_localNorm.bedGraph" )))
      colnames(tb_ESC_dTAG) <- c("chrom", "start", "end", "score")
      bg_ESC_dTAG <- makeGRangesFromDataFrame(tb_ESC_dTAG, keep.extra.columns = TRUE)
      keep <- countOverlaps(bg_ESC_dTAG) == 1
      bg_ESC_dTAG <- bg_ESC_dTAG[keep]
      # Import bedgraph
      tb_EpiLC_dTAG <- fread(here(bgDir, paste0("v4C_", vp, "__", "EpiG1dTAG_pooled", "__", res, "kb_localNorm.bedGraph" )))
      colnames(tb_EpiLC_dTAG) <- c("chrom", "start", "end", "score")
      bg_EpiLC_dTAG <- makeGRangesFromDataFrame(tb_EpiLC_dTAG, keep.extra.columns = TRUE)
      keep <- countOverlaps(bg_EpiLC_dTAG) == 1
      bg_EpiLC_dTAG <- bg_EpiLC_dTAG[keep]
      
      # Make differential bedgraph
      tb_diff_ESCvsEpiLC <- tb_ESC_DMSO %>% left_join(tb_EpiLC_DMSO, by = c("chrom", "start", "end")) %>%
        dplyr::mutate(score = score.y - score.x) %>%
        dplyr::select(chrom, start, end, score)
      bg_diff_ESCvsEpiLC <- makeGRangesFromDataFrame(tb_diff_ESCvsEpiLC, keep.extra.columns = TRUE)
      keep <- countOverlaps(bg_diff_ESCvsEpiLC) == 1
      bg_diff_ESCvsEpiLC <- bg_diff_ESCvsEpiLC[keep]
      
      # Make differential bedgraph
      tb_diff_ESC_dTAGvsDMSO <- tb_ESC_DMSO %>% left_join(tb_ESC_dTAG, by = c("chrom", "start", "end")) %>%
        dplyr::mutate(score = score.y - score.x) %>%
        dplyr::select(chrom, start, end, score)
      bg_diff_ESC_dTAGvsDMSO <- makeGRangesFromDataFrame(tb_diff_ESC_dTAGvsDMSO, keep.extra.columns = TRUE)
      keep <- countOverlaps(bg_diff_ESC_dTAGvsDMSO) == 1
      bg_diff_ESC_dTAGvsDMSO <- bg_diff_ESC_dTAGvsDMSO[keep]
      
      # Make differential bedgraph
      tb_diff_EpiLC_dTAGvsDMSO <- tb_EpiLC_DMSO %>% left_join(tb_EpiLC_dTAG, by = c("chrom", "start", "end")) %>%
        dplyr::mutate(score = score.y - score.x) %>%
        dplyr::select(chrom, start, end, score)
      bg_diff_EpiLC_dTAGvsDMSO <- makeGRangesFromDataFrame(tb_diff_EpiLC_dTAGvsDMSO, keep.extra.columns = TRUE)
      keep <- countOverlaps(bg_diff_EpiLC_dTAGvsDMSO) == 1
      bg_diff_EpiLC_dTAGvsDMSO <- bg_diff_EpiLC_dTAGvsDMSO[keep]
      
      # CHIP TRACKS
      bw_H3K4me3_ESC <- here(refDir, "ChIP", "H3K4me3", "H3K4me3_ESC_avg_depthNorm.bw")
      bed_H3K4me3_ESCspecific <- here(refDir, "ChIP", "differential", "H3K4me3_diff_specific_ESC.bed")
      bw_H3K27ac_ESC <- here(refDir, "ChIP", "H3K27ac", "H3K27ac_ESC_avg_depthNorm.bw")
      bed_H3K27ac_ESCspecific <- here(refDir, "ChIP", "differential", "H3K27ac_diff_specific_ESC.bed")
      bw_H3K4me3_EpiLC <- here(refDir, "ChIP", "H3K4me3", "H3K4me3_EpiLC_avg_depthNorm.bw")
      bed_H3K4me3_EpiLCspecific <- here(refDir, "ChIP", "differential", "H3K4me3_diff_specific_EpiLC.bed")
      bw_H3K27ac_EpiLC <- here(refDir, "ChIP", "H3K27ac", "H3K27ac_EpiLC_avg_depthNorm.bw")
      bed_H3K27ac_EpiLCspecific <- here(refDir, "ChIP", "differential", "H3K27ac_diff_specific_EpiLC.bed")
      bw_RAD21_ESC <- here(refDir, "ChIP", "RAD21-HA", "RAD21_ESC-DMSO_avg_spikeNorm_inputNorm.bw") 
      bw_RAD21_EpiLC <- here(refDir, "ChIP", "RAD21-HA", "RAD21_EpiLC_avg_spikeNorm_inputNorm.bw") 
      
      # PROMOTER LOOPS
      bedpe_promoterLoops <- here(loopDir, "chromosight_score_complex_allP.bedpe")
      ################################################################################
      # DEFINING RANGES
      ################################################################################
      # Y ranges for virtual 4C
      ymin <- 0
      ymax <- if_else(res == 10, 1000, 800)
      diffYmax <- if_else(res == 10, 400, 300)
      
      # X ranges 
      data <- as_tibble(fread(here(bgDir, paste0("v4C_", vp, "__", "G1DMSO_pooled", "__", res, "kb_localNorm.bedGraph" ))))
      
      xmin <- min((data %>% dplyr::filter(V4 > cutoff))$V2)
      xmax <- max((data %>% dplyr::filter(V4 > cutoff))$V2)
      chr <- data$V1[1]
      
      # Y ranges for ChIP
      region_gr <- GRanges(seqnames = chr,
                           ranges = IRanges(start = xmin,
                                            end = xmax))
      ## H3K4me3 range
      bw_subset1 <- rtracklayer::import.bw(bw_H3K4me3_ESC, which = region_gr)
      bw_subset2 <- rtracklayer::import.bw(bw_H3K4me3_EpiLC, which = region_gr)
      max_score <- max(bw_subset1$score, bw_subset2$score)
      common_range_H3K4me3 <- c(0, ceiling(max_score / 0.1) * 0.1)
      
      ## H3K27ac range
      bw_subset1 <- rtracklayer::import.bw(bw_H3K27ac_ESC, which = region_gr)
      bw_subset2 <- rtracklayer::import.bw(bw_H3K27ac_EpiLC, which = region_gr)
      max_score <- max(bw_subset1$score, bw_subset2$score)
      common_range_H3K27ac <- c(0, ceiling(max_score / 0.1) * 0.1)
      
      
      ## RAD21 range
      bw_subset1 <- rtracklayer::import.bw(bw_RAD21_ESC, which = region_gr)
      bw_subset2 <- rtracklayer::import.bw(bw_RAD21_EpiLC, which = region_gr)
      max_score <- max(bw_subset1$score, bw_subset2$score)
      common_range_RAD21 <- c(0, ceiling(max_score / 0.1) * 0.1)
      
      ## Defining the genomic region to plot
      plot_region <- pgParams(
        chrom = chr,
        chromstart = xmin,
        chromend = xmax,
        assembly = "mm10"
      )
      # Plotting parameters
      xPosition <- 0.25
      yPosition <- 0.4
      width <- 3.5
      height.v4C <- 0.5
      height.v4CDiff <- 0.3
      height.chip <- 0.1
      height.chipPeak <- 0.05
      yPositionShift <- "0.15b"
      yPositionShiftSmall <- "0.05b"
      yPositionShiftMini <- "0.025b"
      
      yPositionShiftNum <- 0.15
      yPositionShiftSmallNum <- 0.05
      yPositionShiftMiniNum <- 0.025
      textShift <- 0.05
      fontfamily <- "Helvetica"
      
      ################################################################################
      # Plot 1: Drawing ESC vs EpiLC DMSO
      ################################################################################
      pdf(here(figDir, paste0("v4C_", vp, "_", res, "kb_EpiLCvsESC_cutoff", cutoff, ".pdf")), width = 4, height = 8)
      pageCreate(width = 4, height = 8, default.units = "inches", showGuides = FALSE)
      # --- TITLE ---
      plotText(
        label = paste0("Virtual 4C, ", unlist(strsplit(vp, "\\."))[2], " (", res, "kb resolution)"),
        x = 0.2, y = 0.2,
        just = "left",
        fontface = "bold",
        fontsize = 8, fontfamily = fontfamily
      )
      # --- v4C G1 ESC DMSO---
      plotText(label = "G1.ESC.DMSO", x = xPosition + textShift, y = yPosition, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      # Plot the signal track
      v4C.ESC.DMSO <- plotSignal(
        data = bg_ESC_DMSO,
        params = plot_region,
        range = c(ymin, ymax),
        x = xPosition, y = yPosition, width = width, height = height.v4C,
        just = c("left", "top"),
        linecolor = color_esc_dark, lwd = 0.25*mmToLinePlotgarden,
        fill = color_esc_dark
      )
      annoYaxis(
        plot = v4C.ESC.DMSO,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- ChIP H3K4me3 ESC ---
      plotText(label = "H3K4me3", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.h3k4me3.ESC <- plotSignal(
        data = bw_H3K4me3_ESC,
        params = plot_region, range = common_range_H3K4me3,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = strong_red,
        fill = strong_red
      )
      annoYaxis(
        plot = chip.h3k4me3.ESC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- ESC specific peaks ---
      plotText(label = "ESC specific peak", x = xPosition + textShift, y = yPositionShiftMini, just = "left",
               fontsize = 4, fontfamily = fontfamily)
      plotRanges(
        data = bed_H3K4me3_ESCspecific,
        params = plot_region,
        x = xPosition, y = yPositionShiftMini, width = width, height = height.chip,
        just = c("left", "top"),
        fill = strong_red
      )
      # --- ChIP H3K27ac ESC ---
      plotText(label = "H3K27ac", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.H3K27ac.ESC <- plotSignal(
        data = bw_H3K27ac_ESC,
        params = plot_region, range = common_range_H3K27ac,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = strong_blue,
        fill = strong_blue
      )
      annoYaxis(
        plot = chip.H3K27ac.ESC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- ESC specific peaks ---
      plotText(label = "ESC specific peak", x = xPosition + textShift, y = yPositionShiftMini, just = "left",
               fontsize = 4, fontfamily = fontfamily)
      plotRanges(
        data = bed_H3K27ac_ESCspecific,
        params = plot_region,
        x = xPosition, y = yPositionShiftMini, width = width, height = height.chip,
        just = c("left", "top"),
        fill = strong_blue
      )
      # --- ChIP RAD21 ESC ---
      plotText(label = "RAD21", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.RAD21.ESC <- plotSignal(
        data = bw_RAD21_ESC,
        params = plot_region, range = common_range_RAD21,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = "grey30",
        fill = "grey30"
      )
      annoYaxis(
        plot = chip.RAD21.ESC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      
      # --- v4C G1 EpiLC DMSO---
      # The y-coordinate "0.2b" places this plot 0.2 inches below the previous one
      plotText(label = "G1.EpiLC.DMSO", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      v4C.EpiLC.DMSO <- plotSignal(
        data = bg_EpiLC_DMSO,
        params = plot_region,
        range = c(ymin, ymax),
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C,
        just = c("left", "top"),
        linecolor = color_epi_dark, lwd = 0.25*mmToLinePlotgarden,
        fill = color_epi_dark
      )
      annoYaxis(
        plot = v4C.EpiLC.DMSO,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      # --- ChIP H3K4me3 EpiLC ---
      plotText(label = "H3K4me3", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.h3k4me3.EpiLC <- plotSignal(
        data = bw_H3K4me3_EpiLC,
        params = plot_region, range = common_range_H3K4me3,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = strong_red,
        fill = strong_red
      )
      annoYaxis(
        plot = chip.h3k4me3.EpiLC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- EpiLC specific peaks ---
      plotText(label = "EpiLC specific peak", x = xPosition + textShift, y = yPositionShiftMini, just = "left",
               fontsize = 4, fontfamily = fontfamily)
      plotRanges(
        data = bed_H3K4me3_EpiLCspecific,
        params = plot_region,
        x = xPosition, y = yPositionShiftMini, width = width, height = height.chip,
        just = c("left", "top"),
        fill = strong_red
      )
      # --- ChIP H3K27ac EpiLC ---
      plotText(label = "H3K27ac", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.H3K27ac.EpiLC <- plotSignal(
        data = bw_H3K27ac_EpiLC,
        params = plot_region, range = common_range_H3K27ac,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = strong_blue,
        fill = strong_blue
      )
      annoYaxis(
        plot = chip.H3K27ac.EpiLC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- EpiLC specific peaks ---
      plotText(label = "EpiLC specific peak", x = xPosition + textShift, y = yPositionShiftMini, just = "left",
               fontsize = 4, fontfamily = fontfamily)
      plotRanges(
        data = bed_H3K27ac_EpiLCspecific,
        params = plot_region,
        x = xPosition, y = yPositionShiftMini, width = width, height = height.chip,
        just = c("left", "top"),
        fill = strong_blue
      )
      # --- ChIP RAD21 EpiLC ---
      plotText(label = "RAD21", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.RAD21.EpiLC <- plotSignal(
        data = bw_RAD21_EpiLC,
        params = plot_region, range = common_range_RAD21,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = "grey30",
        fill = "grey30"
      )
      annoYaxis(
        plot = chip.RAD21.EpiLC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      
      # --- v4C G1 Differential --
      positive_scores_gr <- bg_diff_ESCvsEpiLC[bg_diff_ESCvsEpiLC$score > 0]
      negative_scores_gr <- bg_diff_ESCvsEpiLC[bg_diff_ESCvsEpiLC$score < 0]
      
      
      plotText(label = "Diff (EpiLC - ESC)", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      v4C.diff.EpiLCvsESC <- plotSignal(
        data = negative_scores_gr,
        params = plot_region,
        range = c(-diffYmax, diffYmax),
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4CDiff,
        just = c("left", "top"),
        linecolor = color_esc_dark, lwd = 0.25*mmToLinePlotgarden, baseline = FALSE,
        fill = color_esc_dark
      )
      plotSignal(
        data = positive_scores_gr,
        params = plot_region,
        range = c(-diffYmax, diffYmax),
        x = xPosition, y = yPosition + yPositionShiftSmallNum*8 + yPositionShiftMiniNum*4 + height.v4C*2 + height.chip*10 , width = width, height = height.v4CDiff,
        just = c("left", "top"),
        linecolor = color_epi_dark, lwd = 0.25*mmToLinePlotgarden, baseline = FALSE,
        fill = color_epi_dark
      )
      annoYaxis(
        plot = v4C.diff.EpiLCvsESC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      # --- Promoter loops ---
      plotText(label = "All promoter loops", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      plotPairsArches(
        data = bedpe_promoterLoops,
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C,
        just = c("left", "top"),
        fill = "black",
        linecolor = NA,
        alpha = 0.5,
        clip = TRUE
      )
      # --- genome axis ---
      plotGenes(
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C/2,
        just = c("left", "top"),
        fontsize = 4, fontcolor = c("black", "black"), fill = c("black", "black")
        
      )
      plotGenomeLabel(
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, length = width,
        fontsize = 4, fontfamily = fontfamily
      )
      dev.off()
      
      ################################################################################
      # Plot 2: Drawing dTAG vs DMSO ESC
      ################################################################################
      pdf(here(figDir, paste0("v4C_", vp, "_", res, "kb_dTAGvsDMSO_ESC_cutoff", cutoff, ".pdf")), width = 4, height = 8)
      pageCreate(width = 4, height = 8, default.units = "inches", showGuides = FALSE)
      # --- TITLE ---
      plotText(
        label = paste0("Virtual 4C, ", unlist(strsplit(vp, "\\."))[2], " (", res, "kb resolution)"),
        x = 0.2, y = 0.2,
        just = "left",
        fontface = "bold",
        fontsize = 8, fontfamily = fontfamily
      )
      # --- v4C G1 ESC DMSO---
      plotText(label = "G1.ESC.DMSO", x = xPosition + textShift, y = yPosition, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      # Plot the signal track
      v4C.ESC.DMSO <- plotSignal(
        data = bg_ESC_DMSO,
        params = plot_region,
        range = c(ymin, ymax),
        x = xPosition, y = yPosition, width = width, height = height.v4C,
        just = c("left", "top"),
        linecolor = color_esc_dark, lwd = 0.25*mmToLinePlotgarden,
        fill = color_esc_dark
      )
      annoYaxis(
        plot = v4C.ESC.DMSO,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- v4C G1 ESC dTAG---
      # The y-coordinate "0.2b" places this plot 0.2 inches below the previous one
      plotText(label = "G1.ESC.dTAG", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      v4C.ESC.dTAG <- plotSignal(
        data = bg_ESC_dTAG,
        params = plot_region,
        range = c(ymin, ymax),
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C,
        just = c("left", "top"),
        linecolor = color_esc_light, lwd = 0.25*mmToLinePlotgarden,
        fill = color_esc_light
      )
      annoYaxis(
        plot = v4C.ESC.dTAG,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      # --- ChIP H3K4me3 ESC ---
      plotText(label = "H3K4me3", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.h3k4me3.ESC <- plotSignal(
        data = bw_H3K4me3_ESC,
        params = plot_region, range = common_range_H3K4me3,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = strong_red,
        fill = strong_red
      )
      annoYaxis(
        plot = chip.h3k4me3.ESC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- ESC specific peaks ---
      plotText(label = "ESC specific peak", x = xPosition + textShift, y = yPositionShiftMini, just = "left",
               fontsize = 4, fontfamily = fontfamily)
      plotRanges(
        data = bed_H3K4me3_ESCspecific,
        params = plot_region,
        x = xPosition, y = yPositionShiftMini, width = width, height = height.chip,
        just = c("left", "top"),
        fill = strong_red
      )
      # --- ChIP H3K27ac ESC ---
      plotText(label = "H3K27ac", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.H3K27ac.ESC <- plotSignal(
        data = bw_H3K27ac_ESC,
        params = plot_region, range = common_range_H3K27ac,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = strong_blue,
        fill = strong_blue
      )
      annoYaxis(
        plot = chip.H3K27ac.ESC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- ESC specific peaks ---
      plotText(label = "ESC specific peak", x = xPosition + textShift, y = yPositionShiftMini, just = "left",
               fontsize = 4, fontfamily = fontfamily)
      plotRanges(
        data = bed_H3K27ac_ESCspecific,
        params = plot_region,
        x = xPosition, y = yPositionShiftMini, width = width, height = height.chip,
        just = c("left", "top"),
        fill = strong_blue
      )
      # --- ChIP RAD21 ESC ---
      plotText(label = "RAD21", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.RAD21.ESC <- plotSignal(
        data = bw_RAD21_ESC,
        params = plot_region, range = common_range_RAD21,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = "grey30",
        fill = "grey30"
      )
      annoYaxis(
        plot = chip.RAD21.ESC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      
      
      # --- v4C G1 Differential --
      positive_scores_gr <- bg_diff_ESC_dTAGvsDMSO[bg_diff_ESC_dTAGvsDMSO$score > 0]
      negative_scores_gr <- bg_diff_ESC_dTAGvsDMSO[bg_diff_ESC_dTAGvsDMSO$score < 0]
      
      
      plotText(label = "Diff (dTAG - DMSO)", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      v4C.diff.dTAGvsDMSO <- plotSignal(
        data = negative_scores_gr,
        params = plot_region,
        range = c(-diffYmax, diffYmax),
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4CDiff,
        just = c("left", "top"),
        linecolor = color_esc_dark, lwd = 0.25*mmToLinePlotgarden, baseline = FALSE,
        fill = color_esc_dark
      )
      plotSignal(
        data = positive_scores_gr,
        params = plot_region,
        range = c(-diffYmax, diffYmax),
        x = xPosition, y = yPosition + yPositionShiftSmallNum*5 + yPositionShiftMiniNum*2 + height.v4C*2 + height.chip*5 , width = width, height = height.v4CDiff,
        just = c("left", "top"),
        linecolor = color_esc_light, lwd = 0.25*mmToLinePlotgarden, baseline = FALSE,
        fill = color_esc_light
      )
      annoYaxis(
        plot = v4C.diff.dTAGvsDMSO,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      # --- Promoter loops ---
      plotText(label = "All promoter loops", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      plotPairsArches(
        data = bedpe_promoterLoops,
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C,
        just = c("left", "top"),
        fill = "black",
        linecolor = NA,
        alpha = 0.5,
        clip = TRUE
      )
      # --- genome axis ---
      plotGenes(
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C/2,
        just = c("left", "top"),
        fontsize = 4, fontcolor = c("black", "black"), fill = c("black", "black")
        
      )
      plotGenomeLabel(
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, length = width,
        fontsize = 4, fontfamily = fontfamily
      )
      dev.off()
      
      ################################################################################
      # Plot 3: Drawing dTAG vs DMSO EpiLC
      ################################################################################
      pdf(here(figDir, paste0("v4C_", vp, "_", res, "kb_dTAGvsDMSO_EpiLC_cutoff", cutoff, ".pdf")), width = 4, height = 8)
      pageCreate(width = 4, height = 8, default.units = "inches", showGuides = FALSE)
      # --- TITLE ---
      plotText(
        label = paste0("Virtual 4C, ", unlist(strsplit(vp, "\\."))[2], " (", res, "kb resolution)"),
        x = 0.2, y = 0.2,
        just = "left",
        fontface = "bold",
        fontsize = 8, fontfamily = fontfamily
      )
      # --- v4C G1 EpiLC DMSO---
      plotText(label = "G1.EpiLC.DMSO", x = xPosition + textShift, y = yPosition, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      # Plot the signal track
      v4C.EpiLC.DMSO <- plotSignal(
        data = bg_EpiLC_DMSO,
        params = plot_region,
        range = c(ymin, ymax),
        x = xPosition, y = yPosition, width = width, height = height.v4C,
        just = c("left", "top"),
        linecolor = color_epi_dark, lwd = 0.25*mmToLinePlotgarden,
        fill = color_epi_dark
      )
      annoYaxis(
        plot = v4C.EpiLC.DMSO,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- v4C G1 EpiLC dTAG---
      # The y-coordinate "0.2b" places this plot 0.2 inches below the previous one
      plotText(label = "G1.EpiLC.dTAG", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      v4C.EpiLC.dTAG <- plotSignal(
        data = bg_EpiLC_dTAG,
        params = plot_region,
        range = c(ymin, ymax),
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C,
        just = c("left", "top"),
        linecolor = color_epi_light, lwd = 0.25*mmToLinePlotgarden,
        fill = color_epi_light
      )
      annoYaxis(
        plot = v4C.EpiLC.dTAG,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      # --- ChIP H3K4me3 EpiLC ---
      plotText(label = "H3K4me3", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.h3k4me3.EpiLC <- plotSignal(
        data = bw_H3K4me3_EpiLC,
        params = plot_region, range = common_range_H3K4me3,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = strong_red,
        fill = strong_red
      )
      annoYaxis(
        plot = chip.h3k4me3.EpiLC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- EpiLC specific peaks ---
      plotText(label = "EpiLC specific peak", x = xPosition + textShift, y = yPositionShiftMini, just = "left",
               fontsize = 4, fontfamily = fontfamily)
      plotRanges(
        data = bed_H3K4me3_EpiLCspecific,
        params = plot_region,
        x = xPosition, y = yPositionShiftMini, width = width, height = height.chip,
        just = c("left", "top"),
        fill = strong_red
      )
      # --- ChIP H3K27ac EpiLC ---
      plotText(label = "H3K27ac", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.H3K27ac.EpiLC <- plotSignal(
        data = bw_H3K27ac_EpiLC,
        params = plot_region, range = common_range_H3K27ac,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = strong_blue,
        fill = strong_blue
      )
      annoYaxis(
        plot = chip.H3K27ac.EpiLC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- EpiLC specific peaks ---
      plotText(label = "EpiLC specific peak", x = xPosition + textShift, y = yPositionShiftMini, just = "left",
               fontsize = 4, fontfamily = fontfamily)
      plotRanges(
        data = bed_H3K27ac_EpiLCspecific,
        params = plot_region,
        x = xPosition, y = yPositionShiftMini, width = width, height = height.chip,
        just = c("left", "top"),
        fill = strong_blue
      )
      # --- ChIP RAD21 EpiLC ---
      plotText(label = "RAD21", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.RAD21.EpiLC <- plotSignal(
        data = bw_RAD21_EpiLC,
        params = plot_region, range = common_range_RAD21,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = "grey30",
        fill = "grey30"
      )
      annoYaxis(
        plot = chip.RAD21.EpiLC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      
      
      # --- v4C G1 Differential --
      positive_scores_gr <- bg_diff_EpiLC_dTAGvsDMSO[bg_diff_EpiLC_dTAGvsDMSO$score > 0]
      negative_scores_gr <- bg_diff_EpiLC_dTAGvsDMSO[bg_diff_EpiLC_dTAGvsDMSO$score < 0]
      
      
      plotText(label = "Diff (dTAG - DMSO)", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      v4C.diff.dTAGvsDMSO <- plotSignal(
        data = negative_scores_gr,
        params = plot_region,
        range = c(-diffYmax, diffYmax),
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4CDiff,
        just = c("left", "top"),
        linecolor = color_epi_dark, lwd = 0.25*mmToLinePlotgarden, baseline = FALSE,
        fill = color_epi_dark
      )
      plotSignal(
        data = positive_scores_gr,
        params = plot_region,
        range = c(-diffYmax, diffYmax),
        x = xPosition, y = yPosition + yPositionShiftSmallNum*5 + yPositionShiftMiniNum*2 + height.v4C*2 + height.chip*5 , width = width, height = height.v4CDiff,
        just = c("left", "top"),
        linecolor = color_epi_light, lwd = 0.25*mmToLinePlotgarden, baseline = FALSE,
        fill = color_epi_light
      )
      annoYaxis(
        plot = v4C.diff.dTAGvsDMSO,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      # --- Promoter loops ---
      plotText(label = "All promoter loops", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      plotPairsArches(
        data = bedpe_promoterLoops,
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C,
        just = c("left", "top"),
        fill = "black",
        linecolor = NA,
        alpha = 0.5,
        clip = TRUE
      )
      # --- genome axis ---
      plotGenes(
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C/2,
        just = c("left", "top"),
        fontsize = 4, fontcolor = c("black", "black"), fill = c("black", "black")
        
      )
      plotGenomeLabel(
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, length = width,
        fontsize = 4, fontfamily = fontfamily
      )
      dev.off()
      
    }
  }
}
```
### Custom range
```{r}

draw_v4C <- function(vp, res, xmin, xmax, ymax, diffYmax){
  
      ################################################################################
      # IMPORTING TRACKS
      ################################################################################
      # Import bedgraph
      tb_ESC_DMSO <- fread(here(bgDir, paste0("v4C_", vp, "__", "G1DMSO_pooled", "__", res, "kb_localNorm.bedGraph" )))
      colnames(tb_ESC_DMSO) <- c("chrom", "start", "end", "score")
      bg_ESC_DMSO <- makeGRangesFromDataFrame(tb_ESC_DMSO, keep.extra.columns = TRUE)
      keep <- countOverlaps(bg_ESC_DMSO) == 1
      bg_ESC_DMSO <- bg_ESC_DMSO[keep]
      # Import bedgraph
      tb_EpiLC_DMSO <- fread(here(bgDir, paste0("v4C_", vp, "__", "EpiG1DMSO_pooled", "__", res, "kb_localNorm.bedGraph" )))
      colnames(tb_EpiLC_DMSO) <- c("chrom", "start", "end", "score")
      bg_EpiLC_DMSO <- makeGRangesFromDataFrame(tb_EpiLC_DMSO, keep.extra.columns = TRUE)
      keep <- countOverlaps(bg_EpiLC_DMSO) == 1
      bg_EpiLC_DMSO <- bg_EpiLC_DMSO[keep]
      
      # Import bedgraph
      tb_ESC_dTAG <- fread(here(bgDir, paste0("v4C_", vp, "__", "G1dTAG_pooled", "__", res, "kb_localNorm.bedGraph" )))
      colnames(tb_ESC_dTAG) <- c("chrom", "start", "end", "score")
      bg_ESC_dTAG <- makeGRangesFromDataFrame(tb_ESC_dTAG, keep.extra.columns = TRUE)
      keep <- countOverlaps(bg_ESC_dTAG) == 1
      bg_ESC_dTAG <- bg_ESC_dTAG[keep]
      # Import bedgraph
      tb_EpiLC_dTAG <- fread(here(bgDir, paste0("v4C_", vp, "__", "EpiG1dTAG_pooled", "__", res, "kb_localNorm.bedGraph" )))
      colnames(tb_EpiLC_dTAG) <- c("chrom", "start", "end", "score")
      bg_EpiLC_dTAG <- makeGRangesFromDataFrame(tb_EpiLC_dTAG, keep.extra.columns = TRUE)
      keep <- countOverlaps(bg_EpiLC_dTAG) == 1
      bg_EpiLC_dTAG <- bg_EpiLC_dTAG[keep]
      
      # Make differential bedgraph
      tb_diff_ESCvsEpiLC <- tb_ESC_DMSO %>% left_join(tb_EpiLC_DMSO, by = c("chrom", "start", "end")) %>%
        dplyr::mutate(score = score.y - score.x) %>%
        dplyr::select(chrom, start, end, score)
      bg_diff_ESCvsEpiLC <- makeGRangesFromDataFrame(tb_diff_ESCvsEpiLC, keep.extra.columns = TRUE)
      keep <- countOverlaps(bg_diff_ESCvsEpiLC) == 1
      bg_diff_ESCvsEpiLC <- bg_diff_ESCvsEpiLC[keep]
      
      # Make differential bedgraph
      tb_diff_ESC_dTAGvsDMSO <- tb_ESC_DMSO %>% left_join(tb_ESC_dTAG, by = c("chrom", "start", "end")) %>%
        dplyr::mutate(score = score.y - score.x) %>%
        dplyr::select(chrom, start, end, score)
      bg_diff_ESC_dTAGvsDMSO <- makeGRangesFromDataFrame(tb_diff_ESC_dTAGvsDMSO, keep.extra.columns = TRUE)
      keep <- countOverlaps(bg_diff_ESC_dTAGvsDMSO) == 1
      bg_diff_ESC_dTAGvsDMSO <- bg_diff_ESC_dTAGvsDMSO[keep]
      
      # Make differential bedgraph
      tb_diff_EpiLC_dTAGvsDMSO <- tb_EpiLC_DMSO %>% left_join(tb_EpiLC_dTAG, by = c("chrom", "start", "end")) %>%
        dplyr::mutate(score = score.y - score.x) %>%
        dplyr::select(chrom, start, end, score)
      bg_diff_EpiLC_dTAGvsDMSO <- makeGRangesFromDataFrame(tb_diff_EpiLC_dTAGvsDMSO, keep.extra.columns = TRUE)
      keep <- countOverlaps(bg_diff_EpiLC_dTAGvsDMSO) == 1
      bg_diff_EpiLC_dTAGvsDMSO <- bg_diff_EpiLC_dTAGvsDMSO[keep]
      
      # CHIP TRACKS
      bw_H3K4me3_ESC <- here(refDir, "ChIP", "H3K4me3", "H3K4me3_ESC_avg_depthNorm.bw")
      bed_H3K4me3_ESCspecific <- here(refDir, "ChIP", "differential", "H3K4me3_diff_specific_ESC.bed")
      bw_H3K27ac_ESC <- here(refDir, "ChIP", "H3K27ac", "H3K27ac_ESC_avg_depthNorm.bw")
      bed_H3K27ac_ESCspecific <- here(refDir, "ChIP", "differential", "H3K27ac_diff_specific_ESC.bed")
      bw_H3K4me3_EpiLC <- here(refDir, "ChIP", "H3K4me3", "H3K4me3_EpiLC_avg_depthNorm.bw")
      bed_H3K4me3_EpiLCspecific <- here(refDir, "ChIP", "differential", "H3K4me3_diff_specific_EpiLC.bed")
      bw_H3K27ac_EpiLC <- here(refDir, "ChIP", "H3K27ac", "H3K27ac_EpiLC_avg_depthNorm.bw")
      bed_H3K27ac_EpiLCspecific <- here(refDir, "ChIP", "differential", "H3K27ac_diff_specific_EpiLC.bed")
      bw_RAD21_ESC <- here(refDir, "ChIP", "RAD21-HA", "RAD21_ESC-DMSO_avg_spikeNorm_inputNorm.bw") 
      bw_RAD21_EpiLC <- here(refDir, "ChIP", "RAD21-HA", "RAD21_EpiLC_avg_spikeNorm_inputNorm.bw") 
      
      # PROMOTER LOOPS
      bedpe_promoterLoops <- here(loopDir, "chromosight_score_complex_allP.bedpe")
      ################################################################################
      # DEFINING RANGES
      ################################################################################
      # Y ranges for virtual 4C
      ymin <- 0
      
      # X ranges 
      data <- as_tibble(fread(here(bgDir, paste0("v4C_", vp, "__", "G1DMSO_pooled", "__", res, "kb_localNorm.bedGraph" ))))
      
      chr <- data$V1[1]
      
      # Y ranges for ChIP
      region_gr <- GRanges(seqnames = chr,
                           ranges = IRanges(start = xmin,
                                            end = xmax))
      ## H3K4me3 range
      bw_subset1 <- rtracklayer::import.bw(bw_H3K4me3_ESC, which = region_gr)
      bw_subset2 <- rtracklayer::import.bw(bw_H3K4me3_EpiLC, which = region_gr)
      max_score <- max(bw_subset1$score, bw_subset2$score)
      common_range_H3K4me3 <- c(0, ceiling(max_score / 0.1) * 0.1)
      
      ## H3K27ac range
      bw_subset1 <- rtracklayer::import.bw(bw_H3K27ac_ESC, which = region_gr)
      bw_subset2 <- rtracklayer::import.bw(bw_H3K27ac_EpiLC, which = region_gr)
      max_score <- max(bw_subset1$score, bw_subset2$score)
      common_range_H3K27ac <- c(0, ceiling(max_score / 0.1) * 0.1)
      
      
      ## RAD21 range
      bw_subset1 <- rtracklayer::import.bw(bw_RAD21_ESC, which = region_gr)
      bw_subset2 <- rtracklayer::import.bw(bw_RAD21_EpiLC, which = region_gr)
      max_score <- max(bw_subset1$score, bw_subset2$score)
      common_range_RAD21 <- c(0, ceiling(max_score / 0.1) * 0.1)
      
      ## Defining the genomic region to plot
      plot_region <- pgParams(
        chrom = chr,
        chromstart = xmin,
        chromend = xmax,
        assembly = "mm10"
      )
      # Plotting parameters
      xPosition <- 0.25
      yPosition <- 0.4
      width <- 3.5
      height.v4C <- 0.5
      height.v4CDiff <- 0.3
      height.chip <- 0.1
      height.chipPeak <- 0.05
      yPositionShift <- "0.15b"
      yPositionShiftSmall <- "0.05b"
      yPositionShiftMini <- "0.025b"
      
      yPositionShiftNum <- 0.15
      yPositionShiftSmallNum <- 0.05
      yPositionShiftMiniNum <- 0.025
      textShift <- 0.05
      fontfamily <- "Helvetica"
      
      ################################################################################
      # Plot 1: Drawing ESC vs EpiLC DMSO
      ################################################################################
      pdf(here(figDir, paste0("v4C_", vp, "_", res, "kb_EpiLCvsESC_", xmin, "_", xmax, ".pdf")), width = 4, height = 8)
      pageCreate(width = 4, height = 8, default.units = "inches", showGuides = FALSE)
      # --- TITLE ---
      plotText(
        label = paste0("Virtual 4C, ", unlist(strsplit(vp, "\\."))[2], " (", res, "kb resolution)"),
        x = 0.2, y = 0.2,
        just = "left",
        fontface = "bold",
        fontsize = 8, fontfamily = fontfamily
      )
      # --- v4C G1 ESC DMSO---
      plotText(label = "G1.ESC.DMSO", x = xPosition + textShift, y = yPosition, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      # Plot the signal track
      v4C.ESC.DMSO <- plotSignal(
        data = bg_ESC_DMSO,
        params = plot_region,
        range = c(ymin, ymax),
        x = xPosition, y = yPosition, width = width, height = height.v4C,
        just = c("left", "top"),
        linecolor = color_esc_dark, lwd = 0.25*mmToLinePlotgarden,
        fill = color_esc_dark
      )
      annoYaxis(
        plot = v4C.ESC.DMSO,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- ChIP H3K4me3 ESC ---
      plotText(label = "H3K4me3", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.h3k4me3.ESC <- plotSignal(
        data = bw_H3K4me3_ESC,
        params = plot_region, range = common_range_H3K4me3,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = strong_red,
        fill = strong_red
      )
      annoYaxis(
        plot = chip.h3k4me3.ESC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- ESC specific peaks ---
      plotText(label = "ESC specific peak", x = xPosition + textShift, y = yPositionShiftMini, just = "left",
               fontsize = 4, fontfamily = fontfamily)
      plotRanges(
        data = bed_H3K4me3_ESCspecific,
        params = plot_region,
        x = xPosition, y = yPositionShiftMini, width = width, height = height.chip,
        just = c("left", "top"),
        fill = strong_red
      )
      # --- ChIP H3K27ac ESC ---
      plotText(label = "H3K27ac", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.H3K27ac.ESC <- plotSignal(
        data = bw_H3K27ac_ESC,
        params = plot_region, range = common_range_H3K27ac,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = strong_blue,
        fill = strong_blue
      )
      annoYaxis(
        plot = chip.H3K27ac.ESC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- ESC specific peaks ---
      plotText(label = "ESC specific peak", x = xPosition + textShift, y = yPositionShiftMini, just = "left",
               fontsize = 4, fontfamily = fontfamily)
      plotRanges(
        data = bed_H3K27ac_ESCspecific,
        params = plot_region,
        x = xPosition, y = yPositionShiftMini, width = width, height = height.chip,
        just = c("left", "top"),
        fill = strong_blue
      )
      # --- ChIP RAD21 ESC ---
      plotText(label = "RAD21", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.RAD21.ESC <- plotSignal(
        data = bw_RAD21_ESC,
        params = plot_region, range = common_range_RAD21,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = "grey30",
        fill = "grey30"
      )
      annoYaxis(
        plot = chip.RAD21.ESC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      
      # --- v4C G1 EpiLC DMSO---
      # The y-coordinate "0.2b" places this plot 0.2 inches below the previous one
      plotText(label = "G1.EpiLC.DMSO", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      v4C.EpiLC.DMSO <- plotSignal(
        data = bg_EpiLC_DMSO,
        params = plot_region,
        range = c(ymin, ymax),
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C,
        just = c("left", "top"),
        linecolor = color_epi_dark, lwd = 0.25*mmToLinePlotgarden,
        fill = color_epi_dark
      )
      annoYaxis(
        plot = v4C.EpiLC.DMSO,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      # --- ChIP H3K4me3 EpiLC ---
      plotText(label = "H3K4me3", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.h3k4me3.EpiLC <- plotSignal(
        data = bw_H3K4me3_EpiLC,
        params = plot_region, range = common_range_H3K4me3,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = strong_red,
        fill = strong_red
      )
      annoYaxis(
        plot = chip.h3k4me3.EpiLC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- EpiLC specific peaks ---
      plotText(label = "EpiLC specific peak", x = xPosition + textShift, y = yPositionShiftMini, just = "left",
               fontsize = 4, fontfamily = fontfamily)
      plotRanges(
        data = bed_H3K4me3_EpiLCspecific,
        params = plot_region,
        x = xPosition, y = yPositionShiftMini, width = width, height = height.chip,
        just = c("left", "top"),
        fill = strong_red
      )
      # --- ChIP H3K27ac EpiLC ---
      plotText(label = "H3K27ac", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.H3K27ac.EpiLC <- plotSignal(
        data = bw_H3K27ac_EpiLC,
        params = plot_region, range = common_range_H3K27ac,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = strong_blue,
        fill = strong_blue
      )
      annoYaxis(
        plot = chip.H3K27ac.EpiLC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- EpiLC specific peaks ---
      plotText(label = "EpiLC specific peak", x = xPosition + textShift, y = yPositionShiftMini, just = "left",
               fontsize = 4, fontfamily = fontfamily)
      plotRanges(
        data = bed_H3K27ac_EpiLCspecific,
        params = plot_region,
        x = xPosition, y = yPositionShiftMini, width = width, height = height.chip,
        just = c("left", "top"),
        fill = strong_blue
      )
      # --- ChIP RAD21 EpiLC ---
      plotText(label = "RAD21", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.RAD21.EpiLC <- plotSignal(
        data = bw_RAD21_EpiLC,
        params = plot_region, range = common_range_RAD21,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = "grey30",
        fill = "grey30"
      )
      annoYaxis(
        plot = chip.RAD21.EpiLC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      
      # --- v4C G1 Differential --
      positive_scores_gr <- bg_diff_ESCvsEpiLC[bg_diff_ESCvsEpiLC$score > 0]
      negative_scores_gr <- bg_diff_ESCvsEpiLC[bg_diff_ESCvsEpiLC$score < 0]
      
      
      plotText(label = "Diff (EpiLC - ESC)", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      v4C.diff.EpiLCvsESC <- plotSignal(
        data = negative_scores_gr,
        params = plot_region,
        range = c(-diffYmax, diffYmax),
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4CDiff,
        just = c("left", "top"),
        linecolor = color_esc_dark, lwd = 0.25*mmToLinePlotgarden, baseline = FALSE,
        fill = color_esc_dark
      )
      plotSignal(
        data = positive_scores_gr,
        params = plot_region,
        range = c(-diffYmax, diffYmax),
        x = xPosition, y = yPosition + yPositionShiftSmallNum*8 + yPositionShiftMiniNum*4 + height.v4C*2 + height.chip*10 , width = width, height = height.v4CDiff,
        just = c("left", "top"),
        linecolor = color_epi_dark, lwd = 0.25*mmToLinePlotgarden, baseline = FALSE,
        fill = color_epi_dark
      )
      annoYaxis(
        plot = v4C.diff.EpiLCvsESC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      # --- Promoter loops ---
      plotText(label = "All promoter loops", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      plotPairsArches(
        data = bedpe_promoterLoops,
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C,
        just = c("left", "top"),
        fill = "black",
        linecolor = NA,
        alpha = 0.5,
        clip = TRUE
      )
      # --- genome axis ---
      plotGenes(
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C/2,
        just = c("left", "top"),
        fontsize = 4, fontcolor = c("black", "black"), fill = c("black", "black")
        
      )
      plotGenomeLabel(
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, length = width,
        fontsize = 4, fontfamily = fontfamily
      )
      dev.off()
      
      ################################################################################
      # Plot 2: Drawing dTAG vs DMSO ESC
      ################################################################################
      pdf(here(figDir, paste0("v4C_", vp, "_", res, "kb_dTAGvsDMSO_ESC_", xmin, "_", xmax, ".pdf")), width = 4, height = 8)
      pageCreate(width = 4, height = 8, default.units = "inches", showGuides = FALSE)
      # --- TITLE ---
      plotText(
        label = paste0("Virtual 4C, ", unlist(strsplit(vp, "\\."))[2], " (", res, "kb resolution)"),
        x = 0.2, y = 0.2,
        just = "left",
        fontface = "bold",
        fontsize = 8, fontfamily = fontfamily
      )
      # --- v4C G1 ESC DMSO---
      plotText(label = "G1.ESC.DMSO", x = xPosition + textShift, y = yPosition, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      # Plot the signal track
      v4C.ESC.DMSO <- plotSignal(
        data = bg_ESC_DMSO,
        params = plot_region,
        range = c(ymin, ymax),
        x = xPosition, y = yPosition, width = width, height = height.v4C,
        just = c("left", "top"),
        linecolor = color_esc_dark, lwd = 0.25*mmToLinePlotgarden,
        fill = color_esc_dark
      )
      annoYaxis(
        plot = v4C.ESC.DMSO,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- v4C G1 ESC dTAG---
      # The y-coordinate "0.2b" places this plot 0.2 inches below the previous one
      plotText(label = "G1.ESC.dTAG", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      v4C.ESC.dTAG <- plotSignal(
        data = bg_ESC_dTAG,
        params = plot_region,
        range = c(ymin, ymax),
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C,
        just = c("left", "top"),
        linecolor = color_esc_light, lwd = 0.25*mmToLinePlotgarden,
        fill = color_esc_light
      )
      annoYaxis(
        plot = v4C.ESC.dTAG,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      # --- ChIP H3K4me3 ESC ---
      plotText(label = "H3K4me3", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.h3k4me3.ESC <- plotSignal(
        data = bw_H3K4me3_ESC,
        params = plot_region, range = common_range_H3K4me3,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = strong_red,
        fill = strong_red
      )
      annoYaxis(
        plot = chip.h3k4me3.ESC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- ESC specific peaks ---
      plotText(label = "ESC specific peak", x = xPosition + textShift, y = yPositionShiftMini, just = "left",
               fontsize = 4, fontfamily = fontfamily)
      plotRanges(
        data = bed_H3K4me3_ESCspecific,
        params = plot_region,
        x = xPosition, y = yPositionShiftMini, width = width, height = height.chip,
        just = c("left", "top"),
        fill = strong_red
      )
      # --- ChIP H3K27ac ESC ---
      plotText(label = "H3K27ac", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.H3K27ac.ESC <- plotSignal(
        data = bw_H3K27ac_ESC,
        params = plot_region, range = common_range_H3K27ac,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = strong_blue,
        fill = strong_blue
      )
      annoYaxis(
        plot = chip.H3K27ac.ESC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- ESC specific peaks ---
      plotText(label = "ESC specific peak", x = xPosition + textShift, y = yPositionShiftMini, just = "left",
               fontsize = 4, fontfamily = fontfamily)
      plotRanges(
        data = bed_H3K27ac_ESCspecific,
        params = plot_region,
        x = xPosition, y = yPositionShiftMini, width = width, height = height.chip,
        just = c("left", "top"),
        fill = strong_blue
      )
      # --- ChIP RAD21 ESC ---
      plotText(label = "RAD21", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.RAD21.ESC <- plotSignal(
        data = bw_RAD21_ESC,
        params = plot_region, range = common_range_RAD21,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = "grey30",
        fill = "grey30"
      )
      annoYaxis(
        plot = chip.RAD21.ESC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      
      
      # --- v4C G1 Differential --
      positive_scores_gr <- bg_diff_ESC_dTAGvsDMSO[bg_diff_ESC_dTAGvsDMSO$score > 0]
      negative_scores_gr <- bg_diff_ESC_dTAGvsDMSO[bg_diff_ESC_dTAGvsDMSO$score < 0]
      
      
      plotText(label = "Diff (dTAG - DMSO)", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      v4C.diff.dTAGvsDMSO <- plotSignal(
        data = negative_scores_gr,
        params = plot_region,
        range = c(-diffYmax, diffYmax),
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4CDiff,
        just = c("left", "top"),
        linecolor = color_esc_dark, lwd = 0.25*mmToLinePlotgarden, baseline = FALSE,
        fill = color_esc_dark
      )
      plotSignal(
        data = positive_scores_gr,
        params = plot_region,
        range = c(-diffYmax, diffYmax),
        x = xPosition, y = yPosition + yPositionShiftSmallNum*5 + yPositionShiftMiniNum*2 + height.v4C*2 + height.chip*5 , width = width, height = height.v4CDiff,
        just = c("left", "top"),
        linecolor = color_esc_light, lwd = 0.25*mmToLinePlotgarden, baseline = FALSE,
        fill = color_esc_light
      )
      annoYaxis(
        plot = v4C.diff.dTAGvsDMSO,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      # --- Promoter loops ---
      plotText(label = "All promoter loops", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      plotPairsArches(
        data = bedpe_promoterLoops,
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C,
        just = c("left", "top"),
        fill = "black",
        linecolor = NA,
        alpha = 0.5,
        clip = TRUE
      )
      # --- genome axis ---
      plotGenes(
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C/2,
        just = c("left", "top"),
        fontsize = 4, fontcolor = c("black", "black"), fill = c("black", "black")
        
      )
      plotGenomeLabel(
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, length = width,
        fontsize = 4, fontfamily = fontfamily
      )
      dev.off()
      
      ################################################################################
      # Plot 3: Drawing dTAG vs DMSO EpiLC
      ################################################################################
      pdf(here(figDir, paste0("v4C_", vp, "_", res, "kb_dTAGvsDMSO_EpiLC_", xmin, "_", xmax, ".pdf")), width = 4, height = 8)
      pageCreate(width = 4, height = 8, default.units = "inches", showGuides = FALSE)
      # --- TITLE ---
      plotText(
        label = paste0("Virtual 4C, ", unlist(strsplit(vp, "\\."))[2], " (", res, "kb resolution)"),
        x = 0.2, y = 0.2,
        just = "left",
        fontface = "bold",
        fontsize = 8, fontfamily = fontfamily
      )
      # --- v4C G1 EpiLC DMSO---
      plotText(label = "G1.EpiLC.DMSO", x = xPosition + textShift, y = yPosition, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      # Plot the signal track
      v4C.EpiLC.DMSO <- plotSignal(
        data = bg_EpiLC_DMSO,
        params = plot_region,
        range = c(ymin, ymax),
        x = xPosition, y = yPosition, width = width, height = height.v4C,
        just = c("left", "top"),
        linecolor = color_epi_dark, lwd = 0.25*mmToLinePlotgarden,
        fill = color_epi_dark
      )
      annoYaxis(
        plot = v4C.EpiLC.DMSO,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- v4C G1 EpiLC dTAG---
      # The y-coordinate "0.2b" places this plot 0.2 inches below the previous one
      plotText(label = "G1.EpiLC.dTAG", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      v4C.EpiLC.dTAG <- plotSignal(
        data = bg_EpiLC_dTAG,
        params = plot_region,
        range = c(ymin, ymax),
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C,
        just = c("left", "top"),
        linecolor = color_epi_light, lwd = 0.25*mmToLinePlotgarden,
        fill = color_epi_light
      )
      annoYaxis(
        plot = v4C.EpiLC.dTAG,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      # --- ChIP H3K4me3 EpiLC ---
      plotText(label = "H3K4me3", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.h3k4me3.EpiLC <- plotSignal(
        data = bw_H3K4me3_EpiLC,
        params = plot_region, range = common_range_H3K4me3,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = strong_red,
        fill = strong_red
      )
      annoYaxis(
        plot = chip.h3k4me3.EpiLC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- EpiLC specific peaks ---
      plotText(label = "EpiLC specific peak", x = xPosition + textShift, y = yPositionShiftMini, just = "left",
               fontsize = 4, fontfamily = fontfamily)
      plotRanges(
        data = bed_H3K4me3_EpiLCspecific,
        params = plot_region,
        x = xPosition, y = yPositionShiftMini, width = width, height = height.chip,
        just = c("left", "top"),
        fill = strong_red
      )
      # --- ChIP H3K27ac EpiLC ---
      plotText(label = "H3K27ac", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.H3K27ac.EpiLC <- plotSignal(
        data = bw_H3K27ac_EpiLC,
        params = plot_region, range = common_range_H3K27ac,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = strong_blue,
        fill = strong_blue
      )
      annoYaxis(
        plot = chip.H3K27ac.EpiLC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- EpiLC specific peaks ---
      plotText(label = "EpiLC specific peak", x = xPosition + textShift, y = yPositionShiftMini, just = "left",
               fontsize = 4, fontfamily = fontfamily)
      plotRanges(
        data = bed_H3K27ac_EpiLCspecific,
        params = plot_region,
        x = xPosition, y = yPositionShiftMini, width = width, height = height.chip,
        just = c("left", "top"),
        fill = strong_blue
      )
      # --- ChIP RAD21 EpiLC ---
      plotText(label = "RAD21", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.RAD21.EpiLC <- plotSignal(
        data = bw_RAD21_EpiLC,
        params = plot_region, range = common_range_RAD21,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = "grey30",
        fill = "grey30"
      )
      annoYaxis(
        plot = chip.RAD21.EpiLC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      
      
      # --- v4C G1 Differential --
      positive_scores_gr <- bg_diff_EpiLC_dTAGvsDMSO[bg_diff_EpiLC_dTAGvsDMSO$score > 0]
      negative_scores_gr <- bg_diff_EpiLC_dTAGvsDMSO[bg_diff_EpiLC_dTAGvsDMSO$score < 0]
      
      
      plotText(label = "Diff (dTAG - DMSO)", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      v4C.diff.dTAGvsDMSO <- plotSignal(
        data = negative_scores_gr,
        params = plot_region,
        range = c(-diffYmax, diffYmax),
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4CDiff,
        just = c("left", "top"),
        linecolor = color_epi_dark, lwd = 0.25*mmToLinePlotgarden, baseline = FALSE,
        fill = color_epi_dark
      )
      plotSignal(
        data = positive_scores_gr,
        params = plot_region,
        range = c(-diffYmax, diffYmax),
        x = xPosition, y = yPosition + yPositionShiftSmallNum*5 + yPositionShiftMiniNum*2 + height.v4C*2 + height.chip*5 , width = width, height = height.v4CDiff,
        just = c("left", "top"),
        linecolor = color_epi_light, lwd = 0.25*mmToLinePlotgarden, baseline = FALSE,
        fill = color_epi_light
      )
      annoYaxis(
        plot = v4C.diff.dTAGvsDMSO,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      # --- Promoter loops ---
      plotText(label = "All promoter loops", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      plotPairsArches(
        data = bedpe_promoterLoops,
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C,
        just = c("left", "top"),
        fill = "black",
        linecolor = NA,
        alpha = 0.5,
        clip = TRUE
      )
      # --- genome axis ---
      plotGenes(
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C/2,
        just = c("left", "top"),
        fontsize = 4, fontcolor = c("black", "black"), fill = c("black", "black")
        
      )
      plotGenomeLabel(
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, length = width,
        fontsize = 4, fontfamily = fontfamily
      )
      dev.off()
      
}

draw_v4C(vp = "ENSMUSG00000003032.Klf4", res = 5,  
         xmin = 55400000, xmax = 55650000, 
         ymax = 1000, diffYmax = 300)

draw_v4C(vp = "ENSMUSG00000018604.Tbx3", res = 10,  
         xmin = 119400000, xmax = 119800000, 
         ymax = 1000, diffYmax = 400)


draw_v4C(vp = "ENSMUSG00000022346.Myc", res = 10,  
         xmin = 61850000, xmax = 62400000, 
         ymax = 1500, diffYmax = 500)

draw_v4C(vp = "ENSMUSG00000029135.Fosl2", res = 5,  
         xmin = 31500000, xmax = 32300000, 
         ymax = 800, diffYmax = 300)


draw_v4C(vp = "ENSMUSG00000007880.Arid1a", res = 10,  
         xmin = 133320000, xmax = 134000000, 
         ymax = 1500, diffYmax = 600)



```

### Plotting with 4C
```{r}

draw_v4C(vp = "ENSMUSG00000003032.Klf4", res = 5,  
         xmin = 55400000, xmax = 55650000, 
         ymax = 1000, diffYmax = 300)


draw_v4C <- function(vp, res, xmin, xmax, ymax, diffYmax){
  
      ################################################################################
      # IMPORTING TRACKS
      ################################################################################
      # 4C TRACKS
      bw_H3K4me3_ESC <- here(refDir, "ChIP", "H3K4me3", "H3K4me3_ESC_avg_depthNorm.bw")
      bed_H3K4me3_ESCspecific <- here(refDir, "ChIP", "differential", "H3K4me3_diff_specific_ESC.bed")
      
      # Import bedgraph
      tb_ESC_DMSO <- fread(here(bgDir, paste0("v4C_", vp, "__", "G1DMSO_pooled", "__", res, "kb_localNorm.bedGraph" )))
      colnames(tb_ESC_DMSO) <- c("chrom", "start", "end", "score")
      bg_ESC_DMSO <- makeGRangesFromDataFrame(tb_ESC_DMSO, keep.extra.columns = TRUE)
      keep <- countOverlaps(bg_ESC_DMSO) == 1
      bg_ESC_DMSO <- bg_ESC_DMSO[keep]
      # Import bedgraph
      tb_EpiLC_DMSO <- fread(here(bgDir, paste0("v4C_", vp, "__", "EpiG1DMSO_pooled", "__", res, "kb_localNorm.bedGraph" )))
      colnames(tb_EpiLC_DMSO) <- c("chrom", "start", "end", "score")
      bg_EpiLC_DMSO <- makeGRangesFromDataFrame(tb_EpiLC_DMSO, keep.extra.columns = TRUE)
      keep <- countOverlaps(bg_EpiLC_DMSO) == 1
      bg_EpiLC_DMSO <- bg_EpiLC_DMSO[keep]
      
      # Import bedgraph
      tb_ESC_dTAG <- fread(here(bgDir, paste0("v4C_", vp, "__", "G1dTAG_pooled", "__", res, "kb_localNorm.bedGraph" )))
      colnames(tb_ESC_dTAG) <- c("chrom", "start", "end", "score")
      bg_ESC_dTAG <- makeGRangesFromDataFrame(tb_ESC_dTAG, keep.extra.columns = TRUE)
      keep <- countOverlaps(bg_ESC_dTAG) == 1
      bg_ESC_dTAG <- bg_ESC_dTAG[keep]
      # Import bedgraph
      tb_EpiLC_dTAG <- fread(here(bgDir, paste0("v4C_", vp, "__", "EpiG1dTAG_pooled", "__", res, "kb_localNorm.bedGraph" )))
      colnames(tb_EpiLC_dTAG) <- c("chrom", "start", "end", "score")
      bg_EpiLC_dTAG <- makeGRangesFromDataFrame(tb_EpiLC_dTAG, keep.extra.columns = TRUE)
      keep <- countOverlaps(bg_EpiLC_dTAG) == 1
      bg_EpiLC_dTAG <- bg_EpiLC_dTAG[keep]
      
      # Make differential bedgraph
      tb_diff_ESCvsEpiLC <- tb_ESC_DMSO %>% left_join(tb_EpiLC_DMSO, by = c("chrom", "start", "end")) %>%
        dplyr::mutate(score = score.y - score.x) %>%
        dplyr::select(chrom, start, end, score)
      bg_diff_ESCvsEpiLC <- makeGRangesFromDataFrame(tb_diff_ESCvsEpiLC, keep.extra.columns = TRUE)
      keep <- countOverlaps(bg_diff_ESCvsEpiLC) == 1
      bg_diff_ESCvsEpiLC <- bg_diff_ESCvsEpiLC[keep]
      
      # Make differential bedgraph
      tb_diff_ESC_dTAGvsDMSO <- tb_ESC_DMSO %>% left_join(tb_ESC_dTAG, by = c("chrom", "start", "end")) %>%
        dplyr::mutate(score = score.y - score.x) %>%
        dplyr::select(chrom, start, end, score)
      bg_diff_ESC_dTAGvsDMSO <- makeGRangesFromDataFrame(tb_diff_ESC_dTAGvsDMSO, keep.extra.columns = TRUE)
      keep <- countOverlaps(bg_diff_ESC_dTAGvsDMSO) == 1
      bg_diff_ESC_dTAGvsDMSO <- bg_diff_ESC_dTAGvsDMSO[keep]
      
      # Make differential bedgraph
      tb_diff_EpiLC_dTAGvsDMSO <- tb_EpiLC_DMSO %>% left_join(tb_EpiLC_dTAG, by = c("chrom", "start", "end")) %>%
        dplyr::mutate(score = score.y - score.x) %>%
        dplyr::select(chrom, start, end, score)
      bg_diff_EpiLC_dTAGvsDMSO <- makeGRangesFromDataFrame(tb_diff_EpiLC_dTAGvsDMSO, keep.extra.columns = TRUE)
      keep <- countOverlaps(bg_diff_EpiLC_dTAGvsDMSO) == 1
      bg_diff_EpiLC_dTAGvsDMSO <- bg_diff_EpiLC_dTAGvsDMSO[keep]
      
      # CHIP TRACKS
      bw_H3K4me3_ESC <- here(refDir, "ChIP", "H3K4me3", "H3K4me3_ESC_avg_depthNorm.bw")
      bed_H3K4me3_ESCspecific <- here(refDir, "ChIP", "differential", "H3K4me3_diff_specific_ESC.bed")
      bw_H3K27ac_ESC <- here(refDir, "ChIP", "H3K27ac", "H3K27ac_ESC_avg_depthNorm.bw")
      bed_H3K27ac_ESCspecific <- here(refDir, "ChIP", "differential", "H3K27ac_diff_specific_ESC.bed")
      bw_H3K4me3_EpiLC <- here(refDir, "ChIP", "H3K4me3", "H3K4me3_EpiLC_avg_depthNorm.bw")
      bed_H3K4me3_EpiLCspecific <- here(refDir, "ChIP", "differential", "H3K4me3_diff_specific_EpiLC.bed")
      bw_H3K27ac_EpiLC <- here(refDir, "ChIP", "H3K27ac", "H3K27ac_EpiLC_avg_depthNorm.bw")
      bed_H3K27ac_EpiLCspecific <- here(refDir, "ChIP", "differential", "H3K27ac_diff_specific_EpiLC.bed")
      bw_RAD21_ESC <- here(refDir, "ChIP", "RAD21-HA", "RAD21_ESC-DMSO_avg_spikeNorm_inputNorm.bw") 
      bw_RAD21_EpiLC <- here(refDir, "ChIP", "RAD21-HA", "RAD21_EpiLC_avg_spikeNorm_inputNorm.bw") 
      
      # PROMOTER LOOPS
      bedpe_promoterLoops <- here(loopDir, "chromosight_score_complex_allP.bedpe")
      ################################################################################
      # DEFINING RANGES
      ################################################################################
      # Y ranges for virtual 4C
      ymin <- 0
      
      # X ranges 
      data <- as_tibble(fread(here(bgDir, paste0("v4C_", vp, "__", "G1DMSO_pooled", "__", res, "kb_localNorm.bedGraph" ))))
      
      chr <- data$V1[1]
      
      # Y ranges for ChIP
      region_gr <- GRanges(seqnames = chr,
                           ranges = IRanges(start = xmin,
                                            end = xmax))
      ## H3K4me3 range
      bw_subset1 <- rtracklayer::import.bw(bw_H3K4me3_ESC, which = region_gr)
      bw_subset2 <- rtracklayer::import.bw(bw_H3K4me3_EpiLC, which = region_gr)
      max_score <- max(bw_subset1$score, bw_subset2$score)
      common_range_H3K4me3 <- c(0, ceiling(max_score / 0.1) * 0.1)
      
      ## H3K27ac range
      bw_subset1 <- rtracklayer::import.bw(bw_H3K27ac_ESC, which = region_gr)
      bw_subset2 <- rtracklayer::import.bw(bw_H3K27ac_EpiLC, which = region_gr)
      max_score <- max(bw_subset1$score, bw_subset2$score)
      common_range_H3K27ac <- c(0, ceiling(max_score / 0.1) * 0.1)
      
      
      ## RAD21 range
      bw_subset1 <- rtracklayer::import.bw(bw_RAD21_ESC, which = region_gr)
      bw_subset2 <- rtracklayer::import.bw(bw_RAD21_EpiLC, which = region_gr)
      max_score <- max(bw_subset1$score, bw_subset2$score)
      common_range_RAD21 <- c(0, ceiling(max_score / 0.1) * 0.1)
      
      ## Defining the genomic region to plot
      plot_region <- pgParams(
        chrom = chr,
        chromstart = xmin,
        chromend = xmax,
        assembly = "mm10"
      )
      # Plotting parameters
      xPosition <- 0.25
      yPosition <- 0.4
      width <- 3.5
      height.v4C <- 0.5
      height.v4CDiff <- 0.3
      height.chip <- 0.1
      height.chipPeak <- 0.05
      yPositionShift <- "0.15b"
      yPositionShiftSmall <- "0.05b"
      yPositionShiftMini <- "0.025b"
      
      yPositionShiftNum <- 0.15
      yPositionShiftSmallNum <- 0.05
      yPositionShiftMiniNum <- 0.025
      textShift <- 0.05
      fontfamily <- "Helvetica"
      
      ################################################################################
      # Plot 1: Drawing ESC vs EpiLC DMSO
      ################################################################################
      pdf(here(figDir, paste0("v4C_", vp, "_", res, "kb_EpiLCvsESC_", xmin, "_", xmax, ".pdf")), width = 4, height = 8)
      pageCreate(width = 4, height = 8, default.units = "inches", showGuides = FALSE)
      # --- TITLE ---
      plotText(
        label = paste0("Virtual 4C, ", unlist(strsplit(vp, "\\."))[2], " (", res, "kb resolution)"),
        x = 0.2, y = 0.2,
        just = "left",
        fontface = "bold",
        fontsize = 8, fontfamily = fontfamily
      )
      # --- v4C G1 ESC DMSO---
      plotText(label = "G1.ESC.DMSO", x = xPosition + textShift, y = yPosition, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      # Plot the signal track
      v4C.ESC.DMSO <- plotSignal(
        data = bg_ESC_DMSO,
        params = plot_region,
        range = c(ymin, ymax),
        x = xPosition, y = yPosition, width = width, height = height.v4C,
        just = c("left", "top"),
        linecolor = color_esc_dark, lwd = 0.25*mmToLinePlotgarden,
        fill = color_esc_dark
      )
      annoYaxis(
        plot = v4C.ESC.DMSO,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- ChIP H3K4me3 ESC ---
      plotText(label = "H3K4me3", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.h3k4me3.ESC <- plotSignal(
        data = bw_H3K4me3_ESC,
        params = plot_region, range = common_range_H3K4me3,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = strong_red,
        fill = strong_red
      )
      annoYaxis(
        plot = chip.h3k4me3.ESC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- ESC specific peaks ---
      plotText(label = "ESC specific peak", x = xPosition + textShift, y = yPositionShiftMini, just = "left",
               fontsize = 4, fontfamily = fontfamily)
      plotRanges(
        data = bed_H3K4me3_ESCspecific,
        params = plot_region,
        x = xPosition, y = yPositionShiftMini, width = width, height = height.chip,
        just = c("left", "top"),
        fill = strong_red
      )
      # --- ChIP H3K27ac ESC ---
      plotText(label = "H3K27ac", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.H3K27ac.ESC <- plotSignal(
        data = bw_H3K27ac_ESC,
        params = plot_region, range = common_range_H3K27ac,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = strong_blue,
        fill = strong_blue
      )
      annoYaxis(
        plot = chip.H3K27ac.ESC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- ESC specific peaks ---
      plotText(label = "ESC specific peak", x = xPosition + textShift, y = yPositionShiftMini, just = "left",
               fontsize = 4, fontfamily = fontfamily)
      plotRanges(
        data = bed_H3K27ac_ESCspecific,
        params = plot_region,
        x = xPosition, y = yPositionShiftMini, width = width, height = height.chip,
        just = c("left", "top"),
        fill = strong_blue
      )
      # --- ChIP RAD21 ESC ---
      plotText(label = "RAD21", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.RAD21.ESC <- plotSignal(
        data = bw_RAD21_ESC,
        params = plot_region, range = common_range_RAD21,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = "grey30",
        fill = "grey30"
      )
      annoYaxis(
        plot = chip.RAD21.ESC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      
      # --- v4C G1 EpiLC DMSO---
      # The y-coordinate "0.2b" places this plot 0.2 inches below the previous one
      plotText(label = "G1.EpiLC.DMSO", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      v4C.EpiLC.DMSO <- plotSignal(
        data = bg_EpiLC_DMSO,
        params = plot_region,
        range = c(ymin, ymax),
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C,
        just = c("left", "top"),
        linecolor = color_epi_dark, lwd = 0.25*mmToLinePlotgarden,
        fill = color_epi_dark
      )
      annoYaxis(
        plot = v4C.EpiLC.DMSO,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      # --- ChIP H3K4me3 EpiLC ---
      plotText(label = "H3K4me3", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.h3k4me3.EpiLC <- plotSignal(
        data = bw_H3K4me3_EpiLC,
        params = plot_region, range = common_range_H3K4me3,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = strong_red,
        fill = strong_red
      )
      annoYaxis(
        plot = chip.h3k4me3.EpiLC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- EpiLC specific peaks ---
      plotText(label = "EpiLC specific peak", x = xPosition + textShift, y = yPositionShiftMini, just = "left",
               fontsize = 4, fontfamily = fontfamily)
      plotRanges(
        data = bed_H3K4me3_EpiLCspecific,
        params = plot_region,
        x = xPosition, y = yPositionShiftMini, width = width, height = height.chip,
        just = c("left", "top"),
        fill = strong_red
      )
      # --- ChIP H3K27ac EpiLC ---
      plotText(label = "H3K27ac", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.H3K27ac.EpiLC <- plotSignal(
        data = bw_H3K27ac_EpiLC,
        params = plot_region, range = common_range_H3K27ac,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = strong_blue,
        fill = strong_blue
      )
      annoYaxis(
        plot = chip.H3K27ac.EpiLC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- EpiLC specific peaks ---
      plotText(label = "EpiLC specific peak", x = xPosition + textShift, y = yPositionShiftMini, just = "left",
               fontsize = 4, fontfamily = fontfamily)
      plotRanges(
        data = bed_H3K27ac_EpiLCspecific,
        params = plot_region,
        x = xPosition, y = yPositionShiftMini, width = width, height = height.chip,
        just = c("left", "top"),
        fill = strong_blue
      )
      # --- ChIP RAD21 EpiLC ---
      plotText(label = "RAD21", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.RAD21.EpiLC <- plotSignal(
        data = bw_RAD21_EpiLC,
        params = plot_region, range = common_range_RAD21,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = "grey30",
        fill = "grey30"
      )
      annoYaxis(
        plot = chip.RAD21.EpiLC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      
      # --- v4C G1 Differential --
      positive_scores_gr <- bg_diff_ESCvsEpiLC[bg_diff_ESCvsEpiLC$score > 0]
      negative_scores_gr <- bg_diff_ESCvsEpiLC[bg_diff_ESCvsEpiLC$score < 0]
      
      
      plotText(label = "Diff (EpiLC - ESC)", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      v4C.diff.EpiLCvsESC <- plotSignal(
        data = negative_scores_gr,
        params = plot_region,
        range = c(-diffYmax, diffYmax),
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4CDiff,
        just = c("left", "top"),
        linecolor = color_esc_dark, lwd = 0.25*mmToLinePlotgarden, baseline = FALSE,
        fill = color_esc_dark
      )
      plotSignal(
        data = positive_scores_gr,
        params = plot_region,
        range = c(-diffYmax, diffYmax),
        x = xPosition, y = yPosition + yPositionShiftSmallNum*8 + yPositionShiftMiniNum*4 + height.v4C*2 + height.chip*10 , width = width, height = height.v4CDiff,
        just = c("left", "top"),
        linecolor = color_epi_dark, lwd = 0.25*mmToLinePlotgarden, baseline = FALSE,
        fill = color_epi_dark
      )
      annoYaxis(
        plot = v4C.diff.EpiLCvsESC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      # --- Promoter loops ---
      plotText(label = "All promoter loops", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      plotPairsArches(
        data = bedpe_promoterLoops,
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C,
        just = c("left", "top"),
        fill = "black",
        linecolor = NA,
        alpha = 0.5,
        clip = TRUE
      )
      # --- genome axis ---
      plotGenes(
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C/2,
        just = c("left", "top"),
        fontsize = 4, fontcolor = c("black", "black"), fill = c("black", "black")
        
      )
      plotGenomeLabel(
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, length = width,
        fontsize = 4, fontfamily = fontfamily
      )
      dev.off()
      
      ################################################################################
      # Plot 2: Drawing dTAG vs DMSO ESC
      ################################################################################
      pdf(here(figDir, paste0("v4C_", vp, "_", res, "kb_dTAGvsDMSO_ESC_", xmin, "_", xmax, ".pdf")), width = 4, height = 8)
      pageCreate(width = 4, height = 8, default.units = "inches", showGuides = FALSE)
      # --- TITLE ---
      plotText(
        label = paste0("Virtual 4C, ", unlist(strsplit(vp, "\\."))[2], " (", res, "kb resolution)"),
        x = 0.2, y = 0.2,
        just = "left",
        fontface = "bold",
        fontsize = 8, fontfamily = fontfamily
      )
      # --- v4C G1 ESC DMSO---
      plotText(label = "G1.ESC.DMSO", x = xPosition + textShift, y = yPosition, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      # Plot the signal track
      v4C.ESC.DMSO <- plotSignal(
        data = bg_ESC_DMSO,
        params = plot_region,
        range = c(ymin, ymax),
        x = xPosition, y = yPosition, width = width, height = height.v4C,
        just = c("left", "top"),
        linecolor = color_esc_dark, lwd = 0.25*mmToLinePlotgarden,
        fill = color_esc_dark
      )
      annoYaxis(
        plot = v4C.ESC.DMSO,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- v4C G1 ESC dTAG---
      # The y-coordinate "0.2b" places this plot 0.2 inches below the previous one
      plotText(label = "G1.ESC.dTAG", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      v4C.ESC.dTAG <- plotSignal(
        data = bg_ESC_dTAG,
        params = plot_region,
        range = c(ymin, ymax),
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C,
        just = c("left", "top"),
        linecolor = color_esc_light, lwd = 0.25*mmToLinePlotgarden,
        fill = color_esc_light
      )
      annoYaxis(
        plot = v4C.ESC.dTAG,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      # --- ChIP H3K4me3 ESC ---
      plotText(label = "H3K4me3", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.h3k4me3.ESC <- plotSignal(
        data = bw_H3K4me3_ESC,
        params = plot_region, range = common_range_H3K4me3,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = strong_red,
        fill = strong_red
      )
      annoYaxis(
        plot = chip.h3k4me3.ESC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- ESC specific peaks ---
      plotText(label = "ESC specific peak", x = xPosition + textShift, y = yPositionShiftMini, just = "left",
               fontsize = 4, fontfamily = fontfamily)
      plotRanges(
        data = bed_H3K4me3_ESCspecific,
        params = plot_region,
        x = xPosition, y = yPositionShiftMini, width = width, height = height.chip,
        just = c("left", "top"),
        fill = strong_red
      )
      # --- ChIP H3K27ac ESC ---
      plotText(label = "H3K27ac", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.H3K27ac.ESC <- plotSignal(
        data = bw_H3K27ac_ESC,
        params = plot_region, range = common_range_H3K27ac,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = strong_blue,
        fill = strong_blue
      )
      annoYaxis(
        plot = chip.H3K27ac.ESC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- ESC specific peaks ---
      plotText(label = "ESC specific peak", x = xPosition + textShift, y = yPositionShiftMini, just = "left",
               fontsize = 4, fontfamily = fontfamily)
      plotRanges(
        data = bed_H3K27ac_ESCspecific,
        params = plot_region,
        x = xPosition, y = yPositionShiftMini, width = width, height = height.chip,
        just = c("left", "top"),
        fill = strong_blue
      )
      # --- ChIP RAD21 ESC ---
      plotText(label = "RAD21", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.RAD21.ESC <- plotSignal(
        data = bw_RAD21_ESC,
        params = plot_region, range = common_range_RAD21,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = "grey30",
        fill = "grey30"
      )
      annoYaxis(
        plot = chip.RAD21.ESC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      
      
      # --- v4C G1 Differential --
      positive_scores_gr <- bg_diff_ESC_dTAGvsDMSO[bg_diff_ESC_dTAGvsDMSO$score > 0]
      negative_scores_gr <- bg_diff_ESC_dTAGvsDMSO[bg_diff_ESC_dTAGvsDMSO$score < 0]
      
      
      plotText(label = "Diff (dTAG - DMSO)", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      v4C.diff.dTAGvsDMSO <- plotSignal(
        data = negative_scores_gr,
        params = plot_region,
        range = c(-diffYmax, diffYmax),
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4CDiff,
        just = c("left", "top"),
        linecolor = color_esc_dark, lwd = 0.25*mmToLinePlotgarden, baseline = FALSE,
        fill = color_esc_dark
      )
      plotSignal(
        data = positive_scores_gr,
        params = plot_region,
        range = c(-diffYmax, diffYmax),
        x = xPosition, y = yPosition + yPositionShiftSmallNum*5 + yPositionShiftMiniNum*2 + height.v4C*2 + height.chip*5 , width = width, height = height.v4CDiff,
        just = c("left", "top"),
        linecolor = color_esc_light, lwd = 0.25*mmToLinePlotgarden, baseline = FALSE,
        fill = color_esc_light
      )
      annoYaxis(
        plot = v4C.diff.dTAGvsDMSO,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      # --- Promoter loops ---
      plotText(label = "All promoter loops", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      plotPairsArches(
        data = bedpe_promoterLoops,
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C,
        just = c("left", "top"),
        fill = "black",
        linecolor = NA,
        alpha = 0.5,
        clip = TRUE
      )
      # --- genome axis ---
      plotGenes(
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C/2,
        just = c("left", "top"),
        fontsize = 4, fontcolor = c("black", "black"), fill = c("black", "black")
        
      )
      plotGenomeLabel(
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, length = width,
        fontsize = 4, fontfamily = fontfamily
      )
      dev.off()
      
      ################################################################################
      # Plot 3: Drawing dTAG vs DMSO EpiLC
      ################################################################################
      pdf(here(figDir, paste0("v4C_", vp, "_", res, "kb_dTAGvsDMSO_EpiLC_", xmin, "_", xmax, ".pdf")), width = 4, height = 8)
      pageCreate(width = 4, height = 8, default.units = "inches", showGuides = FALSE)
      # --- TITLE ---
      plotText(
        label = paste0("Virtual 4C, ", unlist(strsplit(vp, "\\."))[2], " (", res, "kb resolution)"),
        x = 0.2, y = 0.2,
        just = "left",
        fontface = "bold",
        fontsize = 8, fontfamily = fontfamily
      )
      # --- v4C G1 EpiLC DMSO---
      plotText(label = "G1.EpiLC.DMSO", x = xPosition + textShift, y = yPosition, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      # Plot the signal track
      v4C.EpiLC.DMSO <- plotSignal(
        data = bg_EpiLC_DMSO,
        params = plot_region,
        range = c(ymin, ymax),
        x = xPosition, y = yPosition, width = width, height = height.v4C,
        just = c("left", "top"),
        linecolor = color_epi_dark, lwd = 0.25*mmToLinePlotgarden,
        fill = color_epi_dark
      )
      annoYaxis(
        plot = v4C.EpiLC.DMSO,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- v4C G1 EpiLC dTAG---
      # The y-coordinate "0.2b" places this plot 0.2 inches below the previous one
      plotText(label = "G1.EpiLC.dTAG", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      v4C.EpiLC.dTAG <- plotSignal(
        data = bg_EpiLC_dTAG,
        params = plot_region,
        range = c(ymin, ymax),
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C,
        just = c("left", "top"),
        linecolor = color_epi_light, lwd = 0.25*mmToLinePlotgarden,
        fill = color_epi_light
      )
      annoYaxis(
        plot = v4C.EpiLC.dTAG,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      # --- ChIP H3K4me3 EpiLC ---
      plotText(label = "H3K4me3", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.h3k4me3.EpiLC <- plotSignal(
        data = bw_H3K4me3_EpiLC,
        params = plot_region, range = common_range_H3K4me3,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = strong_red,
        fill = strong_red
      )
      annoYaxis(
        plot = chip.h3k4me3.EpiLC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- EpiLC specific peaks ---
      plotText(label = "EpiLC specific peak", x = xPosition + textShift, y = yPositionShiftMini, just = "left",
               fontsize = 4, fontfamily = fontfamily)
      plotRanges(
        data = bed_H3K4me3_EpiLCspecific,
        params = plot_region,
        x = xPosition, y = yPositionShiftMini, width = width, height = height.chip,
        just = c("left", "top"),
        fill = strong_red
      )
      # --- ChIP H3K27ac EpiLC ---
      plotText(label = "H3K27ac", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.H3K27ac.EpiLC <- plotSignal(
        data = bw_H3K27ac_EpiLC,
        params = plot_region, range = common_range_H3K27ac,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = strong_blue,
        fill = strong_blue
      )
      annoYaxis(
        plot = chip.H3K27ac.EpiLC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      # --- EpiLC specific peaks ---
      plotText(label = "EpiLC specific peak", x = xPosition + textShift, y = yPositionShiftMini, just = "left",
               fontsize = 4, fontfamily = fontfamily)
      plotRanges(
        data = bed_H3K27ac_EpiLCspecific,
        params = plot_region,
        x = xPosition, y = yPositionShiftMini, width = width, height = height.chip,
        just = c("left", "top"),
        fill = strong_blue
      )
      # --- ChIP RAD21 EpiLC ---
      plotText(label = "RAD21", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      chip.RAD21.EpiLC <- plotSignal(
        data = bw_RAD21_EpiLC,
        params = plot_region, range = common_range_RAD21,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.chip,
        just = c("left", "top"),
        linecolor = "grey30",
        fill = "grey30"
      )
      annoYaxis(
        plot = chip.RAD21.EpiLC,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      
      
      # --- v4C G1 Differential --
      positive_scores_gr <- bg_diff_EpiLC_dTAGvsDMSO[bg_diff_EpiLC_dTAGvsDMSO$score > 0]
      negative_scores_gr <- bg_diff_EpiLC_dTAGvsDMSO[bg_diff_EpiLC_dTAGvsDMSO$score < 0]
      
      
      plotText(label = "Diff (dTAG - DMSO)", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      v4C.diff.dTAGvsDMSO <- plotSignal(
        data = negative_scores_gr,
        params = plot_region,
        range = c(-diffYmax, diffYmax),
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4CDiff,
        just = c("left", "top"),
        linecolor = color_epi_dark, lwd = 0.25*mmToLinePlotgarden, baseline = FALSE,
        fill = color_epi_dark
      )
      plotSignal(
        data = positive_scores_gr,
        params = plot_region,
        range = c(-diffYmax, diffYmax),
        x = xPosition, y = yPosition + yPositionShiftSmallNum*5 + yPositionShiftMiniNum*2 + height.v4C*2 + height.chip*5 , width = width, height = height.v4CDiff,
        just = c("left", "top"),
        linecolor = color_epi_light, lwd = 0.25*mmToLinePlotgarden, baseline = FALSE,
        fill = color_epi_light
      )
      annoYaxis(
        plot = v4C.diff.dTAGvsDMSO,
        axisLine = TRUE, lineend = "square",
        fontsize = 4, fontfamily = fontfamily
      )
      
      # --- Promoter loops ---
      plotText(label = "All promoter loops", x = xPosition + textShift, y = yPositionShiftSmall, just = "left",
               fontsize = 6, fontfamily = fontfamily)
      plotPairsArches(
        data = bedpe_promoterLoops,
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C,
        just = c("left", "top"),
        fill = "black",
        linecolor = NA,
        alpha = 0.5,
        clip = TRUE
      )
      # --- genome axis ---
      plotGenes(
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, width = width, height = height.v4C/2,
        just = c("left", "top"),
        fontsize = 4, fontcolor = c("black", "black"), fill = c("black", "black")
        
      )
      plotGenomeLabel(
        params = plot_region,
        x = xPosition, y = yPositionShiftSmall, length = width,
        fontsize = 4, fontfamily = fontfamily
      )
      dev.off()
      
}


```

