---
title: "R Notebook"
editor_options: 
  chunk_output_type: console
---
# LOADING REQUIREMENTS
## PACKAGES
```{r message=FALSE, warning=FALSE, include=FALSE}
# LIST OF PACKAGES
pkgs = c("tidyverse", "here", "svglite", "ggplot2", "ggrepel",
         "stringr", "BiocManager", "RColorBrewer", "viridis", "magick",
         "nlme", "factoextra", "devtools", "beepr", "remotes",
         "cowplot", "data.table", "strawr", "rtracklayer", "utils",
         "reshape2", "VennDiagram","gridExtra", "eulerr", "scales",
         "InteractionSet", "gplots", "locfdr", "future", "ggbeeswarm", "glossary", "ComplexUpset", "ggnewscale")
bio_pkgs = c("biomaRt", "DESeq2", "ComplexHeatmap", "apeglm", "vsn",
             "rhdf5", "InteractionSet", "plotgardener", "HiCDCPlus",
             "GenomicInteractions", "LOLA", "AnnotationHub",
             "org.Mm.eg.db", "enrichplot", "DOSE", "clusterProfiler",
             "HiCExperiment", "HiCool", "HiContacts", "HiContactsData", "fourDNData", "DNAZooData",
             "MotifDb", "Biostrings")

# INSTALL PACKAGES
# install.packages(pkgs)
# BiocManager::install(bio_pkgs)

# INSTALL HiC related packages
# remotes::install_github("robinweide/GENOVA")
# devtools::install_github("thomasp85/scico")
# devtools::install_github("psyteachr/introdataviz")

# LOAD PACKAGES
lapply(pkgs, require, character.only = TRUE)
lapply(bio_pkgs, require, character.only = TRUE)
require(GENOVA)
require(scico)
require(introdataviz)

# CLEANING
rm(pkgs, bio_pkgs)

options(scipen=999)



```
## ENSEMBL
```{r}
# ENSEMBL DATABASE
# v102 is for the latest mm10 version
# ensembl.v102 <- useMart(host = "https://nov2020.archive.ensembl.org",
#                        biomart = "ENSEMBL_MART_ENSEMBL",
#                        dataset = "mmusculus_gene_ensembl")
```
## COLOR PALETTE
```{r}
palette_1 = list(red = "#E9002D",
                 amber = "#FFAA00",
                 green = "#00B000")
palette_2 = list(red = "#FF1F5B",
                 green = "#00CD6C",
                 blue = "#009ADE",
                 purple = "#AF58BA",
                 yellow = "#FFC61E",
                 orange = "#F28522",
                 grey = "#A0B1BA",
                 brown = "#A6761D")
palette_3 = list(grey1 = "#a0b1ba",
                 grey2 = "#c5d0d5",
                 grey3 = "#eceff1")

colorListLoop <-  c(palette_3[["grey3"]], palette_3[["grey2"]], palette_3[["grey1"]],
                    "#C5E1EF", "#6CB0D6", "#226E9C",
                    "#06592A", 
                    "#FED976", "#FD8D3C", "#E31A1C")
colorListPromoter <- c(palette_3[["grey3"]],"#E88587", "#E31A1C")
colorListEnhancer <- c(palette_3[["grey3"]],"#87AFC7", "#226E9C")
colorListStructure <- c(palette_3[["grey3"]], palette_3[["grey2"]], palette_3[["grey1"]])
```

## DIR LIST
```{r}
refDir <- here("../reference")
figDir <- here("../figure/loop_analysis")
dataDir <- here("..", "data", "loop_analysis")
commonLoopDir <- here("../data/loop_analysis")

```
## FIGURE PARAMETERS
```{r}
library(colorspace)

fontType <- "Helvetica"

fontSizeL <- 10 # pt
fontSizeM <- 8
fontSizeS <- 6

lineThick <- 0.75 # pt
lineMedium <- 0.5
lineThin <- 0.25

panelUnit <- 30 # mm
panelMargin <- 1.5

mmToInch <- 0.03937007874
mmToLineUnit <- 1/2.13
mmToLinePlotgarden <- 1/0.75
ptToMM <- 1/2.845


strong_red <- "#CB333A"
strong_blue <- "#4851A0"
weak_red <- lighten(strong_red, amount = 0.4)   # FF7D81
weak_blue <- lighten(strong_blue, amount = 0.4) # 8A91DD
no_grey <- "#A8A8A8"

strong_teel <- "#0892A5"
strong_green <- "#23CE6B" # A485
strong_darkgreen <- "#054A29"
strong_yellow <- "#FFBA49"
strong_orange <- "#F18F01" # dTAG 
strong_lightpurple <- "#BD93D8"
strong_purple <- "#9E33CB" # Epi

panelSize <- function(num, unit = panelUnit, margin = panelMargin){
  return(num*unit - 2*margin)
}

```
## FUNCTIONS
```{r}
importBedpe = function(bedpe){
  a1 = makeGRangesFromDataFrame(data.frame(
    chr = bedpe$V1,
    start = bedpe$V2 +1,
    end = bedpe$V3))
  a2 = makeGRangesFromDataFrame(data.frame(
    chr = bedpe$V4,
    start = bedpe$V5 +1,
    end = bedpe$V6))
  GInteractions(a1, a2)
}

get_density <- function(x, y, ...) {
  dens <- MASS::kde2d(x, y, ...)
  ix <- findInterval(x, dens$x)
  iy <- findInterval(y, dens$y)
  ii <- cbind(ix, iy)
  return(dens$z[ii])
}
label_kb_mb <- function(x) {
  ifelse(x >= 1000000, paste0(x / 1000000, "Mb"), paste0(x / 1000, "kb"))
}

importPeak = function(fileName){
  df = fread(fileName)
  gr = makeGRangesFromDataFrame(data.frame(
    chr = df$V1, start = df$V2, end = df$V3
  ))
}

convPvalue <- function(pvalue){
  out <- ifelse(pvalue < 0.0001, "****",
                ifelse(pvalue < 0.001, "***",
                       ifelse(pvalue < 0.01, "**",
                              ifelse(pvalue < 0.05, "*", "ns"))))
  return(out)
}

loadLoopAnnoData <- function(fileName){
  data <- as_tibble(fread(fileName)) %>%
    dplyr::mutate(gene = strsplit(gene, '\\|'))
  return(data)
}
```

# PREPROCESSING
## Adding +ooe +geneAnno
Loop should be dropped from specific comparison if ooe in any of the condition are NA
If ooe is 0, fixed pseudo value of 0.01 will be used.
log2(ooe) should be plotted in range (-5, 8) if plotted in scatterplot
### Common stuffs
```{r}
# Parameters
commonLoopDir <- here("../data/loop_analysis")

# Defining gene neighborhood
flankSize <- 2500
gene.TSS.tb <- fread(here(refDir, "mm10", "mm10_GRCm38.p6_gene_sorted.bed")) %>%
  dplyr::mutate(TSS = ifelse(V4 == "+", V2, V3),
                TSSstart = TSS - flankSize,
                TSSend = TSS + flankSize) %>%
  dplyr::select(V1, TSSstart, TSSend, V4, V5, V6)
colnames(gene.TSS.tb) <- c("chr", "start", "end", "strand", "gene", "ensembl")
fwrite(gene.TSS.tb, here(refDir, "mm10", "mm10_GRCm38.p6_TSS2.5kb.bed"), sep = "\t", col.names = FALSE, row.names = FALSE)

# Functions for calculating gene overlap
findOverlapGene <- function(gene.TSS.tb, chrom1, start1, end1){
  temp <- gene.TSS.tb %>% dplyr::filter(chr == chrom1) %>%
    dplyr::filter((start <= end1) & (end >= start1))
  return(temp$gene)
}
findOverlapEnsembl <- function(gene.TSS.tb, chrom1, start1, end1){
  temp <- gene.TSS.tb %>% dplyr::filter(chr == chrom1) %>%
    dplyr::filter((start <= end1) & (end >= start1))
  return(temp$ensembl)
}
```
### Preparing 
#### [1] Chromosight
```{r}
################################################################################
# Import annotated postprocessed loops with chromosight score
loopDir <- here("../data/loop_chromosight")
data <- as_tibble(fread(here(loopDir, "chromo_union_allRes_postprocessed_annotated.tsv")))

################################################################################
# Integrate chromosight score from asynchronous experiment
resolutions  <- c(25000, 10000, 5000)

# helper: read, filter, select and rename for one experiment
annotate_scores <- function(experiment, suffix) {
  map_dfr(resolutions, function(res) {
    fread(here(loopDir,
               sprintf("GSE178982_%s_pooled_union_%dbp_pu100pz100.tsv",
                       experiment, res))
    ) %>%
      mutate(id = paste(chrom1, start1, end1, chrom2, start2, end2, sep = "_")) %>%
      filter(start2 - start1 <= 2e6) %>%
      select(id, score, pvalue, qvalue) %>%
      rename_with(~ paste0(., "_", suffix),
                  c(score, pvalue, qvalue))
  })
}

data <- data %>%
  left_join(annotate_scores("AsyncUT", "UT"), by = "id")
data <- data %>%
  left_join(annotate_scores("AsyncAID", "AID"), by = "id")
data <- data %>%
  relocate(score_UT:qvalue_AID, .after = qvalue_EpidTAG)
data <- data %>% dplyr::select(-seq(29, 41))
data <- data %>% relocate(A1:Anno_simple, .after = id)

################################################################################
# Adding obs/exp
import_obsexp <- function(experiment, suffix) {
  map_dfr(resolutions, function(res) {
    fread(here(loopDir,
               sprintf("chromo_union_allRes_postprocessed_%dbp_obxexp_%s_pooled.tsv",
                       res, experiment))
    ) %>%
      mutate(id = paste(chrom1, start1, end1, chrom2, start2, end2, sep = "_")) %>%
      select(id, ooe) %>%
      rename_with(~ paste0(., "_", suffix),
                  c(ooe))
  })
}

data <- data %>%
  left_join(import_obsexp("G1DMSO", "DMSO"), by = "id")
data <- data %>%
  left_join(import_obsexp("G1dTAG", "dTAG"), by = "id")
data <- data %>%
  left_join(import_obsexp("G1A485", "A485"), by = "id")
data <- data %>%
  left_join(import_obsexp("EpiG1DMSO", "EpiDMSO"), by = "id")
data <- data %>%
  left_join(import_obsexp("EpiG1dTAG", "EpidTAG"), by = "id")
data <- data %>%
  left_join(import_obsexp("GSE178982_AsyncUT", "UT"), by = "id")
data <- data %>%
  left_join(import_obsexp("GSE178982_AsyncAID", "AID"), by = "id")

fwrite(data, here(commonLoopDir, "loopScore_chromosight.tsv"),
       sep = "\t", col.names = TRUE)
# fwrite(data, here(commonLoopDir, "loopScore_chromosight_extra.tsv"),
#        sep = "\t", col.names = TRUE)
################################################################################
# Annotate genes for p-n loops
name <- "chromosight"
gene.TSS.tb <- fread(here(refDir, "mm10", "mm10_GRCm38.p6_TSS2.5kb.bed"))
colnames(gene.TSS.tb) <- c("chr", "start", "end", "strand", "gene", "ensembl")

# Simple annotation
temp <- data %>% rowwise() %>%
  dplyr::mutate(A1_gene = ifelse(A1_simple == "P",
                                 list(findOverlapGene(gene.TSS.tb, chrom1, start1, end1)),
                                 NA),
                A2_gene = ifelse(A2_simple == "P",
                                 list(findOverlapGene(gene.TSS.tb, chrom2, start2, end2)),
                                 NA),
                gene = list(unique(c(A1_gene, A2_gene))))

temp_p_n <- temp %>% dplyr::filter(Anno_simple %in% c("P-P", "P-E", "P-S", "P-X")) %>% dplyr::select(-c("A1_gene", "A2_gene"))
fwrite(temp_p_n, here(commonLoopDir, paste0("loopScore_", name, "_p-n_simple_geneList.tsv")), sep = "\t")

temp <- data %>% rowwise() %>%
  dplyr::mutate(A1_gene = ifelse(A1_simple == "P",
                                 list(findOverlapEnsembl(gene.TSS.tb, chrom1, start1, end1)),
                                 NA),
                A2_gene = ifelse(A2_simple == "P",
                                 list(findOverlapEnsembl(gene.TSS.tb, chrom2, start2, end2)),
                                 NA),
                gene = list(unique(c(A1_gene, A2_gene))))

temp_p_n <- temp %>% dplyr::filter(Anno_simple %in% c("P-P", "P-E", "P-S", "P-X")) %>% dplyr::select(-c("A1_gene", "A2_gene"))
fwrite(temp_p_n, here(commonLoopDir, paste0("loopScore_", name, "_p-n_simple_ensemblList.tsv")), sep = "\t")

# complex annotation
temp <- data %>% rowwise() %>%
  dplyr::mutate(A1_gene = ifelse(A1 %in% c("P", "SP"),
                                 list(findOverlapGene(gene.TSS.tb, chrom1, start1, end1)),
                                 NA),
                A2_gene = ifelse(A2 %in% c("P", "SP"),
                                 list(findOverlapGene(gene.TSS.tb, chrom2, start2, end2)),
                                 NA),
                gene = list(unique(c(A1_gene, A2_gene))))

temp_p_n <- temp %>% dplyr::filter(str_detect(Anno, "P")) %>%
  dplyr::select(-c("A1_gene", "A2_gene"))
fwrite(temp_p_n, here(commonLoopDir, paste0("loopScore_", name, "_p-n_complex_geneList.tsv")), sep = "\t")

temp <- data %>% rowwise() %>%
  dplyr::mutate(A1_gene = ifelse(A1 %in% c("P", "SP"),
                                 list(findOverlapEnsembl(gene.TSS.tb, chrom1, start1, end1)),
                                 NA),
                A2_gene = ifelse(A2 %in% c("P", "SP"),
                                 list(findOverlapEnsembl(gene.TSS.tb, chrom2, start2, end2)),
                                 NA),
                gene = list(unique(c(A1_gene, A2_gene))))

temp_p_n <- temp %>% dplyr::filter(str_detect(Anno, "P")) %>%
  dplyr::select(-c("A1_gene", "A2_gene"))
fwrite(temp_p_n, here(commonLoopDir, paste0("loopScore_", name, "_p-n_complex_ensemblList.tsv")), sep = "\t")

```

#### [2] hicdcp
```{r}
################################################################################
loopDir <- here("../data/loop_hicdcp")

################################################################################
# Import annotated postprocessed loops with chromosight score
data <- as_tibble(fread(here(loopDir, "hicdcp_union_10000bp_annotated.tsv"))) %>%
  mutate(id = paste(chrom1, start1, end1, chrom2, start2, end2, sep = "_")) %>%
  relocate(id, .after = end2) %>%
  dplyr::select(-c("anchor_id_1", "anchor_id_2"))

resolutions  <- c(10000)

# Adding obs/exp
import_obsexp <- function(experiment, suffix) {
  map_dfr(resolutions, function(res) {
    fread(here(loopDir,
               sprintf("hicdcp_union_%dbp_obxexp_%s_pooled.tsv",
                       res, experiment))
    ) %>%
      mutate(id = paste(chrom1, start1, end1, chrom2, start2, end2, sep = "_")) %>%
      select(id, ooe) %>%
      rename_with(~ paste0(., "_", suffix),
                  c(ooe))
  })
}

data <- data %>%
  left_join(import_obsexp("G1DMSO", "DMSO"), by = "id")
data <- data %>%
  left_join(import_obsexp("G1dTAG", "dTAG"), by = "id")
data <- data %>%
  left_join(import_obsexp("G1A485", "A485"), by = "id")
data <- data %>%
  left_join(import_obsexp("EpiG1DMSO", "EpiDMSO"), by = "id")
data <- data %>%
  left_join(import_obsexp("EpiG1dTAG", "EpidTAG"), by = "id")
data <- data %>%
  left_join(import_obsexp("GSE178982_AsyncUT", "UT"), by = "id")
data <- data %>%
  left_join(import_obsexp("GSE178982_AsyncAID", "AID"), by = "id")

fwrite(data, here(commonLoopDir, "loopScore_hicdcp.tsv"),
       sep = "\t", col.names = TRUE)

################################################################################
# Annotate genes for p-n loops
name <- "hicdcp"
gene.TSS.tb <- fread(here(refDir, "mm10", "mm10_GRCm38.p6_TSS2.5kb.bed"))
colnames(gene.TSS.tb) <- c("chr", "start", "end", "strand", "gene", "ensembl")


# Simple annotation
temp <- data %>% rowwise() %>%
  dplyr::mutate(A1_gene = ifelse(A1_simple == "P",
                                 list(findOverlapGene(gene.TSS.tb, chrom1, start1, end1)),
                                 NA),
                A2_gene = ifelse(A2_simple == "P",
                                 list(findOverlapGene(gene.TSS.tb, chrom2, start2, end2)),
                                 NA),
                gene = list(unique(c(A1_gene, A2_gene))))

temp_p_n <- temp %>% dplyr::filter(Anno_simple %in% c("P-P", "P-E", "P-S", "P-X")) %>% dplyr::select(-c("A1_gene", "A2_gene"))
fwrite(temp_p_n, here(commonLoopDir, paste0("loopScore_", name, "_p-n_simple_geneList.tsv")), sep = "\t")

temp <- data %>% rowwise() %>%
  dplyr::mutate(A1_gene = ifelse(A1_simple == "P",
                                 list(findOverlapEnsembl(gene.TSS.tb, chrom1, start1, end1)),
                                 NA),
                A2_gene = ifelse(A2_simple == "P",
                                 list(findOverlapEnsembl(gene.TSS.tb, chrom2, start2, end2)),
                                 NA),
                gene = list(unique(c(A1_gene, A2_gene))))

temp_p_n <- temp %>% dplyr::filter(Anno_simple %in% c("P-P", "P-E", "P-S", "P-X")) %>% dplyr::select(-c("A1_gene", "A2_gene"))
fwrite(temp_p_n, here(commonLoopDir, paste0("loopScore_", name, "_p-n_simple_ensemblList.tsv")), sep = "\t")

# complex annotation
temp <- data %>% rowwise() %>%
  dplyr::mutate(A1_gene = ifelse(A1 %in% c("P", "SP"),
                                 list(findOverlapGene(gene.TSS.tb, chrom1, start1, end1)),
                                 NA),
                A2_gene = ifelse(A2 %in% c("P", "SP"),
                                 list(findOverlapGene(gene.TSS.tb, chrom2, start2, end2)),
                                 NA),
                gene = list(unique(c(A1_gene, A2_gene))))

temp_p_n <- temp %>% dplyr::filter(str_detect(Anno, "P")) %>%
  dplyr::select(-c("A1_gene", "A2_gene"))
fwrite(temp_p_n, here(commonLoopDir, paste0("loopScore_", name, "_p-n_complex_geneList.tsv")), sep = "\t")

temp <- data %>% rowwise() %>%
  dplyr::mutate(A1_gene = ifelse(A1 %in% c("P", "SP"),
                                 list(findOverlapEnsembl(gene.TSS.tb, chrom1, start1, end1)),
                                 NA),
                A2_gene = ifelse(A2 %in% c("P", "SP"),
                                 list(findOverlapEnsembl(gene.TSS.tb, chrom2, start2, end2)),
                                 NA),
                gene = list(unique(c(A1_gene, A2_gene))))

temp_p_n <- temp %>% dplyr::filter(str_detect(Anno, "P")) %>%
  dplyr::select(-c("A1_gene", "A2_gene"))
fwrite(temp_p_n, here(commonLoopDir, paste0("loopScore_", name, "_p-n_complex_ensemblList.tsv")), sep = "\t")


```
#### [3] Hansen
Loops with 
```{r}
################################################################################
loopDir <- here("../data/loop_Hansen")

################################################################################
# Import annotated postprocessed loops with chromosight score
data <- as_tibble(fread(here(loopDir, "Hansen_union_allRes_postprocessed_annotated.tsv"))) %>%
  mutate(id = paste(chrom1, start1, end1, chrom2, start2, end2, sep = "_")) %>%
  relocate(id, .after = end2) %>%
  dplyr::select(-c("anchor_id_1", "anchor_id_2"))

resolutions  <- c(5000, 2000, 1000)

# helper: read, filter, select and rename for one experiment
annotate_scores <- function(experiment, suffix) {
  map_dfr(resolutions, function(res) {
    fread(here(loopDir,
               sprintf("%s_pooled_union_%dbp_pu100pz100.tsv",
                       experiment, res))
    ) %>%
      mutate(id = paste(chrom1, start1, end1, chrom2, start2, end2, sep = "_")) %>%
      filter(start2 - start1 <= 2e6) %>%
      select(id, score, pvalue, qvalue) %>%
      rename_with(~ paste0(., "_", suffix),
                  c(score, pvalue, qvalue))
  })
}

data <- data %>%
  left_join(annotate_scores("G1DMSO", "DMSO"), by = "id")
data <- data %>%
  left_join(annotate_scores("G1dTAG", "dTAG"), by = "id")
data <- data %>%
  left_join(annotate_scores("G1A485", "A485"), by = "id")
data <- data %>%
  left_join(annotate_scores("EpiG1DMSO", "EpiDMSO"), by = "id")
data <- data %>%
  left_join(annotate_scores("EpiG1dTAG", "EpidTAG"), by = "id")
data <- data %>%
  left_join(annotate_scores("GSE178982_AsyncUT", "UT"), by = "id")
data <- data %>%
  left_join(annotate_scores("GSE178982_AsyncAID", "AID"), by = "id")

# Adding obs/exp
import_obsexp <- function(experiment, suffix) {
  map_dfr(resolutions, function(res) {
    fread(here(loopDir,
               sprintf("Hansen_union_allRes_postprocessed_%dbp_obxexp_%s_pooled.tsv",
                       res, experiment))
    ) %>%
      mutate(id = paste(chrom1, start1, end1, chrom2, start2, end2, sep = "_")) %>%
      select(id, ooe) %>%
      rename_with(~ paste0(., "_", suffix),
                  c(ooe))
  })
}

data <- data %>%
  left_join(import_obsexp("G1DMSO", "DMSO") %>% distinct(), by = "id")
data <- data %>%
  left_join(import_obsexp("G1dTAG", "dTAG") %>% distinct(), by = "id")
data <- data %>%
  left_join(import_obsexp("G1A485", "A485") %>% distinct(), by = "id")
data <- data %>%
  left_join(import_obsexp("EpiG1DMSO", "EpiDMSO") %>% distinct(), by = "id")
data <- data %>%
  left_join(import_obsexp("EpiG1dTAG", "EpidTAG") %>% distinct(), by = "id")
data <- data %>%
  left_join(import_obsexp("GSE178982_AsyncUT", "UT") %>% distinct(), by = "id")
data <- data %>%
  left_join(import_obsexp("GSE178982_AsyncAID", "AID") %>% distinct(), by = "id")

data <- data %>% 
  filter(
    !if_all(starts_with("ooe"), ~ .x == 0)   # drop rows where every ooe value is 0 or NA
  )

fwrite(data, here(commonLoopDir, "loopScore_Hansen.tsv"),
       sep = "\t", col.names = TRUE)

################################################################################
# Annotate genes for p-n loops
name <- "Hansen"
gene.TSS.tb <- fread(here(refDir, "mm10", "mm10_GRCm38.p6_TSS2.5kb.bed"))
colnames(gene.TSS.tb) <- c("chr", "start", "end", "strand", "gene", "ensembl")


# Simple annotation
temp <- data %>% rowwise() %>%
  dplyr::mutate(A1_gene = ifelse(A1_simple == "P",
                                 list(findOverlapGene(gene.TSS.tb, chrom1, start1, end1)),
                                 NA),
                A2_gene = ifelse(A2_simple == "P",
                                 list(findOverlapGene(gene.TSS.tb, chrom2, start2, end2)),
                                 NA),
                gene = list(unique(c(A1_gene, A2_gene))))

temp_p_n <- temp %>% dplyr::filter(Anno_simple %in% c("P-P", "P-E", "P-S", "P-X")) %>% dplyr::select(-c("A1_gene", "A2_gene"))
fwrite(temp_p_n, here(commonLoopDir, paste0("loopScore_", name, "_p-n_simple_geneList.tsv")), sep = "\t")

temp <- data %>% rowwise() %>%
  dplyr::mutate(A1_gene = ifelse(A1_simple == "P",
                                 list(findOverlapEnsembl(gene.TSS.tb, chrom1, start1, end1)),
                                 NA),
                A2_gene = ifelse(A2_simple == "P",
                                 list(findOverlapEnsembl(gene.TSS.tb, chrom2, start2, end2)),
                                 NA),
                gene = list(unique(c(A1_gene, A2_gene))))

temp_p_n <- temp %>% dplyr::filter(Anno_simple %in% c("P-P", "P-E", "P-S", "P-X")) %>% dplyr::select(-c("A1_gene", "A2_gene"))
fwrite(temp_p_n, here(commonLoopDir, paste0("loopScore_", name, "_p-n_simple_ensemblList.tsv")), sep = "\t")

# complex annotation
temp <- data %>% rowwise() %>%
  dplyr::mutate(A1_gene = ifelse(A1 %in% c("P", "SP"),
                                 list(findOverlapGene(gene.TSS.tb, chrom1, start1, end1)),
                                 NA),
                A2_gene = ifelse(A2 %in% c("P", "SP"),
                                 list(findOverlapGene(gene.TSS.tb, chrom2, start2, end2)),
                                 NA),
                gene = list(unique(c(A1_gene, A2_gene))))

temp_p_n <- temp %>% dplyr::filter(str_detect(Anno, "P")) %>%
  dplyr::select(-c("A1_gene", "A2_gene"))
fwrite(temp_p_n, here(commonLoopDir, paste0("loopScore_", name, "_p-n_complex_geneList.tsv")), sep = "\t")

temp <- data %>% rowwise() %>%
  dplyr::mutate(A1_gene = ifelse(A1 %in% c("P", "SP"),
                                 list(findOverlapEnsembl(gene.TSS.tb, chrom1, start1, end1)),
                                 NA),
                A2_gene = ifelse(A2 %in% c("P", "SP"),
                                 list(findOverlapEnsembl(gene.TSS.tb, chrom2, start2, end2)),
                                 NA),
                gene = list(unique(c(A1_gene, A2_gene))))

temp_p_n <- temp %>% dplyr::filter(str_detect(Anno, "P")) %>%
  dplyr::select(-c("A1_gene", "A2_gene"))
fwrite(temp_p_n, here(commonLoopDir, paste0("loopScore_", name, "_p-n_complex_ensemblList.tsv")), sep = "\t")

```

#### [4] FitHiC
```{r}
################################################################################
loopDir <- here("../data/loop_fithic")

################################################################################
# Import annotated postprocessed loops with chromosight score
data <- as_tibble(fread(here(loopDir, "fithic_union_10000bp_annotated.tsv"))) %>%
  mutate(id = paste(chrom1, start1, end1, chrom2, start2, end2, sep = "_")) %>%
  relocate(id, .after = end2) %>%
  dplyr::select(-c("anchor_id_1", "anchor_id_2"))

resolutions  <- c(10000)


# Adding obs/exp
import_obsexp <- function(experiment, suffix) {
  map_dfr(resolutions, function(res) {
    fread(here(loopDir,
               sprintf("fithic_union_%dbp_obxexp_%s_pooled.tsv",
                       res, experiment))
    ) %>%
      mutate(id = paste(chrom1, start1, end1, chrom2, start2, end2, sep = "_")) %>%
      select(id, ooe) %>%
      rename_with(~ paste0(., "_", suffix),
                  c(ooe))
  })
}

data <- data %>%
  left_join(import_obsexp("G1DMSO", "DMSO") %>% distinct(), by = "id")
data <- data %>%
  left_join(import_obsexp("G1dTAG", "dTAG") %>% distinct(), by = "id")
data <- data %>%
  left_join(import_obsexp("G1A485", "A485") %>% distinct(), by = "id")
data <- data %>%
  left_join(import_obsexp("EpiG1DMSO", "EpiDMSO") %>% distinct(), by = "id")
data <- data %>%
  left_join(import_obsexp("EpiG1dTAG", "EpidTAG") %>% distinct(), by = "id")
data <- data %>%
  left_join(import_obsexp("GSE178982_AsyncUT", "UT") %>% distinct(), by = "id")
data <- data %>%
  left_join(import_obsexp("GSE178982_AsyncAID", "AID") %>% distinct(), by = "id")

data <- data %>% 
  filter(
    !if_all(starts_with("ooe"), ~ .x == 0)   # drop rows where every ooe value is 0 or NA
  )

fwrite(data, here(commonLoopDir, "loopScore_fithic.tsv"),
       sep = "\t", col.names = TRUE)

################################################################################
# Annotate genes for p-n loops
name <- "fithic"
gene.TSS.tb <- fread(here(refDir, "mm10", "mm10_GRCm38.p6_TSS2.5kb.bed"))
colnames(gene.TSS.tb) <- c("chr", "start", "end", "strand", "gene", "ensembl")


# Simple annotation
temp <- data %>% rowwise() %>%
  dplyr::mutate(A1_gene = ifelse(A1_simple == "P",
                                 list(findOverlapGene(gene.TSS.tb, chrom1, start1, end1)),
                                 NA),
                A2_gene = ifelse(A2_simple == "P",
                                 list(findOverlapGene(gene.TSS.tb, chrom2, start2, end2)),
                                 NA),
                gene = list(unique(c(A1_gene, A2_gene))))

temp_p_n <- temp %>% dplyr::filter(Anno_simple %in% c("P-P", "P-E", "P-S", "P-X")) %>% dplyr::select(-c("A1_gene", "A2_gene"))
fwrite(temp_p_n, here(commonLoopDir, paste0("loopScore_", name, "_p-n_simple_geneList.tsv")), sep = "\t")

temp <- data %>% rowwise() %>%
  dplyr::mutate(A1_gene = ifelse(A1_simple == "P",
                                 list(findOverlapEnsembl(gene.TSS.tb, chrom1, start1, end1)),
                                 NA),
                A2_gene = ifelse(A2_simple == "P",
                                 list(findOverlapEnsembl(gene.TSS.tb, chrom2, start2, end2)),
                                 NA),
                gene = list(unique(c(A1_gene, A2_gene))))

temp_p_n <- temp %>% dplyr::filter(Anno_simple %in% c("P-P", "P-E", "P-S", "P-X")) %>% dplyr::select(-c("A1_gene", "A2_gene"))
fwrite(temp_p_n, here(commonLoopDir, paste0("loopScore_", name, "_p-n_simple_ensemblList.tsv")), sep = "\t")

# complex annotation
temp <- data %>% rowwise() %>%
  dplyr::mutate(A1_gene = ifelse(A1 %in% c("P", "SP"),
                                 list(findOverlapGene(gene.TSS.tb, chrom1, start1, end1)),
                                 NA),
                A2_gene = ifelse(A2 %in% c("P", "SP"),
                                 list(findOverlapGene(gene.TSS.tb, chrom2, start2, end2)),
                                 NA),
                gene = list(unique(c(A1_gene, A2_gene))))

temp_p_n <- temp %>% dplyr::filter(str_detect(Anno, "P")) %>%
  dplyr::select(-c("A1_gene", "A2_gene"))
fwrite(temp_p_n, here(commonLoopDir, paste0("loopScore_", name, "_p-n_complex_geneList.tsv")), sep = "\t")

temp <- data %>% rowwise() %>%
  dplyr::mutate(A1_gene = ifelse(A1 %in% c("P", "SP"),
                                 list(findOverlapEnsembl(gene.TSS.tb, chrom1, start1, end1)),
                                 NA),
                A2_gene = ifelse(A2 %in% c("P", "SP"),
                                 list(findOverlapEnsembl(gene.TSS.tb, chrom2, start2, end2)),
                                 NA),
                gene = list(unique(c(A1_gene, A2_gene))))

temp_p_n <- temp %>% dplyr::filter(str_detect(Anno, "P")) %>%
  dplyr::select(-c("A1_gene", "A2_gene"))
fwrite(temp_p_n, here(commonLoopDir, paste0("loopScore_", name, "_p-n_complex_ensemblList.tsv")), sep = "\t")
```


#### [5] Tjian
```{r}
################################################################################
# Import annotated postprocessed loops with chromosight score
loopDir <- here("../data/loop_Tjian")
data <- as_tibble(fread(here(loopDir, "TjianChromo_union_allRes_annotated.tsv"))) %>%
  mutate(id = paste(chrom1, start1, end1, chrom2, start2, end2, sep = "_")) %>%
  relocate(id, .after = end2) %>%
  dplyr::select(-c("anchor_id_1", "anchor_id_2"))

################################################################################
# Integrate chromosight score from asynchronous experiment
resolutions  <- c(20000, 10000, 4000, 2000, 1000, 800, 600, 400)

# helper: read, filter, select and rename for one experiment
annotate_scores <- function(experiment, suffix) {
  map_dfr(resolutions, function(res) {
    fread(here(loopDir,
               sprintf("%s_pooled_union_%dbp_pu100pz100.tsv",
                       experiment, res))
    ) %>%
      mutate(id = paste(chrom1, start1, end1, chrom2, start2, end2, sep = "_")) %>%
      filter(start2 - start1 <= 2e6) %>%
      select(id, score, pvalue, qvalue) %>%
      rename_with(~ paste0(., "_", suffix),
                  c(score, pvalue, qvalue))
  })
}

data <- data %>%
  left_join(annotate_scores("G1DMSO", "DMSO"), by = "id")
data <- data %>%
  left_join(annotate_scores("G1dTAG", "dTAG"), by = "id")
data <- data %>%
  left_join(annotate_scores("G1A485", "A485"), by = "id")
data <- data %>%
  left_join(annotate_scores("EpiG1DMSO", "EpiDMSO"), by = "id")
data <- data %>%
  left_join(annotate_scores("EpiG1dTAG", "EpidTAG"), by = "id")
data <- data %>%
  left_join(annotate_scores("GSE178982_AsyncUT", "UT"), by = "id")
data <- data %>%
  left_join(annotate_scores("GSE178982_AsyncAID", "AID"), by = "id")

################################################################################
fwrite(data, here(commonLoopDir, "loopScore_TjianChromo.tsv"),
       sep = "\t", col.names = TRUE)


```

# ANALYSIS
## [1] Comparing loops
### Common stuffs
```{r}
# Parameters
commonLoopDir <- here("../data/loop_analysis")
pseudoOoe <- 0.01

figDir <- here("../figure/loop_analysis")

```

### Import loops
```{r}
loop_chromo <- as_tibble(fread(here(commonLoopDir, "loopScore_chromosight.tsv")))
loop_hicdcp <- as_tibble(fread(here(commonLoopDir, "loopScore_hicdcp.tsv")))
loop_Hansen <- as_tibble(fread(here(commonLoopDir, "loopScore_Hansen.tsv")))
loop_fithic <- as_tibble(fread(here(commonLoopDir, "loopScore_fithic.tsv")))
loop_Tjian_chromo <- as_tibble(fread(here(commonLoopDir, "../loop_Tjian/TjianChromo_union_allRes.bedpe"))) %>%
  dplyr::filter(abs(V5-V2) <= 2000000)
colnames(loop_Tjian_chromo) <- c("chrom1", "start1", "end1", "chrom2", "start2", "end2")
loop_Tjian_mustache <- as_tibble(fread(here(commonLoopDir, "../loop_Tjian/TjianMustache_union_allRes.bedpe"))) %>%
  dplyr::filter(abs(V5-V2) <= 2000000)
colnames(loop_Tjian_mustache) <- c("chrom1", "start1", "end1", "chrom2", "start2", "end2")

```
### [FIG] Offset plot
#### Function
```{r}
# helper that only uses the two anchors
make_gi <- function(df) {
  a1 <- GRanges(df$chrom1, IRanges(df$start1, df$end1))
  a2 <- GRanges(df$chrom2, IRanges(df$start2, df$end2))
  GInteractions(a1, a2)
}

# Expand each anchor to 10 kb for comparison
expand_anchor <- function(gi) {
  a1 <- anchors(gi, "first")
  a2 <- anchors(gi, "second")
  a1r <- resize(a1, width = pmax(width(a1), expandSize), fix = "center")
  a2r <- resize(a2, width = pmax(width(a2), expandSize), fix = "center")
  GInteractions(a1r, a2r)
}
```
#### HiCDC+, Hansen
```{r}
for(expandSize in c(5000, 10000, 25000)){
  gi1 <- expand_anchor(make_gi(loop_chromo))
  gi2 <- expand_anchor(make_gi(loop_hicdcp))
  gi3 <- expand_anchor(make_gi(loop_Hansen))
  
  gi_all <- c(gi1, gi2, gi3)
  
  # 4) determine membership via Chebyshev distance = 1 bin(10 kb)
  in1 <- overlapsAny(gi_all, gi1,    maxgap=expandSize, type="equal")
  in2 <- overlapsAny(gi_all, gi2,    maxgap=expandSize, type="equal")
  in3 <- overlapsAny(gi_all, gi3,    maxgap=expandSize, type="equal")
  
  membership_df <- tibble(
    loop = seq_along(gi_all),
    set1 = in1,
    set2 = in2,
    set3 = in3
  )
  
  membership_df <- membership_df %>%
    rename(
      Chromosight     = set1,
      HiCDCP     = set2,
      Hansen = set3
    )
  
  # 5) UpSet plot of all combinations
  p <- upset(
    membership_df,
    intersect = c("Chromosight","HiCDCP","Hansen"),
    name = "",
    base_annotations = list(
      'Loop #' = intersection_size()
    )
  ) + ggtitle(paste0(expandSize, "bp expanded, 1 Chebychev dist"))
  
  
  fileName <- here(figDir, paste0("offSet_", expandSize, "bp_expanded"))
  width <- panelSize(5.5)*mmToInch
  height <- panelSize(3)*mmToInch
  svglite(paste0(fileName, ".svg"), width = width, height = height)
  print(p)
  dev.off()
}
```
#### Tjian
```{r}
for(expandSize in c(5000, 10000, 25000)){
  gi1 <- expand_anchor(make_gi(loop_chromo))
  gi2 <- expand_anchor(make_gi(loop_Tjian_chromo))
  gi3 <- expand_anchor(make_gi(loop_Tjian_mustache))
  
  gi_all <- c(gi1, gi2, gi3)
  
  # 4) determine membership via Chebyshev distance = 1 bin(10 kb)
  in1 <- overlapsAny(gi_all, gi1,    maxgap=expandSize, type="equal")
  in2 <- overlapsAny(gi_all, gi2,    maxgap=expandSize, type="equal")
  in3 <- overlapsAny(gi_all, gi3,    maxgap=expandSize, type="equal")
  
  membership_df <- tibble(
    loop = seq_along(gi_all),
    set1 = in1,
    set2 = in2,
    set3 = in3
  )
  
  membership_df <- membership_df %>%
    rename(
      Lee_chromo     = set1,
      Tjian_chromo     = set2,
      Tjian_mustache = set3
    )
  
  # 5) UpSet plot of all combinations
  p <- upset(
    membership_df,
    intersect = c("Lee_chromo","Tjian_chromo","Tjian_mustache"),
    name = "",
    base_annotations = list(
      'Loop #' = intersection_size(text = list(size = fontSizeS*ptToMM,
                                               family = fontType,
                                               color = "#000000"))
    ),
    matrix = intersection_matrix(
      geom = geom_point(size = 1.2)                         # shrink dots
    ),
    themes = upset_default_themes(
      text = element_text(
        size = fontSizeS,
        family = fontType,
        color = "#000000"
      ),
      axis.text.x = element_text(
        size = fontSizeS,
        family = fontType,
        color = "#000000"
      ),
      axis.text.y = element_text(
        size = fontSizeS,
        family = fontType,
        color = "#000000"
      )
    )
  ) + ggtitle(paste0(expandSize, "bp expanded, 1 Chebychev dist")) +
    theme(
      text = element_text(
        size = fontSizeS,
        family = fontType,
        color = "#000000"
      ),
      legend.position = "none",
      plot.title = element_text(
        hjust = 0.5,
        size = fontSizeS,
        family = fontType,
        color = "#000000"
        
      ),
      axis.title = element_text(
        size = fontSizeS,
        family = fontType,
        color = "#000000"
      ),
      axis.text = element_text(
        size = fontSizeS,
        family = fontType,
        color = "#000000"
      ),
      axis.line = element_line(
        color = "#000000",
        size = lineThick*mmToLineUnit,
        lineend = "square"
      ),
      axis.ticks = element_line(
        color = "#000000",
        size = lineThick*mmToLineUnit,
        lineend = "square"
      ),
      panel.background = element_rect(fill = "transparent"),
    )
  
  
  fileName <- here(figDir, paste0("offSet2_", expandSize, "bp_expanded"))
  width <- panelSize(2.4)*mmToInch
  height <- panelSize(1.7)*mmToInch  
  svglite(paste0(fileName, ".svg"), width = width, height = height)
  print(p)
  dev.off()
}
```
## [2] Visualizing loop annotation
### [FIG] Loop Anno pie chart
```{r}

draw_donut <- function(loop, name){
  n <- nrow(loop)
  temp <- loop %>% dplyr::select(id, Anno) %>%
    dplyr::mutate(AnnoGroup = case_when(
      Anno %in% c("P-P", "P-E", "E-E") ~ "reg_reg",
      Anno %in% c("P-SP", "P-SE", "E-SP", "E-SE") ~ "reg_strReg",
      Anno %in% c("P-S", "E-S") ~ "reg_str",
      Anno %in% c("SP-SP", "SP-SE", "SE-SE") ~ "strReg_strReg",
      Anno %in% c("SP-S", "SE-S") ~ "strReg_str",
      Anno %in% c("P-X", "E-X") ~ "reg_other",
      Anno %in% c("SP-X", "SE-X") ~ "strReg_other",
      Anno %in% c("S-S", "S-X") ~ "str",
      Anno %in% c("X-X") ~ "other",
      TRUE ~ NA
    ))
  
  
  temp$AnnoGroup <- factor(temp$AnnoGroup, levels = c("reg_reg",
                                                      "reg_strReg", "reg_str",
                                                      "strReg_strReg", "strReg_str",
                                                      "reg_other", "strReg_other", 
                                                      "str", "other"))
  
  
  pie_df <- temp %>%
    as_tibble() %>%                          # if temp was a list, this will error early
    group_by(AnnoGroup) %>%
    summarise(n = n(), .groups = "drop") %>%
    mutate(label = paste0(AnnoGroup, "\n(", n, ")"))
  
  p <- ggplot(pie_df, aes(x = 2, y = n, fill = AnnoGroup)) +
    geom_col(color = "black", width = 1) +
    coord_polar(theta = "y") +
    xlim(0.5, 2.5) +                # <- creates the “hole” by clipping the center
    geom_text(aes(label = label),
              position = position_stack(vjust = 0.5),
              size = 3) +
    theme_void() +
    theme(legend.position = "none",
          plot.title = element_text(hjust = 0.5)
    ) +
    ggtitle(paste0(name, "\n", n, " loops"))
  
  
  width = 3
  height = 3
  svglite(here(figDir,
               paste0("donut_loopAnno_", name, ".svg")),
          width = width, height = height)
  plot(p)
  dev.off()
}

draw_donut2 <- function(loop, name){
  n <- nrow(loop)
  temp <- loop %>% dplyr::select(id, Anno) %>%
    dplyr::mutate(AnnoGroup = case_when(
      Anno %in% c("P-P", "P-E", "E-E") ~ "Reg (P-P, P-E, E-E)",
      Anno %in% c("P-S", "E-S") ~ "Reg/Str mixed",
      Anno %in% c("P-SP", "P-SE", "E-SP", "E-SE") ~ "Reg/Str mixed",
      Anno %in% c("SP-SP", "SP-SE", "SE-SE") ~ "Reg/Str mixed",
      Anno %in% c("SP-S", "SE-S") ~ "Reg/Str mixed",
      Anno %in% c("S-S", "S-X") ~ "Str (S-S, S-X)",
      Anno %in% c("P-X", "E-X") ~ "Reg-Other",
      Anno %in% c("SP-X", "SE-X") ~ "StrReg-Other",
      Anno %in% c("X-X") ~ "Other",
      TRUE ~ NA
    ))
  
  
  temp$AnnoGroup <- factor(temp$AnnoGroup, levels = c("Reg (P-P, P-E, E-E)",
                                                      "Reg/Str mixed",
                                                      "Str (S-S, S-X)",
                                                      "Reg-Other",
                                                      "StrReg-Other",
                                                      "Other"))
  
  
  pie_df <- temp %>%
    as_tibble() %>%                          # if temp was a list, this will error early
    group_by(AnnoGroup) %>%
    summarise(n = n(), .groups = "drop") %>%
    mutate(label = paste0(AnnoGroup, "\n(", n, ")"))
  
  p <- ggplot(pie_df, aes(x = 2, y = n, fill = AnnoGroup)) +
    geom_col(color = "black", width = 1) +
    coord_polar(theta = "y") +
    xlim(0.5, 2.5) +                # <- creates the “hole” by clipping the center
    geom_text(aes(label = label),
              position = position_stack(vjust = 0.5),
              size = 3) +
    theme_void() +
    theme(legend.position = "none",
          plot.title = element_text(hjust = 0.5)
    ) +
    ggtitle(paste0(name, "\n", n, " loops"))
  
  
  width = 3
  height = 3
  svglite(here(figDir,
               paste0("donut_loopAnno2_", name, ".svg")),
          width = width, height = height)
  plot(p)
  dev.off()
  
  
  
}

draw_donut(loop_chromo, "chromosight")
draw_donut(loop_hicdcp, "hicdcp")
draw_donut(loop_Hansen, "Hansen")

draw_donut2(loop_chromo, "chromosight")
draw_donut2(loop_hicdcp, "hicdcp")
draw_donut2(loop_Hansen, "Hansen")

```
### [FIG] Loop Anno stacked barplot
```{r}
loop_chromo <- as_tibble(fread(here(commonLoopDir, "loopScore_chromosight.tsv")))
loop_hicdcp <- as_tibble(fread(here(commonLoopDir, "loopScore_hicdcp.tsv")))
loop_fithic <- as_tibble(fread(here(commonLoopDir, "loopScore_fithic.tsv")))

make_table <- function(loop, name){
  n <- nrow(loop)
  temp <- loop %>% dplyr::select(id, Anno) %>%
    dplyr::mutate(AnnoGroup = case_when(
      
      
      Anno %in% c("P-P", "P-E", "E-E") ~ "Reg (P-P, P-E, E-E)",
      
      Anno %in% c("P-S", "E-S") ~ "Reg/Str mixed",
      Anno %in% c("P-SP", "P-SE", "E-SP", "E-SE") ~ "Reg/Str mixed",
      Anno %in% c("SP-SP", "SP-SE", "SE-SE") ~ "Reg/Str mixed",
      Anno %in% c("SP-S", "SE-S") ~ "Reg/Str mixed",
      
      Anno %in% c("S-S", "S-X") ~ "Str (S-S, S-X)",
      Anno %in% c("P-X", "E-X") ~ "Reg-Other",
      Anno %in% c("SP-X", "SE-X") ~ "StrReg-Other",
      Anno %in% c("X-X") ~ "Other",
      TRUE ~ NA
    ))
  
  
  temp$AnnoGroup <- factor(temp$AnnoGroup, levels = c("Reg (P-P, P-E, E-E)",
                                                      "Reg/Str mixed",
                                                      "Str (S-S, S-X)",
                                                      "Reg-Other",
                                                      "StrReg-Other",
                                                      "Other"))
  
  
  pie_df <- temp %>%
    as_tibble() %>%                          # if temp was a list, this will error early
    group_by(AnnoGroup) %>%
    summarise(n = n(), .groups = "drop")
  colnames(pie_df) <- c("AnnoGroup", name)
  return(pie_df)
}

data <- make_table(loop_chromo, "chromosight") %>%
  dplyr::full_join(make_table(loop_hicdcp, "hicdcp"), by = c("AnnoGroup")) %>%
  dplyr::full_join(make_table(loop_fithic, "fithic"), by = c("AnnoGroup"))

data_long <- data %>% pivot_longer(cols = c("chromosight", "hicdcp", "fithic"))

p <- ggplot(data_long, aes(x = name, y = value, fill = AnnoGroup)) +
  geom_bar(position="fill", stat="identity") +
  labs(
    x = NULL,         # Remove x-axis title
    y = "Ratio"       # Change y-axis title
  ) +
  theme(
    axis.title = element_text(
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45, hjust = 1,          # Rotate x labels
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text.y = element_text(
      size = fontSizeS, family = fontType, color = "#000000"
    ),
    axis.line = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) + 
  annotate(
    "text", x = 1, y = 1,
    label = paste0("n = ", nrow(loop_chromo)),
    color = "black", hjust = 0, vjust = -1, size = fontSizeS*ptToMM, angle = 45
  ) + 
  annotate(
    "text", x = 2, y = 1,
    label = paste0("n = ", nrow(loop_fithic)),
    color = "black", hjust = 0, vjust = -1, size = fontSizeS*ptToMM, angle = 45
  )+ 
  annotate(
    "text", x = 3, y = 1,
    label = paste0("n = ", nrow(loop_hicdcp)),
    color = "black", hjust = 0, vjust = -1, size = fontSizeS*ptToMM, angle = 45
  )+   guides(
    fill = guide_legend(
      keywidth = 0.2,  # Adjust the width of the legend keys
      keyheight = 0.2  # Adjust the height of the legend keys
    )
  )

fileName <- paste0("Barplot_loopAnno_comparison")
width <- panelSize(2.2)*mmToInch
height <- panelSize(1.75)*mmToInch
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()


### Single barplot
data <- make_table(loop_chromo, "chromosight")
data_long <- data %>% pivot_longer(cols = c("chromosight"))

p <- ggplot(data_long, aes(x = name, y = value, fill = AnnoGroup)) +
  geom_bar(position="stack", stat="identity") +
  labs(
    x = NULL,         # Remove x-axis title
    y = "Counts"       # Change y-axis title
  ) +
  theme(
    axis.title = element_text(
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45, hjust = 1,          # Rotate x labels
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text.y = element_text(
      size = fontSizeS, family = fontType, color = "#000000"
    ),
    axis.line = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) +   guides(
    fill = guide_legend(
      keywidth = 0.2,  # Adjust the width of the legend keys
      keyheight = 0.2  # Adjust the height of the legend keys
    )
  )

fileName <- paste0("Barplot_loopAnno_Chromosight")
width <- panelSize(1.75)*mmToInch
height <- panelSize(1.75)*mmToInch
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()
```


## [3] Loop analysis
### Chromosight
#### [FUNCTIONS]
```{r}
# FUNCTIONS
extractAnchor <- function(loop){
  anchor1 <- loop %>% dplyr::select(c(1, 2, 3))
  colnames(anchor1) <- c("chr", "start", "end")
  anchor2 <- loop %>% dplyr::select(c(4, 5, 6))
  colnames(anchor2) <- c("chr", "start", "end")
  anchors <- reduce(makeGRangesFromDataFrame(bind_rows(anchor1, anchor2)))
  return(anchors)
}



getOverlapLoopNum <- function(loop, peak){
  anchor <- extractAnchor(loop)
  a <- unique(queryHits(findOverlaps(anchor, peak)))
  return(length(a))
}
```

#### [FIG] Loop size, strength
```{r}
commonLoopDir <- here("../data/loop_analysis")
figDir <- here("../figure/loop_analysis")
dir.create(figDir, showWarnings = FALSE, recursive = TRUE)
data <- fread(here(commonLoopDir, "loopScore_chromosight.tsv")) %>%
  dplyr::mutate(size = start2 - start1,
                group = case_when(
                  Anno %in% c("P-P", "P-E", "E-E") ~ "pure regulatory",
                  Anno %in% c("P-SP", "P-SE", "E-SP", "E-SE") ~ "reg_strReg",
                  Anno %in% c("P-S", "E-S") ~ "reg_str",
                  Anno %in% c("SP-SP", "SP-SE", "SE-SE") ~ "strReg_strReg",
                  Anno %in% c("SP-S", "SE-S") ~ "strReg_str",
                  Anno %in% c("P-X", "E-X") ~ "reg_other",
                  Anno %in% c("SP-X", "SE-X") ~ "strReg_other",
                  Anno %in% c("S-S", "S-X") ~ "str",
                  Anno %in% c("X-X") ~ "other",
                  TRUE ~ NA
                ))


ggplot(data, aes(y = size, x = group))  + geom_violin() +
  geom_boxplot(width = 0.3, color = "black",
               linewidth = lineThick * mmToLineUnit, lineend = "square",
               outlier.shape = NA,  alpha = 0.6, show.legend = FALSE) +
  coord_cartesian(ylim = c(0, 800*1000))


ggplot(data, aes(y = ooe_DMSO, x = Anno))  + geom_violin() +
  geom_boxplot(width = 0.3, color = "black",
               linewidth = lineThick * mmToLineUnit, lineend = "square",
               outlier.shape = NA,  alpha = 0.6, show.legend = FALSE)+
  coord_cartesian(ylim = c(0, 7))

```


#### [FIG] Scatterplot
```{r}
plot_loop_scores_and_save_bedpe <- function(df, col_x, col_y, 
                                            col_x_ooe, col_y_ooe, pseudo_ooe = 0.01,
                                            annoList,
                                            name, title, figDir, loopDir,
                                            diffCutoff = 0.2, density_n = 200,
                                            xmin = -0.5, xmax = 1,
                                            ymin = -0.5, ymax = 1, anno_col = "Anno_simple") {
  col_x_q <- enquo(col_x)
  col_y_q <- enquo(col_y)
  
  col_x_ooe_q <- enquo(col_x_ooe)
  col_y_ooe_q <- enquo(col_y_ooe)
  
  anno_col_sym <- sym(anno_col)
  
  
  # Create a single, unified dataframe for plotting and saving as bedpe
  df_temp <- df %>%
    dplyr::filter(!!anno_col_sym %in% annoList) %>%
    dplyr::select(chrom1, start1, end1, chrom2, start2, end2, id, x = !!col_x_q, y = !!col_y_q,
                  x_ooe = !!col_x_ooe_q, y_ooe = !!col_y_ooe_q) %>%
    dplyr::mutate(diff = y-x,
                  density = get_density(x, y, n = density_n)) %>%
    dplyr::mutate(updown = case_when(
      diff > diffCutoff  ~ "UP",
      diff < -diffCutoff ~ "DOWN",
      TRUE               ~ "NO"))
  
  # Export as bedpe
  bedpeUP <- df_temp %>% dplyr::filter(updown == "UP") %>% dplyr::select(chrom1, start1, end1, chrom2, start2, end2)
  fwrite(bedpeUP, here(loopDir, paste0(name, "_diffCutoff", diffCutoff, "_UP.bedpe")), sep = "\t", col.names = FALSE)
  bedpeNO <- df_temp %>% dplyr::filter(updown == "NO") %>% dplyr::select(chrom1, start1, end1, chrom2, start2, end2)
  fwrite(bedpeNO, here(loopDir, paste0(name, "_diffCutoff", diffCutoff, "_NO.bedpe")), sep = "\t", col.names = FALSE)
  bedpeDOWN <- df_temp %>% dplyr::filter(updown == "DOWN") %>% dplyr::select(chrom1, start1, end1, chrom2, start2, end2)
  fwrite(bedpeDOWN, here(loopDir, paste0(name, "_diffCutoff", diffCutoff, "_DOWN.bedpe")), sep = "\t", col.names = FALSE)
  
  
  plot_df <- df_temp %>%
    dplyr::select(id, x, y, updown, density) %>%
    arrange(density)
  
  plot_df$updown <- factor(plot_df$updown, c("UP", "NO", "DOWN"))
  
  # Calculate a global density range to make scales comparable
  density_limits <- range(plot_df$density, na.rm = TRUE)
  
  # ------ Scatterplot ------ 
  p1 <- ggplot(plot_df, aes(x = x, y = y)) +
    # Layer 1: Plot "NO" points with a grey scale
    geom_point(data = . %>% filter(updown == "NO") %>% arrange(density),
               aes(color = density), size = 0.1) +
    scale_color_gradient(
      low = "grey85", high = "grey30",
      limits = density_limits, name = "Density (NO)"
    ) +
    
    # Introduce a new scale for the "UP" points
    new_scale_color() +
    # Layer 2: Plot "UP" points with a red scale
    geom_point(data = . %>% filter(updown == "UP") %>% arrange(density),
               aes(color = density), size = 0.1) +
    scale_color_gradient(
      low = weak_red, high = strong_red, # Light to dark red
      limits = density_limits, name = "Density (UP)"
    ) +
    
    # Introduce a new scale for the "DOWN" points
    new_scale_color() +
    # Layer 3: Plot "DOWN" points with a blue scale
    geom_point(data = . %>% filter(updown == "DOWN") %>% arrange(density),
               aes(color = density), size = 0.1) +
    scale_color_gradient(
      low = weak_blue, high = strong_blue, # Light to dark blue
      limits = density_limits, name = "Density (DOWN)"
    ) +
    
    # Add annotations and theme elements
    xlim(-0.5, 1) + ylim(-0.5, 1) + coord_fixed() +
    geom_abline(slope = 1, intercept = 0, col = "grey50", linetype = "dashed",
                color = "black",
                size = lineThick*mmToLineUnit,
                lineend = "square") +
    geom_abline(slope = 1, intercept = -diffCutoff, col = "grey", linetype = "dotted",
                color = "black",
                size = lineThick*mmToLineUnit,
                lineend = "square") +
    geom_abline(slope = 1, intercept = diffCutoff, col = "grey", linetype = "dotted",
                color = "black",
                size = lineThick*mmToLineUnit,
                lineend = "square") +
    geom_hline(yintercept = 0, alpha = 0.5, color = "grey",
               color = "black",
               size = lineThick*mmToLineUnit,
               lineend = "square") +
    geom_vline(xintercept = 0, alpha = 0.5, color = "grey",
               color = "black",
               size = lineThick*mmToLineUnit,
               lineend = "square") +
    theme_classic() +
    theme(
      legend.position = "none",
      plot.title = element_text(
        hjust = 0.5,
        size = fontSizeS,
        family = fontType
      ),
      axis.title = element_text(
        size = fontSizeS,
        family = fontType,
        color = "#000000"
      ),
      axis.text = element_text(
        size = fontSizeS,
        family = fontType,
        color = "#000000"
      ),
      axis.line = element_line(
        color = "#000000",
        size = lineThick*mmToLineUnit,
        lineend = "square"
      ),
      axis.ticks = element_line(
        color = "#000000",
        size = lineThick*mmToLineUnit,
        lineend = "square"
      ),
      panel.background = element_rect(fill = "transparent"),
    ) +
    labs(x = quo_name(col_x_q), y = quo_name(col_y_q)) +
    ggtitle(title)
  
  width <- panelSize(1.5)*mmToInch
  height <- panelSize(1.5)*mmToInch  
  
  fileName <- paste0("scatterplot_", name)
  svglite(here(figDir, paste0(fileName, ".svg")), 
          width = width, height = height)
  print(p1)
  dev.off()
  
  png(here(figDir, paste0(fileName, ".png")), 
      width = width, height = height, units = "in", res = 600)
  print(p1)
  dev.off()
  
  # ------ Pie chart ------ 
  pie_df <- plot_df %>%
    dplyr::count(updown) %>%
    mutate(
      pct   = n / sum(n) * 100,
      label = paste0(updown, "\n", n, "\n(", round(pct, 1), "%)")
    )
  
  p2 <- ggplot(pie_df, aes(x = "", y = n, fill = updown)) +
    geom_col(color = "white") +
    coord_polar(theta = "y") +
    geom_text(
      aes(label = label),
      position = position_stack(vjust = 0.5),
      size     = fontSizeM / 2
    ) +
    theme_void() + theme(legend.position = "none") +
    scale_fill_manual(values = c("UP" = strong_red, "NO" = no_grey, "DOWN" = strong_blue))
  
  fileName <- here(figDir, paste0("piechart_", name))
  width <- panelSize(2)*mmToInch
  height <- panelSize(2)*mmToInch
  svglite(paste0(fileName, ".svg"), width = width, height = height)
  print(p2)
  dev.off()
  
  
  # ------ Violin plot of ooe ------ 
  plot_df_ooe <- df_temp %>%
    dplyr::select(id, x_ooe, y_ooe, updown) %>%
    dplyr::filter(!is.na(x_ooe), !is.na(y_ooe)) %>%
    dplyr::mutate(
      x_ooe = if_else(x_ooe == 0, pseudo_ooe, x_ooe),
      y_ooe = if_else(y_ooe == 0, pseudo_ooe, y_ooe),
      fc    = log2(y_ooe / x_ooe)
    )
  
  plot_df_ooe$updown <- factor(plot_df_ooe$updown, levels = c("UP", "NO", "DOWN"))
  
  p3 <- ggplot(plot_df_ooe, aes(x = updown, y = fc, fill = updown)) +
    geom_violin(linewidth = lineThick * mmToLineUnit, lineend = "square", alpha = .4, show.legend = FALSE,
                color = "black") + 
    geom_boxplot(width = 0.3, color = "black",
                 linewidth = lineThick * mmToLineUnit, lineend = "square",
                 outlier.shape = NA,  alpha = 0.6, show.legend = FALSE) + 
    theme_classic() + 
    stat_summary(
      aes(group = updown), fun = mean,
      geom = "point", shape = 21, size = 0.5,
      fill = "black", color = "black", position = position_dodge(.3)
    ) +
    coord_cartesian(ylim = c(-3, 3)) +
    scale_fill_manual(values = c(strong_red, "grey30", strong_blue)) +
    theme(
      axis.title.x = element_blank(),
      axis.title.y = element_text(
        size = fontSizeM,
        family = fontType,
        color = "#000000"
      ),
      axis.text = element_text(
        size = fontSizeS,
        family = fontType,
        color = "#000000"
      ),
      axis.text.x = element_text(
        size = fontSizeM,
        angle = 45,      # Rotate x-axis labels 45 degrees
        hjust = 1,       # Adjust horizontal justification
        vjust = 1        # Adjust vertical justification
      ),
      axis.line = element_line(
        color = "#000000",
        size = lineThick*mmToLineUnit,
        lineend = "square"
      ),
      axis.ticks = element_line(
        color = "#000000",
        size = lineThick*mmToLineUnit,
        lineend = "square"
      ),
      panel.background = element_rect(fill = "transparent"),
      legend.text = element_text(family = fontType, size = fontSizeS),
      legend.title = element_text(family = fontType, size = fontSizeS)
    ) +labs(y = "log2 fc of obs/exp")  +
    geom_hline(yintercept = 0,
               alpha = 1,
               color = "black",
               size = lineThick*mmToLineUnit,
               lineend = "square") +
    geom_hline(yintercept = c(-0.5, 0.5),
               alpha = 0.5,
               color = "black",
               size = lineThick*mmToLineUnit,
               lineend = "square", linetype = "dashed")
  
  
  width <- panelSize(1.1)*mmToInch
  height <- panelSize(1.5)*mmToInch
  fileName <- paste0("violin_", name)
  svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
  print(p3)
  dev.off()  
}


commonLoopDir <- here("../data/loop_analysis")
figDir <- here("../figure/loop_analysis")
dir.create(figDir, showWarnings = FALSE, recursive = TRUE)




data <- fread(here(commonLoopDir, "loopScore_chromosight.tsv"))
plot_loop_scores_and_save_bedpe(data, score_DMSO, score_dTAG, 
                                ooe_DMSO, ooe_dTAG, pseudo_ooe = 0.01,
                                c("P-P", "P-E", "P-S", "P-X",
                                  "E-E", "E-S", "E-X",
                                  "S-S", "S-X",
                                  "X-X"),
                                name = "chromosight_score_all_dTAGvsDMSO",
                                title = "chromo_score_all_dTAGvsDMSO", figDir, commonLoopDir)


plot_loop_scores_and_save_bedpe(data, score_DMSO, score_dTAG, 
                                ooe_DMSO, ooe_dTAG, pseudo_ooe = 0.01,
                                c("P-P", "P-E", "E-E"),
                                name = "chromosight_score_complex_reg_dTAGvsDMSO",
                                title = "chromo_score_complex_reg_dTAGvsDMSO", figDir, commonLoopDir,
                                anno_col = "Anno")

plot_loop_scores_and_save_bedpe(data, score_DMSO, score_dTAG, 
                                ooe_DMSO, ooe_dTAG, pseudo_ooe = 0.01,
                                c("S-S", "S-X"),
                                name = "chromosight_score_complex_str_dTAGvsDMSO",
                                title = "chromo_score_complex_str_dTAGvsDMSO", figDir, commonLoopDir,
                                anno_col = "Anno")

plot_loop_scores_and_save_bedpe(data, score_DMSO, score_dTAG, 
                                ooe_DMSO, ooe_dTAG, pseudo_ooe = 0.01,
                                c("S-S"),
                                name = "chromosight_score_complex_str2_dTAGvsDMSO",
                                title = "chromo_score_complex_str2__dTAGvsDMSO", figDir, commonLoopDir,
                                anno_col = "Anno")


plot_loop_scores_and_save_bedpe(data, score_DMSO, score_A485, 
                                ooe_DMSO, ooe_A485, pseudo_ooe = 0.01,
                                c("P-P", "P-E", "P-S", "P-X",
                                  "E-E", "E-S", "E-X",
                                  "S-S", "S-X",
                                  "X-X"),
                                name = "chromosight_score_all_A485vsDMSO",
                                title = "chromo_score_all_A485vsDMSO", figDir, commonLoopDir)

# EpiLC Micro-C analysis
plot_loop_scores_and_save_bedpe(data, score_DMSO, score_EpiDMSO, 
                                ooe_DMSO, ooe_EpiDMSO, pseudo_ooe = 0.01,
                                c("P-P", "P-E", "P-S", "P-X",
                                  "E-E", "E-S", "E-X",
                                  "S-S", "S-X",
                                  "X-X"),
                                name = "chromosight_score_all_EpiDMSOvsDMSO",
                                title = "chromo_score_all_EpiDMSOvsDMSO", figDir, commonLoopDir)
plot_loop_scores_and_save_bedpe(data, score_DMSO, score_EpiDMSO, 
                                ooe_DMSO, ooe_EpiDMSO, pseudo_ooe = 0.01,
                                c("P-P", "P-E", "E-E"),
                                name = "chromosight_score_complex_reg_EpiDMSOvsDMSO",
                                title = "chromo_score_complex_reg_EpiDMSOvsDMSO", figDir, commonLoopDir)
plot_loop_scores_and_save_bedpe(data, score_DMSO, score_EpiDMSO, 
                                ooe_DMSO, ooe_EpiDMSO, pseudo_ooe = 0.01,
                                c("S-S", "S-X"),
                                name = "chromosight_score_complex_str_EpiDMSOvsDMSO",
                                title = "chromo_score_complex_str_EpiDMSOvsDMSO", figDir, commonLoopDir)

plot_loop_scores_and_save_bedpe(data, score_EpiDMSO, score_EpidTAG, 
                                ooe_EpiDMSO, ooe_EpidTAG, pseudo_ooe = 0.01,
                                c("P-P", "P-E", "P-S", "P-X",
                                  "E-E", "E-S", "E-X",
                                  "S-S", "S-X",
                                  "X-X"),
                                name = "chromosight_score_all_EpidTAGvsEpiDMSO",
                                title = "chromo_score_all_EpidTAGvsEpiDMSO", figDir, commonLoopDir)
plot_loop_scores_and_save_bedpe(data, score_EpiDMSO, score_EpidTAG, 
                                ooe_EpiDMSO, ooe_EpidTAG, pseudo_ooe = 0.01,
                                c("P-P", "P-E", "E-E"),
                                name = "chromosight_score_complex_reg_EpidTAGvsEpiDMSO",
                                title = "chromo_score_complex_reg_EpidTAGvsEpiDMSO", figDir, commonLoopDir,
                                anno_col = "Anno")

plot_loop_scores_and_save_bedpe(data, score_EpiDMSO, score_EpidTAG, 
                                ooe_EpiDMSO, ooe_EpidTAG, pseudo_ooe = 0.01,
                                c("S-S", "S-X"),
                                name = "chromosight_score_complex_str_EpidTAGvsEpiDMSO",
                                title = "chromo_score_complex_str_EpidTAGvsEpiDMSO", figDir, commonLoopDir,
                                anno_col = "Anno")
# Exporting loops of interest
temp <- data %>%
  dplyr::filter(str_detect(Anno, "P")) %>% dplyr::select(chrom1, start1, end1, chrom2, start2, end2)

fwrite(temp, here(commonLoopDir, paste0("chromosight_score_complex_allP.bedpe")), sep = "\t", col.names = FALSE)



```
#### [FIG] Comparing OOE acrosso condition
```{r}
commonLoopDir <- here("../data/loop_analysis")
diffCutoff <- 0.2
pseudo_ooe <- 0.01
loop.all <- fread(here(commonLoopDir, "loopScore_chromosight.tsv")) %>%
  dplyr::filter(Anno_simple %in% c("P-P", "P-E", "E-E")) %>%
  dplyr::mutate(diff = score_dTAG - score_DMSO) %>%
  dplyr::mutate(updown = case_when(
    diff > diffCutoff  ~ "UP/NO",
    diff < -diffCutoff ~ "DOWN",
    TRUE               ~ "UP/NO"))

plot_df_ooe <- loop.all %>% dplyr::select(id, ooe_DMSO, ooe_dTAG, ooe_A485, updown) %>%
  dplyr::filter(!is.na(ooe_DMSO), !is.na(ooe_dTAG), !is.na(ooe_A485)) %>%
  dplyr::mutate(
    ooe_DMSO = if_else(ooe_DMSO == 0, pseudo_ooe, ooe_DMSO),
    ooe_dTAG = if_else(ooe_dTAG == 0, pseudo_ooe, ooe_dTAG),
    ooe_A485 = if_else(ooe_A485 == 0, pseudo_ooe, ooe_A485),
    fc_dTAG    = log2(ooe_dTAG / ooe_DMSO),
    fc_A485    = log2(ooe_A485 / ooe_DMSO)
  )


plot_df_ooe$updown <- factor(plot_df_ooe$updown, levels = c("UP/NO", "DOWN"))
plot_df_long <- plot_df_ooe %>%
  pivot_longer(
    cols      = c(fc_dTAG, fc_A485),
    names_to  = "treatment",
    values_to = "fc"
  )
plot_df_long$treatment <- factor(plot_df_long$treatment, levels = c("fc_dTAG", "fc_A485"))
p1 <- ggplot(plot_df_long, aes(x = updown, y = fc, fill = updown)) +
  geom_violin(
    linewidth   = lineThick * mmToLineUnit,
    lineend     = "square",
    alpha       = .4,
    show.legend = FALSE
  ) +
  geom_boxplot(
    width       = 0.3,
    color       = "black",
    linewidth   = lineThick * mmToLineUnit,
    lineend     = "square",
    outlier.shape = NA,
    alpha       = 0.6,
    show.legend = FALSE
  ) +
  stat_summary(
    aes(group = updown),
    fun     = mean,
    geom    = "point",
    shape   = 21,
    size    = 0.5,
    fill    = "black",
    color   = "black",
    position = position_dodge(.3)
  ) +
  scale_fill_manual(values = c("grey30", strong_blue)) +
  coord_cartesian(ylim = c(-3, 3)) +
  geom_hline(
    yintercept = 0,
    size       = lineThick * mmToLineUnit,
    lineend    = "square"
  ) +
  geom_hline(
    yintercept = c(-0.5, 0.5),
    linetype   = "dashed",
    alpha      = 0.5,
    size       = lineThick * mmToLineUnit,
    lineend    = "square"
  ) +
  facet_wrap(~ treatment, ncol = 2) +
  labs(y = "log2 fc of obs/exp") +
  theme_classic() +
  theme(
    axis.title.x  = element_blank(),
    axis.text.x   = element_text(angle = 45, hjust = 1, vjust = 1, size = fontSizeM),
    axis.title.y  = element_text(size = fontSizeM, family = fontType),
    axis.text     = element_text(size = fontSizeS, family = fontType),
    axis.line     = element_line(
      color = "#000000",
      size  = lineThick * mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks    = element_line(
      color = "#000000",
      size  = lineThick * mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent")
  )

name <- "chromosight_score_complex_reg_dTAGvsDMSO_compareA485"
width <- panelSize(2.5)*mmToInch
height <- panelSize(2)*mmToInch
fileName <- paste0("violin_", name)
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p1)
dev.off()  

```

#### [FIG] LOLA on anchor
##### ESC
```{r}
#-------------------------------------------------------------------------------
# LOADING LOLA DB
library("simpleCache")
library("LOLA")
lolaDB = loadRegionDB("/Volumes/UKJIN_SSD/reference/LOLACore_cached/mm10")

atac <- fread(here(refDir, "mm10", "GSM3106257_ATAC_ESC_1.bed")) %>% dplyr::select(V1, V2, V3)
colnames(atac) <- c("chr", "start", "end")
atac.gr <- makeGRangesFromDataFrame(atac)

lolaDir <- here(commonLoopDir, "LOLA")
dir.create(lolaDir, showWarnings = FALSE, recursive = TRUE)

```
###### All loops
```{r}
#-------------------------------------------------------------------------------
# LOADING LOOPS
### Importing differential regulatory loops & extract anchor
commonLoopDir <- here("../data/loop_analysis")
loop.all <- fread(here(commonLoopDir, "loopScore_chromosight.tsv"))
anchor.all <- (extractAnchor(loop.all))
overlaps <- findOverlaps(anchor.all, atac.gr)
anchor.all <- pintersect(anchor.all[queryHits(overlaps)], atac.gr[subjectHits(overlaps)])

loop.up <- fread(here(commonLoopDir, "chromosight_score_all_dTAGvsDMSO_diffCutoff0.2_UP.bedpe"))
anchor.up <- (extractAnchor(loop.up))
overlaps <- findOverlaps(anchor.up, atac.gr)
anchor.up <- pintersect(anchor.up[queryHits(overlaps)], atac.gr[subjectHits(overlaps)])

loop.no <- fread(here(commonLoopDir, "chromosight_score_all_dTAGvsDMSO_diffCutoff0.2_NO.bedpe"))
anchor.no <- (extractAnchor(loop.no))
overlaps <- findOverlaps(anchor.no, atac.gr)
anchor.no <- pintersect(anchor.no[queryHits(overlaps)], atac.gr[subjectHits(overlaps)])

loop.upno <- bind_rows(loop.up, loop.no)
anchor.upno <- (extractAnchor(loop.upno))
overlaps <- findOverlaps(anchor.upno, atac.gr)
anchor.upno <- pintersect(anchor.upno[queryHits(overlaps)], atac.gr[subjectHits(overlaps)])

loopNum <- nrow(loop.upno)

# Selecting same number of extreme down loops
diffCutoff <- 0.2
loop.down <- loop.all %>% dplyr::mutate(diff = score_dTAG - score_DMSO) %>%
  dplyr::filter(diff < -diffCutoff) %>% dplyr::arrange(diff) %>% slice_head(n = loopNum)
anchor.down <- (extractAnchor(loop.down))
overlaps <- findOverlaps(anchor.down, atac.gr)
anchor.down <- pintersect(anchor.down[queryHits(overlaps)], atac.gr[subjectHits(overlaps)])


#-------------------------------------------------------------------------------
# RUNNING LOLA
# UP
name <- "chromosight_score_all_dTAGvsDMSO_diffCutoff0.2"
result <- runLOLA(anchor.up, anchor.all, lolaDB)
tb <- as_tibble(result)
fwrite(tb, here(lolaDir, paste0("LOLA_", name, "UP_ATAC.tsv")), sep = "\t")

# NO
result <- runLOLA(anchor.no, anchor.all, lolaDB)
tb <- as_tibble(result)
fwrite(tb, here(lolaDir, paste0("LOLA_", name, "NO_ATAC.tsv")), sep = "\t")

# UPNO
result <- runLOLA(anchor.upno, anchor.all, lolaDB)
tb <- as_tibble(result)
fwrite(tb, here(lolaDir, paste0("LOLA_", name, "UPNO_ATAC.tsv")), sep = "\t")

# DOWN
result <- runLOLA(anchor.down, anchor.all, lolaDB)
tb <- as_tibble(result)
fwrite(tb, here(lolaDir, paste0("LOLA_", name, "DOWN_ATAC.tsv")), sep = "\t")


### Visualizing p-value and OR
alpha <- 0.05
tb.up <- fread(here(lolaDir, paste0("LOLA_", name, "UP_ATAC.tsv"))) %>%
  dplyr::mutate(target = toupper(antibody),
                qValueLog = -log2(qValue),
                group = "UP") %>%
  filter(str_to_lower(cellType) == "embryonic stem cell") %>%
  dplyr::filter(qValue < alpha) %>% dplyr::group_by(target) %>%
  slice_min(meanRnk, with_ties = FALSE)
tb.no <- fread(here(lolaDir, paste0("LOLA_", name, "NO_ATAC.tsv"))) %>%
  dplyr::mutate(target = toupper(antibody),
                qValueLog = -log2(qValue),
                group = "NO") %>%
  filter(str_to_lower(cellType) == "embryonic stem cell") %>%
  dplyr::filter(qValue < alpha) %>% dplyr::group_by(target) %>%
  slice_min(meanRnk, with_ties = FALSE)

tb.upno <- fread(here(lolaDir, paste0("LOLA_", name, "UPNO_ATAC.tsv"))) %>%
  dplyr::mutate(target = toupper(antibody),
                qValueLog = -log10(qValue),
                group = "UP/NO") %>%
  filter(str_to_lower(cellType) == "embryonic stem cell") %>%
  dplyr::filter(qValue < alpha) %>% dplyr::group_by(target) %>%
  slice_min(meanRnk, with_ties = FALSE)

fwrite(tb.upno %>% dplyr::select(c(24, 16, 20,
                                   4, 25, 5,
                                   7, 8, 9, 10, 11,
                                   12, 13, 14)), here(lolaDir, paste0("LOLA_", name, "UPNO_ATAC_pub.tsv")), sep = "\t")

tb.down <- fread(here(lolaDir, paste0("LOLA_", name, "DOWN_ATAC.tsv"))) %>%
  dplyr::mutate(target = toupper(antibody),
                qValueLog = -log10(qValue),
                group = "DOWN") %>%
  filter(str_to_lower(cellType) == "embryonic stem cell") %>%
  dplyr::filter(qValue < alpha) %>% dplyr::group_by(target) %>%
  slice_min(meanRnk, with_ties = FALSE)
fwrite(tb.down %>% dplyr::select(c(24, 16, 20,
                                   4, 25, 5,
                                   7, 8, 9, 10, 11,
                                   12, 13, 14)), here(lolaDir, paste0("LOLA_", name, "DOWN_ATAC__pub.tsv")), sep = "\t")

### For replotting without calculation
name <- "chromosight_score_all_dTAGvsDMSO_diffCutoff0.2"
tb.upno <- fread(here(lolaDir, paste0("LOLA_", name, "UPNO_ATAC_pub.tsv"))) %>%
  dplyr::mutate(group = "UP/NO")
tb.down <- fread(here(lolaDir, paste0("LOLA_", name, "DOWN_ATAC__pub.tsv"))) %>%
  dplyr::mutate(group = "DOWN")

temp.upno <- tb.upno %>% dplyr::select(target, oddsRatio, qValueLog, group)
temp.down <- tb.down %>% dplyr::select(target, oddsRatio, qValueLog, group)

temp <- bind_rows(temp.upno, temp.down)

order <- c((temp %>% dplyr::filter(group == "UP/NO") %>% arrange(desc(oddsRatio)))$target,
           (temp %>% dplyr::filter(group == "DOWN") %>% arrange(desc(oddsRatio)))$target)

temp$target <- factor(temp$target, levels = rev(order))
temp$group <- factor(temp$group, levels = c("UP/NO", "DOWN"))

targetList <- c("POLR2A", "CTR9",
                "AFF4", "ELL2",
                "MED1", "MED12",
                "TBP", "TAF1", "TAF3",
                "ESRRB", "KLF4", "NANOG", "POU5F1", "SOX2", "STAT3", "E2F1", "YY1",
                "EP300", "DPY30", "RBBP5", "KDM2B", "KDB4B", "KDM4C", "KDM6B", 
                "EZH2", "JARID2", "SUZ12", 
                "RAD21", "SMC1A", "SMC3", "CTCF"
)

temp <- temp %>% dplyr::filter(target %in% targetList)
temp$target <- factor(temp$target, levels = rev(targetList))


qValueLogMax <- 50
temp2 <- temp %>% dplyr::mutate(qValueLog = min(qValueLog, qValueLogMax))

p <- ggplot(temp2, aes(x = group, y = target, fill = oddsRatio, size = qValueLog)) +
  geom_point(shape = 21,        # Ensures a point with an outline
             stroke = 1*ptToMM      # Line width for the border
  ) + theme_bw() + 
  scale_size_continuous(range = c(0.5, 2)) +  # Set min and max point sizes here
  scale_fill_gradient(low = "white", high = "#CB333A",
                      limits = c(1, 3),
                      oob = scales::squish, # Define gradient colors
                      guide = guide_colorbar(
                        barwidth = 1.5/5.08,  # Adjust width of the color bar
                        barheight = 15/5.08   # Adjust height of the color bar
                      )
  ) + 
  labs(x = NULL, y = NULL)  +
  theme(
    axis.title = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  )

figDir <- here("../figure/loop_analysis")

fileName <- paste0("LOLA_", name)
width <- panelSize(1.5)*mmToInch
height <- panelSize(1.9)*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()
```

###### Reg loops
```{r}
#-------------------------------------------------------------------------------
# LOADING LOOPS
### Importing differential regulatory loops & extract anchor
commonLoopDir <- here("../data/loop_analysis")
loop.all <- fread(here(commonLoopDir, "loopScore_chromosight.tsv"))
anchor.all <- (extractAnchor(loop.all))
overlaps <- findOverlaps(anchor.all, atac.gr)
anchor.all <- pintersect(anchor.all[queryHits(overlaps)], atac.gr[subjectHits(overlaps)])

loop.up <- fread(here(commonLoopDir, "chromosight_score_complex_reg_dTAGvsDMSO_diffCutoff0.2_UP.bedpe"))
anchor.up <- (extractAnchor(loop.up))
overlaps <- findOverlaps(anchor.up, atac.gr)
anchor.up <- pintersect(anchor.up[queryHits(overlaps)], atac.gr[subjectHits(overlaps)])

loop.no <- fread(here(commonLoopDir, "chromosight_score_complex_reg_dTAGvsDMSO_diffCutoff0.2_NO.bedpe"))
anchor.no <- (extractAnchor(loop.no))
overlaps <- findOverlaps(anchor.no, atac.gr)
anchor.no <- pintersect(anchor.no[queryHits(overlaps)], atac.gr[subjectHits(overlaps)])

loop.upno <- bind_rows(loop.up, loop.no)
anchor.upno <- (extractAnchor(loop.upno))
overlaps <- findOverlaps(anchor.upno, atac.gr)
anchor.upno <- pintersect(anchor.upno[queryHits(overlaps)], atac.gr[subjectHits(overlaps)])

loop.down <- fread(here(commonLoopDir, "chromosight_score_complex_reg_dTAGvsDMSO_diffCutoff0.2_DOWN.bedpe"))
anchor.down <- (extractAnchor(loop.down))
overlaps <- findOverlaps(anchor.down, atac.gr)
anchor.down <- pintersect(anchor.down[queryHits(overlaps)], atac.gr[subjectHits(overlaps)])

#-------------------------------------------------------------------------------
# RUNNING LOLA
# UP
lolaDir <- here(commonLoopDir, "LOLA")
dir.create(lolaDir, showWarnings = FALSE, recursive = TRUE)


name <- "chromosight_score_complex_reg_dTAGvsDMSO_diffCutoff0.2"
result <- runLOLA(anchor.up, anchor.all, lolaDB)
tb <- as_tibble(result)
fwrite(tb, here(lolaDir, paste0("LOLA_", name, "UP_ATAC.tsv")), sep = "\t")

# NO
result <- runLOLA(anchor.no, anchor.all, lolaDB)
tb <- as_tibble(result)
fwrite(tb, here(lolaDir, paste0("LOLA_", name, "NO_ATAC.tsv")), sep = "\t")

# UPNO
result <- runLOLA(anchor.upno, anchor.all, lolaDB)
tb <- as_tibble(result)
fwrite(tb, here(lolaDir, paste0("LOLA_", name, "UPNO_ATAC.tsv")), sep = "\t")

# DOWN
result <- runLOLA(anchor.down, anchor.all, lolaDB)
tb <- as_tibble(result)
fwrite(tb, here(lolaDir, paste0("LOLA_", name, "DOWN_ATAC.tsv")), sep = "\t")


### Visualizing p-value and OR
alpha <- 0.05
tb.up <- fread(here(lolaDir, paste0("LOLA_", name, "UP_ATAC.tsv"))) %>%
  dplyr::mutate(target = toupper(antibody),
                qValueLog = -log2(qValue),
                group = "UP") %>%
  filter(str_to_lower(cellType) == "embryonic stem cell") %>%
  dplyr::filter(qValue < alpha) %>% dplyr::group_by(target) %>%
  slice_min(meanRnk, with_ties = FALSE)
tb.no <- fread(here(lolaDir, paste0("LOLA_", name, "NO_ATAC.tsv"))) %>%
  dplyr::mutate(target = toupper(antibody),
                qValueLog = -log2(qValue),
                group = "NO") %>%
  filter(str_to_lower(cellType) == "embryonic stem cell") %>%
  dplyr::filter(qValue < alpha) %>% dplyr::group_by(target) %>%
  slice_min(meanRnk, with_ties = FALSE)

tb.upno <- fread(here(lolaDir, paste0("LOLA_", name, "UPNO_ATAC.tsv"))) %>%
  dplyr::mutate(target = toupper(antibody),
                qValueLog = -log10(qValue),
                group = "UP/NO") %>%
  filter(str_to_lower(cellType) == "embryonic stem cell") %>%
  dplyr::filter(qValue < alpha) %>% dplyr::group_by(target) %>%
  slice_min(meanRnk, with_ties = FALSE)

fwrite(tb.upno %>% dplyr::select(c(24, 16, 20,
                                   4, 25, 5,
                                   7, 8, 9, 10, 11,
                                   12, 13, 14)), here(lolaDir, paste0("LOLA_", name, "UPNO_ATAC_pub.tsv")), sep = "\t")

tb.down <- fread(here(lolaDir, paste0("LOLA_", name, "DOWN_ATAC.tsv"))) %>%
  dplyr::mutate(target = toupper(antibody),
                qValueLog = -log10(qValue),
                group = "DOWN") %>%
  filter(str_to_lower(cellType) == "embryonic stem cell") %>%
  dplyr::filter(qValue < alpha) %>% dplyr::group_by(target) %>%
  slice_min(meanRnk, with_ties = FALSE)
fwrite(tb.down %>% dplyr::select(c(24, 16, 20,
                                   4, 25, 5,
                                   7, 8, 9, 10, 11,
                                   12, 13, 14)), here(lolaDir, paste0("LOLA_", name, "DOWN_ATAC__pub.tsv")), sep = "\t")


temp.upno <- tb.upno %>% dplyr::select(target, oddsRatio, qValueLog, group)
temp.down <- tb.down %>% dplyr::select(target, oddsRatio, qValueLog, group)

temp <- bind_rows(temp.upno, temp.down)

order <- c((temp %>% dplyr::filter(group == "UP/NO") %>% arrange(desc(oddsRatio)))$target,
           (temp %>% dplyr::filter(group == "DOWN") %>% arrange(desc(oddsRatio)))$target)
order <- unique(order)

temp$target <- factor(temp$target, levels = rev(order))
temp$group <- factor(temp$group, levels = c("UP/NO", "DOWN"))

targetList <- c("POLR2A", "CTR9", "ARR4", "ELL2",
                "MED1", "MED12", "TBP", "TAF1", "TAF2",
                "ESRRB", "KLF4", "NANOG", "POU5F1", "SOX2", "MYCN", "GATA4", "STAT3", "E2F1", "YY1",
                "P300", "DPY30", "RBBP5", "KDM2B", "KDM4B", "KDM6B", "TET1",
                "CBX7", "RING1B", "RNF2", "EZH2", "SUZ12", "JARID2", "MTF2", "PHF19")

temp <- temp %>% dplyr::filter(target %in% targetList)
temp$target <- factor(temp$target, levels = rev(targetList))


qValueLogMax <- 50
temp2 <- temp %>% dplyr::mutate(qValueLog = min(qValueLog, qValueLogMax))

p <- ggplot(temp2, aes(x = group, y = target, fill = oddsRatio, size = qValueLog)) +
  geom_point(shape = 21,        # Ensures a point with an outline
             stroke = 1*ptToMM      # Line width for the border
  ) + theme_bw() + 
  scale_size_continuous(range = c(0.5, 2)) +  # Set min and max point sizes here
  scale_fill_gradient(low = "white", high = "#CB333A",
                      limits = c(1, 3),
                      oob = scales::squish, # Define gradient colors
                      guide = guide_colorbar(
                        barwidth = 1.5/5.08,  # Adjust width of the color bar
                        barheight = 15/5.08   # Adjust height of the color bar
                      )
  ) + 
  labs(x = NULL, y = NULL)  +
  theme(
    axis.title = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  )

figDir <- here("../figure/loop_analysis")

fileName <- paste0("LOLA_", name)
width <- panelSize(1.5)*mmToInch
height <- panelSize(1.9)*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()
```
##### EpiLC
###### All loops
```{r}
#-------------------------------------------------------------------------------
# LOADING LOOPS
### Importing differential regulatory loops & extract anchor
commonLoopDir <- here("../data/loop_analysis")
loop.all <- fread(here(commonLoopDir, "loopScore_chromosight.tsv"))
anchor.all <- (extractAnchor(loop.all))
overlaps <- findOverlaps(anchor.all, atac.gr)
anchor.all <- pintersect(anchor.all[queryHits(overlaps)], atac.gr[subjectHits(overlaps)])

loop.up <- fread(here(commonLoopDir, "chromosight_score_all_EpidTAGvsEpiDMSO_diffCutoff0.2_UP.bedpe"))
anchor.up <- (extractAnchor(loop.up))
overlaps <- findOverlaps(anchor.up, atac.gr)
anchor.up <- pintersect(anchor.up[queryHits(overlaps)], atac.gr[subjectHits(overlaps)])

loop.no <- fread(here(commonLoopDir, "chromosight_score_all_EpidTAGvsEpiDMSO_diffCutoff0.2_NO.bedpe"))
anchor.no <- (extractAnchor(loop.no))
overlaps <- findOverlaps(anchor.no, atac.gr)
anchor.no <- pintersect(anchor.no[queryHits(overlaps)], atac.gr[subjectHits(overlaps)])

loop.upno <- bind_rows(loop.up, loop.no)
anchor.upno <- (extractAnchor(loop.upno))
overlaps <- findOverlaps(anchor.upno, atac.gr)
anchor.upno <- pintersect(anchor.upno[queryHits(overlaps)], atac.gr[subjectHits(overlaps)])

loopNum <- nrow(loop.upno)

# Selecting same number of extreme down loops
loop.down <- fread(here(commonLoopDir, "chromosight_score_all_EpidTAGvsEpiDMSO_diffCutoff0.2_DOWN.bedpe"))
anchor.down <- (extractAnchor(loop.down))
overlaps <- findOverlaps(anchor.down, atac.gr)
anchor.down <- pintersect(anchor.down[queryHits(overlaps)], atac.gr[subjectHits(overlaps)])


#-------------------------------------------------------------------------------
# RUNNING LOLA
# UP
lolaDir <- here(commonLoopDir, "LOLA")
dir.create(lolaDir, showWarnings = FALSE, recursive = TRUE)


name <- "chromosight_score_all_EpidTAGvsEpiDMSO_diffCutoff0.2"
result <- runLOLA(anchor.up, anchor.all, lolaDB)
tb <- as_tibble(result)
fwrite(tb, here(lolaDir, paste0("LOLA_", name, "UP_ATAC.tsv")), sep = "\t")

# NO
result <- runLOLA(anchor.no, anchor.all, lolaDB)
tb <- as_tibble(result)
fwrite(tb, here(lolaDir, paste0("LOLA_", name, "NO_ATAC.tsv")), sep = "\t")

# UPNO
result <- runLOLA(anchor.upno, anchor.all, lolaDB)
tb <- as_tibble(result)
fwrite(tb, here(lolaDir, paste0("LOLA_", name, "UPNO_ATAC.tsv")), sep = "\t")

# DOWN
result <- runLOLA(anchor.down, anchor.all, lolaDB)
tb <- as_tibble(result)
fwrite(tb, here(lolaDir, paste0("LOLA_", name, "DOWN_ATAC.tsv")), sep = "\t")


### Visualizing p-value and OR
alpha <- 0.05
tb.up <- fread(here(lolaDir, paste0("LOLA_", name, "UP_ATAC.tsv"))) %>%
  dplyr::mutate(target = toupper(antibody),
                qValueLog = -log2(qValue),
                group = "UP") %>%
  filter(str_to_lower(cellType) == "embryonic stem cell") %>%
  dplyr::filter(qValue < alpha) %>% dplyr::group_by(target) %>%
  slice_min(meanRnk, with_ties = FALSE)
tb.no <- fread(here(lolaDir, paste0("LOLA_", name, "NO_ATAC.tsv"))) %>%
  dplyr::mutate(target = toupper(antibody),
                qValueLog = -log2(qValue),
                group = "NO") %>%
  filter(str_to_lower(cellType) == "embryonic stem cell") %>%
  dplyr::filter(qValue < alpha) %>% dplyr::group_by(target) %>%
  slice_min(meanRnk, with_ties = FALSE)

tb.upno <- fread(here(lolaDir, paste0("LOLA_", name, "UPNO_ATAC.tsv"))) %>%
  dplyr::mutate(target = toupper(antibody),
                qValueLog = -log10(qValue),
                group = "UP/NO") %>%
  filter(str_to_lower(cellType) == "embryonic stem cell") %>%
  dplyr::filter(qValue < alpha) %>% dplyr::group_by(target) %>%
  slice_min(meanRnk, with_ties = FALSE)

fwrite(tb.upno %>% dplyr::select(c(24, 16, 20,
                                   4, 25, 5,
                                   7, 8, 9, 10, 11,
                                   12, 13, 14)), here(lolaDir, paste0("LOLA_", name, "UPNO_ATAC_pub.tsv")), sep = "\t")

tb.down <- fread(here(lolaDir, paste0("LOLA_", name, "DOWN_ATAC.tsv"))) %>%
  dplyr::mutate(target = toupper(antibody),
                qValueLog = -log10(qValue),
                group = "DOWN") %>%
  filter(str_to_lower(cellType) == "embryonic stem cell") %>%
  dplyr::filter(qValue < alpha) %>% dplyr::group_by(target) %>%
  slice_min(meanRnk, with_ties = FALSE)
fwrite(tb.down %>% dplyr::select(c(24, 16, 20,
                                   4, 25, 5,
                                   7, 8, 9, 10, 11,
                                   12, 13, 14)), here(lolaDir, paste0("LOLA_", name, "DOWN_ATAC__pub.tsv")), sep = "\t")


temp.upno <- tb.upno %>% dplyr::select(target, oddsRatio, qValueLog, group)
temp.down <- tb.down %>% dplyr::select(target, oddsRatio, qValueLog, group)

temp <- bind_rows(temp.upno, temp.down)

order <- unique(c((temp %>% dplyr::filter(group == "UP/NO") %>% arrange(desc(oddsRatio)))$target,
           (temp %>% dplyr::filter(group == "DOWN") %>% arrange(desc(oddsRatio)))$target))

temp$target <- factor(temp$target, levels = rev(order))
temp$group <- factor(temp$group, levels = c("UP/NO", "DOWN"))

targetList <- c("POLR2A", "CTR9",
                "AFF4", "ELL2",
                "MED1", "MED12",
                "TBP", "TAF1", "TAF3",
                "ESRRB", "KLF4", "NANOG", "POU5F1", "E2F1", "YY1", "MYCN", "GATA4",
                "EP300", "DPY30", "RBBP5", "KDM2B", "KDM4B", "KDM4C", "KDM6B", 
                "PRDM14", "TET1", "RNF2", "EZH2",  "SUZ12", "JARID2", 
                "RAD21", "SMC1A", "SMC3", "CTCF"
)



temp <- temp %>% dplyr::filter(target %in% targetList)
temp$target <- factor(temp$target, levels = rev(targetList))


qValueLogMax <- 50
temp2 <- temp %>% dplyr::mutate(qValueLog = min(qValueLog, qValueLogMax))

p <- ggplot(temp2, aes(x = group, y = target, fill = oddsRatio, size = qValueLog)) +
  geom_point(shape = 21,        # Ensures a point with an outline
             stroke = 1*ptToMM      # Line width for the border
  ) + theme_bw() + 
  scale_size_continuous(range = c(0.5, 2)) +  # Set min and max point sizes here
  scale_fill_gradient(low = "white", high = "#CB333A",
                      limits = c(1, 3),
                      oob = scales::squish, # Define gradient colors
                      guide = guide_colorbar(
                        barwidth = 1.5/5.08,  # Adjust width of the color bar
                        barheight = 15/5.08   # Adjust height of the color bar
                      )
  ) + 
  labs(x = NULL, y = NULL)  +
  theme(
    axis.title = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  )

figDir <- here("../figure/loop_analysis")

fileName <- paste0("LOLA_", name)
width <- panelSize(1.5)*mmToInch
height <- panelSize(1.9)*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()
```
###### Reg loops
```{r}
#-------------------------------------------------------------------------------
# LOADING LOOPS
### Importing differential regulatory loops & extract anchor
commonLoopDir <- here("../data/loop_analysis")
loop.all <- fread(here(commonLoopDir, "loopScore_chromosight.tsv"))
anchor.all <- (extractAnchor(loop.all))
overlaps <- findOverlaps(anchor.all, atac.gr)
anchor.all <- pintersect(anchor.all[queryHits(overlaps)], atac.gr[subjectHits(overlaps)])

loop.up <- fread(here(commonLoopDir, "chromosight_score_complex_reg_EpidTAGvsEpiDMSO_diffCutoff0.2_UP.bedpe"))
anchor.up <- (extractAnchor(loop.up))
overlaps <- findOverlaps(anchor.up, atac.gr)
anchor.up <- pintersect(anchor.up[queryHits(overlaps)], atac.gr[subjectHits(overlaps)])

loop.no <- fread(here(commonLoopDir, "chromosight_score_complex_reg_EpidTAGvsEpiDMSO_diffCutoff0.2_NO.bedpe"))
anchor.no <- (extractAnchor(loop.no))
overlaps <- findOverlaps(anchor.no, atac.gr)
anchor.no <- pintersect(anchor.no[queryHits(overlaps)], atac.gr[subjectHits(overlaps)])

loop.upno <- bind_rows(loop.up, loop.no)
anchor.upno <- (extractAnchor(loop.upno))
overlaps <- findOverlaps(anchor.upno, atac.gr)
anchor.upno <- pintersect(anchor.upno[queryHits(overlaps)], atac.gr[subjectHits(overlaps)])

loop.down <- fread(here(commonLoopDir, "chromosight_score_complex_reg_EpidTAGvsEpiDMSO_diffCutoff0.2_DOWN.bedpe"))
anchor.down <- (extractAnchor(loop.down))
overlaps <- findOverlaps(anchor.down, atac.gr)
anchor.down <- pintersect(anchor.down[queryHits(overlaps)], atac.gr[subjectHits(overlaps)])

#-------------------------------------------------------------------------------
# RUNNING LOLA
# UP
lolaDir <- here(commonLoopDir, "LOLA")
dir.create(lolaDir, showWarnings = FALSE, recursive = TRUE)


name <- "chromosight_score_complex_reg_EpidTAGvsEpiDMSO_diffCutoff0.2"
result <- runLOLA(anchor.up, anchor.all, lolaDB)
tb <- as_tibble(result)
fwrite(tb, here(lolaDir, paste0("LOLA_", name, "UP_ATAC.tsv")), sep = "\t")

# NO
result <- runLOLA(anchor.no, anchor.all, lolaDB)
tb <- as_tibble(result)
fwrite(tb, here(lolaDir, paste0("LOLA_", name, "NO_ATAC.tsv")), sep = "\t")

# UPNO
result <- runLOLA(anchor.upno, anchor.all, lolaDB)
tb <- as_tibble(result)
fwrite(tb, here(lolaDir, paste0("LOLA_", name, "UPNO_ATAC.tsv")), sep = "\t")

# DOWN
result <- runLOLA(anchor.down, anchor.all, lolaDB)
tb <- as_tibble(result)
fwrite(tb, here(lolaDir, paste0("LOLA_", name, "DOWN_ATAC.tsv")), sep = "\t")


### Visualizing p-value and OR
alpha <- 0.05
tb.up <- fread(here(lolaDir, paste0("LOLA_", name, "UP_ATAC.tsv"))) %>%
  dplyr::mutate(target = toupper(antibody),
                qValueLog = -log2(qValue),
                group = "UP") %>%
  filter(str_to_lower(cellType) == "embryonic stem cell") %>%
  dplyr::filter(qValue < alpha) %>% dplyr::group_by(target) %>%
  slice_min(meanRnk, with_ties = FALSE)
tb.no <- fread(here(lolaDir, paste0("LOLA_", name, "NO_ATAC.tsv"))) %>%
  dplyr::mutate(target = toupper(antibody),
                qValueLog = -log2(qValue),
                group = "NO") %>%
  filter(str_to_lower(cellType) == "embryonic stem cell") %>%
  dplyr::filter(qValue < alpha) %>% dplyr::group_by(target) %>%
  slice_min(meanRnk, with_ties = FALSE)

tb.upno <- fread(here(lolaDir, paste0("LOLA_", name, "UPNO_ATAC.tsv"))) %>%
  dplyr::mutate(target = toupper(antibody),
                qValueLog = -log10(qValue),
                group = "UP/NO") %>%
  filter(str_to_lower(cellType) == "embryonic stem cell") %>%
  dplyr::filter(qValue < alpha) %>% dplyr::group_by(target) %>%
  slice_min(meanRnk, with_ties = FALSE)

fwrite(tb.upno %>% dplyr::select(c(24, 16, 20,
                                   4, 25, 5,
                                   7, 8, 9, 10, 11,
                                   12, 13, 14)), here(lolaDir, paste0("LOLA_", name, "UPNO_ATAC_pub.tsv")), sep = "\t")

tb.down <- fread(here(lolaDir, paste0("LOLA_", name, "DOWN_ATAC.tsv"))) %>%
  dplyr::mutate(target = toupper(antibody),
                qValueLog = -log10(qValue),
                group = "DOWN") %>%
  filter(str_to_lower(cellType) == "embryonic stem cell") %>%
  dplyr::filter(qValue < alpha) %>% dplyr::group_by(target) %>%
  slice_min(meanRnk, with_ties = FALSE)
fwrite(tb.down %>% dplyr::select(c(24, 16, 20,
                                   4, 25, 5,
                                   7, 8, 9, 10, 11,
                                   12, 13, 14)), here(lolaDir, paste0("LOLA_", name, "DOWN_ATAC__pub.tsv")), sep = "\t")


temp.upno <- tb.upno %>% dplyr::select(target, oddsRatio, qValueLog, group)
temp.down <- tb.down %>% dplyr::select(target, oddsRatio, qValueLog, group)

temp <- bind_rows(temp.upno, temp.down)

order <- c((temp %>% dplyr::filter(group == "UP/NO") %>% arrange(desc(oddsRatio)))$target,
           (temp %>% dplyr::filter(group == "DOWN") %>% arrange(desc(oddsRatio)))$target)
order <- unique(order)

temp$target <- factor(temp$target, levels = rev(order))
temp$group <- factor(temp$group, levels = c("UP/NO", "DOWN"))

targetList <- c("POLR2A", "AFF4", "SUPT5", "SUPT5H",
                "MED1", "MED12", "TBP", "TAF1", "TAF3", 
                "ESRRB", "KLF4", "NANOG", "POU5F1", "SOX2", "MYCN", "GATA4", "STAT3", "E2F1", "YY1", "TRP53",
                "P300", "RBBP5", "KDM2B", "KDM4B", "KDM6B", "TET1", "PRDM14", "CTNNB1",
                "CBX7", "RING1B", "RNF2", "EZH2", "SUZ12", "MTF2", "PHF19")

temp <- temp %>% dplyr::filter(target %in% targetList)
temp$target <- factor(temp$target, levels = rev(targetList))


qValueLogMax <- 50
temp2 <- temp %>% dplyr::mutate(qValueLog = min(qValueLog, qValueLogMax))

p <- ggplot(temp2, aes(x = group, y = target, fill = oddsRatio, size = qValueLog)) +
  geom_point(shape = 21,        # Ensures a point with an outline
             stroke = 1*ptToMM      # Line width for the border
  ) + theme_bw() + 
  scale_size_continuous(range = c(0.5, 2)) +  # Set min and max point sizes here
  scale_fill_gradient(low = "white", high = "#CB333A",
                      limits = c(1, 3),
                      oob = scales::squish, # Define gradient colors
                      guide = guide_colorbar(
                        barwidth = 1.5/5.08,  # Adjust width of the color bar
                        barheight = 15/5.08   # Adjust height of the color bar
                      )
  ) + 
  labs(x = NULL, y = NULL)  +
  theme(
    axis.title = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  )

figDir <- here("../figure/loop_analysis")

fileName <- paste0("LOLA_", name)
width <- panelSize(1.5)*mmToInch
height <- panelSize(1.9)*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()
```


#### [FIG] Distance vs OOE plot per group
```{r}
# Import annotation
commonLoopDir <- here("../data/loop_analysis")
figDir <- here("../figure/loop_analysis")

pseudo_ooe <- 0.01

data <- fread(here(commonLoopDir, "loopScore_chromosight.tsv")) %>%
  dplyr::filter(!is.na(ooe_DMSO), !is.na(ooe_dTAG)) %>%
  dplyr::mutate(
    x_ooe = if_else(ooe_DMSO == 0, pseudo_ooe, ooe_DMSO),
    y_ooe = if_else(ooe_dTAG == 0, pseudo_ooe, ooe_dTAG),
    fc    = log2(y_ooe / x_ooe),
    distance = start2 - start1,
    dist_bin = cut(
      distance,
      breaks = c(seq(0, 1e6, by = 1e5), Inf),
      labels = c(
        paste0(seq(0, 9e5, 1e5) / 1e3, "-", seq(1e5, 1e6, 1e5) / 1e3, "kb"),
        ">1Mb"
      ),
      right = FALSE
    )) %>%
  dplyr::select(distance, dist_bin, fc, Anno)

data <- data %>% dplyr::mutate(
  group = case_when(
    Anno %in% c("P-P", "P-E", "E-E") ~ "reg_reg",
    Anno %in% c("P-SP", "P-SE", "E-SP", "E-SE") ~ "reg_strReg",
    Anno %in% c("P-S", "E-S") ~ "reg_str",
    Anno %in% c("SP-SP", "SP-SE", "SE-SE") ~ "strReg_strReg",
    Anno %in% c("SP-S", "SE-S") ~ "strReg_str",
    Anno %in% c("P-X", "E-X") ~ "reg_other",
    Anno %in% c("SP-X", "SE-X") ~ "strReg_other",
    Anno %in% c("S-S", "S-X") ~ "str",
    Anno %in% c("X-X") ~ "other",
    
    TRUE ~ "XXX")
) %>%
  dplyr::mutate(group2 = case_when(
    group %in% c("reg_reg") ~ "reg_reg",
    group %in% c("reg_strReg", "reg_str", "reg_other") ~ "reg_nonReg",
    group %in% c("strReg_strReg", "strReg_str", "strReg_other") ~ "strReg_nonReg",
    group %in% c("str") ~ "str",
    TRUE ~ "XXXXX"
  ))





data$group2 <- factor(data$group2, levels = c("reg_reg", "reg_nonReg", "strReg_nonReg", "str"))

p <- ggplot(
  data %>% dplyr::filter(group2 %in% c("reg_reg", "str")),
  aes(x = dist_bin, y = fc, fill = group2)
) +
  # remove outliers with outlier.shape = NA (there is no "outliers" argument)
  geom_boxplot(outlier.shape = NA) +
  # keep empty bins
  scale_x_discrete(drop = FALSE) +
  # add red‐filled mean points, dodged by group
  stat_summary(
    aes(group = group2),
    fun = mean,
    geom = "point",
    shape = 21,
    size = 0.5,
    fill = "red",
    color = "red",
    position = position_dodge(0.8)
  ) +
  labs(
    x = "Distance bin",
    y = "Δ log2(obs/exp)",
  ) +
  theme_classic() +
  theme(
    axis.text.x        = element_text(angle = 45, hjust = 1),
    strip.background   = element_rect(fill = "white", color = "black"),
    strip.text         = element_text(face = "bold")
  ) +
  geom_hline(yintercept = 0) +
  coord_cartesian(ylim = c(-3, 2)) +
  theme(
    axis.title       = element_text(size = fontSizeS, family = fontType),
    axis.text        = element_text(size = fontSizeS, family = fontType),
    axis.line        = element_line(size = lineThick * mmToLineUnit, lineend = "square"),
    axis.ticks       = element_line(size = lineThick * mmToLineUnit, lineend = "square"),
    panel.background = element_rect(fill = "transparent"),
    legend.title     = element_text(family = fontType, size = fontSizeS),
    legend.text      = element_text(family = fontType, size = fontSizeS)
  ) +
  # close the guides(...) call properly and map the correct aesthetic
  guides(
    fill = guide_legend(
      keywidth  = 0.5,
      keyheight = 0.5
    )
  )

name <- "chromosight_score_all_dTAGvsDMSO_diffCutoff0.2"


fileName <- paste0("dist_vs_score_binned_reg_str_", name)

width <- panelSize(2.5)*mmToInch
height <- panelSize(1.75)*mmToInch
png(here(figDir, paste0(fileName, ".png")), 
    res = 600, units = "in", width = width, height = height)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")), 
        width = width, height = height)
print(p)
dev.off()


```

#### [FIG] ChIP levels on loops
```{r}
commonLoopDir <- here("../data/loop_analysis")
figDir <- here("../figure/loop_analysis")

getPvalWilcox <- function(data, group1, group2){
  distance1 <- (data %>% dplyr::filter(loop ==group1) )$sumScore
  distance2 <- (data %>% dplyr::filter(loop ==group2) )$sumScore
  wil <- wilcox.test(distance1, distance2)
  return(wil$p.value)
}



getSumScores <- function(track, anchor) {
  # Find overlaps between all anchors and track regions at once
  overlaps <- findOverlaps(anchor, track)
  
  # Extract the scores and corresponding anchor indices
  anchor_indices <- queryHits(overlaps)
  track_scores <- score(track)[subjectHits(overlaps)]
  
  # Use tapply to calculate the median scores for each anchor
  median_scores <- tapply(track_scores, anchor_indices, mean, na.rm = TRUE)
  
  # Initialize a numeric vector to store the median scores for each anchor
  all_median_scores <- rep(NA, length(anchor))
  
  # Populate the median scores for the anchors that have overlaps
  all_median_scores[as.numeric(names(median_scores))] <- median_scores
  
  return(all_median_scores)
}

plotSumScoresBinary <- function(track, peak, name, note = "", anchor.upno, anchor.down){
  peakTrack <- track[unique(queryHits(findOverlaps(track, peak)))]
  b <- getSumScores(peakTrack, anchor.upno)
  c <- getSumScores(peakTrack, anchor.down)
  b.tb <- tibble(loop = "UP/NO",
                 sumScore = b)
  c.tb <- tibble(loop = "DOWN",
                 sumScore = c)
  data <- bind_rows(b.tb, c.tb) %>% drop_na()
  data$loop <- factor(data$loop, levels = c("UP/NO", "DOWN"))
  
  p12 <- getPvalWilcox(data, "UP/NO", "DOWN")
  p <-  ggplot(data, aes(x = name, fill = loop, y = sumScore)) + 
    labs(x = NULL, y = paste0("Average ChIP peak score at anchor)")) +
    introdataviz::geom_split_violin(linewidth = lineThick * mmToLineUnit, lineend = "square",
                                    alpha = .4) +
    geom_boxplot(width = 0.3, color = "black",
                 linewidth = lineThick * mmToLineUnit, lineend = "square",
                 outlier.shape = NA,  alpha = 0.6, show.legend = FALSE) + theme_classic() +
    scale_fill_manual(
      name = "Reg loop type",
      values = c("grey30", strong_blue)
    ) +
    stat_summary(
      aes(group = loop), fun = mean,
      geom = "point", shape = 21, size = 0.5,
      fill = "black", color = "black", position = position_dodge(.3)
    ) + theme(
      axis.title = element_text(
        size = fontSizeM, family = fontType, color = "#000000"
      ),
      axis.text = element_text(
        size = fontSizeM, family = fontType, color = "#000000"
      ),
      axis.text.x = element_text(
        angle = 0,
      ),
      axis.text.y = element_text(
        size = fontSizeS,
      ),
      axis.line = element_line(
        color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
      ),
      axis.ticks = element_line(
        color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
      ),
      panel.background = element_rect(fill = "transparent"),
      legend.text = element_text(family = fontType, size = fontSizeS),
      legend.title = element_text(family = fontType, size = fontSizeS)
    ) + 
    coord_cartesian(ylim = c(quantile(data$sumScore, 0.0), quantile(data$sumScore, 0.95))) + 
    annotate(
      "text", x = 1, y = quantile(data$sumScore, 0.5),
      label = paste0("p12: ", convPvalue(p12)),
      color = "black", hjust = 0, size = 1
    ) +   guides(
      fill = guide_legend(
        keywidth = 0.2,  # Adjust the width of the legend keys
        keyheight = 0.2  # Adjust the height of the legend keys
      )
    )
  
  fileName <- paste0("ChIP_peak_avgPeakScore_", name, note)
  width <- panelSize(1.5)*mmToInch
  height <- panelSize(1.2)*mmToInch
  png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
  print(p)
  dev.off()
  svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
  print(p)
  dev.off()
}
```

##### ESC
```{r}
# Sum peak score
#### ESC
name <- "chromosight_score_complex_reg_dTAGvsDMSO_diffCutoff0.2"

# UP loop
loop.up <- fread(here(commonLoopDir, paste0(name, "_UP.bedpe")))
anchor.up <- (extractAnchor(loop.up))
# NO loop
loop.no <- fread(here(commonLoopDir, paste0(name, "_NO.bedpe")))
anchor.no <- (extractAnchor(loop.no))
# UP NO
loop.upno <- bind_rows(loop.up, loop.no)
anchor.upno <- (extractAnchor(loop.upno))
# DOWN loop
loop.down <- fread(here(commonLoopDir, paste0(name, "_DOWN.bedpe")))
anchor.down <- (extractAnchor(loop.down))

track <- import(here(refDir, "ChIP", "H3K4me3", "H3K4me3_ESC_avg_depthNorm.bw"), format = "BigWig")
peak <- importPeak(here(refDir, "ChIP", "H3K4me3", "H3K4me3_ESC_unionPeaks.bed"))
plotSumScoresBinary(track, peak, "H3K4me3", anchor.upno, anchor.down)

track <- import(here(refDir, "ChIP", "H3K27ac", "H3K27ac_ESC_avg_depthNorm.bw"), format = "BigWig")
peak <- importPeak(here(refDir, "ChIP", "H3K27ac", "H3K27ac_ESC_unionPeaks.bed"))
plotSumScoresBinary(track, peak, "H3K27ac", anchor.upno, anchor.down)

track <- import(here(refDir, "ChIP", "CTCF", "CTCF_ESC_Rep1_33248_depthNorm.bw"), format = "BigWig")
peak <- importPeak(here(refDir, "ChIP", "CTCF", "CTCF_ESC_unionPeaks.bed"))
plotSumScoresBinary(track, peak, "CTCF", anchor.upno, anchor.down)

track <- import(here(refDir, "ChIP", "RAD21-HA", "RAD21_ESC-DMSO_Rep1_depthNorm.bw"), format = "BigWig")
peak <- importPeak(here(refDir, "ChIP", "RAD21-HA", "RAD21_ESC-DMSO_unionPeaks.bed"))
plotSumScoresBinary(track, peak, "RAD21", anchor.upno, anchor.down)
```

##### EpiLC
```{r}
#### EpiLC
name <- "chromosight_score_complex_reg_EpidTAGvsEpiDMSO_diffCutoff0.2"

# UP loop
loop.up <- fread(here(commonLoopDir, paste0(name, "_UP.bedpe")))
anchor.up <- (extractAnchor(loop.up))
# NO loop
loop.no <- fread(here(commonLoopDir, paste0(name, "_NO.bedpe")))
anchor.no <- (extractAnchor(loop.no))
# UP NO
loop.upno <- bind_rows(loop.up, loop.no)
anchor.upno <- (extractAnchor(loop.upno))
# DOWN loop
loop.down <- fread(here(commonLoopDir, paste0(name, "_DOWN.bedpe")))
anchor.down <- (extractAnchor(loop.down))

track <- import(here(refDir, "ChIP", "H3K4me3", "H3K4me3_ESC_avg_depthNorm.bw"), format = "BigWig")
peak <- importPeak(here(refDir, "ChIP", "H3K4me3", "H3K4me3_ESC_unionPeaks.bed"))
plotSumScoresBinary(track, peak, "H3K4me3", "_EpiLC", anchor.upno, anchor.down)

track <- import(here(refDir, "ChIP", "H3K27ac", "H3K27ac_ESC_avg_depthNorm.bw"), format = "BigWig")
peak <- importPeak(here(refDir, "ChIP", "H3K27ac", "H3K27ac_ESC_unionPeaks.bed"))
plotSumScoresBinary(track, peak, "H3K27ac", "_EpiLC",anchor.upno, anchor.down)

track <- import(here(refDir, "ChIP", "CTCF", "CTCF_ESC_Rep1_33248_depthNorm.bw"), format = "BigWig")
peak <- importPeak(here(refDir, "ChIP", "CTCF", "CTCF_ESC_unionPeaks.bed"))
plotSumScoresBinary(track, peak, "CTCF", "_EpiLC",anchor.upno, anchor.down)

track <- import(here(refDir, "ChIP", "RAD21-HA", "RAD21_ESC-DMSO_Rep1_depthNorm.bw"), format = "BigWig")
peak <- importPeak(here(refDir, "ChIP", "RAD21-HA", "RAD21_ESC-DMSO_unionPeaks.bed"))
plotSumScoresBinary(track, peak, "RAD21", "_EpiLC",anchor.upno, anchor.down)

```

#### [FIG] SE overlap & enrichment
```{r}
peak.Whyte.SE <- importPeak(here(refDir, "mm10", "superEnhancer_Whyte_ESC_mm10.bed"))
peak.Dylan.SE <- importPeak(here(refDir, "mm10", "superEnhancer_Dylan_ESC.bed"))
```

Q1. How many SEs are covered with Micro-C loops?

```{r}
name <- "chromosight_score_complex_reg_dTAGvsDMSO_diffCutoff0.2"
loop <- fread(here(commonLoopDir, "loopScore_chromosight.tsv")) %>%
  dplyr::filter(Anno_simple %in% c("P-P", "P-E", "E-E"))
anchor <- extractAnchor(loop)

### Whyte
n <- length(unique(subjectHits(findOverlaps(anchor, peak.Whyte.SE))))
total <- length(peak.Whyte.SE)
counts <- c(n, total - n)
labels <- paste(c("covered", "NOT covered"), counts)
pie(counts, labels = labels,  main = paste0("Whyte SE coverage \n", total), col = c("grey", "white"))

fileName <- here(figDir, "pie_SE_Whyte")
width <- panelSize(4)*mmToInch
height <- panelSize(4)*mmToInch
svglite(paste0(fileName, ".svg"), width = width, height =height)
pie(counts, labels = labels,  main = paste0("Whyte SE coverage \n", total), col = c("grey", "white"),
    cex = 1, cex.main = 1)
dev.off()



### Dylan
n <- length(unique(subjectHits(findOverlaps(anchor, peak.Dylan.SE))))
total <- length(peak.Dylan.SE)
counts <- c(n, total - n)
labels <- paste(c("covered", "NOT covered"), counts)
pie(counts, labels = labels, main = paste0("Dylan SE coverage \n", total), col = c("grey", "white"))

fileName <- here(figDir, "pie_SE_Murphy")
width <- panelSize(4)*mmToInch
height <- panelSize(4)*mmToInch
svglite(paste0(fileName, ".svg"), width = width, height =height)
pie(counts, labels = labels, main = paste0("Murphy SE coverage \n", total), col = c("grey", "white"))

dev.off()
```

Q2. Fisher's exact test?
```{r}
name <- "chromosight_score_complex_reg_dTAGvsDMSO_diffCutoff0.2"
loop <- fread(here(commonLoopDir, "loopScore_chromosight.tsv")) %>%
  dplyr::filter(Anno_simple %in% c("P-P", "P-E", "E-E"))


getSEOverlapFisher <- function(allLoop, subsetLoop, peak){
  all.overlap <- getOverlapLoopNum(allLoop, peak)
  all.notOverlap <- nrow(allLoop) - all.overlap
  
  subset.overlap <- getOverlapLoopNum(subsetLoop, peak)
  subset.notOverlap <- nrow(subsetLoop) - subset.overlap
  
  contingency_table <- matrix(c(subset.overlap, subset.notOverlap,
                                all.overlap, all.notOverlap), nrow = 2, byrow = TRUE)
  colnames(contingency_table) <- c("Overlapping", "Not_Overlapping")
  rownames(contingency_table) <- c("All loops", "Subset loops")
  
  # Perform Fisher's Exact Test
  fisher_test_result <- fisher.test(contingency_table)
  return(fisher_test_result)
}

loop.up <- fread(here(commonLoopDir, paste0(name, "_UP.bedpe")))
loop.no <- fread(here(commonLoopDir, paste0(name, "_NO.bedpe")))
loop.upno <- bind_rows(loop.up, loop.no)
loop.down <- fread(here(commonLoopDir, paste0(name, "_DOWN.bedpe")))


### Dylan
# Seeding
temp <- getSEOverlapFisher(loop, loop.upno, peak.Dylan.SE)
result.tb <- tibble(loopType = "UP/NO",
                    target = "Dylan SE",
                    pvalue = temp$p.value,
                    oddsRatio = temp$estimate)

temp <- getSEOverlapFisher(loop, loop.down, peak.Dylan.SE)
result.tb <- result.tb %>% 
  add_row(loopType = "DOWN",
          target = "Dylan SE",
          pvalue = temp$p.value,
          oddsRatio = temp$estimate)

### Whyte
# Add row
temp <- getSEOverlapFisher(loop, loop.upno, peak.Whyte.SE)
result.tb <- result.tb %>% 
  add_row(loopType = "UP/NO",
          target = "Whyte SE",
          pvalue = temp$p.value,
          oddsRatio = temp$estimate)

temp <- getSEOverlapFisher(loop, loop.down, peak.Whyte.SE)
result.tb <- result.tb %>% 
  add_row(loopType = "DOWN",
          target = "Whyte SE",
          pvalue = temp$p.value,
          oddsRatio = temp$estimate)


# Visualization
library(circlize)
data <- result.tb
heatmap_data <- data %>% dplyr::select(target, loopType, oddsRatio) %>%
  pivot_wider(names_from = loopType, values_from = oddsRatio) %>%
  column_to_rownames(var = "target")

pvalue_data <- data %>% dplyr::select(target, loopType, pvalue) %>%
  pivot_wider(names_from = loopType, values_from = pvalue) %>%
  column_to_rownames(var = "target")

col_fun <- colorRamp2(c(0, 1, 2), 
                      c("#4852A0", "white", "#CB333A"))

data$loopType <- factor(data$loopType, levels = c("UP/NO", "DOWN"))
data$target <- factor(data$target, levels = c("Whyte SE", "Dylan SE"))

p <- ggplot(data, aes(x = loopType, y = target, size = -log10(pvalue), fill = oddsRatio)) +
  geom_point(shape = 21,        # Ensures a point with an outline
             stroke = 1*ptToMM      # Line width for the border
  ) + theme_bw() + 
  scale_size_continuous(range = c(1, 3)) +  # Set min and max point sizes here
  scale_fill_gradientn(colors = c("#4852A0", "white", "#CB333A"),  # Define gradient colors
                       values = scales::rescale(c(0.5, 1, 1.5)), limits = c(0.5, 1.5), 
                       #low = "white", high = "#CB333A",
                       #                  limits = c(1, 3),
                       oob = scales::squish, # Define gradient colors
                       guide = guide_colorbar(
                         barwidth = 1.5/5.08,  # Adjust width of the color bar
                         barheight = 15/5.08   # Adjust height of the color bar
                       )
  ) +
  labs(x = NULL, y = NULL)  +
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  )

fileName <- here(figDir, "dotplot_SE_enrichment_dotplot")
width <- panelSize(1.8)*mmToInch
height <- panelSize(1.1)*mmToInch
svglite(paste0(fileName, ".svg"),  height = height, width = width)
print(p)
dev.off()
```

#### [FIG] Hub enrichment
```{r}
loop <- fread(here(commonLoopDir, "loopScore_chromosight.tsv")) %>%
  dplyr::filter(Anno_simple %in% c("P-P", "P-E", "E-E"))

name <- "chromosight_score_complex_reg_dTAGvsDMSO_diffCutoff0.2"
loop.up <- fread(here(commonLoopDir, paste0(name, "_UP.bedpe")))
loop.no <- fread(here(commonLoopDir, paste0(name, "_NO.bedpe")))
anchor.upno <- extractAnchor(bind_rows(loop.up, loop.no))

loop.down <- fread(here(commonLoopDir, paste0(name, "_DOWN.bedpe")))
anchor.down <- extractAnchor(loop.down)


getSEOverlapFisher <- function(allLoop, subsetLoop, peak){
  all.overlap <- getOverlapLoopNum(allLoop, peak)
  all.notOverlap <- nrow(allLoop) - all.overlap
  
  subset.overlap <- getOverlapLoopNum(subsetLoop, peak)
  subset.notOverlap <- nrow(subsetLoop) - subset.overlap
  
  contingency_table <- matrix(c(subset.overlap, subset.notOverlap,
                                all.overlap, all.notOverlap), nrow = 2, byrow = TRUE)
  colnames(contingency_table) <- c("Overlapping", "Not_Overlapping")
  rownames(contingency_table) <- c("All loops", "Subset loops")
  
  # Perform Fisher's Exact Test
  fisher_test_result <- fisher.test(contingency_table)
  return(fisher_test_result)
}


i <- 2
data <- fread(here(refDir, "mm10", "Dylan_hub_esc.csv"))

data.hub <- data %>% dplyr::filter(all_lcon > i)
temp <- data.hub %>% dplyr::select(c(1, 2, 3))
colnames(temp) <- c("chr", "start", "end")
hub.anchor <- makeGRangesFromDataFrame(temp)


temp <- getSEOverlapFisher(loop, loop.upno, hub.anchor)
result.tb <- tibble(loopType = "UP/NO",
                    target = paste0("hub_", i),
                    pvalue = temp$p.value,
                    oddsRatio = temp$estimate)
temp <- getSEOverlapFisher(loop, loop.down, hub.anchor)
result.tb <- result.tb %>% 
  add_row(loopType = "DOWN",
          target =  paste0("hub_", i),
          pvalue = temp$p.value,
          oddsRatio = temp$estimate)


for(i in c(5)){
  data <- fread(here(refDir, "mm10", "Dylan_hub_esc.csv"))
  
  data.hub <- data %>% dplyr::filter(all_lcon > i)
  temp <- data.hub %>% dplyr::select(c(1, 2, 3))
  colnames(temp) <- c("chr", "start", "end")
  hub.anchor <- makeGRangesFromDataFrame(temp)
  
  
  temp <- getSEOverlapFisher(loop, loop.upno, hub.anchor)
  result.tb <- result.tb %>% add_row(loopType = "UP/NO",
                                     target = paste0("hub_", i),
                                     pvalue = temp$p.value,
                                     oddsRatio = temp$estimate)
  temp <- getSEOverlapFisher(loop, loop.down, hub.anchor)
  result.tb <- result.tb %>% 
    add_row(loopType = "DOWN",
            target =  paste0("hub_", i),
            pvalue = temp$p.value,
            oddsRatio = temp$estimate)
  
  
}

library(circlize)
data <- result.tb
heatmap_data <- data %>% dplyr::select(target, loopType, oddsRatio) %>%
  pivot_wider(names_from = loopType, values_from = oddsRatio) %>%
  column_to_rownames(var = "target")

pvalue_data <- data %>% dplyr::select(target, loopType, pvalue) %>%
  pivot_wider(names_from = loopType, values_from = pvalue) %>%
  column_to_rownames(var = "target")

# col_fun <- colorRamp2(c(0, 1, 2), 
#                       c("blue", "white", "red"))

data$target <- factor(data$target, levels = c("hub_5", "hub_2"))
data$loopType <- factor(data$loopType, levels = c("UP/NO", "DOWN"))
p <- ggplot(data, aes(x = loopType, y = target, size = -log10(pvalue), fill = oddsRatio)) +
  geom_point(shape = 21,        # Ensures a point with an outline
             stroke = 1*ptToMM      # Line width for the border
  ) + theme_bw() + 
  scale_size_continuous(range = c(1, 3)) +  # Set min and max point sizes here
  scale_fill_gradientn(colors = c("#4852A0", "white", "#CB333A"),  # Define gradient colors
                       values = scales::rescale(c(0.5, 1, 1.5)), limits = c(0.5, 1.5), 
                       #low = "white", high = "#CB333A",
                       #                  limits = c(1, 3),
                       oob = scales::squish, # Define gradient colors
                       guide = guide_colorbar(
                         barwidth = 1.5/5.08,  # Adjust width of the color bar
                         barheight = 15/5.08   # Adjust height of the color bar
                       )
  ) + labs(x = NULL, y = NULL)  +
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  )


fileName <- here(figDir, "dotplot_hub_enrichment_dotplot_all")
width <- panelSize(1.65)*mmToInch
height <- panelSize(1.1)*mmToInch
svglite(paste0(fileName, ".svg"),  height = height, width = width)
print(p)
dev.off()
```
#### [FIG] Bookmarking enrichment
```{r}
runFisherExact <- function(interest.gr, background.gr, query.gr){
  overlaps_interest <- countOverlaps(interest.gr, query.gr)
  a <- sum(overlaps_interest > 0)
  total_interest <- length(interest.gr)
  c <- total_interest - a
  
  overlaps_background <- countOverlaps(background.gr, query.gr)
  b <- sum(overlaps_background > 0)
  total_background <- length(background.gr)
  d <- total_background - b
  
  # Construct contingency table
  contingency_table <- matrix(c(a, c, b, d), nrow=2, byrow=TRUE,
                              dimnames=list("Region" = c("interest.gr", "background.gr"),
                                            "Overlap" = c("Yes", "No")))
  
  # Perform Fisher's exact test
  fisher_result <- fisher.test(contingency_table)
  
  return(fisher_result)
}

runFisherExactCombination <- function(interestName,
                                      loop.up.gr, loop.no.gr, loop.down.gr,
                                      background.gr,
                                      targetName,
                                      retained.gr, lost.gr){
  # SEEDING
  temp <- runFisherExact(loop.up.gr, background.gr, retained.gr)
  result.tb <- tibble(interest = paste0(interestName, "_UP"),
                      target = paste0(targetName, "_retained"),
                      pvalue = temp$p.value,
                      oddsRatio = temp$estimate)
  
  # ADDING ROWS
  temp <- runFisherExact(loop.no.gr, background.gr, retained.gr)
  result.tb <- result.tb %>% 
    add_row(interest = paste0(interestName, "_NO"),
            target = paste0(targetName, "_retained"),
            pvalue = temp$p.value,
            oddsRatio = temp$estimate)
  
  # ADDING ROWS
  temp <- runFisherExact(loop.down.gr, background.gr, retained.gr)
  result.tb <- result.tb %>% 
    add_row(interest = paste0(interestName, "_DOWN"),
            target = paste0(targetName, "_retained"),
            pvalue = temp$p.value,
            oddsRatio = temp$estimate)
  
  # ADDING ROWS
  temp <- runFisherExact(loop.up.gr, background.gr, lost.gr)
  result.tb <- result.tb %>% 
    add_row(interest = paste0(interestName, "_UP"),
            target = paste0(targetName, "_lost"),
            pvalue = temp$p.value,
            oddsRatio = temp$estimate)
  
  # ADDING ROWS
  temp <- runFisherExact(loop.no.gr, background.gr, lost.gr)
  result.tb <- result.tb %>% 
    add_row(interest = paste0(interestName, "_NO"),
            target = paste0(targetName, "_lost"),
            pvalue = temp$p.value,
            oddsRatio = temp$estimate)
  
  # ADDING ROWS
  temp <- runFisherExact(loop.down.gr, background.gr, lost.gr)
  result.tb <- result.tb %>% 
    add_row(interest = paste0(interestName, "_DOWN"),
            target = paste0(targetName, "_lost"),
            pvalue = temp$p.value,
            oddsRatio = temp$estimate)
  
  return(result.tb)
}

runFisherExactCombinationBinary <- function(interestName,
                                            loop.upno.gr, loop.down.gr,
                                            background.gr,
                                            targetName,
                                            retained.gr, lost.gr){
  # SEEDING
  temp <- runFisherExact(loop.upno.gr, background.gr, retained.gr)
  result.tb <- tibble(interest = paste0(interestName, "_UP/NO"),
                      target = paste0(targetName, "_retained"),
                      pvalue = temp$p.value,
                      oddsRatio = temp$estimate)
  
  
  # ADDING ROWS
  temp <- runFisherExact(loop.down.gr, background.gr, retained.gr)
  result.tb <- result.tb %>% 
    add_row(interest = paste0(interestName, "_DOWN"),
            target = paste0(targetName, "_retained"),
            pvalue = temp$p.value,
            oddsRatio = temp$estimate)
  
  # ADDING ROWS
  temp <- runFisherExact(loop.upno.gr, background.gr, lost.gr)
  result.tb <- result.tb %>% 
    add_row(interest = paste0(interestName, "_UP/NO"),
            target = paste0(targetName, "_lost"),
            pvalue = temp$p.value,
            oddsRatio = temp$estimate)
  
  
  # ADDING ROWS
  temp <- runFisherExact(loop.down.gr, background.gr, lost.gr)
  result.tb <- result.tb %>% 
    add_row(interest = paste0(interestName, "_DOWN"),
            target = paste0(targetName, "_lost"),
            pvalue = temp$p.value,
            oddsRatio = temp$estimate)
  
  return(result.tb)
}

runFisherExactCombinationRetained <- function(interestName,
                                              loop.up.gr, loop.no.gr, loop.down.gr,
                                              background.gr,
                                              targetName,
                                              retained.gr){
  # SEEDING
  temp <- runFisherExact(loop.up.gr, background.gr, retained.gr)
  result.tb <- tibble(interest = paste0(interestName, "_UP"),
                      target = paste0(targetName, "_retained"),
                      pvalue = temp$p.value,
                      oddsRatio = temp$estimate)
  
  # ADDING ROWS
  temp <- runFisherExact(loop.no.gr, background.gr, retained.gr)
  result.tb <- result.tb %>% 
    add_row(interest = paste0(interestName, "_NO"),
            target = paste0(targetName, "_retained"),
            pvalue = temp$p.value,
            oddsRatio = temp$estimate)
  
  # ADDING ROWS
  temp <- runFisherExact(loop.down.gr, background.gr, retained.gr)
  result.tb <- result.tb %>% 
    add_row(interest = paste0(interestName, "_DOWN"),
            target = paste0(targetName, "_retained"),
            pvalue = temp$p.value,
            oddsRatio = temp$estimate)
  
  
  return(result.tb)
}

runFisherExactCombinationTarget <- function(targetName, interestName,
                                            loop.up.gr, loop.no.gr, loop.down.gr,
                                            background.gr){
  # RETAINED loop
  retained <- fread(here(refDir, "TF_bookmarking", targetName, "am_a.bed")) %>% dplyr::select(c(1, 2, 3))
  colnames(retained) <- c("chr", "start", "end")
  retained.gr <- makeGRangesFromDataFrame(retained)
  # LOST loop
  lost <- fread(here(refDir, "TF_bookmarking", targetName, "oa.bed")) %>% dplyr::select(c(1, 2, 3))
  colnames(lost) <- c("chr", "start", "end")
  lost.gr <- makeGRangesFromDataFrame(lost)
  
  result <- runFisherExactCombination(interestName,
                                      loop.up.gr, loop.no.gr, loop.down.gr,
                                      background.gr,
                                      targetName,
                                      retained.gr, lost.gr)
  
  return(result)
}

runFisherExactCombinationTargetBinary <- function(targetName, interestName,
                                                  loop.upno.gr, loop.down.gr,
                                                  background.gr){
  # RETAINED loop
  retained <- fread(here(refDir, "TF_bookmarking", targetName, "am_a.bed")) %>% dplyr::select(c(1, 2, 3))
  colnames(retained) <- c("chr", "start", "end")
  retained.gr <- makeGRangesFromDataFrame(retained)
  # LOST loop
  lost <- fread(here(refDir, "TF_bookmarking", targetName, "oa.bed")) %>% dplyr::select(c(1, 2, 3))
  colnames(lost) <- c("chr", "start", "end")
  lost.gr <- makeGRangesFromDataFrame(lost)
  
  result <- runFisherExactCombinationBinary(interestName,
                                            loop.upno.gr, loop.down.gr,
                                            background.gr,
                                            targetName,
                                            retained.gr, lost.gr)
  
  return(result)
}

runFisherExactCombinationTargetRetained <- function(targetName, interestName,
                                                    loop.up.gr, loop.no.gr, loop.down.gr,
                                                    background.gr){
  # RETAINED loop
  retained <- fread(here(refDir, "TF_bookmarking", targetName, "am_a.bed")) %>% dplyr::select(c(1, 2, 3))
  colnames(retained) <- c("chr", "start", "end")
  retained.gr <- makeGRangesFromDataFrame(retained)
  # LOST loop
  result <- runFisherExactCombinationRetained(interestName,
                                              loop.up.gr, loop.no.gr, loop.down.gr,
                                              background.gr,
                                              targetName,
                                              retained.gr)
  
  return(result)
}
```

```{r}
#### Importing loops of interest
background <- fread(here(commonLoopDir, "loopScore_chromosight.tsv")) %>%
  dplyr::filter(Anno_simple %in% c("P-P", "P-E", "E-E"))
background.gr <- (extractAnchor(background))

name <- "chromosight_score_complex_reg_dTAGvsDMSO_diffCutoff0.2"
loop.up <- fread(here(commonLoopDir, paste0(name, "_UP.bedpe")))
loop.up.gr <- (extractAnchor(loop.up))

loop.no <- fread(here(commonLoopDir, paste0(name, "_NO.bedpe")))
loop.no.gr <- (extractAnchor(loop.no))

loop.upno.gr <- extractAnchor(bind_rows(loop.up, loop.no))

loop.down <- fread(here(commonLoopDir, paste0(name, "_DOWN.bedpe")))
loop.down.gr <- (extractAnchor(loop.down))

interestName <- "regLoop"

```
###### H3K27ac only
```{r}
targetName <- "H3K27ac_effie"
# RETAINED loop
retained <- fread(here(refDir, "TF_bookmarking", targetName, "am_a.bed")) %>% dplyr::select(c(1, 2, 3))
colnames(retained) <- c("chr", "start", "end")
retained.gr <- makeGRangesFromDataFrame(retained)
# LOST loop
lost <- fread(here(refDir, "TF_bookmarking", targetName, "oa.bed")) %>% dplyr::select(c(1, 2, 3))
colnames(lost) <- c("chr", "start", "end")
lost.gr <- makeGRangesFromDataFrame(lost)


# SEEDING
temp <- runFisherExact(loop.upno.gr, background.gr, retained.gr)
result.tb <- tibble(interest = paste0(interestName, "_UP/NO"),
                    target = paste0(targetName, "_retained"),
                    pvalue = temp$p.value,
                    oddsRatio = temp$estimate)

# ADDING ROWS
temp <- runFisherExact(loop.down.gr, background.gr, retained.gr)
result.tb <- result.tb %>% 
  add_row(interest = paste0(interestName, "_DOWN"),
          target = paste0(targetName, "_retained"),
          pvalue = temp$p.value,
          oddsRatio = temp$estimate)

# ADDING ROWS
temp <- runFisherExact(loop.upno.gr, background.gr, lost.gr)
result.tb <- result.tb %>% 
  add_row(interest = paste0(interestName, "_UP/NO"),
          target = paste0(targetName, "_lost"),
          pvalue = temp$p.value,
          oddsRatio = temp$estimate)

# ADDING ROWS
temp <- runFisherExact(loop.down.gr, background.gr, lost.gr)
result.tb <- result.tb %>% 
  add_row(interest = paste0(interestName, "_DOWN"),
          target = paste0(targetName, "_lost"),
          pvalue = temp$p.value,
          oddsRatio = temp$estimate)


result.tb <- result.tb %>%
  mutate(interest = recode(interest, 
                           "regLoop_UP/NO" = "UP/NO",
                           "regLoop_DOWN" = "DOWN"),
         target = recode(target,
                         "H3K27ac_effie_retained" = "Retained",
                         "H3K27ac_effie_lost" = "Lost"))

result.tb$interest <- factor(result.tb$interest, levels = c("UP/NO", "DOWN"))

p <- ggplot(result.tb, aes(x = interest, y = target, size = -log10(pvalue), fill = oddsRatio)) +
  geom_point(shape = 21,        # Ensures a point with an outline
             stroke = 0.5*ptToMM      # Line width for the border
  ) + theme_bw() + 
  scale_size_continuous(range = c(1, 3)) +  # Set min and max point sizes here
  scale_fill_gradientn(colors = c("#4852A0", "white", "#CB333A"),  # Define gradient colors
                       values = scales::rescale(c(0.5, 1, 1.5)), limits = c(0.5, 1.5), 
                       #low = "white", high = "#CB333A",
                       #                  limits = c(1, 3),
                       oob = scales::squish, # Define gradient colors
                       guide = guide_colorbar(
                         barwidth = 1.5/5.08,  # Adjust width of the color bar
                         barheight = 15/5.08   # Adjust height of the color bar
                       )
  ) +
  labs(x = NULL, y = NULL)  +
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  )

fileName <- here(figDir, "dotplot_bookmarking_H3K27ac_reg")
width <- panelSize(1.7)*mmToInch
height <- panelSize(1.1)*mmToInch
svglite(paste0(fileName, ".svg"),  height = height, width = width)
print(p)
dev.off()


```
###### all targets
```{r}
#### Importing loops of target

temp1 <- runFisherExactCombinationTargetBinary("ESRRB_dsg", interestName,
                                               loop.upno.gr, loop.down.gr,
                                               background.gr)
temp2 <- runFisherExactCombinationTargetBinary("H3K27ac_effie", interestName,
                                               loop.upno.gr, loop.down.gr,
                                               background.gr)
temp3 <- runFisherExactCombinationTargetBinary("KLF4_effie", interestName,
                                               loop.upno.gr, loop.down.gr,
                                               background.gr)
temp4 <- runFisherExactCombinationTargetBinary("OCT4_effie", interestName,
                                               loop.upno.gr, loop.down.gr,
                                               background.gr)
temp5 <- runFisherExactCombinationTargetBinary("SOX2_effie", interestName,
                                               loop.upno.gr, loop.down.gr,
                                               background.gr)


data <- bind_rows(temp1, temp2, temp3, temp4, temp5)



result.tb <- data %>%
  mutate(interest = recode(interest, 
                           "regLoop_UP/NO" = "UP/NO",
                           "regLoop_DOWN" = "DOWN"),
         target = recode(target,
                         "H3K27ac_effie_retained" = "H3K27ac_retained",
                         "H3K27ac_effie_lost" = "H3K27ac_lost",
                         "ESRRB_dsg_retained" = "ESRRB_retained",
                         "ESRRB_dsg_lost" = "ESRRB_lost",
                         "KLF4_effie_retained" = "KLF4_retained",
                         "KLF4_effie_lost" = "KLF4_lost",
                         "OCT4_effie_retained" = "OCT4_retained",
                         "OCT4_effie_lost" = "OCT4_lost",
                         "SOX2_effie_retained" = "SOX2_retained",
                         "SOX2_effie_lost" = "SOX2_lost"))

result.tb$interest <- factor(result.tb$interest, levels = c("UP/NO", "DOWN"))
result.tb$target <- factor(result.tb$target, levels = rev(c("H3K27ac_retained", "H3K27ac_lost",
                                                            "ESRRB_retained", "ESRRB_lost",
                                                            "KLF4_retained", "KLF4_lost",
                                                            "OCT4_retained", "OCT4_lost",
                                                            "SOX2_retained","SOX2_lost")))

p <- ggplot(result.tb, aes(x = interest, y = target, size = -log10(pvalue), fill = oddsRatio)) +
  geom_point(shape = 21,        # Ensures a point with an outline
             stroke = 1*ptToMM      # Line width for the border
  ) + theme_bw() + 
  scale_size_continuous(range = c(1, 3)) +  # Set min and max point sizes here
  scale_fill_gradientn(colors = c("#4852A0", "white", "#CB333A"),  # Define gradient colors
                       values = scales::rescale(c(0.5, 1, 1.5)), limits = c(0.5, 1.5), 
                       #low = "white", high = "#CB333A",
                       #                  limits = c(1, 3),
                       oob = scales::squish, # Define gradient colors
                       guide = guide_colorbar(
                         barwidth = 1.5/5.08,  # Adjust width of the color bar
                         barheight = 15/5.08   # Adjust height of the color bar
                       )
  ) +
  labs(x = NULL, y = NULL)  +
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  )

fileName <- here(figDir, "dotplot_bookmarking_alltarget_reg")
width <- panelSize(2.2)*mmToInch
height <- panelSize(2)*mmToInch
svglite(paste0(fileName, ".svg"),  height = height, width = width)
print(p)
dev.off()








```

## [4] Loop + Transcription (Chromosight) ESC
### [FIG] RNA-seq loopscore scatterplot
```{r}
loadLoopAnnoData <- function(fileName){
  data <- as_tibble(fread(fileName)) %>%
    dplyr::mutate(gene = strsplit(gene, '\\|'))
  return(data)
}
```

```{r}
alpha <- 0.05
fcCutoff <- 0.5
diffCutoff <- 0.2
commonLoopDir <- here("../data/loop_analysis")
diffDir <- here("../data/RNA_diff")
name <- "chromosight"

geneAnnoData <- loadLoopAnnoData(here(commonLoopDir, paste0("loopScore_", name, "_p-n_complex_ensemblList.tsv"))) %>%
  dplyr::mutate(diff = score_dTAG - score_DMSO)

diff.RNA <- fread(here(diffDir, "diff_G1.dTAG_G1.2i.dTAG_vs_G1.2i.DMSO.tsv")) %>%
  dplyr::select(ensembl_gene_id, log2FoldChange, shrinked_log2FC, padj, external_gene_name)
geneList.down.RNA <- (diff.RNA %>% dplyr::filter(shrinked_log2FC <= 0, padj < alpha))$ensembl_gene_id
geneList.up.RNA <- (diff.RNA %>% dplyr::filter(shrinked_log2FC > 0, padj < alpha))$ensembl_gene_id


maxLog2FC <- 2

temp <- geneAnnoData %>% dplyr::select(diff, gene) %>% 
  unnest(gene) %>% group_by(gene) %>%
  summarize(mean_diff_score = mean(diff), .groups = 'drop')

temp <- left_join(temp, diff.RNA, by = c("gene" = "ensembl_gene_id")) %>% 
  dplyr::mutate(flag = ifelse((gene %in% geneList.down.RNA) & (mean_diff_score < 0) , "Group1",
                              ifelse((mean_diff_score < 0) & !(gene %in% c(geneList.up.RNA, geneList.down.RNA)), "Group2",
                                     "Group3")),
                maxFlag = (abs(shrinked_log2FC) > maxLog2FC),
                shrlog2fcMax = pmax(pmin(shrinked_log2FC, maxLog2FC), -maxLog2FC)) %>% 
  dplyr::arrange(flag)

dataDir <- here("..", "data", "loop_analysis")
fwrite(temp %>% dplyr::filter(flag == "Group1"), here(dataDir, paste0("geneList_", name, "_dTAG_vs_DMSO_p-n_complex_Group1.tsv")), sep = "\t")
fwrite(temp %>% dplyr::filter(flag == "Group2"), here(dataDir, paste0("geneList_", name, "_dTAG_vs_DMSO_p-n_complex_Group2.tsv")), sep = "\t")

p <- ggplot(temp, aes(x = shrlog2fcMax, y = mean_diff_score, color = flag,
                      label = ifelse(flag == "Group1", external_gene_name, NA),
                      shape = as.factor(maxFlag))) +
  geom_point(size = 1) + 
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = 0) +
  geom_text_repel(size = fontSizeS/2.845, family = fontType) + theme_classic() +
  
  scale_color_manual(values = c("Group1" = strong_orange, "Group2" = no_grey, "Group3" = "grey90")) +  # Corrected color mapping
  scale_shape_manual(values = c("TRUE" = 2, "FALSE" = 19)) +
  theme(
    legend.position = "none",
    plot.title = element_text(
      hjust = 0.5,
      size = fontSizeS,
      family = fontType
    ),
    axis.title = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
  ) +
  labs(y = "Avg. Δ loop score", x = "log2(shrunken FC)")

fileName <- paste0("RNAvsLoop_", name, "_score_complexAnno")
width <- panelSize(2.2)*mmToInch
height <- panelSize(1.7)*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()



```

```{r}
alpha <- 0.05
fcCutoff <- 0.5
diffCutoff <- 0.5
commonLoopDir <- here("../data/loop_analysis")
diffDir <- here("../data/RNA_diff")
name <- "chromosight"
geneAnnoData <- loadLoopAnnoData(here(commonLoopDir, paste0("loopScore_", name, "_p-n_complex_ensemblList.tsv"))) %>%
  rowwise() %>%                                # work one row at a time
  mutate(
    minpos = min(c(ooe_dTAG, ooe_DMSO)[c(ooe_dTAG, ooe_DMSO) > 0], na.rm = TRUE),
    eps    = ifelse(is.finite(minpos), minpos * 0.1, 0.01),      # fallback if both are 0
    diff   = log2((ooe_dTAG + eps) / (ooe_DMSO + eps))
  ) %>% 
  ungroup() %>% 
  select(-minpos, -eps) 
#dplyr::mutate(diff = log2(ooe_dTAG) - log2(ooe_DMSO))

diff.RNA <- fread(here(diffDir, "diff_G1.dTAG_G1.2i.dTAG_vs_G1.2i.DMSO.tsv")) %>%
  dplyr::select(ensembl_gene_id, log2FoldChange, shrinked_log2FC, padj, external_gene_name)
geneList.down.RNA <- (diff.RNA %>% dplyr::filter(shrinked_log2FC <= 0, padj < alpha))$ensembl_gene_id
geneList.up.RNA <- (diff.RNA %>% dplyr::filter(shrinked_log2FC > 0, padj < alpha))$ensembl_gene_id


maxLog2FC <- 2

temp <- geneAnnoData %>% dplyr::select(diff, gene) %>% 
  unnest(gene) %>% group_by(gene) %>%
  summarize(mean_diff_score = mean(diff), .groups = 'drop')

temp <- left_join(temp, diff.RNA, by = c("gene" = "ensembl_gene_id")) %>% 
  dplyr::mutate(flag = ifelse(gene %in% geneList.down.RNA, "2DOWN",
                              ifelse(gene %in% geneList.up.RNA, "1UP", "0NO")),
                maxFlag = (abs(shrinked_log2FC) > maxLog2FC),
                shrlog2fcMax = pmax(pmin(shrinked_log2FC, maxLog2FC), -maxLog2FC)) %>% 
  dplyr::arrange(flag)

p <- ggplot(temp, aes(x = shrlog2fcMax, y = mean_diff_score, color = flag,
                      label = ifelse(flag != "0NO", external_gene_name, NA),
                      shape = as.factor(maxFlag))) +
  geom_point() + geom_text_repel() + theme_classic() +
  ggtitle(paste0(name, "_ooe_complexAnno")) +
  geom_hline(yintercept = 0) +
  geom_hline(yintercept = diffCutoff, linetype = "dashed") +
  geom_hline(yintercept = - diffCutoff, linetype = "dashed") +
  geom_vline(xintercept = 0) +
  geom_vline(xintercept = fcCutoff, linetype = "dashed") +
  geom_vline(xintercept = -fcCutoff, linetype = "dashed") +
  scale_color_manual(values = c("0NO" = "grey", "1UP" = "red", "2DOWN" = "blue")) +  # Corrected color mapping
  scale_shape_manual(values = c("TRUE" = 2, "FALSE" = 19))

fileName <- paste0("RNAvsLoop_", name, "_ooe_complexAnno")
height <- 4
width <- 7
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()
```

### Link TAD to gene
```{r}
dataDir <- here("../data")

# Importing gene
gene.TSS.tb <- fread(here(refDir, "mm10", "mm10_GRCm38.p6_gene_sorted.bed")) %>%
  dplyr::mutate(TSS = ifelse(V4 == "+", V2, V3)) %>%
  dplyr::select(V6, V5, V1, TSS)
colnames(gene.TSS.tb) <- c("ensembl", "gene", "chr", "TSS")

# Importing TAD boundary with insulation score for each sample
## Boundaries are sized 25kb
tad_boundary <- fread(here(dataDir, "TAD", "TAD_25kb_125kb_otsu_boundaries_G1DMSO_pooled.bed"))
colnames(tad_boundary) <- c("chr", "start", "end")
tad_boundary <- tad_boundary %>% dplyr::mutate(boundaryID = paste(chr, start, end, sep = "_"),
                                               center = (start + end)/2,
                                               size = end - start)
insScore.DMSO <- fread(here(dataDir, "TAD", "insulationScore_25000bp_G1DMSO_pooled.tsv")) %>%
  dplyr::mutate(binID = paste(chrom, start, end, sep = "_")) %>%
  dplyr::select(binID, log2InsScore_DMSO = log2_insulation_score_125000)
insScore.dTAG <- fread(here(dataDir, "TAD", "insulationScore_25000bp_G1dTAG_pooled.tsv")) %>%
  dplyr::mutate(binID = paste(chrom, start, end, sep = "_")) %>%
  dplyr::select(binID, log2InsScore_dTAG = log2_insulation_score_125000)
tad_boundary <- list(insScore.DMSO, insScore.dTAG) %>%
  purrr::reduce(
    left_join,
    .init = tad_boundary,
    by   = c("boundaryID" = "binID")
  )
tad_boundary <- tad_boundary %>%
  mutate(negDiffInsScore = (2^log2InsScore_DMSO) - (2^log2InsScore_dTAG))


# Importing TAD
tad <- fread(here(dataDir, "TAD",
                  "TAD_25kb_125kb_otsu_G1DMSO_pooled.bedpe"))
colnames(tad) <- c("chr1", "start1", "end1", "chr2", "start2", "end2")
tad <- tad %>% dplyr::mutate(tadID = paste(chr1, start1, end1, sep = "_"),
                             tadSize = end1 - start1)

################################################################################
# Functions to link gene to TADs
findItsTAD <- function(chrom, TSS, tad){
  temp <- tad %>% dplyr::filter(chr1 == chrom,
                                start1 < TSS,
                                end1 > TSS)
  if(nrow(temp) == 1){
    return(temp$tadID)
  }else{
    return(NA)
  }
}
findClosestTADBoundary <- function(chrom, TSS, tad_boundary){
  temp <- tad_boundary %>% dplyr::filter(chr == chrom) %>%
    dplyr::mutate(distance = abs(center - TSS)) %>%
    slice_min(distance)
  if(nrow(temp) == 1){
    return(temp$boundaryID)
  }else{
    return(NA)
  }
}

gene.TSS.tb <- gene.TSS.tb %>% rowwise() %>%
  dplyr::mutate(tadID = findItsTAD(chr, TSS, tad),
                boundaryID = findClosestTADBoundary(chr, TSS, tad_boundary))



data <- gene.TSS.tb %>% left_join(tad %>% dplyr::select(tadID, tadSize), by = c("tadID")) %>%
  left_join(tad_boundary %>% dplyr::select(c(4, 5, 7, 8, 9)), by = c("boundaryID")) %>%
  mutate(distanceToBoundary = abs(center - TSS))

```
#### [FIG] TAD distance/size/insulation score
```{r}
group1 <- fread(here(dataDir, "loop_analysis",
                     "geneList_chromosight_dTAG_vs_DMSO_p-n_complex_Group1.tsv"))$gene
group2 <- fread(here(dataDir, "loop_analysis",
                     "geneList_chromosight_dTAG_vs_DMSO_p-n_complex_Group2.tsv"))$gene


dataToPlot <- data %>% rowwise() %>% dplyr::mutate(
  group = ifelse(ensembl %in% group1, "group1",
                 ifelse(ensembl %in% group2, "group2", NA))) %>% 
  dplyr::filter(!is.na(group)) %>%
  ungroup()


## Plot distance
getPvalWilcox <- function(data, group1, group2){
  distance1 <- (data %>% dplyr::filter(group == group1) )$distanceToBoundary
  distance2 <- (data %>% dplyr::filter(group == group2) )$distanceToBoundary
  wil <- wilcox.test(distance1, distance2)
  return(wil$p.value)
}


plot_TADdistance <- function(temp.tb, note, ymin = 0, ymax = 2000000){
  p12 <- round(getPvalWilcox(temp.tb, "group1", "group2"), 5)
  
  p <- ggplot(temp.tb, aes(x = ".", fill = group, y = distanceToBoundary)) +
    # Set axis labels (no x-axis title)
    scale_fill_manual(values = ((rev(c("#777777", "#F28E2C"))))) +
    labs(x = NULL, y = "Distance from TAD boundary") +
    # Violin plot with black outline, customized line width and end
    introdataviz::geom_split_violin(
      color = "black", alpha = 0.4,
      linewidth = lineMedium * mmToLineUnit, lineend = "square"
    ) +
    # Box plot with customized line width, square end, and outlier size
    geom_boxplot(
      width = 0.3, color = "black", alpha = 0.6,
      linewidth = lineMedium * mmToLineUnit, lineend = "square",
      outlier.shape = NA,  alpha = 0.6, show.legend = FALSE
    ) +
    
    # Mean point in each group
    stat_summary(
      aes(group = group), fun = mean,
      geom = "point", shape = 21, size = 0.5,
      fill = "black", color = "black", position = position_dodge(.3)
    ) +
    
    # Horizontal line at y = 0
    geom_hline(yintercept = 0, linewidth = lineThick * mmToLineUnit) +
    
    # Coordinate limits and custom y-axis labels
    coord_cartesian(ylim = c(ymin, ymax)) +
    scale_y_continuous(labels = label_kb_mb) +
    
    # Annotate p-value text
    annotate(
      "text", x = 1, y = ymin + 1,
      label = paste0("p12: ", convPvalue(p12)),
      color = "black", hjust = 0, size = 1
    ) +
    
    # Theme customization
    theme_classic() +
    theme(
      axis.title = element_text(
        size = fontSizeM, family = fontType, color = "#000000"
      ),
      axis.text = element_text(
        size = fontSizeS, family = fontType, color = "#000000"
      ),
      axis.text.x = element_text(
        angle = 45, hjust = 1, vjust = 1
      ),
      axis.line = element_line(
        color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
      ),
      axis.ticks = element_line(
        color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
      ),
      panel.background = element_rect(fill = "transparent"),
      legend.text = element_text(family = fontType, size = fontSizeS),
      legend.title = element_text(family = fontType, size = fontSizeS)) +
    guides(
      fill = guide_legend(
        keywidth = 0.2,  # Adjust the width of the legend keys
        keyheight = 0.2  # Adjust the height of the legend keys
      )
    )
  
  
  fileName <- paste0("TAD_distance_to_boundary_", note)
  width <- panelSize(1.55)*mmToInch
  height <- panelSize(1.2)*mmToInch
  png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
  print(p)
  dev.off()
  svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
  print(p)
  dev.off()
}

plot_TADdistance(dataToPlot, "group_binayGroup", ymax = 1000000)


## Plot TAD size
getPvalWilcox <- function(data, group1, group2){
  distance1 <- (data %>% dplyr::filter(group == group1) )$tadSize
  distance2 <- (data %>% dplyr::filter(group == group2) )$tadSize
  wil <- wilcox.test(distance1, distance2)
  return(wil$p.value)
}


plot_TADsize <- function(temp.tb, note, ymin = 0, ymax = 2000000){
  p12 <- round(getPvalWilcox(temp.tb, "group1", "group2"), 5)
  
  p <- ggplot(temp.tb, aes(x = ".", fill = group, y = tadSize)) +
    # Set axis labels (no x-axis title)
    labs(x = NULL, y = "Size of TAD") +
    scale_fill_manual(values = ((rev(c("#777777", "#F28E2C"))))) +
    
    # Violin plot with black outline, customized line width and end
    introdataviz::geom_split_violin(
      color = "black",
      linewidth = lineMedium* mmToLineUnit, lineend = "square", alpha = .4
    ) +
    
    # Box plot with customized line width, square end, and outlier size
    geom_boxplot(
      width = 0.3, color = "black",
      linewidth = lineMedium* mmToLineUnit, lineend = "square",
      outlier.shape = NA,  alpha = 0.6, show.legend = FALSE
    ) +
    
    # Mean point in each group
    stat_summary(
      aes(group = group), fun = mean,
      geom = "point", shape = 21, size = 0.5,
      fill = "black", color = "black", position = position_dodge(.3)
    ) +
    
    # Horizontal line at y = 0
    geom_hline(yintercept = 0, linewidth = lineThick * mmToLineUnit) +
    
    # Coordinate limits and custom y-axis labels
    coord_cartesian(ylim = c(ymin, ymax)) +
    scale_y_continuous(labels = label_kb_mb) +
    
    # Annotate p-value text
    annotate(
      "text", x = 1, y = ymin + 1,
      label = paste0("p12: ", convPvalue(p12)),
      color = "black", hjust = 0, size = 1
    ) +
    
    # Theme customization
    theme_classic() +
    theme(
      axis.title = element_text(
        size = fontSizeM, family = fontType, color = "#000000"
      ),
      axis.text = element_text(
        size = fontSizeS, family = fontType, color = "#000000"
      ),
      axis.text.x = element_text(
        angle = 45, hjust = 1, vjust = 1
      ),
      axis.line = element_line(
        color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
      ),
      axis.ticks = element_line(
        color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
      ),
      panel.background = element_rect(fill = "transparent"),
      legend.text = element_text(family = fontType, size = fontSizeS),
      legend.title = element_text(family = fontType, size = fontSizeS)
    ) + guides(
      fill = guide_legend(
        keywidth = 0.2,  # Adjust the width of the legend keys
        keyheight = 0.2  # Adjust the height of the legend keys
      ))
  
  
  
  fileName <- paste0("TAD_size_", note)
  width <- panelSize(1.55)*mmToInch
  height <- panelSize(1.2)*mmToInch
  png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
  print(p)
  dev.off()
  svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
  print(p)
  dev.off()
}
plot_TADsize(dataToPlot, "group_binaryGroup", ymax = 3e6)


getPvalWilcox <- function(data, group1, group2){
  distance1 <- (data %>% dplyr::filter(group ==group1) )$negDiffInsScore
  distance2 <- (data %>% dplyr::filter(group ==group2) )$negDiffInsScore
  wil <- wilcox.test(distance1, distance2)
  return(wil$p.value)
}


plot_insScore <- function(temp.tb, note, ymin = -1.5, ymax = 0){
  p12 <- round(getPvalWilcox(temp.tb, "group1", "group2"), 5)
  
  p <- ggplot(temp.tb, aes(x = ".", fill = group, y = negDiffInsScore)) + 
    labs(x = NULL, y = "- Insulation score") +  # Remove x-axis title
    scale_fill_manual(values = rev((c("#777777", "#F28E2C")))) +
    
    introdataviz::geom_split_violin(linewidth = lineMedium * mmToLineUnit, lineend = "square",
                                    alpha = .4) +
    geom_boxplot(width = 0.3, color = "black",
                 linewidth = lineMedium * mmToLineUnit, lineend = "square",
                 outlier.shape = NA,  alpha = 0.6, show.legend = FALSE) + theme_classic() +
    stat_summary(
      aes(group = group), fun = mean,
      geom = "point", shape = 21, size = 0.5,
      fill = "black", color = "black", position = position_dodge(.3)
    ) +
    geom_hline(yintercept = 0, linewidth = lineThick*mmToLineUnit) +
    theme_classic() + coord_cartesian(ylim = c(ymin, ymax)) +
    annotate("text", x = 1, y = ymin + 0.1, label = paste0("p12: ", convPvalue(p12)),
             color = "black", hjust = 0, size = 1) +
    theme(
      axis.title = element_text(
        size = fontSizeM,
        family = fontType,
        color = "#000000"
      ),
      axis.text = element_text(
        size = fontSizeM,
        family = fontType,
        color = "#000000"
      ),
      axis.text.x = element_text(
        angle = 45,      # Rotate x-axis labels 45 degrees
        hjust = 1,       # Adjust horizontal justification
        vjust = 1        # Adjust vertical justification
      ),
      axis.line = element_line(
        color = "#000000",
        size = lineThick*mmToLineUnit,
        lineend = "square"
      ),
      axis.ticks = element_line(
        color = "#000000",
        size = lineThick*mmToLineUnit,
        lineend = "square"
      ),
      panel.background = element_rect(fill = "transparent"),
      legend.text = element_text(family = fontType, size = fontSizeS),
      legend.title = element_text(family = fontType, size = fontSizeS)
    )+   guides(
      fill = guide_legend(
        keywidth = 0.2,  # Adjust the width of the legend keys
        keyheight = 0.2  # Adjust the height of the legend keys
      ))
  
  
  fileName <- paste0("TAD_insulation_score_binaryGroup_", note)
  width <- panelSize(1.55)*mmToInch
  height <- panelSize(1.2)*mmToInch
  png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
  print(p)
  dev.off()
  svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
  print(p)
  dev.off()
}

# DMSO
plot_insScore(dataToPlot, "nearestBoundary_diff_dTAGvsDMSO_binaryGroup", ymin = -0.3, ymax = 0.2)


getPvalWilcox <- function(data, group1, group2){
  distance1 <- (data %>% dplyr::filter(group ==group1) )$negInsScore_DMSO
  distance2 <- (data %>% dplyr::filter(group ==group2) )$negInsScore_DMSO
  wil <- wilcox.test(distance1, distance2)
  return(wil$p.value)
}

plot_insScoreSingle <- function(temp.tb, note, ymin = -1.5, ymax = 0){
  p12 <- round(getPvalWilcox(temp.tb, "group1", "group2"), 5)
  
  p <- ggplot(temp.tb, aes(x = ".", fill = group, y =  negInsScore_DMSO)) + 
    labs(x = NULL, y = "- Insulation score") +  # Remove x-axis title
    scale_fill_manual(values = rev((c("#777777", "#F28E2C")))) +
    
    introdataviz::geom_split_violin(linewidth = lineMedium * mmToLineUnit, lineend = "square",
                                    alpha = .4) +
    geom_boxplot(width = 0.3, color = "black",
                 linewidth = lineMedium * mmToLineUnit, lineend = "square",
                 outlier.shape = NA,  alpha = 0.6, show.legend = FALSE) + theme_classic() +
    stat_summary(
      aes(group = group), fun = mean,
      geom = "point", shape = 21, size = 0.5,
      fill = "black", color = "black", position = position_dodge(.3)
    ) +
    geom_hline(yintercept = 0, linewidth = lineThick*mmToLineUnit) +
    theme_classic() + coord_cartesian(ylim = c(ymin, ymax)) +
    annotate("text", x = 1, y = ymin + 0.1, label = paste0("p12: ", convPvalue(p12)),
             color = "black", hjust = 0, size = 1) +
    theme(
      axis.title = element_text(
        size = fontSizeM,
        family = fontType,
        color = "#000000"
      ),
      axis.text = element_text(
        size = fontSizeM,
        family = fontType,
        color = "#000000"
      ),
      axis.text.x = element_text(
        angle = 45,      # Rotate x-axis labels 45 degrees
        hjust = 1,       # Adjust horizontal justification
        vjust = 1        # Adjust vertical justification
      ),
      axis.line = element_line(
        color = "#000000",
        size = lineThick*mmToLineUnit,
        lineend = "square"
      ),
      axis.ticks = element_line(
        color = "#000000",
        size = lineThick*mmToLineUnit,
        lineend = "square"
      ),
      panel.background = element_rect(fill = "transparent"),
      legend.text = element_text(family = fontType, size = fontSizeS),
      legend.title = element_text(family = fontType, size = fontSizeS)
    )+   guides(
      fill = guide_legend(
        keywidth = 0.2,  # Adjust the width of the legend keys
        keyheight = 0.2  # Adjust the height of the legend keys
      ))
  
  
  fileName <- paste0("TAD_insulation_score_binaryGroup_", note)
  width <- panelSize(1.55)*mmToInch
  height <- panelSize(1.2)*mmToInch
  png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
  print(p)
  dev.off()
  svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
  print(p)
  dev.off()
}

dataToPlot <- dataToPlot %>% mutate(negInsScore_DMSO = -(2^log2InsScore_DMSO))
plot_insScoreSingle(dataToPlot, "nearestBoundary_DMSO_dTAGvsDMSO_binaryGroup", ymin = -1, ymax = 0)


```


#### [FIG] TAD gene density
##### Making TAD gene count
```{r}
# Counting gene number for each TAD
tadGeneCount <- data %>% rowwise() %>%
  drop_na(tadID) %>%
  group_by(tadID) %>%
  summarize(tadGeneCount = n())

# Merging with data
dataToPlot <- data %>% left_join(tadGeneCount, by = c("tadID")) %>%
  mutate(tadGeneDensity = tadGeneCount/tadSize*100e3)


group1 <- fread(here(dataDir, "loop_analysis",
                     "geneList_chromosight_dTAG_vs_DMSO_p-n_complex_Group1.tsv"))$gene
group2 <- fread(here(dataDir, "loop_analysis",
                     "geneList_chromosight_dTAG_vs_DMSO_p-n_complex_Group2.tsv"))$gene
```
##### Plot gene density
```{r}
dataToPlot2 <- dataToPlot %>% rowwise() %>% dplyr::mutate(
  group = ifelse(ensembl %in% group1, "group1",
                 ifelse(ensembl %in% group2, "group2", NA))) %>% 
  dplyr::filter(!is.na(group))


## geneDensity
getPvalWilcox <- function(data, group1, group2){
  distance1 <- (data %>% dplyr::filter(group == group1) )$tadGeneDensity
  distance2 <- (data %>% dplyr::filter(group == group2) )$tadGeneDensity
  wil <- wilcox.test(distance1, distance2)
  return(wil$p.value)
}


plot_geneDensity <- function(temp.tb, note, ymin = 0, ymax = 10){
  p12 <- round(getPvalWilcox(temp.tb, "group1", "group2"), 5)
  
  p <- ggplot(temp.tb, aes(x = ".", fill = group, y = tadGeneDensity)) +
    # Set axis labels (no x-axis title)
    labs(x = NULL, y = "Gene density within TAD") +
    scale_fill_manual(values = (rev(c("#777777","#F28E2C")))) +
    # Violin plot with black outline, customized line width and end
    introdataviz::geom_split_violin(linewidth = lineMedium * mmToLineUnit, lineend = "square",
                                    alpha = .4
    ) +
    
    # Box plot with customized line width, square end, and outlier size
    geom_boxplot(
      width = 0.3, color = "black",
      linewidth = lineMedium * mmToLineUnit, lineend = "square",
      outlier.shape = NA,  alpha = 0.6, show.legend = FALSE
    ) +
    
    # Mean point in each group
    stat_summary(
      aes(group = group), fun = mean,
      geom = "point", shape = 21, size = 0.5,
      fill = "black", color = "black", position = position_dodge(.3)
    ) +
    
    # Horizontal line at y = 0
    geom_hline(yintercept = 0, linewidth = lineThick * mmToLineUnit) +
    
    # Coordinate limits and custom y-axis labels
    coord_cartesian(ylim = c(ymin, ymax)) +
    
    # Annotate p-value text
    annotate(
      "text", x = 1, y = ymin + 1,
      label = paste0("p12: ", convPvalue(p12)),
      color = "black", hjust = 0, size = 1
    ) +
    
    # Theme customization
    theme_classic() +
    theme(
      axis.title = element_text(
        size = fontSizeM, family = fontType, color = "#000000"
      ),
      axis.text = element_text(
        size = fontSizeS, family = fontType, color = "#000000"
      ),
      axis.text.x = element_text(
        angle = 45, hjust = 1, vjust = 1
      ),
      axis.line = element_line(
        color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
      ),
      axis.ticks = element_line(
        color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
      ),
      panel.background = element_rect(fill = "transparent"),
      legend.text = element_text(family = fontType, size = fontSizeS),
      legend.title = element_text(family = fontType, size = fontSizeS)
    )+   
    guides(
      fill = guide_legend(
        keywidth = 0.2,  # Adjust the width of the legend keys
        keyheight = 0.2  # Adjust the height of the legend keys
      ))
  
  
  
  fileName <- paste0("TAD_geneDensity_", note)
  width <- panelSize(1.5)*mmToInch
  height <- panelSize(1.2)*mmToInch
  png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
  print(p)
  dev.off()
  svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
  print(p)
  dev.off()
}
plot_geneDensity(dataToPlot2, "group_binaryGroup", ymax = 8)

```

##### Plot TAD avg log2FC & CDF
```{r}
diffDir <- here(dataDir, "RNA_diff")
diff.RNA <- fread(here(diffDir, "diff_G1.dTAG_G1.2i.dTAG_vs_G1.2i.DMSO.tsv")) %>%
  dplyr::select(ensembl_gene_id, log2FoldChange, shrinked_log2FC, padj)

temp <- left_join(dataToPlot, diff.RNA, by = c("ensembl" = "ensembl_gene_id")) %>% dplyr::filter(!is.na(log2FoldChange),
                                                                                                 !is.na(tadID))


dataToPlot2 <- temp %>% group_by(tadID) %>%
  dplyr::summarise(absAvgLog2FC = abs(mean(log2FoldChange)),
                   avgLog2FC = mean(log2FoldChange),
                   tadGeneDensity = mean(tadGeneDensity))

dataToPlot2$geneDensityGroup <- cut(
  dataToPlot2$tadGeneDensity,
  breaks = quantile(dataToPlot2$tadGeneDensity, probs = seq(0, 1, 0.2), na.rm = TRUE),
  include.lowest = TRUE,
  labels = paste0(seq(0, 80, 20), "-", seq(20, 100, 20), "%")
)



### Ploting  grouping

p <- ggplot(dataToPlot2, aes(x = geneDensityGroup, y = tadGeneDensity, fill = geneDensityGroup)) + 
  geom_boxplot(color = "black",
               linewidth = lineThick * mmToLineUnit, lineend = "square",
               outlier.shape = NA) + theme_classic() +
  coord_cartesian(ylim = c(0, quantile(dataToPlot2$tadGeneDensity, 0.99))) +
  labs(x = "Gene density group" , y = "TAD gene density") +
  scale_fill_manual(values = rev(c("#777777", "#8B7E65", "#A28452", "#C2884D", "#F28E2C"))) +
  theme(
    legend.position = "none",
    axis.title = element_text(
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS, family = fontType, color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45, hjust = 1, vjust = 1
    ),
    axis.line = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  )

fileName <- paste0("TAD_tadGroup_geneDensity")
width <- panelSize(1.2)*mmToInch
height <- panelSize(1.7)*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()


### CDF plot

group1_data <- dataToPlot2$absAvgLog2FC[dataToPlot2$geneDensityGroup == "0-20%"]
group2_data <- dataToPlot2$absAvgLog2FC[dataToPlot2$geneDensityGroup == "20-40%"]
ks.test(group1_data, group2_data)

group1_data <- dataToPlot2$absAvgLog2FC[dataToPlot2$geneDensityGroup == "20-40%"]
group2_data <- dataToPlot2$absAvgLog2FC[dataToPlot2$geneDensityGroup == "40-60%"]
ks.test(group1_data, group2_data)

group1_data <- dataToPlot2$absAvgLog2FC[dataToPlot2$geneDensityGroup == "40-60%"]
group2_data <- dataToPlot2$absAvgLog2FC[dataToPlot2$geneDensityGroup == "60-80%"]
ks.test(group1_data, group2_data)

group1_data <- dataToPlot2$absAvgLog2FC[dataToPlot2$geneDensityGroup == "60-80%"]
group2_data <- dataToPlot2$absAvgLog2FC[dataToPlot2$geneDensityGroup == "80-100%"]
ks.test(group1_data, group2_data)

p <- ggplot(dataToPlot2, aes(x = absAvgLog2FC, color = geneDensityGroup)) +
  scale_color_manual(values = rev(c("#777777", "#8B7E65", "#A28452", "#C2884D", "#F28E2C"))) +
  stat_ecdf(size = 0.4, linewidth = lineMedium * mmToLineUnit, lineend = "square" ) + # Use stat_ecdf to plot the empirical CDF
  labs(
    x = "Abs. log2(fold change)",
    y = "Cumulative Probability"
  ) + coord_cartesian(xlim = c(0, 1.5)) +
  theme_classic() + # Clean theme
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.position = "none",
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS),
  ) + scale_y_continuous(labels = scales::number_format(accuracy = 0.1))


fileName <- paste0("TAD_tadGroup_geneDensity_avgLog2FC_cdf")
width <- 33*mmToInch
height <-33*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()




```

### ChIP peak density
#### [FIG] H3K4me3 peak number per 10kb
```{r}
process_group <- function(group_name, gene_data, bins) {
  # Filter for the specific group
  TSS_group <- makeGRangesFromDataFrame(
    gene_data %>% filter(ensembl %in% get(group_name)), 
    keep.extra.columns = TRUE
  )
  
  # Process each TSS in the group
  all_results <- map_dfr(seq_len(length(TSS_group)), function(i) {
    # Find overlapping bins
    overlapping_bins <- findOverlaps(TSS_group[i], bins)
    bins_near_tss <- as_tibble(bins[subjectHits(overlapping_bins)])
    
    # Calculate distance bin
    TSScenter <- (as_tibble(TSS_group[i]) %>%
                    mutate(center = (start + end) / 2))$center
    centerBinStart <- floor(TSScenter / 10e3) * 10e3 + 1
    bins_near_tss <- bins_near_tss %>%
      mutate(distanceBin = abs((start - centerBinStart)) / 10e3)
    
    # Summarize results
    result <- bins_near_tss %>%
      group_by(distanceBin) %>%
      summarise(mean_peak_counts = mean(peak_counts, na.rm = TRUE), .groups = "drop") %>%
      mutate(gene = TSS_group[i]$ensembl)
    return(result)
  })
  
  # Calculate mean and SD for the group
  mean_data <- all_results %>%
    group_by(distanceBin) %>%
    summarise(mean_peak_counts = mean(mean_peak_counts, na.rm = TRUE))
  
  sd_data <- all_results %>%
    group_by(distanceBin) %>%
    summarise(sd_peak_counts = sd(mean_peak_counts, na.rm = TRUE))
  
  # Join and add group name
  summary_data <- left_join(mean_data, sd_data, by = "distanceBin") %>%
    mutate(group = group_name)
  
  return(summary_data)
}

##########################################################


## Import peak
peak <- importPeak(here(refDir, "ChIP", "H3K4me3", "H3K4me3_ESC_unionPeaks.bed"))

## Import 10kb bin
temp <- fread(here(refDir, "mm10", "mm10.bin.10kb.bed")) %>% dplyr::select(c(1, 2, 3))
colnames(temp) <- c("chr", "start", "end")
temp <- temp %>% dplyr::mutate(start = start+1)
bins.10kb <- makeGRangesFromDataFrame(temp)

## Count overlap
counts <- countOverlaps(bins.10kb, peak)
mcols(bins.10kb)$peak_counts <- counts

## Getting TSS
flankSize <- 1e6
gene.tb <- fread(here(refDir, "mm10", "mm10_GRCm38.p6_gene_sorted.bed")) %>%
  dplyr::mutate(TSS = ifelse(V4 == "+", V2, V3),
                TSSstart = TSS - flankSize,
                TSSend = TSS + flankSize) %>%
  dplyr::select(V1, TSSstart, TSSend, V6)
colnames(gene.tb) <- c("chr", "start", "end", "ensembl")


group1 <- fread(here(dataDir, "loop_analysis",
                     "geneList_chromosight_dTAG_vs_DMSO_p-n_complex_Group1.tsv"))$gene
group2 <- fread(here(dataDir, "loop_analysis",
                     "geneList_chromosight_dTAG_vs_DMSO_p-n_complex_Group2.tsv"))$gene

##########################################################
# List of groups to process
groups <- c("group1", "group2")

# Process each group and combine results
summary_data <- map_dfr(groups, ~ process_group(.x, gene.tb, bins.10kb))

# Add the additional rows for distanceBin = 0
summary_data <- summary_data %>%
  mutate(distanceBin = distanceBin + 1) %>%
  add_row(distanceBin = 0, mean_peak_counts = 0, sd_peak_counts = 0, group = "group1") %>%
  add_row(distanceBin = 0, mean_peak_counts = 0, sd_peak_counts = 0, group = "group2")

n1 <- length(group1)
n2 <- length(group2)

# summary_data <- summary_data %>%
#   mutate(
#     se = if_else(group == "group1", sd_peak_counts / sqrt(n1), sd_peak_counts/sqrt(n2)),
#     ci_low  = mean_peak_counts - 2*se,
#     ci_high = mean_peak_counts + 2*se
#   )

# Plot the results
p <- ggplot(summary_data, aes(x = distanceBin, y = mean_peak_counts, fill = group, color = group, group = group)) + 
  geom_smooth(
    linewidth = lineMedium,
    method = "loess", 
    formula = y ~ x, 
    span   = 0.2,    # smaller = wigglier, larger = smoother
    se     = TRUE,   # show confidence band
    alpha  = 0.2, 
  ) +
  labs(
    x = "Distance bin (10kb)",
    y = "H3K4me3 peak count"
  ) +
  theme_classic() +
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    # legend.text = element_text(family = fontType, size = fontSizeS),
    # legend.title = element_text(family = fontType, size = fontSizeS),
    legend.position = "none"  # Moves legend inside the plot (x, y relative coordinates)
  ) +
  guides(
    fill = guide_legend(
      keywidth = 0.2,  # Adjust the width of the legend keys
      keyheight = 0.2  # Adjust the height of the legend keys
    ))+
  scale_color_manual(values = rev(c("#777777", "#F28E2C"))) +
  scale_fill_manual(values = rev(c("#777777", "#F28E2C")))


##########################################################

fileName <- paste0("binPeakDensity_H3K4me3_binaryGroup")
width <- 38*mmToInch
height <- 35*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()
```

#### [FIG] RAD21 peak number at promoter
```{r}
dataDir <- here("..", "data")
peak.RAD21 <- importPeak(here(refDir, "ChIP", "RAD21-HA", "RAD21_ESC-DMSO_unionPeaks.bed"))

################################################################################
flankSize <- 5000
gene.tb <- fread(here(refDir, "mm10", "mm10_GRCm38.p6_gene_sorted.bed")) %>%
  dplyr::mutate(TSS = ifelse(V4 == "+", V2, V3),
                TSSstart = TSS - flankSize,
                TSSend = TSS + flankSize) %>%
  dplyr::select(V6, V5, V1, TSSstart, TSSend)
colnames(gene.tb) <- c("ensembl_gene_id", "external_gene_name", "chr", "start", "end")


# Convert gene.tb to a GRanges object
gene_gr <- GRanges(
  seqnames = gene.tb$chr,
  ranges = IRanges(start = gene.tb$start, end = gene.tb$end),
  gene_id = gene.tb$ensembl_gene_id
)

# Count overlaps between peaks and genes
overlap_counts <- countOverlaps(gene_gr, peak.RAD21)

# Add overlap counts to the original gene.tb data
gene.tb$peak_count <- overlap_counts


group1 <- fread(here(dataDir, "loop_analysis",
                     "geneList_chromosight_dTAG_vs_DMSO_p-n_complex_Group1.tsv"))$gene
group2 <- fread(here(dataDir, "loop_analysis",
                     "geneList_chromosight_dTAG_vs_DMSO_p-n_complex_Group2.tsv"))$gene


gene.tb <- gene.tb %>% dplyr::mutate(group = case_when(
  ensembl_gene_id %in% group1 ~ "group1",
  ensembl_gene_id %in% group2 ~ "group2",
  TRUE ~ NA
)) %>%
  dplyr::filter(!is.na(group))

###############################################################################

dataToPlot <- gene.tb %>%
  mutate(peakNum = case_when(
    peak_count == 0 ~ "peakNum0",
    peak_count == 1 ~ "peakNum1",
    peak_count >- 2 ~ "peakNumOver2",
    TRUE ~ NA
  ))

# Create the stacked barplot
dataToPlot_short <- dataToPlot %>%
  group_by(group, peakNum) %>%
  summarize(count = n(), .groups = "drop") %>%
  group_by(group) %>%
  mutate(ratio = count / sum(count))

# Define the gradient colors
gradient_colors <- c("#D4D4D4", "#E48D8A", "#F44641")

# Create the stacked barplot with gradient colors
p <- ggplot(dataToPlot_short, aes(x = group, y = ratio, fill = peakNum)) +
  geom_bar(stat = "identity", position = "stack",
           linewidth = lineMedium * mmToLineUnit, lineend = "square") +
  theme_classic() + labs(x = NULL , y = "Ratio") +
  theme(
    axis.title = element_text(
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS, family = fontType, color = "#000000"
    ),
    axis.line = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) +
  scale_fill_manual(values = gradient_colors, 
                    guide = guide_legend(
                      override.aes = list(size = 0.5), # Adjust legend key symbol size
                      keywidth = 2 / 2.54, # Convert 2mm to cm
                      keyheight = 2 / 2.54 # Convert 2mm to cm
                    )) 

width <- panelSize(1.7)*mmToInch
height <- panelSize(1.3)*mmToInch
fileName <- paste0("ChIP_peak_RAD21occupancy_promoter")
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()  

```
### Loop analysis per gene group
```{r}
commonLoopDir <- here("../data/loop_analysis")
diffDir <- here("../data/RNA_diff")
name <- "chromosight"

geneAnnoData <- loadLoopAnnoData(here(commonLoopDir, paste0("loopScore_", name, "_p-n_complex_ensemblList.tsv"))) %>%
  dplyr::mutate(distance = start2 - start1,
                loopID = paste(chrom1, start1, start2, sep = "_"))
```

##### [FIG] Loop number per gene: P-N
```{r}
dataDir <- here("../data")
group1 <- fread(here(dataDir, "loop_analysis",
                     "geneList_chromosight_dTAG_vs_DMSO_p-n_complex_Group1.tsv"))$gene
group2 <- fread(here(dataDir, "loop_analysis",
                     "geneList_chromosight_dTAG_vs_DMSO_p-n_complex_Group2.tsv"))$gene

# Counting number of loop per genes
tempSum <- geneAnnoData %>% dplyr::select(loopID, gene, Anno) %>% unnest(gene) %>% group_by(gene) %>%  summarize(
  loop = list(loopID),
  anno = list(Anno),
  count = n())

tempSum <- tempSum %>% dplyr::mutate(
  group = ifelse(gene %in% group1, "group1",
                 ifelse(gene %in% group2, "group2", NA))
) %>% dplyr::filter(!is.na(group)) %>% dplyr::filter(group %in% c("group1", "group2"))

# 
getPvalWilcox <- function(data, group1, group2){
  distance1 <- (data %>% dplyr::filter(group ==group1) )$count
  distance2 <- (data %>% dplyr::filter(group ==group2) )$count
  wil <- wilcox.test(distance1, distance2)
  return(wil$p.value)
}


pv12 <- round(getPvalWilcox(tempSum,"group1", "group2"), 5)


p <- ggplot(tempSum, aes(x = group, y = count, fill = group)) + 
  geom_boxplot(width = 0.5, outlier.shape = NA, color = "black", alpha = 0.6,
               linewidth = lineThick * mmToLineUnit, lineend = "square", show.legend = FALSE) + theme_classic() +       
  scale_y_continuous(breaks = seq(0, 10, by = 2)) + 
  labs(x = NULL, y = "# of P-N loops per gene") +
  coord_cartesian(ylim = c(0, 8)) +
  stat_summary(
    aes(group = group), fun = mean,
    geom = "point", shape = 21, size = 0.5,
    fill = "black", color = "black"
  )+
  annotate("text", x = 1, y = 3, label = paste0("pv12: ", convPvalue(pv12)),
           color = "black", hjust = 0, size = 1) +
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) +  scale_fill_manual(values = rev(c("#777777", "#F28E2C")))



fileName <- paste0("count_barplot_diffGroup_dTAG_vs_DMSO_binaryGroup")
width <- panelSize(0.8)*mmToInch
height <- panelSize(1.5)*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()
```

##### [FIG] Counting loop types
```{r}
#######
temp2 <- tempSum %>% rowwise() %>% mutate(total = length(anno),
                                          num_pp = sum(anno == "P-P"),
                                          num_pe = sum(anno == "P-E"),
                                          num_px = sum(anno == "P-X"),
                                          num_ps = sum(anno %in% c("P-S", "P-SP", "P-SE")),
                                          num_sps = sum(anno %in% c("SP-SP", "SP-SE", "SP-S")),
                                          num_spn = sum(anno %in% c("E-SP", "SP-X")),
                                          num_str = num_ps + num_sps + num_spn
)


loopType <- temp2 %>% group_by(group) %>% summarise(num_pp = sum(num_pp),
                                                    num_pe = sum(num_pe),
                                                    # num_str = sum(num_str),
                                                    num_ps = sum(num_ps),
                                                    num_sps = sum(num_sps),
                                                    num_spn = sum(num_spn),
                                                    num_px = sum(num_px))


loopTypeLong <- loopType %>% pivot_longer(-group, names_to = "type", values_to = "count")

loopTypeLong$type <- factor(loopTypeLong$type, levels = c("num_pp", "num_pe", "num_px",
                                                          "num_ps", "num_spn", 
                                                          "num_sps" ))

# Plotting
p <- ggplot(loopTypeLong, aes(fill=type, y=count, x=group)) + 
  geom_bar(position="fill", stat="identity") + theme_classic() +
  scale_fill_manual(values = c(strong_red, strong_orange, strong_yellow, strong_blue, "grey30", "grey10")) +theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  )+   
  guides(
    fill = guide_legend(
      keywidth = 0.2,  # Adjust the width of the legend keys
      keyheight = 0.2  # Adjust the height of the legend keys
    ))

fileName <- paste0("count_barplot_diffGroup_dTAG_vs_DMSO_binaryGroup_alltypes")
width <- panelSize(1.5)*mmToInch
height <- panelSize(1.5)*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()
```

##### [FIG] Loop number per gene: structural
```{r}
########
data <- temp2 %>% dplyr::select(group, num_str)
colnames(data) <- c("group", "count")
pv12 <- round(getPvalWilcox(data,"group1", "group2"), 5)

p <- ggplot(data, aes(x = group, y = count, fill = group)) + 
  geom_boxplot(width = 0.5, outlier.shape = NA, color = "black", alpha = 0.6,
               linewidth = lineThick * mmToLineUnit, lineend = "square", show.legend = FALSE) + theme_classic() +       
  scale_y_continuous(breaks = seq(0, 10, by = 2)) + 
  labs(x = NULL, y = "# of PromStr loops per gene") +
  coord_cartesian(ylim = c(0, 4)) +
  stat_summary(
    aes(group = group), fun = mean,
    geom = "point", shape = 21, size = 0.5,
    fill = "black", color = "black"
  )+
  annotate("text", x = 1, y = 3, label = paste0("pv12: ", convPvalue(pv12)),
           color = "black", hjust = 0, size = 1) +
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) +  scale_fill_manual(values = rev(c("#777777", "#F28E2C")))

fileName <- paste0("count_barplot_diffGroup_dTAG_vs_DMSO_str_binaryGroup")
width <- panelSize(0.8)*mmToInch
height <- panelSize(1.5)*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()
```
###### (WIP) GO
```{r}
group1 <- fread(here(dataDir, "loop_analysis",
                     "geneList_chromosight_dTAG_vs_DMSO_p-n_complex_Group1.tsv"))$gene
group2 <- fread(here(dataDir, "loop_analysis",
                     "geneList_chromosight_dTAG_vs_DMSO_p-n_complex_Group2.tsv"))$gene


GO1 <- enrichGO(gene = group1, OrgDb = org.Mm.eg.db, keyType = "ENSEMBL", ont = "BP", readable = TRUE)
GO2 <- enrichGO(gene = group2, OrgDb = org.Mm.eg.db, keyType = "ENSEMBL", ont = "BP", readable = TRUE)
GO1.df <- as.data.frame(GO1)
GO2.df <- as.data.frame(GO2)
# fwrite(GO1.df, here("GO_binary_group1.tsv"), sep = "\t")
# fwrite(GO2.df, here("GO_binary_group2.tsv"), sep = "\t")
# GO1.df <- fread(here("GO_binary_group1.tsv"))
# GO2.df <- fread(here("GO_binary_group2.tsv"))

subset1 <- GO1.df %>% dplyr::select(ID, Description, GeneRatio, p.adjust) %>% dplyr::mutate(group = "Group 1") %>%
  dplyr::arrange(p.adjust)
subset1$GeneRatio <- sapply(strsplit(subset1$GeneRatio, "/"), function(x) as.numeric(x[1]) / as.numeric(x[2]))

subset2 <- GO2.df %>% dplyr::select(ID, Description, GeneRatio, p.adjust) %>% dplyr::mutate(group = "Group 2") %>%
  dplyr::arrange(p.adjust)
subset2$GeneRatio <- sapply(strsplit(subset2$GeneRatio, "/"), function(x) as.numeric(x[1]) / as.numeric(x[2]))


GOlist <- factor(c("GO:0001824", "GO:0030879", "GO:0021953", "GO:0019827",
                   "GO:0022613", "GO:0050767", "GO:0034470", "GO:0016055",
                   "GO:0006397", "GO:0030900", "GO:0008380"))

data <- bind_rows(subset1, subset2) %>%
  dplyr::filter(ID %in% GOlist)

descOrder <- sort(unique(data$Description))[c(4, 10, 2, 1, 3, 11, 7, 8, 6, 5, 9)]
pValueLogMax <- 10
data <- data %>% dplyr::rowwise() %>% dplyr::mutate(pValueLog = min(-log10(p.adjust), pValueLogMax))
p <- ggplot(data, aes(x = group, y = factor(Description, levels = descOrder), size = pValueLog, fill = GeneRatio)) +
  geom_point(shape = 21,        # Ensures a point with an outline
             stroke = 1*ptToMM) +
  scale_size_continuous(range = c(0.5, 2)) +  # Set min and max point sizes here
  scale_fill_gradient(low = "white", high = "#CB333A",
                      # limits = c(0, 1),
                      oob = scales::squish, # Define gradient colors
                      guide = guide_colorbar(
                        barwidth = 1.5/5.08,  # Adjust width of the color bar
                        barheight = 15/5.08   # Adjust height of the color bar
                      )
  ) +
  labs(x = NULL, y = NULL) +
  theme_bw() +  # Apply theme_bw first, so custom theme settings come after
  theme(
    panel.background = element_rect(fill = "transparent"),  # Override theme_bw panel
    axis.title = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),
    axis.text.x = element_text(
      size = fontSizeS,         # Ensure size is set for x-axis text
      family = fontType,
      color = "#000000",
    ),
    axis.text.y = element_text(
      size = fontSizeS,         # Ensure size is set for y-axis text
      family = fontType,
      color = "#000000",
      lineheight = 0.9          # Allows wrapping for y-axis labels to fit into 2 lines
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick * mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick * mmToLineUnit,
      lineend = "square"
    ),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  )



fileName <- here(figDir, "..", "GO", "GO_groups_binaryGrouping")
width <- panelSize(2.5)*mmToInch
height <- panelSize(1.2)*mmToInch
svglite(paste0(fileName, ".svg"), height = height, width = width)
print(p)
dev.off()
```

##### [FIG] Loop number per gene: P-P
```{r}
########
data <- temp2 %>% dplyr::select(group, num_pp)
colnames(data) <- c("group", "count")
pv12 <- round(getPvalWilcox(data,"group1", "group2"), 5)

p <- ggplot(data, aes(x = group, y = count, fill = group)) + 
  geom_boxplot(width = 0.5, outlier.shape = NA, color = "black", alpha = 0.6,
               linewidth = lineThick * mmToLineUnit, lineend = "square", show.legend = FALSE) + theme_classic() +       
  scale_y_continuous(breaks = seq(0, 10, by = 2)) + 
  labs(x = NULL, y = "# of P-P loops per gene") +
  coord_cartesian(ylim = c(0, 4)) +
  stat_summary(
    aes(group = group), fun = mean,
    geom = "point", shape = 21, size = 0.5,
    fill = "black", color = "black"
  )+
  annotate("text", x = 1, y = 3, label = paste0("pv12: ", convPvalue(pv12)),
           color = "black", hjust = 0, size = 1) +
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) +  scale_fill_manual(values = rev(c("#777777", "#F28E2C")))

fileName <- paste0("count_barplot_diffGroup_dTAG_vs_DMSO_pp_binaryGroup")
width <- panelSize(0.8)*mmToInch
height <- panelSize(1.5)*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()
```

##### [FIG] Loop number per gene: P-E
```{r}
########
data <- temp2 %>% dplyr::select(group, num_pe)
colnames(data) <- c("group", "count")
pv12 <- round(getPvalWilcox(data,"group1", "group2"), 5)

p <- ggplot(data, aes(x = group, y = count, fill = group)) + 
  geom_boxplot(width = 0.5, outlier.shape = NA, color = "black", alpha = 0.6,
               linewidth = lineThick * mmToLineUnit, lineend = "square", show.legend = FALSE) + theme_classic() +       
  scale_y_continuous(breaks = seq(0, 10, by = 2)) + 
  labs(x = NULL, y = "# of P-E loops per gene") +
  coord_cartesian(ylim = c(0, 4)) +
  stat_summary(
    aes(group = group), fun = mean,
    geom = "point", shape = 21, size = 0.5,
    fill = "black", color = "black"
  )+
  annotate("text", x = 1, y = 3, label = paste0("pv12: ", convPvalue(pv12)),
           color = "black", hjust = 0, size = 1) +
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) +  scale_fill_manual(values = rev(c("#777777", "#F28E2C")))

fileName <- paste0("count_barplot_diffGroup_dTAG_vs_DMSO_pe_binaryGroup")
width <- panelSize(0.8)*mmToInch
height <- panelSize(1.5)*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()
```


##### [FIG] Grouping genes with structural loop number
###### Grouping
```{r}
# temp2 contains genes from group 1, 2, 5, 8 and loop counts
#calculating diff score and log2fc distribution based on p-n numbers

psOver4 <- (temp2 %>% dplyr::filter(num_str >= 4))$gene
ps3 <- (temp2 %>% dplyr::filter(num_str == 3))$gene
ps2 <- (temp2 %>% dplyr::filter(num_str == 2))$gene
ps1 <- (temp2 %>% dplyr::filter(num_str == 1))$gene
ps0 <- (temp2 %>% dplyr::filter(num_str == 0))$gene

# 
# temp2 <- tempSum %>% rowwise() %>% mutate(total = length(anno),
#                                           num_pp = sum(anno == "P-P"),
#                                           num_pe = sum(anno == "P-E"),
#                                           num_px = sum(anno == "P-X"),
#                                           num_ps = sum(anno %in% c("P-S", "P-SP", "P-SE")),
#                                           num_sps = sum(anno %in% c("SP-SP", "SP-SE", "SP-S")),
#                                           num_spn = sum(anno %in% c("E-SP", "SP-X")),
#                                           num_str = num_ps + num_sps + num_spn
# 

## Dividing genes into groups

temp <- geneAnnoData %>% dplyr::mutate(diff = score_dTAG - score_DMSO) %>%
  dplyr::filter(str_detect(Anno, "P")) %>%
  dplyr::select(diff, gene) %>%
  unnest(gene) %>% group_by(gene) %>%
  summarize(mean_diff_score = mean(diff),
            .groups = 'drop')

diff.RNA <- fread(here(diffDir, "diff_G1.dTAG_G1.2i.dTAG_vs_G1.2i.DMSO.tsv")) %>%
  dplyr::select(ensembl_gene_id, log2FoldChange, shrinked_log2FC, padj, external_gene_name)

temp <- left_join(temp, diff.RNA, by = c("gene" = "ensembl_gene_id")) %>% 
  drop_na(shrinked_log2FC)


temp <- temp %>% dplyr::mutate(
  psNum = case_when(
    gene %in% ps0 ~ "ps0",
    gene %in% ps1 ~ "ps1",
    gene %in% ps2 ~ "ps2",
    gene %in% ps3 ~ "ps3",
    gene %in% psOver4 ~ "psOver4",
    TRUE ~ NA
  )) %>%
  drop_na(psNum)

```



###### loop score
```{r}
getPvalWilcox <- function(data, group1, group2){
  distance1 <- (data %>% dplyr::filter(psNum ==group1) )$mean_diff_score
  distance2 <- (data %>% dplyr::filter(psNum ==group2) )$mean_diff_score
  wil <- wilcox.test(distance1, distance2)
  return(wil$p.value)
}

ps01 <- round(getPvalWilcox(temp, "ps0", "ps1"), 5)
ps12 <- round(getPvalWilcox(temp, "ps1", "ps2"), 5)
ps23 <- round(getPvalWilcox(temp, "ps2", "ps3"), 5)
ps34 <- round(getPvalWilcox(temp, "ps3", "psOver4"), 5)
ps24 <- round(getPvalWilcox(temp, "ps2", "psOver4"), 5)
ps14 <- round(getPvalWilcox(temp, "ps1", "psOver4"), 5)
ps04 <- round(getPvalWilcox(temp, "ps0", "psOver4"), 5)


p <- ggplot(temp, aes(x = psNum, y = mean_diff_score)) + 
  geom_violin(aes(fill = psNum), 
              color = "black",
              linewidth = lineMedium * mmToLineUnit, lineend = "square",
              show.legend = FALSE) +
  geom_boxplot(width = 0.3, color = "black",
               linewidth = lineMedium * mmToLineUnit, lineend = "square",
               outlier.shape = NA
  ) + theme_classic() + labs(x = NULL , y = "Average Δ loop score") +
  stat_summary(
    aes(group = psNum), fun = mean,
    geom = "point", shape = 21, size = 0.5,
    fill = "black", color = "black"
  ) +
  geom_hline(yintercept = 0, linewidth = lineThick * mmToLineUnit) +
  theme(
    axis.title = element_text(
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS, family = fontType, color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45, hjust = 1, vjust = 1
    ),
    axis.line = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  )+
  annotate("text", x = 1, y = -0.3, label = paste0("ps01: ", convPvalue(ps01), "\n",
                                                   "ps12: ", convPvalue(ps12), "\n",
                                                   "ps23: ", convPvalue(ps23), "\n",
                                                   "ps34: ", convPvalue(ps34), "\n",
                                                   "ps24: ", convPvalue(ps24), "\n", 
                                                   "ps14: ", convPvalue(ps14), "\n",
                                                   "ps04: ", convPvalue(ps04), "\n"), 
           color = "black", hjust = 0, size = 2) +
  scale_fill_manual(values = c("#777777", "#8B7E65", "#A28452", "#C2884D", "#F28E2C")) +
  coord_cartesian(ylim = c(-0.7, 0))



fileName <- paste0("psGroup_loopDiffScorre_dTAGvsDMSO")
width <- 33*mmToInch
height <-38*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()
```

###### log2FC
```{r}
temp <- temp %>% dplyr::mutate(absLog2FC = abs(log2FoldChange))

ks_result1 <- ks.test(
  temp %>% dplyr::filter(psNum == "psOver4") %>% pull(absLog2FC),
  temp %>% dplyr::filter(psNum == "ps3") %>% pull(absLog2FC)
)
ks_result2 <- ks.test(
  temp %>% dplyr::filter(psNum == "psOver4") %>% pull(absLog2FC),
  temp %>% dplyr::filter(psNum == "ps2") %>% pull(absLog2FC)
)
ks_result3 <- ks.test(
  temp %>% dplyr::filter(psNum == "psOver4") %>% pull(absLog2FC),
  temp %>% dplyr::filter(psNum == "ps1") %>% pull(absLog2FC)
)
ks_result4 <- ks.test(
  temp %>% dplyr::filter(psNum == "psOver4") %>% pull(absLog2FC),
  temp %>% dplyr::filter(psNum == "ps0") %>% pull(absLog2FC)
)

p <- ggplot(temp, aes(x = absLog2FC, color = psNum)) +
  scale_color_manual(values = (c("#777777", "#8B7E65", "#A28452", "#C2884D", "#F28E2C"))) +
  stat_ecdf(size = 0.4, linewidth = lineMedium * mmToLineUnit, lineend = "square" ) + # Use stat_ecdf to plot the empirical CDF
  labs(
    x = "Abs. log2(fold change)",
    y = "Cumulative Probability"
  ) + coord_cartesian(xlim = c(0, 1.5)) +
  theme_classic() + # Clean theme
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.position = "none",
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) + scale_y_continuous(labels = scales::number_format(accuracy = 0.1))


fileName <- paste0("psGroup_abslog2FC_dTAGvsDMSO")
width <- 33*mmToInch
height <-33*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()



```


###### (WIP) GO
```{r}
# GOfigDir <- here(figDir, "../GO")
# getGO("psOver4", GOfigDir, psOver4)
# getGO("psOver3", GOfigDir, psOver3)
# getGO("psOver2", GOfigDir, psOver2)
# getGO("psOver1", GOfigDir, psOver1)
# getGO("psOver0", GOfigDir, psOver0)
# 
# #####################
GO0.df <- as.data.frame(enrichGO(gene = ps0, OrgDb = org.Mm.eg.db, keyType = "ENSEMBL", ont = "BP"))
GO1.df <- as.data.frame(enrichGO(gene = ps1, OrgDb = org.Mm.eg.db, keyType = "ENSEMBL", ont = "BP"))
GO2.df <- as.data.frame(enrichGO(gene = ps2, OrgDb = org.Mm.eg.db, keyType = "ENSEMBL", ont = "BP"))
GO3.df <- as.data.frame(enrichGO(gene = ps3, OrgDb = org.Mm.eg.db, keyType = "ENSEMBL", ont = "BP"))
GO4.df <- as.data.frame(enrichGO(gene = psOver4, OrgDb = org.Mm.eg.db, keyType = "ENSEMBL", ont = "BP"))


subset0 <- GO0.df %>% dplyr::select(ID, Description, GeneRatio, p.adjust) %>% dplyr::mutate(group = "psOver0") %>%
  dplyr::mutate(
    gr = sapply(GeneRatio, function(x) {
      # Split the string by "/"
      parts <- unlist(strsplit(x, "/"))
      # Convert to numeric and perform the division
      as.numeric(parts[1]) / as.numeric(parts[2])
    })
  ) %>% dplyr::arrange(desc(gr))
subset1 <- GO1.df %>% dplyr::select(ID, Description, GeneRatio, p.adjust) %>% dplyr::mutate(group = "psOver1") %>%
  dplyr::mutate(
    gr = sapply(GeneRatio, function(x) {
      # Split the string by "/"
      parts <- unlist(strsplit(x, "/"))
      # Convert to numeric and perform the division
      as.numeric(parts[1]) / as.numeric(parts[2])
    })
  ) %>% dplyr::arrange(desc(gr))
subset2 <- GO2.df %>% dplyr::select(ID, Description, GeneRatio, p.adjust) %>% dplyr::mutate(group = "psOver2") %>%
  dplyr::mutate(
    gr = sapply(GeneRatio, function(x) {
      # Split the string by "/"
      parts <- unlist(strsplit(x, "/"))
      # Convert to numeric and perform the division
      as.numeric(parts[1]) / as.numeric(parts[2])
    })
  ) %>% dplyr::arrange(desc(gr))
subset3 <- GO3.df %>% dplyr::select(ID, Description, GeneRatio, p.adjust) %>% dplyr::mutate(group = "psOver3") %>%
  dplyr::mutate(
    gr = sapply(GeneRatio, function(x) {
      # Split the string by "/"
      parts <- unlist(strsplit(x, "/"))
      # Convert to numeric and perform the division
      as.numeric(parts[1]) / as.numeric(parts[2])
    })
  ) %>% dplyr::arrange(desc(gr))
subset4 <- GO4.df %>% dplyr::select(ID, Description, GeneRatio, p.adjust) %>% dplyr::mutate(group = "psOver4") %>%
  dplyr::mutate(
    gr = sapply(GeneRatio, function(x) {
      # Split the string by "/"
      parts <- unlist(strsplit(x, "/"))
      # Convert to numeric and perform the division
      as.numeric(parts[1]) / as.numeric(parts[2])
    })
  ) %>% dplyr::arrange(desc(gr))


GOlist <- factor(c("GO:0006397", "GO:0008380", "GO:0022613", "GO:0034470",
                   "GO:0016055", "GO:0007389", "GO:0048562", "GO:0045165",
                   "GO:0072001", "GO:0007517", "GO:0048705", "GO:0003002"))

data <- bind_rows(bind_rows(bind_rows(bind_rows(subset0, subset1), subset2), subset3), subset4) %>%
  dplyr::filter(ID %in% GOlist)

p <- ggplot(data, aes(x = group, y = Description, color = p.adjust, size = gr)) +
  geom_point() + theme_bw() +
  scale_color_gradient(low = "red", high = "blue", limits = c(0, 0.05)) +
  scale_size_continuous(range = c(0, 3)) +
  labs(x = NULL, y = NULL) +
  theme(axis.text = element_text(size = 6),  # Set axis text size
        axis.title = element_text(size = 6), # Set axis title size (if not removed)
        legend.text = element_text(size = 6), # Set legend text size
        legend.title = element_text(size = 6))
# 
# fileName <- here(figDir, "..", "GO", "GO_groups_ps")
# height = 2
# width = 3.4
# svglite(paste0(fileName, ".svg"), height = height, width = width)
# print(p)
# dev.off()
```

### Finding largest convergent encompassing S-S
#### Annotating CTCF motif
```{r}
peak.CTCF <- fread(here(refDir, "ChIP", "CTCF_ESC_unionPeaks.bed"))
colnames(peak.CTCF) <- c("chr", "start", "end")
peak.CTCF <- makeGRangesFromDataFrame(peak.CTCF)

library(BSgenome.Mmusculus.UCSC.mm10)

# Extract sequences
sequences <- getSeq(BSgenome.Mmusculus.UCSC.mm10, peak.CTCF)


# Get the CTCF motif (or provide your own PWM)
ctcf_motif <- query(MotifDb, c("CTCF", "Mmusculus"))[[1]]

# Scan for the motif
motif_hits <- lapply(sequences, function(seq) {
  matchPWM(ctcf_motif, seq, min.score = "80%") # Adjust min.score as needed
})

motif_hits_reverse <- lapply(sequences, function(seq) {
  matchPWM(ctcf_motif, reverseComplement(seq), min.score = "80%")
})

# Combine results
result <- mapply(function(fwd, rev) {
  list(forward = fwd, reverse = rev)
}, motif_hits, motif_hits_reverse, SIMPLIFY = FALSE)


result[[5]]


result_tibble <- tibble(
  peak_id = seq_along(result), # Create an identifier for each peak
  alignment = map_chr(result, function(res) {
    has_fwd <- length(res$forward) > 0
    has_rev <- length(res$reverse) > 0
    
    if (has_fwd && has_rev) "fwdrev"
    else if (has_fwd) "fwd"
    else if (has_rev) "rev"
    else "none"
  })
)

temp <- as_tibble(peak.CTCF) %>% bind_cols(result_tibble) %>%
  dplyr::select(seqnames, start, end, alignment)

fwrite(temp, here(refDir, "ChIP", "CTCF_ESC_unionPeaks_motifAnnotated.bed"), sep = "\t", col.names = FALSE)

```

#### Filtering convergent S-S loops
```{r}
commonLoopDir <- here("../data/loop_analysis")
figDir <- here("../figure/loop_analysis")

# IMPORT S-S LOOPS
allLoops <- as_tibble(fread(here(commonLoopDir, "loopScore_chromosight.tsv")))
strLoops <- allLoops %>% dplyr::filter(Anno %in% c("S-S", "SE-SE", "SP-SP", "SP-SE", "SE-S", "SP-S"))
temp <- strLoops %>% dplyr::select(c(1, 2, 3, 5, 6))  # Assuming columns include anchor positions
colnames(temp) <- c("chrom", "start1", "end1", "start2", "end2")
strLoopsGr1 <- makeGRangesFromDataFrame(
  temp %>% dplyr::select(chrom, start = start1, end = end1)
)
strLoopsGr2 <- makeGRangesFromDataFrame(
  temp %>% dplyr::select(chrom, start = start2, end = end2)
)

# IMPORT CTCF MOTIF ANNOTATION
ctcfMotifs <- fread(here(refDir, "ChIP", "CTCF_ESC_unionPeaks_motifAnnotated.bed"))
colnames(ctcfMotifs) <- c("chrom", "start", "end", "motif")
ctcfGr <- makeGRangesFromDataFrame(ctcfMotifs, keep.extra.columns = TRUE)

# Annotate each loop anchor with CTCF motifs
anchor1_with_ctcf <- findOverlaps(strLoopsGr1, ctcfGr)
anchor2_with_ctcf <- findOverlaps(strLoopsGr2, ctcfGr)

# Get motif information for anchor1 overlaps
anchor1_ctcf_motif <- rep(NA, length(strLoopsGr1))  # Initialize motif vector
anchor1_ctcf_motif[queryHits(anchor1_with_ctcf)] <- ctcfGr[subjectHits(anchor1_with_ctcf)]$motif

# Get motif information for anchor2 overlaps
anchor2_ctcf_motif <- rep(NA, length(strLoopsGr2))  # Initialize motif vector
anchor2_ctcf_motif[queryHits(anchor2_with_ctcf)] <- ctcfGr[subjectHits(anchor2_with_ctcf)]$motif

# Add motif information to strLoops
strLoops <- strLoops %>%
  mutate(
    anchor1_motif = anchor1_ctcf_motif,
    anchor2_motif = anchor2_ctcf_motif
  )

# Filter for convergent CTCF interactions (forward in anchor1, reverse in anchor2)
convergent_loops <- strLoops %>%
  filter(anchor1_motif %in% c("fwd", "fwdrev") &
           (anchor2_motif %in% c("rev", "fwdrev")))
```

#### Processing
```{r}
# Extract loop boundary from left to right
temp <- convergent_loops %>% dplyr::select(c(1, 2, 6))
colnames(temp) <- c("chrom", "start", "end")
strLoopsGr <- makeGRangesFromDataFrame(temp)

# IMPORT TSS
flankSize <- 10
gene.tb <- fread(here(refDir, "mm10", "mm10_GRCm38.p6_gene_sorted.bed")) %>%
  dplyr::mutate(TSS = ifelse(V4 == "+", V2, V3),
                TSSstart = TSS - flankSize,
                TSSend = TSS + flankSize) %>%
  dplyr::select(V1, TSSstart, TSSend, V6, V5)
colnames(gene.tb) <- c("chrom", "start", "end", "ensembl", "gene")
geneGr <- makeGRangesFromDataFrame(gene.tb, keep.extra.columns = TRUE)


# Find overlaps between loops and TSS
overlaps <- findOverlaps(geneGr, strLoopsGr)

# Annotate overlaps
tss_with_loops <- geneGr[queryHits(overlaps)]
loops_with_tss <- strLoopsGr[subjectHits(overlaps)]

# Combine into a data frame for processing
loop_data <- data.frame(gene = tss_with_loops$ensembl,
                        loop_chr = seqnames(loops_with_tss),
                        loop_start = start(loops_with_tss),
                        loop_end = end(loops_with_tss),
                        loop_width = width(loops_with_tss))

# Identify the largest loop for each gene
largest_loops <- loop_data[order(loop_data$gene, -loop_data$loop_width), ]
largest_loops <- largest_loops[!duplicated(largest_loops$gene), ]
largest_loops <- largest_loops %>% dplyr::mutate(loopID = paste(loop_chr, loop_start, loop_end, sep = "_"))
largest_loops <- largest_loops[order(largest_loops$loop_chr, largest_loops$loop_start),]
rownames(largest_loops) <- NULL

# Checking smaller loops
allLoopsGr <- makeGRangesFromDataFrame(allLoops, 
                                       seqnames.field = "chrom1", 
                                       start.field = "start1", 
                                       end.field = "end2")
largestLoopsGr <- makeGRangesFromDataFrame(largest_loops, 
                                           seqnames.field = "loop_chr", 
                                           start.field = "loop_start", 
                                           end.field = "loop_end")
largestLoopsGr <- sort(unique(largestLoopsGr))
complete_overlaps <- subsetByOverlaps(allLoopsGr, largestLoopsGr, type = "within")
overlaps <- findOverlaps(complete_overlaps, largestLoopsGr)


completeOverlapDf <- as.data.frame(complete_overlaps) %>% dplyr::mutate(loopIDAll = paste(seqnames, start, end, sep = "_"))
largestLoopsDf <- as.data.frame(largestLoopsGr) %>% dplyr::mutate(loopID = paste(seqnames, start, end, sep = "_"))

loopIDpairs <- data.frame(loopID = completeOverlapDf$loopIDAll[queryHits(overlaps)],
                          loopID2 = largestLoopsDf$loopID[subjectHits(overlaps)])

##################################
name <- "chromosight"
geneAnnoData <- loadLoopAnnoData(here(commonLoopDir, paste0("loopScore_", name, "_p-n_complex_ensemblList.tsv"))) %>% unnest(gene) %>%
  dplyr::mutate(loopID = paste(chrom1, start1, end2, sep = "_"))

temp <- geneAnnoData %>% dplyr::left_join(loopIDpairs, by = c("loopID"))
temp2 <- temp %>% dplyr::select(gene, loopID2) %>%
  group_by(gene) %>%              # Group data by the 'gene' column
  summarize(geneHasSS = !all(is.na(loopID2)))  # Check if all loopID2 values for each gene are NA
temp <- temp %>% left_join(temp2, by = c("gene"))
data.all <- temp %>% dplyr::mutate(group = ifelse(!geneHasSS, "No",
                                                  ifelse(is.na(loopID2), "Outside", "Within")),
                                   size = start2 - start1)

##################################
outfile <- data.all %>% dplyr::select(loopID2, gene) %>% dplyr::filter(!is.na(loopID2)) %>%
  separate(loopID2, into = c("chrom", "start", "end"), sep = "_", convert = TRUE)
fwrite(outfile, here(commonLoopDir, "convergent_ss_domain.bed"), sep = "\t", col.names = FALSE)
```

#### [FIG] Plotting
```{r}
data <- data.all %>% dplyr::filter(Anno %in% c("P-SE", "P-E", "E-SP", "SP-SE"))

temp <- data %>% dplyr::select(chrom1, start1, end1, chrom2, start2, end2, group) %>% distinct()

temp1 <- temp %>% dplyr::filter(group == "No") %>% dplyr::select(-group)
fwrite(temp1, here(commonLoopDir, "insulated_domain_ss_no.bedpe"), sep = "\t", col.names = FALSE)
temp2 <- temp %>% dplyr::filter(group == "Within") %>% dplyr::select(-group)
fwrite(temp2, here(commonLoopDir, "insulated_domain_ss_within.bedpe"), sep = "\t", col.names = FALSE)
temp3 <- temp %>% dplyr::filter(group == "Outside") %>% dplyr::select(-group)
fwrite(temp3, here(commonLoopDir, "insulated_domain_ss_outside.bedpe"), sep = "\t", col.names = FALSE)

temp <- data %>% dplyr::select(id, group, size) %>% distinct()


getPvalWilcox <- function(data, group1, group2){
  distance1 <- (data %>% dplyr::filter(group ==group1) )$size
  distance2 <- (data %>% dplyr::filter(group ==group2) )$size
  wil <- wilcox.test(distance1, distance2)
  return(wil$p.value)
}

convPvalue(getPvalWilcox(temp, "No", "Within"))
convPvalue(getPvalWilcox(temp, "No", "Outside"))
convPvalue(getPvalWilcox(temp, "Outside", "Within"))


p <- ggplot(temp, aes(x = group, y = size, fill = group)) + 
  geom_violin(linewidth = lineMedium* mmToLineUnit, lineend = "square", alpha = .4, , show.legend = FALSE) + 
  geom_boxplot(width = 0.3, color = "black",
               linewidth = lineMedium * mmToLineUnit, lineend = "square",
               outlier.shape = NA,  alpha = 0.6, show.legend = FALSE) + 
  theme_classic() + scale_y_continuous(labels = label_kb_mb) +
  scale_fill_manual(values = c("#777777", "#F28E2C", "#F28E2C")) +
  stat_summary(
    aes(group = group), fun = mean,
    geom = "point", shape = 21, size = 0.5,
    fill = "black", color = "black", position = position_dodge(.3)
  ) +
  theme(
    axis.title.x = element_blank(),
    axis.title.y = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),    
    axis.text.x = element_text(
      size = fontSizeM,
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) +labs(y = "Loop size")

fileName <- paste0("insulationBoundary_size_convSS_PE")
width <-33*mmToInch
height <- 35*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()


###################################


pseudo_ooe <- 0.01

data <- data %>% dplyr::filter(!is.na(ooe_DMSO), !is.na(ooe_dTAG)) %>%
  dplyr::mutate(
    ooe_DMSO = if_else(ooe_DMSO == 0, pseudo_ooe, ooe_DMSO),
    ooe_dTAG = if_else(ooe_dTAG == 0, pseudo_ooe, ooe_dTAG),
    fc_dTAG    = log2(ooe_dTAG / ooe_DMSO),
    diff_dTAG = score_dTAG - score_DMSO
  )

getPvalWilcox <- function(data, group1, group2){
  distance1 <- (data %>% dplyr::filter(group ==group1) )$diff_dTAG
  distance2 <- (data %>% dplyr::filter(group ==group2) )$diff_dTAG
  wil <- wilcox.test(distance1, distance2)
  return(wil$p.value)
}

temp <- data %>% dplyr::select(group, id, diff_dTAG) %>% distinct()

convPvalue(getPvalWilcox(temp, "No", "Within"))
convPvalue(getPvalWilcox(temp, "No", "Outside"))
convPvalue(getPvalWilcox(temp, "Outside", "Within"))

p <- ggplot(data, aes(x = group, y = diff_dTAG, fill = group)) + 
  geom_violin(linewidth = lineMedium * mmToLineUnit, lineend = "square", alpha = .4, , show.legend = FALSE) + 
  geom_boxplot(width = 0.3, color = "black",
               linewidth = lineMedium * mmToLineUnit, lineend = "square",
               outlier.shape = NA,  alpha = 0.6, show.legend = FALSE) + 
  theme_classic() +   
  scale_fill_manual(values = c("#777777", "#F28E2C", "#F28E2C")) +
  stat_summary(
    aes(group = group), fun = mean,
    geom = "point", shape = 21, size = 0.5,
    fill = "black", color = "black", position = position_dodge(.3)
  ) +
  theme(
    axis.title.x = element_blank(),
    axis.title.y = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),    
    axis.text.x = element_text(
      size = fontSizeM,
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) +labs(y = "Δ loop score") +    
  geom_hline(yintercept = 0,
             alpha = 1,
             color = "black",
             size = lineThick*mmToLineUnit,
             lineend = "square") + 
  geom_hline(yintercept = -0.2,
             alpha = 0.5, 
             color = "black",
             size = lineThick*mmToLineUnit,
             lineend = "square", linetype = "dashed") +
  coord_cartesian(ylim = c(-0.7, 0.5))

fileName <- paste0("insulationBoundary_delta_convSS_pe")
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()  


###################### 
convPvalue(getPvalWilcox(temp, "No", "Within"))
convPvalue(getPvalWilcox(temp, "No", "Outside"))
convPvalue(getPvalWilcox(temp, "Outside", "Within"))

temp <- data %>% dplyr::select(group, id, fc_dTAG) %>% distinct()

p <- ggplot(temp, aes(x = group, y = fc_dTAG, fill = group)) + 
  geom_violin(linewidth = lineMedium* mmToLineUnit, lineend = "square", alpha = .4, , show.legend = FALSE) + 
  geom_boxplot(width = 0.3, color = "black",
               linewidth = lineMedium * mmToLineUnit, lineend = "square",
               outlier.shape = NA,  alpha = 0.6, show.legend = FALSE) + 
  theme_classic() + 
  scale_fill_manual(values = c("#777777", "#F28E2C", "#F28E2C")) +
  
  stat_summary(
    aes(group = group), fun = mean,
    geom = "point", shape = 21, size = 0.5,
    fill = "black", color = "black", position = position_dodge(.3)
  ) +
  theme(
    axis.title.x = element_blank(),
    axis.title.y = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),    
    axis.text.x = element_text(
      size = fontSizeM,
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) +labs(y = "log2(fc of obs/exp)") +    
  geom_hline(yintercept = 0,
             alpha = 1,
             color = "black",
             size = lineThick*mmToLineUnit,
             lineend = "square") + 
  coord_cartesian(ylim = c(-2.5, 2))
width <-30*mmToInch
height <- 35*mmToInch
fileName <- paste0("insulationBoundary_obsexp_convSS_pe")
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()  





#####################

data <- data.all
gene.withinBoundary <- unique((data %>% dplyr::filter(group == "Within"))$gene)
gene.outsideBoundary <- unique((data %>% dplyr::filter(group == "Outside"))$gene)
gene.noBoundary <- unique((data %>% dplyr::filter(group == "No"))$gene)

gene.wBoundary <- unique(c(gene.withinBoundary, gene.outsideBoundary))

dataDir <- here("..", "data")
gene.group1 <- fread(here(dataDir, "loop_analysis",
                          "geneList_chromosight_dTAG_vs_DMSO_p-n_complex_Group1.tsv"))$gene
gene.group2 <- fread(here(dataDir, "loop_analysis",
                          "geneList_chromosight_dTAG_vs_DMSO_p-n_complex_Group2.tsv"))$gene


print("perc of gene with loop within Boundary")
sum(gene.group1 %in% gene.withinBoundary)/length(gene.group1)*100
sum(gene.group2 %in% gene.withinBoundary)/length(gene.group2)*100

print("perc of gene with loop outside Boundary")
sum(gene.group1 %in% gene.outsideBoundary)/length(gene.group1)*100
sum(gene.group2 %in% gene.outsideBoundary)/length(gene.group2)*100

print("perc of gene with loop with no Boundary")
sum(gene.group1 %in% gene.noBoundary)/length(gene.group1)*100
sum(gene.group2 %in% gene.noBoundary)/length(gene.group2)*100


diff.RNA <- fread(here(dataDir, "RNA_diff", "diff_G1.dTAG_G1.2i.dTAG_vs_G1.2i.DMSO.tsv")) %>%
  dplyr::select(ensembl_gene_id, log2FoldChange, shrinked_log2FC, padj, external_gene_name) %>%
  dplyr::mutate(group = case_when(
    ensembl_gene_id %in% gene.wBoundary ~ "withinBoundary",
    ensembl_gene_id %in% gene.noBoundary ~ "noBoundary",
    TRUE ~ NA
  )) %>%
  dplyr::filter(!is.na(group)) %>%
  dplyr::mutate(absLog2FC = abs(log2FoldChange))
ks_result <- ks.test(
  diff.RNA %>% dplyr::filter(group == "withinBoundary") %>% pull(absLog2FC),
  diff.RNA %>% dplyr::filter(group == "noBoundary") %>% pull(absLog2FC)
)

# ggplot(diff.RNA, aes(x = group, y = log2FoldChange)) + geom_violin() + geom_boxplot(width = 0.1)

# getPvalWilcox <- function(data, group1, group2){
#   distance1 <- (data %>% dplyr::filter(group ==group1) )$absLog2FC
#   distance2 <- (data %>% dplyr::filter(group ==group2) )$absLog2FC
#   wil <- wilcox.test(distance1, distance2)
#   return(wil$p.value)
# }
# convPvalue(getPvalWilcox(diff.RNA, "withBoundary", "noBoundary"))
# 

# Create the CDF plot
p <- ggplot(diff.RNA, aes(x = absLog2FC, color = group)) +
  scale_color_manual(values = (c("#777777", "#F28E2C"))) +
  
  stat_ecdf(size = 0.4, linewidth = lineMedium * mmToLineUnit, lineend = "square"  ) + # Use stat_ecdf to plot the empirical CDF
  labs(
    x = "Abs. log2(fold change)",
    y = "Cumulative Probability"
  ) + coord_cartesian(xlim = c(0, 1)) +
  theme_classic() + # Clean theme
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.position = "none",
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  )

fileName <- paste0("log2FC_cdf_insulationBoundary_convSS_pe")
width <- 33*mmToInch
height <-33*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()
```


## [5] Loop + Transcription (Chromosight) EpiLC
### [FIG] RNA-seq loopscore scatterplot

```{r}
alpha <- 0.05
fcCutoff <- 0.5
diffCutoff <- 0.2
commonLoopDir <- here("../data/loop_analysis")
diffDir <- here("../data/RNA_diff")
name <- "chromosight"

geneAnnoData <- loadLoopAnnoData(here(commonLoopDir, paste0("loopScore_", name, "_p-n_complex_ensemblList.tsv"))) %>%
  dplyr::mutate(diff = score_EpidTAG - score_EpiDMSO)

diff.RNA <- fread(here(diffDir, "diff_G1.dTAG_G1.Epi.dTAG_vs_G1.Epi.DMSO.tsv")) %>%
  dplyr::select(ensembl_gene_id, log2FoldChange, shrinked_log2FC, padj, external_gene_name)
geneList.down.RNA <- (diff.RNA %>% dplyr::filter(shrinked_log2FC <= 0, padj < alpha))$ensembl_gene_id
geneList.up.RNA <- (diff.RNA %>% dplyr::filter(shrinked_log2FC > 0, padj < alpha))$ensembl_gene_id


maxLog2FC <- 2

temp <- geneAnnoData %>% dplyr::select(diff, gene) %>% 
  unnest(gene) %>% group_by(gene) %>%
  summarize(mean_diff_score = mean(diff), .groups = 'drop')

temp <- left_join(temp, diff.RNA, by = c("gene" = "ensembl_gene_id")) %>% 
  dplyr::mutate(flag = ifelse((gene %in% geneList.down.RNA) & (mean_diff_score < 0) , "Group1",
                              ifelse((mean_diff_score < 0) & !(gene %in% c(geneList.up.RNA, geneList.down.RNA)), "Group2",
                                     "Group3")),
                maxFlag = (abs(shrinked_log2FC) > maxLog2FC),
                shrlog2fcMax = pmax(pmin(shrinked_log2FC, maxLog2FC), -maxLog2FC)) %>% 
  dplyr::arrange(flag)

dataDir <- here("..", "data", "loop_analysis")
fwrite(temp %>% dplyr::filter(flag == "Group1"), here(dataDir, paste0("geneList_", name, "_EpidTAG_vs_EpiDMSO_p-n_complex_Group1.tsv")), sep = "\t")
fwrite(temp %>% dplyr::filter(flag == "Group2"), here(dataDir, paste0("geneList_", name, "_EpidTAG_vs_EpiDMSO_p-n_complex_Group2.tsv")), sep = "\t")

p <- ggplot(temp, aes(x = shrlog2fcMax, y = mean_diff_score, color = flag,
                      label = ifelse(flag == "Group1", external_gene_name, NA),
                      shape = as.factor(maxFlag))) +
  geom_point(size = 1) + 
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = 0) +
  geom_text_repel(size = fontSizeS/2.845, family = fontType) + theme_classic() +
  
  scale_color_manual(values = c("Group1" = strong_orange, "Group2" = no_grey, "Group3" = "grey90")) +  # Corrected color mapping
  scale_shape_manual(values = c("TRUE" = 2, "FALSE" = 19)) +
  theme(
    legend.position = "none",
    plot.title = element_text(
      hjust = 0.5,
      size = fontSizeS,
      family = fontType
    ),
    axis.title = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
  ) +
  labs(y = "Avg. Δ loop score", x = "log2(shrunken FC)")

fileName <- paste0("RNAvsLoop_", name, "_score_complexAnno_EpiLC")
width <- panelSize(2.2)*mmToInch
height <- panelSize(1.7)*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()



```

### Link TAD to gene
```{r}
dataDir <- here("../data")

# Importing gene
gene.TSS.tb <- fread(here(refDir, "mm10", "mm10_GRCm38.p6_gene_sorted.bed")) %>%
  dplyr::mutate(TSS = ifelse(V4 == "+", V2, V3)) %>%
  dplyr::select(V6, V5, V1, TSS)
colnames(gene.TSS.tb) <- c("ensembl", "gene", "chr", "TSS")

# Importing TAD boundary with insulation score for each sample
## Boundaries are sized 25kb
tad_boundary <- fread(here(dataDir, "TAD", "TAD_25kb_125kb_otsu_boundaries_G1DMSO_pooled.bed"))
colnames(tad_boundary) <- c("chr", "start", "end")
tad_boundary <- tad_boundary %>% dplyr::mutate(boundaryID = paste(chr, start, end, sep = "_"),
                                               center = (start + end)/2,
                                               size = end - start)
insScore.DMSO <- fread(here(dataDir, "TAD", "insulationScore_25000bp_EpiG1DMSO_pooled.tsv")) %>%
  dplyr::mutate(binID = paste(chrom, start, end, sep = "_")) %>%
  dplyr::select(binID, log2InsScore_DMSO = log2_insulation_score_125000)
insScore.dTAG <- fread(here(dataDir, "TAD", "insulationScore_25000bp_G1dTAG_pooled.tsv")) %>%
  dplyr::mutate(binID = paste(chrom, start, end, sep = "_")) %>%
  dplyr::select(binID, log2InsScore_dTAG = log2_insulation_score_125000)
tad_boundary <- list(insScore.DMSO, insScore.dTAG) %>%
  purrr::reduce(
    left_join,
    .init = tad_boundary,
    by   = c("boundaryID" = "binID")
  )
tad_boundary <- tad_boundary %>%
  mutate(negDiffInsScore = (2^log2InsScore_DMSO) - (2^log2InsScore_dTAG))


# Importing TAD
tad <- fread(here(dataDir, "TAD",
                  "TAD_25kb_125kb_otsu_G1DMSO_pooled.bedpe"))
colnames(tad) <- c("chr1", "start1", "end1", "chr2", "start2", "end2")
tad <- tad %>% dplyr::mutate(tadID = paste(chr1, start1, end1, sep = "_"),
                             tadSize = end1 - start1)

################################################################################
# Functions to link gene to TADs
findItsTAD <- function(chrom, TSS, tad){
  temp <- tad %>% dplyr::filter(chr1 == chrom,
                                start1 < TSS,
                                end1 > TSS)
  if(nrow(temp) == 1){
    return(temp$tadID)
  }else{
    return(NA)
  }
}
findClosestTADBoundary <- function(chrom, TSS, tad_boundary){
  temp <- tad_boundary %>% dplyr::filter(chr == chrom) %>%
    dplyr::mutate(distance = abs(center - TSS)) %>%
    slice_min(distance)
  if(nrow(temp) == 1){
    return(temp$boundaryID)
  }else{
    return(NA)
  }
}

gene.TSS.tb <- gene.TSS.tb %>% rowwise() %>%
  dplyr::mutate(tadID = findItsTAD(chr, TSS, tad),
                boundaryID = findClosestTADBoundary(chr, TSS, tad_boundary))



data <- gene.TSS.tb %>% left_join(tad %>% dplyr::select(tadID, tadSize), by = c("tadID")) %>%
  left_join(tad_boundary %>% dplyr::select(c(4, 5, 7, 8, 9)), by = c("boundaryID")) %>%
  mutate(distanceToBoundary = abs(center - TSS))

```
#### [FIG] TAD distance/size/insulation score
```{r}
group1 <- fread(here(dataDir, "loop_analysis",
                     "geneList_chromosight_EpidTAG_vs_EpiDMSO_p-n_complex_Group1.tsv"))$gene
group2 <- fread(here(dataDir, "loop_analysis",
                     "geneList_chromosight_EpidTAG_vs_EpiDMSO_p-n_complex_Group2.tsv"))$gene


dataToPlot <- data %>% rowwise() %>% dplyr::mutate(
  group = ifelse(ensembl %in% group1, "group1",
                 ifelse(ensembl %in% group2, "group2", NA))) %>% 
  dplyr::filter(!is.na(group)) %>%
  ungroup()


## Plot distance
getPvalWilcox <- function(data, group1, group2){
  distance1 <- (data %>% dplyr::filter(group == group1) )$distanceToBoundary
  distance2 <- (data %>% dplyr::filter(group == group2) )$distanceToBoundary
  wil <- wilcox.test(distance1, distance2)
  return(wil$p.value)
}


plot_TADdistance <- function(temp.tb, note, ymin = 0, ymax = 2000000){
  p12 <- round(getPvalWilcox(temp.tb, "group1", "group2"), 5)
  
  p <- ggplot(temp.tb, aes(x = ".", fill = group, y = distanceToBoundary)) +
    # Set axis labels (no x-axis title)
    scale_fill_manual(values = ((rev(c("#777777", "#F28E2C"))))) +
    labs(x = NULL, y = "Distance from TAD boundary") +
    # Violin plot with black outline, customized line width and end
    introdataviz::geom_split_violin(
      color = "black", alpha = 0.4,
      linewidth = lineMedium * mmToLineUnit, lineend = "square"
    ) +
    # Box plot with customized line width, square end, and outlier size
    geom_boxplot(
      width = 0.3, color = "black", alpha = 0.6,
      linewidth = lineMedium * mmToLineUnit, lineend = "square",
      outlier.shape = NA,  alpha = 0.6, show.legend = FALSE
    ) +
    
    # Mean point in each group
    stat_summary(
      aes(group = group), fun = mean,
      geom = "point", shape = 21, size = 0.5,
      fill = "black", color = "black", position = position_dodge(.3)
    ) +
    
    # Horizontal line at y = 0
    geom_hline(yintercept = 0, linewidth = lineThick * mmToLineUnit) +
    
    # Coordinate limits and custom y-axis labels
    coord_cartesian(ylim = c(ymin, ymax)) +
    scale_y_continuous(labels = label_kb_mb) +
    
    # Annotate p-value text
    annotate(
      "text", x = 1, y = ymin + 1,
      label = paste0("p12: ", convPvalue(p12)),
      color = "black", hjust = 0, size = 1
    ) +
    
    # Theme customization
    theme_classic() +
    theme(
      axis.title = element_text(
        size = fontSizeM, family = fontType, color = "#000000"
      ),
      axis.text = element_text(
        size = fontSizeS, family = fontType, color = "#000000"
      ),
      axis.text.x = element_text(
        angle = 45, hjust = 1, vjust = 1
      ),
      axis.line = element_line(
        color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
      ),
      axis.ticks = element_line(
        color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
      ),
      panel.background = element_rect(fill = "transparent"),
      legend.text = element_text(family = fontType, size = fontSizeS),
      legend.title = element_text(family = fontType, size = fontSizeS)) +
    guides(
      fill = guide_legend(
        keywidth = 0.2,  # Adjust the width of the legend keys
        keyheight = 0.2  # Adjust the height of the legend keys
      )
    )
  
  
  fileName <- paste0("TAD_distance_to_boundary_", note)
  width <- panelSize(1.55)*mmToInch
  height <- panelSize(1.2)*mmToInch
  png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
  print(p)
  dev.off()
  svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
  print(p)
  dev.off()
}

plot_TADdistance(dataToPlot, "group_binayGroup_EpiLC", ymax = 1000000)


## Plot TAD size
getPvalWilcox <- function(data, group1, group2){
  distance1 <- (data %>% dplyr::filter(group == group1) )$tadSize
  distance2 <- (data %>% dplyr::filter(group == group2) )$tadSize
  wil <- wilcox.test(distance1, distance2)
  return(wil$p.value)
}


plot_TADsize <- function(temp.tb, note, ymin = 0, ymax = 2000000){
  p12 <- round(getPvalWilcox(temp.tb, "group1", "group2"), 5)
  
  p <- ggplot(temp.tb, aes(x = ".", fill = group, y = tadSize)) +
    # Set axis labels (no x-axis title)
    labs(x = NULL, y = "Size of TAD") +
    scale_fill_manual(values = ((rev(c("#777777", "#F28E2C"))))) +
    
    # Violin plot with black outline, customized line width and end
    introdataviz::geom_split_violin(
      color = "black",
      linewidth = lineMedium* mmToLineUnit, lineend = "square", alpha = .4
    ) +
    
    # Box plot with customized line width, square end, and outlier size
    geom_boxplot(
      width = 0.3, color = "black",
      linewidth = lineMedium* mmToLineUnit, lineend = "square",
      outlier.shape = NA,  alpha = 0.6, show.legend = FALSE
    ) +
    
    # Mean point in each group
    stat_summary(
      aes(group = group), fun = mean,
      geom = "point", shape = 21, size = 0.5,
      fill = "black", color = "black", position = position_dodge(.3)
    ) +
    
    # Horizontal line at y = 0
    geom_hline(yintercept = 0, linewidth = lineThick * mmToLineUnit) +
    
    # Coordinate limits and custom y-axis labels
    coord_cartesian(ylim = c(ymin, ymax)) +
    scale_y_continuous(labels = label_kb_mb) +
    
    # Annotate p-value text
    annotate(
      "text", x = 1, y = ymin + 1,
      label = paste0("p12: ", convPvalue(p12)),
      color = "black", hjust = 0, size = 1
    ) +
    
    # Theme customization
    theme_classic() +
    theme(
      axis.title = element_text(
        size = fontSizeM, family = fontType, color = "#000000"
      ),
      axis.text = element_text(
        size = fontSizeS, family = fontType, color = "#000000"
      ),
      axis.text.x = element_text(
        angle = 45, hjust = 1, vjust = 1
      ),
      axis.line = element_line(
        color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
      ),
      axis.ticks = element_line(
        color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
      ),
      panel.background = element_rect(fill = "transparent"),
      legend.text = element_text(family = fontType, size = fontSizeS),
      legend.title = element_text(family = fontType, size = fontSizeS)
    ) + guides(
      fill = guide_legend(
        keywidth = 0.2,  # Adjust the width of the legend keys
        keyheight = 0.2  # Adjust the height of the legend keys
      ))
  
  
  
  fileName <- paste0("TAD_size_", note)
  width <- panelSize(1.55)*mmToInch
  height <- panelSize(1.2)*mmToInch
  png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
  print(p)
  dev.off()
  svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
  print(p)
  dev.off()
}
plot_TADsize(dataToPlot, "group_binaryGroup_EpiLC", ymax = 3e6)


getPvalWilcox <- function(data, group1, group2){
  distance1 <- (data %>% dplyr::filter(group ==group1) )$negDiffInsScore
  distance2 <- (data %>% dplyr::filter(group ==group2) )$negDiffInsScore
  wil <- wilcox.test(distance1, distance2)
  return(wil$p.value)
}


plot_insScore <- function(temp.tb, note, ymin = -1.5, ymax = 0){
  p12 <- round(getPvalWilcox(temp.tb, "group1", "group2"), 5)
  
  p <- ggplot(temp.tb, aes(x = ".", fill = group, y = negDiffInsScore)) + 
    labs(x = NULL, y = "- Insulation score") +  # Remove x-axis title
    scale_fill_manual(values = rev((c("#777777", "#F28E2C")))) +
    
    introdataviz::geom_split_violin(linewidth = lineMedium * mmToLineUnit, lineend = "square",
                                    alpha = .4) +
    geom_boxplot(width = 0.3, color = "black",
                 linewidth = lineMedium * mmToLineUnit, lineend = "square",
                 outlier.shape = NA,  alpha = 0.6, show.legend = FALSE) + theme_classic() +
    stat_summary(
      aes(group = group), fun = mean,
      geom = "point", shape = 21, size = 0.5,
      fill = "black", color = "black", position = position_dodge(.3)
    ) +
    geom_hline(yintercept = 0, linewidth = lineThick*mmToLineUnit) +
    theme_classic() + coord_cartesian(ylim = c(ymin, ymax)) +
    annotate("text", x = 1, y = ymin + 0.1, label = paste0("p12: ", convPvalue(p12)),
             color = "black", hjust = 0, size = 1) +
    theme(
      axis.title = element_text(
        size = fontSizeM,
        family = fontType,
        color = "#000000"
      ),
      axis.text = element_text(
        size = fontSizeM,
        family = fontType,
        color = "#000000"
      ),
      axis.text.x = element_text(
        angle = 45,      # Rotate x-axis labels 45 degrees
        hjust = 1,       # Adjust horizontal justification
        vjust = 1        # Adjust vertical justification
      ),
      axis.line = element_line(
        color = "#000000",
        size = lineThick*mmToLineUnit,
        lineend = "square"
      ),
      axis.ticks = element_line(
        color = "#000000",
        size = lineThick*mmToLineUnit,
        lineend = "square"
      ),
      panel.background = element_rect(fill = "transparent"),
      legend.text = element_text(family = fontType, size = fontSizeS),
      legend.title = element_text(family = fontType, size = fontSizeS)
    )+   guides(
      fill = guide_legend(
        keywidth = 0.2,  # Adjust the width of the legend keys
        keyheight = 0.2  # Adjust the height of the legend keys
      ))
  
  
  fileName <- paste0("TAD_insulation_score_binaryGroup_", note)
  width <- panelSize(1.55)*mmToInch
  height <- panelSize(1.2)*mmToInch
  png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
  print(p)
  dev.off()
  svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
  print(p)
  dev.off()
}

# DMSO
plot_insScore(dataToPlot, "nearestBoundary_diff_EpiLCdTAGvsEpiLCDMSO_binaryGroup", ymin = -0.3, ymax = 0.2)


getPvalWilcox <- function(data, group1, group2){
  distance1 <- (data %>% dplyr::filter(group ==group1) )$negInsScore_DMSO
  distance2 <- (data %>% dplyr::filter(group ==group2) )$negInsScore_DMSO
  wil <- wilcox.test(distance1, distance2)
  return(wil$p.value)
}

plot_insScoreSingle <- function(temp.tb, note, ymin = -1.5, ymax = 0){
  p12 <- round(getPvalWilcox(temp.tb, "group1", "group2"), 5)
  
  p <- ggplot(temp.tb, aes(x = ".", fill = group, y =  negInsScore_DMSO)) + 
    labs(x = NULL, y = "- Insulation score") +  # Remove x-axis title
    scale_fill_manual(values = rev((c("#777777", "#F28E2C")))) +
    
    introdataviz::geom_split_violin(linewidth = lineMedium * mmToLineUnit, lineend = "square",
                                    alpha = .4) +
    geom_boxplot(width = 0.3, color = "black",
                 linewidth = lineMedium * mmToLineUnit, lineend = "square",
                 outlier.shape = NA,  alpha = 0.6, show.legend = FALSE) + theme_classic() +
    stat_summary(
      aes(group = group), fun = mean,
      geom = "point", shape = 21, size = 0.5,
      fill = "black", color = "black", position = position_dodge(.3)
    ) +
    geom_hline(yintercept = 0, linewidth = lineThick*mmToLineUnit) +
    theme_classic() + coord_cartesian(ylim = c(ymin, ymax)) +
    annotate("text", x = 1, y = ymin + 0.1, label = paste0("p12: ", convPvalue(p12)),
             color = "black", hjust = 0, size = 1) +
    theme(
      axis.title = element_text(
        size = fontSizeM,
        family = fontType,
        color = "#000000"
      ),
      axis.text = element_text(
        size = fontSizeM,
        family = fontType,
        color = "#000000"
      ),
      axis.text.x = element_text(
        angle = 45,      # Rotate x-axis labels 45 degrees
        hjust = 1,       # Adjust horizontal justification
        vjust = 1        # Adjust vertical justification
      ),
      axis.line = element_line(
        color = "#000000",
        size = lineThick*mmToLineUnit,
        lineend = "square"
      ),
      axis.ticks = element_line(
        color = "#000000",
        size = lineThick*mmToLineUnit,
        lineend = "square"
      ),
      panel.background = element_rect(fill = "transparent"),
      legend.text = element_text(family = fontType, size = fontSizeS),
      legend.title = element_text(family = fontType, size = fontSizeS)
    )+   guides(
      fill = guide_legend(
        keywidth = 0.2,  # Adjust the width of the legend keys
        keyheight = 0.2  # Adjust the height of the legend keys
      ))
  
  
  fileName <- paste0("TAD_insulation_score_binaryGroup_", note)
  width <- panelSize(1.55)*mmToInch
  height <- panelSize(1.2)*mmToInch
  png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
  print(p)
  dev.off()
  svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
  print(p)
  dev.off()
}

dataToPlot <- dataToPlot %>% mutate(negInsScore_DMSO = -(2^log2InsScore_DMSO))
plot_insScoreSingle(dataToPlot, "nearestBoundary_DMSO_EpidTAGvsEpiDMSO_binaryGroup", ymin = -1, ymax = 0)


```


#### [FIG] TAD gene density
##### Making TAD gene count
```{r}
# Counting gene number for each TAD
tadGeneCount <- data %>% rowwise() %>%
  drop_na(tadID) %>%
  group_by(tadID) %>%
  summarize(tadGeneCount = n())

# Merging with data
dataToPlot <- data %>% left_join(tadGeneCount, by = c("tadID")) %>%
  mutate(tadGeneDensity = tadGeneCount/tadSize*100e3)


group1 <- fread(here(dataDir, "loop_analysis",
                     "geneList_chromosight_EpidTAG_vs_EpiDMSO_p-n_complex_Group1.tsv"))$gene
group2 <- fread(here(dataDir, "loop_analysis",
                     "geneList_chromosight_EpidTAG_vs_EpiDMSO_p-n_complex_Group2.tsv"))$gene
```
##### Plot gene density
```{r}
dataToPlot2 <- dataToPlot %>% rowwise() %>% dplyr::mutate(
  group = ifelse(ensembl %in% group1, "group1",
                 ifelse(ensembl %in% group2, "group2", NA))) %>% 
  dplyr::filter(!is.na(group))


## geneDensity
getPvalWilcox <- function(data, group1, group2){
  distance1 <- (data %>% dplyr::filter(group == group1) )$tadGeneDensity
  distance2 <- (data %>% dplyr::filter(group == group2) )$tadGeneDensity
  wil <- wilcox.test(distance1, distance2)
  return(wil$p.value)
}


plot_geneDensity <- function(temp.tb, note, ymin = 0, ymax = 10){
  p12 <- round(getPvalWilcox(temp.tb, "group1", "group2"), 5)
  
  p <- ggplot(temp.tb, aes(x = ".", fill = group, y = tadGeneDensity)) +
    # Set axis labels (no x-axis title)
    labs(x = NULL, y = "Gene density within TAD") +
    scale_fill_manual(values = (rev(c("#777777","#F28E2C")))) +
    # Violin plot with black outline, customized line width and end
    introdataviz::geom_split_violin(linewidth = lineMedium * mmToLineUnit, lineend = "square",
                                    alpha = .4
    ) +
    
    # Box plot with customized line width, square end, and outlier size
    geom_boxplot(
      width = 0.3, color = "black",
      linewidth = lineMedium * mmToLineUnit, lineend = "square",
      outlier.shape = NA,  alpha = 0.6, show.legend = FALSE
    ) +
    
    # Mean point in each group
    stat_summary(
      aes(group = group), fun = mean,
      geom = "point", shape = 21, size = 0.5,
      fill = "black", color = "black", position = position_dodge(.3)
    ) +
    
    # Horizontal line at y = 0
    geom_hline(yintercept = 0, linewidth = lineThick * mmToLineUnit) +
    
    # Coordinate limits and custom y-axis labels
    coord_cartesian(ylim = c(ymin, ymax)) +
    
    # Annotate p-value text
    annotate(
      "text", x = 1, y = ymin + 1,
      label = paste0("p12: ", convPvalue(p12)),
      color = "black", hjust = 0, size = 1
    ) +
    
    # Theme customization
    theme_classic() +
    theme(
      axis.title = element_text(
        size = fontSizeM, family = fontType, color = "#000000"
      ),
      axis.text = element_text(
        size = fontSizeS, family = fontType, color = "#000000"
      ),
      axis.text.x = element_text(
        angle = 45, hjust = 1, vjust = 1
      ),
      axis.line = element_line(
        color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
      ),
      axis.ticks = element_line(
        color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
      ),
      panel.background = element_rect(fill = "transparent"),
      legend.text = element_text(family = fontType, size = fontSizeS),
      legend.title = element_text(family = fontType, size = fontSizeS)
    )+   
    guides(
      fill = guide_legend(
        keywidth = 0.2,  # Adjust the width of the legend keys
        keyheight = 0.2  # Adjust the height of the legend keys
      ))
  
  
  
  fileName <- paste0("TAD_geneDensity_", note)
  width <- panelSize(1.5)*mmToInch
  height <- panelSize(1.2)*mmToInch
  png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
  print(p)
  dev.off()
  svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
  print(p)
  dev.off()
}
plot_geneDensity(dataToPlot2, "group_binaryGroup_EpiLC", ymax = 8)

```

##### Plot TAD avg log2FC & CDF
```{r}
diffDir <- here(dataDir, "RNA_diff")
diff.RNA <- fread(here(diffDir, "diff_G1.dTAG_G1.Epi.dTAG_vs_G1.Epi.DMSO.tsv")) %>%
  dplyr::select(ensembl_gene_id, log2FoldChange, shrinked_log2FC, padj)

temp <- left_join(dataToPlot, diff.RNA, by = c("ensembl" = "ensembl_gene_id")) %>% dplyr::filter(!is.na(log2FoldChange),
                                                                                                 !is.na(tadID))


dataToPlot2 <- temp %>% group_by(tadID) %>%
  dplyr::summarise(absAvgLog2FC = abs(mean(log2FoldChange)),
                   avgLog2FC = mean(log2FoldChange),
                   tadGeneDensity = mean(tadGeneDensity))

dataToPlot2$geneDensityGroup <- cut(
  dataToPlot2$tadGeneDensity,
  breaks = quantile(dataToPlot2$tadGeneDensity, probs = seq(0, 1, 0.2), na.rm = TRUE),
  include.lowest = TRUE,
  labels = paste0(seq(0, 80, 20), "-", seq(20, 100, 20), "%")
)



### Ploting  grouping

p <- ggplot(dataToPlot2, aes(x = geneDensityGroup, y = tadGeneDensity, fill = geneDensityGroup)) + 
  geom_boxplot(color = "black",
               linewidth = lineThick * mmToLineUnit, lineend = "square",
               outlier.shape = NA) + theme_classic() +
  coord_cartesian(ylim = c(0, quantile(dataToPlot2$tadGeneDensity, 0.99))) +
  labs(x = "Gene density group" , y = "TAD gene density") +
  scale_fill_manual(values = rev(c("#777777", "#8B7E65", "#A28452", "#C2884D", "#F28E2C"))) +
  theme(
    legend.position = "none",
    axis.title = element_text(
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS, family = fontType, color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45, hjust = 1, vjust = 1
    ),
    axis.line = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  )

fileName <- paste0("TAD_tadGroup_geneDensity_EpiLC")
width <- panelSize(1.2)*mmToInch
height <- panelSize(1.7)*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()


### CDF plot

group1_data <- dataToPlot2$absAvgLog2FC[dataToPlot2$geneDensityGroup == "0-20%"]
group2_data <- dataToPlot2$absAvgLog2FC[dataToPlot2$geneDensityGroup == "20-40%"]
ks.test(group1_data, group2_data)

group1_data <- dataToPlot2$absAvgLog2FC[dataToPlot2$geneDensityGroup == "20-40%"]
group2_data <- dataToPlot2$absAvgLog2FC[dataToPlot2$geneDensityGroup == "40-60%"]
ks.test(group1_data, group2_data)

group1_data <- dataToPlot2$absAvgLog2FC[dataToPlot2$geneDensityGroup == "40-60%"]
group2_data <- dataToPlot2$absAvgLog2FC[dataToPlot2$geneDensityGroup == "60-80%"]
ks.test(group1_data, group2_data)

group1_data <- dataToPlot2$absAvgLog2FC[dataToPlot2$geneDensityGroup == "60-80%"]
group2_data <- dataToPlot2$absAvgLog2FC[dataToPlot2$geneDensityGroup == "80-100%"]
ks.test(group1_data, group2_data)

p <- ggplot(dataToPlot2, aes(x = absAvgLog2FC, color = geneDensityGroup)) +
  scale_color_manual(values = rev(c("#777777", "#8B7E65", "#A28452", "#C2884D", "#F28E2C"))) +
  stat_ecdf(size = 0.4, linewidth = lineMedium * mmToLineUnit, lineend = "square" ) + # Use stat_ecdf to plot the empirical CDF
  labs(
    x = "Abs. log2(fold change)",
    y = "Cumulative Probability"
  ) + coord_cartesian(xlim = c(0, 1.5)) +
  theme_classic() + # Clean theme
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.position = "none",
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS),
  ) + scale_y_continuous(labels = scales::number_format(accuracy = 0.1))


fileName <- paste0("TAD_tadGroup_geneDensity_EpiLC_avgLog2FC_cdf")
width <- 33*mmToInch
height <-33*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()




```
### ChIP peak density
#### [FIG] H3K4me3 peak number per 10kb
```{r}
process_group <- function(group_name, gene_data, bins) {
  # Filter for the specific group
  TSS_group <- makeGRangesFromDataFrame(
    gene_data %>% filter(ensembl %in% get(group_name)), 
    keep.extra.columns = TRUE
  )
  
  # Process each TSS in the group
  all_results <- map_dfr(seq_len(length(TSS_group)), function(i) {
    # Find overlapping bins
    overlapping_bins <- findOverlaps(TSS_group[i], bins)
    bins_near_tss <- as_tibble(bins[subjectHits(overlapping_bins)])
    
    # Calculate distance bin
    TSScenter <- (as_tibble(TSS_group[i]) %>%
                    mutate(center = (start + end) / 2))$center
    centerBinStart <- floor(TSScenter / 10e3) * 10e3 + 1
    bins_near_tss <- bins_near_tss %>%
      mutate(distanceBin = abs((start - centerBinStart)) / 10e3)
    
    # Summarize results
    result <- bins_near_tss %>%
      group_by(distanceBin) %>%
      summarise(mean_peak_counts = mean(peak_counts, na.rm = TRUE), .groups = "drop") %>%
      mutate(gene = TSS_group[i]$ensembl)
    return(result)
  })
  
  # Calculate mean and SD for the group
  mean_data <- all_results %>%
    group_by(distanceBin) %>%
    summarise(mean_peak_counts = mean(mean_peak_counts, na.rm = TRUE))
  
  sd_data <- all_results %>%
    group_by(distanceBin) %>%
    summarise(sd_peak_counts = sd(mean_peak_counts, na.rm = TRUE))
  
  # Join and add group name
  summary_data <- left_join(mean_data, sd_data, by = "distanceBin") %>%
    mutate(group = group_name)
  
  return(summary_data)
}

##########################################################


## Import peak
peak <- importPeak(here(refDir, "ChIP", "H3K4me3", "H3K4me3_ESC_unionPeaks.bed"))

## Import 10kb bin
temp <- fread(here(refDir, "mm10", "mm10.bin.10kb.bed")) %>% dplyr::select(c(1, 2, 3))
colnames(temp) <- c("chr", "start", "end")
temp <- temp %>% dplyr::mutate(start = start+1)
bins.10kb <- makeGRangesFromDataFrame(temp)

## Count overlap
counts <- countOverlaps(bins.10kb, peak)
mcols(bins.10kb)$peak_counts <- counts

## Getting TSS
flankSize <- 1e6
gene.tb <- fread(here(refDir, "mm10", "mm10_GRCm38.p6_gene_sorted.bed")) %>%
  dplyr::mutate(TSS = ifelse(V4 == "+", V2, V3),
                TSSstart = TSS - flankSize,
                TSSend = TSS + flankSize) %>%
  dplyr::select(V1, TSSstart, TSSend, V6)
colnames(gene.tb) <- c("chr", "start", "end", "ensembl")


group1 <- fread(here(dataDir, "loop_analysis",
                     "geneList_chromosight_EpidTAG_vs_EpiDMSO_p-n_complex_Group1.tsv"))$gene
group2 <- fread(here(dataDir, "loop_analysis",
                     "geneList_chromosight_EpidTAG_vs_EpiDMSO_p-n_complex_Group2.tsv"))$gene

##########################################################
# List of groups to process
groups <- c("group1", "group2")

# Process each group and combine results
summary_data <- map_dfr(groups, ~ process_group(.x, gene.tb, bins.10kb))

# Add the additional rows for distanceBin = 0
summary_data <- summary_data %>%
  mutate(distanceBin = distanceBin + 1) %>%
  add_row(distanceBin = 0, mean_peak_counts = 0, sd_peak_counts = 0, group = "group1") %>%
  add_row(distanceBin = 0, mean_peak_counts = 0, sd_peak_counts = 0, group = "group2")

n1 <- length(group1)
n2 <- length(group2)

# summary_data <- summary_data %>%
#   mutate(
#     se = if_else(group == "group1", sd_peak_counts / sqrt(n1), sd_peak_counts/sqrt(n2)),
#     ci_low  = mean_peak_counts - 2*se,
#     ci_high = mean_peak_counts + 2*se
#   )

# Plot the results
p <- ggplot(summary_data, aes(x = distanceBin, y = mean_peak_counts, fill = group, color = group, group = group)) + 
  geom_smooth(
    linewidth = lineMedium,
    method = "loess", 
    formula = y ~ x, 
    span   = 0.2,    # smaller = wigglier, larger = smoother
    se     = TRUE,   # show confidence band
    alpha  = 0.2, 
  ) +
  labs(
    x = "Distance bin (10kb)",
    y = "H3K4me3 peak count"
  ) +
  theme_classic() +
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    # legend.text = element_text(family = fontType, size = fontSizeS),
    # legend.title = element_text(family = fontType, size = fontSizeS),
    legend.position = "none"  # Moves legend inside the plot (x, y relative coordinates)
  ) +
  guides(
    fill = guide_legend(
      keywidth = 0.2,  # Adjust the width of the legend keys
      keyheight = 0.2  # Adjust the height of the legend keys
    ))+
  scale_color_manual(values = rev(c("#777777", "#F28E2C"))) +
  scale_fill_manual(values = rev(c("#777777", "#F28E2C")))


##########################################################

fileName <- paste0("binPeakDensity_H3K4me3_binaryGroup_EpiLC")
width <- 38*mmToInch
height <- 35*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()
```

#### [FIG] RAD21 peak number at promoter
```{r}
dataDir <- here("..", "data")
peak.RAD21 <- importPeak(here(refDir, "ChIP", "RAD21-HA", "RAD21_ESC-DMSO_unionPeaks.bed"))

################################################################################
flankSize <- 5000
gene.tb <- fread(here(refDir, "mm10", "mm10_GRCm38.p6_gene_sorted.bed")) %>%
  dplyr::mutate(TSS = ifelse(V4 == "+", V2, V3),
                TSSstart = TSS - flankSize,
                TSSend = TSS + flankSize) %>%
  dplyr::select(V6, V5, V1, TSSstart, TSSend)
colnames(gene.tb) <- c("ensembl_gene_id", "external_gene_name", "chr", "start", "end")


# Convert gene.tb to a GRanges object
gene_gr <- GRanges(
  seqnames = gene.tb$chr,
  ranges = IRanges(start = gene.tb$start, end = gene.tb$end),
  gene_id = gene.tb$ensembl_gene_id
)

# Count overlaps between peaks and genes
overlap_counts <- countOverlaps(gene_gr, peak.RAD21)

# Add overlap counts to the original gene.tb data
gene.tb$peak_count <- overlap_counts


group1 <- fread(here(dataDir, "loop_analysis",
                     "geneList_chromosight_EpidTAG_vs_EpiDMSO_p-n_complex_Group1.tsv"))$gene
group2 <- fread(here(dataDir, "loop_analysis",
                     "geneList_chromosight_EpidTAG_vs_EpiDMSO_p-n_complex_Group2.tsv"))$gene


gene.tb <- gene.tb %>% dplyr::mutate(group = case_when(
  ensembl_gene_id %in% group1 ~ "group1",
  ensembl_gene_id %in% group2 ~ "group2",
  TRUE ~ NA
)) %>%
  dplyr::filter(!is.na(group))

###############################################################################

dataToPlot <- gene.tb %>%
  mutate(peakNum = case_when(
    peak_count == 0 ~ "peakNum0",
    peak_count == 1 ~ "peakNum1",
    peak_count >- 2 ~ "peakNumOver2",
    TRUE ~ NA
  ))

# Create the stacked barplot
dataToPlot_short <- dataToPlot %>%
  group_by(group, peakNum) %>%
  summarize(count = n(), .groups = "drop") %>%
  group_by(group) %>%
  mutate(ratio = count / sum(count))

# Define the gradient colors
gradient_colors <- c("#D4D4D4", "#E48D8A", "#F44641")

# Create the stacked barplot with gradient colors
p <- ggplot(dataToPlot_short, aes(x = group, y = ratio, fill = peakNum)) +
  geom_bar(stat = "identity", position = "stack",
           linewidth = lineMedium * mmToLineUnit, lineend = "square") +
  theme_classic() + labs(x = NULL , y = "Ratio") +
  theme(
    axis.title = element_text(
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS, family = fontType, color = "#000000"
    ),
    axis.line = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) +
  scale_fill_manual(values = gradient_colors, 
                    guide = guide_legend(
                      override.aes = list(size = 0.5), # Adjust legend key symbol size
                      keywidth = 2 / 2.54, # Convert 2mm to cm
                      keyheight = 2 / 2.54 # Convert 2mm to cm
                    )) 

width <- panelSize(1.7)*mmToInch
height <- panelSize(1.3)*mmToInch
fileName <- paste0("ChIP_peak_RAD21occupancy_promoter_EpiLC")
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()  

```

### Loop analysis per gene group
```{r}
commonLoopDir <- here("../data/loop_analysis")
diffDir <- here("../data/RNA_diff")
name <- "chromosight"

geneAnnoData <- loadLoopAnnoData(here(commonLoopDir, paste0("loopScore_", name, "_p-n_complex_ensemblList.tsv"))) %>%
  dplyr::mutate(distance = start2 - start1,
                loopID = paste(chrom1, start1, start2, sep = "_"))
```

##### [FIG] Loop number per gene: P-N
```{r}
dataDir <- here("../data")

group1 <- fread(here(dataDir, "loop_analysis",
                     "geneList_chromosight_EpidTAG_vs_EpiDMSO_p-n_complex_Group1.tsv"))$gene
group2 <- fread(here(dataDir, "loop_analysis",
                     "geneList_chromosight_EpidTAG_vs_EpiDMSO_p-n_complex_Group2.tsv"))$gene

# Counting number of loop per genes
tempSum <- geneAnnoData %>% dplyr::select(loopID, gene, Anno) %>% unnest(gene) %>% group_by(gene) %>%  summarize(
  loop = list(loopID),
  anno = list(Anno),
  count = n())

tempSum <- tempSum %>% dplyr::mutate(
  group = ifelse(gene %in% group1, "group1",
                 ifelse(gene %in% group2, "group2", NA))
) %>% dplyr::filter(!is.na(group)) %>% dplyr::filter(group %in% c("group1", "group2"))

# 
getPvalWilcox <- function(data, group1, group2){
  distance1 <- (data %>% dplyr::filter(group ==group1) )$count
  distance2 <- (data %>% dplyr::filter(group ==group2) )$count
  wil <- wilcox.test(distance1, distance2)
  return(wil$p.value)
}


pv12 <- round(getPvalWilcox(tempSum,"group1", "group2"), 5)


p <- ggplot(tempSum, aes(x = group, y = count, fill = group)) + 
  geom_boxplot(width = 0.5, outlier.shape = NA, color = "black", alpha = 0.6,
               linewidth = lineThick * mmToLineUnit, lineend = "square", show.legend = FALSE) + theme_classic() +       
  scale_y_continuous(breaks = seq(0, 10, by = 2)) + 
  labs(x = NULL, y = "# of P-N loops per gene") +
  coord_cartesian(ylim = c(0, 8)) +
  stat_summary(
    aes(group = group), fun = mean,
    geom = "point", shape = 21, size = 0.5,
    fill = "black", color = "black"
  )+
  annotate("text", x = 1, y = 3, label = paste0("pv12: ", convPvalue(pv12)),
           color = "black", hjust = 0, size = 1) +
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) +  scale_fill_manual(values = rev(c("#777777", "#F28E2C")))



fileName <- paste0("count_barplot_diffGroup_EpidTAG_vs_EpiDMSO_binaryGroup")
width <- panelSize(0.8)*mmToInch
height <- panelSize(1.5)*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()
```

##### [FIG] Counting loop types
```{r}
#######
temp2 <- tempSum %>% rowwise() %>% mutate(total = length(anno),
                                          num_pp = sum(anno == "P-P"),
                                          num_pe = sum(anno == "P-E"),
                                          num_px = sum(anno == "P-X"),
                                          num_ps = sum(anno %in% c("P-S", "P-SP", "P-SE")),
                                          num_sps = sum(anno %in% c("SP-SP", "SP-SE", "SP-S")),
                                          num_spn = sum(anno %in% c("E-SP", "SP-X")),
                                          num_str = num_ps + num_sps + num_spn
)


loopType <- temp2 %>% group_by(group) %>% summarise(num_pp = sum(num_pp),
                                                    num_pe = sum(num_pe),
                                                    # num_str = sum(num_str),
                                                    num_ps = sum(num_ps),
                                                    num_sps = sum(num_sps),
                                                    num_spn = sum(num_spn),
                                                    num_px = sum(num_px))


loopTypeLong <- loopType %>% pivot_longer(-group, names_to = "type", values_to = "count")

loopTypeLong$type <- factor(loopTypeLong$type, levels = c("num_pp", "num_pe", "num_px",
                                                          "num_ps", "num_spn", 
                                                          "num_sps" ))

# Plotting
p <- ggplot(loopTypeLong, aes(fill=type, y=count, x=group)) + 
  geom_bar(position="fill", stat="identity") + theme_classic() +
  scale_fill_manual(values = c(strong_red, strong_orange, strong_yellow, strong_blue, "grey30", "grey10")) +theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  )+   
  guides(
    fill = guide_legend(
      keywidth = 0.2,  # Adjust the width of the legend keys
      keyheight = 0.2  # Adjust the height of the legend keys
    ))

fileName <- paste0("count_barplot_diffGroup_EpidTAG_vs_EpiDMSO_binaryGroup_alltypes")
width <- panelSize(1.5)*mmToInch
height <- panelSize(1.5)*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()
```

##### [FIG] Loop number per gene: structural
```{r}
########
data <- temp2 %>% dplyr::select(group, num_str)
colnames(data) <- c("group", "count")
pv12 <- round(getPvalWilcox(data,"group1", "group2"), 5)

p <- ggplot(data, aes(x = group, y = count, fill = group)) + 
  geom_boxplot(width = 0.5, outlier.shape = NA, color = "black", alpha = 0.6,
               linewidth = lineThick * mmToLineUnit, lineend = "square", show.legend = FALSE) + theme_classic() +       
  scale_y_continuous(breaks = seq(0, 10, by = 2)) + 
  labs(x = NULL, y = "# of PromStr loops per gene") +
  coord_cartesian(ylim = c(0, 4)) +
  stat_summary(
    aes(group = group), fun = mean,
    geom = "point", shape = 21, size = 0.5,
    fill = "black", color = "black"
  )+
  annotate("text", x = 1, y = 3, label = paste0("pv12: ", convPvalue(pv12)),
           color = "black", hjust = 0, size = 1) +
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) +  scale_fill_manual(values = rev(c("#777777", "#F28E2C")))

fileName <- paste0("count_barplot_diffGroup_EpidTAG_vs_EpiDMSO_str_binaryGroup")
width <- panelSize(0.8)*mmToInch
height <- panelSize(1.5)*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()
```
###### (WIP) GO
```{r}
# group1 <- fread(here(dataDir, "loop_analysis",
#                      "geneList_chromosight_dTAG_vs_DMSO_p-n_complex_Group1.tsv"))$gene
# group2 <- fread(here(dataDir, "loop_analysis",
#                      "geneList_chromosight_dTAG_vs_DMSO_p-n_complex_Group2.tsv"))$gene
# 
# 
# GO1 <- enrichGO(gene = group1, OrgDb = org.Mm.eg.db, keyType = "ENSEMBL", ont = "BP", readable = TRUE)
# GO2 <- enrichGO(gene = group2, OrgDb = org.Mm.eg.db, keyType = "ENSEMBL", ont = "BP", readable = TRUE)
# GO1.df <- as.data.frame(GO1)
# GO2.df <- as.data.frame(GO2)
# # fwrite(GO1.df, here("GO_binary_group1.tsv"), sep = "\t")
# # fwrite(GO2.df, here("GO_binary_group2.tsv"), sep = "\t")
# # GO1.df <- fread(here("GO_binary_group1.tsv"))
# # GO2.df <- fread(here("GO_binary_group2.tsv"))
# 
# subset1 <- GO1.df %>% dplyr::select(ID, Description, GeneRatio, p.adjust) %>% dplyr::mutate(group = "Group 1") %>%
#   dplyr::arrange(p.adjust)
# subset1$GeneRatio <- sapply(strsplit(subset1$GeneRatio, "/"), function(x) as.numeric(x[1]) / as.numeric(x[2]))
# 
# subset2 <- GO2.df %>% dplyr::select(ID, Description, GeneRatio, p.adjust) %>% dplyr::mutate(group = "Group 2") %>%
#   dplyr::arrange(p.adjust)
# subset2$GeneRatio <- sapply(strsplit(subset2$GeneRatio, "/"), function(x) as.numeric(x[1]) / as.numeric(x[2]))
# 
# 
# GOlist <- factor(c("GO:0001824", "GO:0030879", "GO:0021953", "GO:0019827",
#                    "GO:0022613", "GO:0050767", "GO:0034470", "GO:0016055",
#                    "GO:0006397", "GO:0030900", "GO:0008380"))
# 
# data <- bind_rows(subset1, subset2) %>%
#   dplyr::filter(ID %in% GOlist)
# 
# descOrder <- sort(unique(data$Description))[c(4, 10, 2, 1, 3, 11, 7, 8, 6, 5, 9)]
# pValueLogMax <- 10
# data <- data %>% dplyr::rowwise() %>% dplyr::mutate(pValueLog = min(-log10(p.adjust), pValueLogMax))
# p <- ggplot(data, aes(x = group, y = factor(Description, levels = descOrder), size = pValueLog, fill = GeneRatio)) +
#   geom_point(shape = 21,        # Ensures a point with an outline
#              stroke = 1*ptToMM) +
#   scale_size_continuous(range = c(0.5, 2)) +  # Set min and max point sizes here
#   scale_fill_gradient(low = "white", high = "#CB333A",
#                       # limits = c(0, 1),
#                       oob = scales::squish, # Define gradient colors
#                       guide = guide_colorbar(
#                         barwidth = 1.5/5.08,  # Adjust width of the color bar
#                         barheight = 15/5.08   # Adjust height of the color bar
#                       )
#   ) +
#   labs(x = NULL, y = NULL) +
#   theme_bw() +  # Apply theme_bw first, so custom theme settings come after
#   theme(
#     panel.background = element_rect(fill = "transparent"),  # Override theme_bw panel
#     axis.title = element_text(
#       size = fontSizeS,
#       family = fontType,
#       color = "#000000"
#     ),
#     axis.text.x = element_text(
#       size = fontSizeS,         # Ensure size is set for x-axis text
#       family = fontType,
#       color = "#000000",
#     ),
#     axis.text.y = element_text(
#       size = fontSizeS,         # Ensure size is set for y-axis text
#       family = fontType,
#       color = "#000000",
#       lineheight = 0.9          # Allows wrapping for y-axis labels to fit into 2 lines
#     ),
#     axis.line = element_line(
#       color = "#000000",
#       size = lineThick * mmToLineUnit,
#       lineend = "square"
#     ),
#     axis.ticks = element_line(
#       color = "#000000",
#       size = lineThick * mmToLineUnit,
#       lineend = "square"
#     ),
#     legend.text = element_text(family = fontType, size = fontSizeS),
#     legend.title = element_text(family = fontType, size = fontSizeS)
#   )
# 
# 
# 
# fileName <- here(figDir, "..", "GO", "GO_groups_binaryGrouping")
# width <- panelSize(2.5)*mmToInch
# height <- panelSize(1.2)*mmToInch
# svglite(paste0(fileName, ".svg"), height = height, width = width)
# print(p)
# dev.off()
```

##### [FIG] Loop number per gene: P-P
```{r}
########
data <- temp2 %>% dplyr::select(group, num_pp)
colnames(data) <- c("group", "count")
pv12 <- round(getPvalWilcox(data,"group1", "group2"), 5)

p <- ggplot(data, aes(x = group, y = count, fill = group)) + 
  geom_boxplot(width = 0.5, outlier.shape = NA, color = "black", alpha = 0.6,
               linewidth = lineThick * mmToLineUnit, lineend = "square", show.legend = FALSE) + theme_classic() +       
  scale_y_continuous(breaks = seq(0, 10, by = 2)) + 
  labs(x = NULL, y = "# of P-P loops per gene") +
  coord_cartesian(ylim = c(0, 4)) +
  stat_summary(
    aes(group = group), fun = mean,
    geom = "point", shape = 21, size = 0.5,
    fill = "black", color = "black"
  )+
  annotate("text", x = 1, y = 3, label = paste0("pv12: ", convPvalue(pv12)),
           color = "black", hjust = 0, size = 1) +
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) +  scale_fill_manual(values = rev(c("#777777", "#F28E2C")))

fileName <- paste0("count_barplot_diffGroup_EpidTAG_vs_EpiDMSO_pp_binaryGroup")
width <- panelSize(0.8)*mmToInch
height <- panelSize(1.5)*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()
```

##### [FIG] Loop number per gene: P-E
```{r}
########
data <- temp2 %>% dplyr::select(group, num_pe)
colnames(data) <- c("group", "count")
pv12 <- round(getPvalWilcox(data,"group1", "group2"), 5)

p <- ggplot(data, aes(x = group, y = count, fill = group)) + 
  geom_boxplot(width = 0.5, outlier.shape = NA, color = "black", alpha = 0.6,
               linewidth = lineThick * mmToLineUnit, lineend = "square", show.legend = FALSE) + theme_classic() +       
  scale_y_continuous(breaks = seq(0, 10, by = 2)) + 
  labs(x = NULL, y = "# of P-E loops per gene") +
  coord_cartesian(ylim = c(0, 4)) +
  stat_summary(
    aes(group = group), fun = mean,
    geom = "point", shape = 21, size = 0.5,
    fill = "black", color = "black"
  )+
  annotate("text", x = 1, y = 3, label = paste0("pv12: ", convPvalue(pv12)),
           color = "black", hjust = 0, size = 1) +
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) +  scale_fill_manual(values = rev(c("#777777", "#F28E2C")))

fileName <- paste0("count_barplot_diffGroup_EpidTAG_vs_EpiDMSO_pe_binaryGroup")
width <- panelSize(0.8)*mmToInch
height <- panelSize(1.5)*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()
```


##### [FIG] Grouping genes with structural loop number
###### Grouping
```{r}
# temp2 contains genes from group 1, 2, 5, 8 and loop counts
#calculating diff score and log2fc distribution based on p-n numbers

psOver4 <- (temp2 %>% dplyr::filter(num_str >= 4))$gene
ps3 <- (temp2 %>% dplyr::filter(num_str == 3))$gene
ps2 <- (temp2 %>% dplyr::filter(num_str == 2))$gene
ps1 <- (temp2 %>% dplyr::filter(num_str == 1))$gene
ps0 <- (temp2 %>% dplyr::filter(num_str == 0))$gene

# 
# temp2 <- tempSum %>% rowwise() %>% mutate(total = length(anno),
#                                           num_pp = sum(anno == "P-P"),
#                                           num_pe = sum(anno == "P-E"),
#                                           num_px = sum(anno == "P-X"),
#                                           num_ps = sum(anno %in% c("P-S", "P-SP", "P-SE")),
#                                           num_sps = sum(anno %in% c("SP-SP", "SP-SE", "SP-S")),
#                                           num_spn = sum(anno %in% c("E-SP", "SP-X")),
#                                           num_str = num_ps + num_sps + num_spn
# 

## Dividing genes into groups


temp <- geneAnnoData %>% dplyr::mutate(diff = score_EpidTAG - score_EpiDMSO) %>%
  dplyr::filter(str_detect(Anno, "P")) %>%
  dplyr::select(diff, gene) %>%
  unnest(gene) %>% group_by(gene) %>%
  summarize(mean_diff_score = mean(diff),
            .groups = 'drop')

diff.RNA <- fread(here(diffDir, "diff_G1.dTAG_G1.Epi.dTAG_vs_G1.Epi.DMSO.tsv")) %>%
  dplyr::select(ensembl_gene_id, log2FoldChange, shrinked_log2FC, padj, external_gene_name)

temp <- left_join(temp, diff.RNA, by = c("gene" = "ensembl_gene_id")) %>% 
  drop_na(shrinked_log2FC)


temp <- temp %>% dplyr::mutate(
  psNum = case_when(
    gene %in% ps0 ~ "ps0",
    gene %in% ps1 ~ "ps1",
    gene %in% ps2 ~ "ps2",
    gene %in% ps3 ~ "ps3",
    gene %in% psOver4 ~ "psOver4",
    TRUE ~ NA
  )) %>%
  drop_na(psNum)

```



###### loop score
```{r}
getPvalWilcox <- function(data, group1, group2){
  distance1 <- (data %>% dplyr::filter(psNum ==group1) )$mean_diff_score
  distance2 <- (data %>% dplyr::filter(psNum ==group2) )$mean_diff_score
  wil <- wilcox.test(distance1, distance2)
  return(wil$p.value)
}

ps01 <- round(getPvalWilcox(temp, "ps0", "ps1"), 5)
ps12 <- round(getPvalWilcox(temp, "ps1", "ps2"), 5)
ps23 <- round(getPvalWilcox(temp, "ps2", "ps3"), 5)
ps34 <- round(getPvalWilcox(temp, "ps3", "psOver4"), 5)
ps24 <- round(getPvalWilcox(temp, "ps2", "psOver4"), 5)
ps14 <- round(getPvalWilcox(temp, "ps1", "psOver4"), 5)
ps04 <- round(getPvalWilcox(temp, "ps0", "psOver4"), 5)


p <- ggplot(temp, aes(x = psNum, y = mean_diff_score)) + 
  geom_violin(aes(fill = psNum), 
              color = "black",
              linewidth = lineMedium * mmToLineUnit, lineend = "square",
              show.legend = FALSE) +
  geom_boxplot(width = 0.3, color = "black",
               linewidth = lineMedium * mmToLineUnit, lineend = "square",
               outlier.shape = NA
  ) + theme_classic() + labs(x = NULL , y = "Average Δ loop score") +
  stat_summary(
    aes(group = psNum), fun = mean,
    geom = "point", shape = 21, size = 0.5,
    fill = "black", color = "black"
  ) +
  geom_hline(yintercept = 0, linewidth = lineThick * mmToLineUnit) +
  theme(
    axis.title = element_text(
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS, family = fontType, color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45, hjust = 1, vjust = 1
    ),
    axis.line = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  )+
  annotate("text", x = 1, y = -0.3, label = paste0("ps01: ", convPvalue(ps01), "\n",
                                                   "ps12: ", convPvalue(ps12), "\n",
                                                   "ps23: ", convPvalue(ps23), "\n",
                                                   "ps34: ", convPvalue(ps34), "\n",
                                                   "ps24: ", convPvalue(ps24), "\n", 
                                                   "ps14: ", convPvalue(ps14), "\n",
                                                   "ps04: ", convPvalue(ps04), "\n"), 
           color = "black", hjust = 0, size = 2) +
  scale_fill_manual(values = c("#777777", "#8B7E65", "#A28452", "#C2884D", "#F28E2C")) +
  coord_cartesian(ylim = c(-0.7, 0))



fileName <- paste0("psGroup_loopDiffScorre_EpidTAGvsEpiDMSO")
width <- 33*mmToInch
height <-38*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()
```

###### log2FC
```{r}
temp <- temp %>% dplyr::mutate(absLog2FC = abs(log2FoldChange))

ks_result1 <- ks.test(
  temp %>% dplyr::filter(psNum == "psOver4") %>% pull(absLog2FC),
  temp %>% dplyr::filter(psNum == "ps3") %>% pull(absLog2FC)
)
ks_result2 <- ks.test(
  temp %>% dplyr::filter(psNum == "psOver4") %>% pull(absLog2FC),
  temp %>% dplyr::filter(psNum == "ps2") %>% pull(absLog2FC)
)
ks_result3 <- ks.test(
  temp %>% dplyr::filter(psNum == "psOver4") %>% pull(absLog2FC),
  temp %>% dplyr::filter(psNum == "ps1") %>% pull(absLog2FC)
)
ks_result4 <- ks.test(
  temp %>% dplyr::filter(psNum == "psOver4") %>% pull(absLog2FC),
  temp %>% dplyr::filter(psNum == "ps0") %>% pull(absLog2FC)
)

p <- ggplot(temp, aes(x = absLog2FC, color = psNum)) +
  scale_color_manual(values = (c("#777777", "#8B7E65", "#A28452", "#C2884D", "#F28E2C"))) +
  stat_ecdf(size = 0.4, linewidth = lineMedium * mmToLineUnit, lineend = "square" ) + # Use stat_ecdf to plot the empirical CDF
  labs(
    x = "Abs. log2(fold change)",
    y = "Cumulative Probability"
  ) + coord_cartesian(xlim = c(0, 1.5)) +
  theme_classic() + # Clean theme
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.position = "none",
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) + scale_y_continuous(labels = scales::number_format(accuracy = 0.1))


fileName <- paste0("psGroup_abslog2FC_EpidTAGvsEpiDMSO")
width <- 33*mmToInch
height <-33*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()



```


###### (WIP) GO
```{r}
# # GOfigDir <- here(figDir, "../GO")
# # getGO("psOver4", GOfigDir, psOver4)
# # getGO("psOver3", GOfigDir, psOver3)
# # getGO("psOver2", GOfigDir, psOver2)
# # getGO("psOver1", GOfigDir, psOver1)
# # getGO("psOver0", GOfigDir, psOver0)
# # 
# # #####################
# GO0.df <- as.data.frame(enrichGO(gene = ps0, OrgDb = org.Mm.eg.db, keyType = "ENSEMBL", ont = "BP"))
# GO1.df <- as.data.frame(enrichGO(gene = ps1, OrgDb = org.Mm.eg.db, keyType = "ENSEMBL", ont = "BP"))
# GO2.df <- as.data.frame(enrichGO(gene = ps2, OrgDb = org.Mm.eg.db, keyType = "ENSEMBL", ont = "BP"))
# GO3.df <- as.data.frame(enrichGO(gene = ps3, OrgDb = org.Mm.eg.db, keyType = "ENSEMBL", ont = "BP"))
# GO4.df <- as.data.frame(enrichGO(gene = psOver4, OrgDb = org.Mm.eg.db, keyType = "ENSEMBL", ont = "BP"))
# 
# 
# subset0 <- GO0.df %>% dplyr::select(ID, Description, GeneRatio, p.adjust) %>% dplyr::mutate(group = "psOver0") %>%
#   dplyr::mutate(
#     gr = sapply(GeneRatio, function(x) {
#       # Split the string by "/"
#       parts <- unlist(strsplit(x, "/"))
#       # Convert to numeric and perform the division
#       as.numeric(parts[1]) / as.numeric(parts[2])
#     })
#   ) %>% dplyr::arrange(desc(gr))
# subset1 <- GO1.df %>% dplyr::select(ID, Description, GeneRatio, p.adjust) %>% dplyr::mutate(group = "psOver1") %>%
#   dplyr::mutate(
#     gr = sapply(GeneRatio, function(x) {
#       # Split the string by "/"
#       parts <- unlist(strsplit(x, "/"))
#       # Convert to numeric and perform the division
#       as.numeric(parts[1]) / as.numeric(parts[2])
#     })
#   ) %>% dplyr::arrange(desc(gr))
# subset2 <- GO2.df %>% dplyr::select(ID, Description, GeneRatio, p.adjust) %>% dplyr::mutate(group = "psOver2") %>%
#   dplyr::mutate(
#     gr = sapply(GeneRatio, function(x) {
#       # Split the string by "/"
#       parts <- unlist(strsplit(x, "/"))
#       # Convert to numeric and perform the division
#       as.numeric(parts[1]) / as.numeric(parts[2])
#     })
#   ) %>% dplyr::arrange(desc(gr))
# subset3 <- GO3.df %>% dplyr::select(ID, Description, GeneRatio, p.adjust) %>% dplyr::mutate(group = "psOver3") %>%
#   dplyr::mutate(
#     gr = sapply(GeneRatio, function(x) {
#       # Split the string by "/"
#       parts <- unlist(strsplit(x, "/"))
#       # Convert to numeric and perform the division
#       as.numeric(parts[1]) / as.numeric(parts[2])
#     })
#   ) %>% dplyr::arrange(desc(gr))
# subset4 <- GO4.df %>% dplyr::select(ID, Description, GeneRatio, p.adjust) %>% dplyr::mutate(group = "psOver4") %>%
#   dplyr::mutate(
#     gr = sapply(GeneRatio, function(x) {
#       # Split the string by "/"
#       parts <- unlist(strsplit(x, "/"))
#       # Convert to numeric and perform the division
#       as.numeric(parts[1]) / as.numeric(parts[2])
#     })
#   ) %>% dplyr::arrange(desc(gr))
# 
# 
# GOlist <- factor(c("GO:0006397", "GO:0008380", "GO:0022613", "GO:0034470",
#                    "GO:0016055", "GO:0007389", "GO:0048562", "GO:0045165",
#                    "GO:0072001", "GO:0007517", "GO:0048705", "GO:0003002"))
# 
# data <- bind_rows(bind_rows(bind_rows(bind_rows(subset0, subset1), subset2), subset3), subset4) %>%
#   dplyr::filter(ID %in% GOlist)
# 
# p <- ggplot(data, aes(x = group, y = Description, color = p.adjust, size = gr)) +
#   geom_point() + theme_bw() +
#   scale_color_gradient(low = "red", high = "blue", limits = c(0, 0.05)) +
#   scale_size_continuous(range = c(0, 3)) +
#   labs(x = NULL, y = NULL) +
#   theme(axis.text = element_text(size = 6),  # Set axis text size
#         axis.title = element_text(size = 6), # Set axis title size (if not removed)
#         legend.text = element_text(size = 6), # Set legend text size
#         legend.title = element_text(size = 6))
# # 
# # fileName <- here(figDir, "..", "GO", "GO_groups_ps")
# # height = 2
# # width = 3.4
# # svglite(paste0(fileName, ".svg"), height = height, width = width)
# # print(p)
# # dev.off()
```

### Finding largest convergent encompassing S-S
#### Annotating CTCF motif
#### Filtering convergent S-S loops
```{r}
commonLoopDir <- here("../data/loop_analysis")
figDir <- here("../figure/loop_analysis")

# IMPORT S-S LOOPS
allLoops <- as_tibble(fread(here(commonLoopDir, "loopScore_chromosight.tsv")))
strLoops <- allLoops %>% dplyr::filter(Anno %in% c("S-S", "SE-SE", "SP-SP", "SP-SE", "SE-S", "SP-S"))
temp <- strLoops %>% dplyr::select(c(1, 2, 3, 5, 6))  # Assuming columns include anchor positions
colnames(temp) <- c("chrom", "start1", "end1", "start2", "end2")
strLoopsGr1 <- makeGRangesFromDataFrame(
  temp %>% dplyr::select(chrom, start = start1, end = end1)
)
strLoopsGr2 <- makeGRangesFromDataFrame(
  temp %>% dplyr::select(chrom, start = start2, end = end2)
)

# IMPORT CTCF MOTIF ANNOTATION
ctcfMotifs <- fread(here(refDir, "ChIP", "CTCF", "CTCF_ESC_unionPeaks_motifAnnotated.bed"))
colnames(ctcfMotifs) <- c("chrom", "start", "end", "motif")
ctcfGr <- makeGRangesFromDataFrame(ctcfMotifs, keep.extra.columns = TRUE)

# Annotate each loop anchor with CTCF motifs
anchor1_with_ctcf <- findOverlaps(strLoopsGr1, ctcfGr)
anchor2_with_ctcf <- findOverlaps(strLoopsGr2, ctcfGr)

# Get motif information for anchor1 overlaps
anchor1_ctcf_motif <- rep(NA, length(strLoopsGr1))  # Initialize motif vector
anchor1_ctcf_motif[queryHits(anchor1_with_ctcf)] <- ctcfGr[subjectHits(anchor1_with_ctcf)]$motif

# Get motif information for anchor2 overlaps
anchor2_ctcf_motif <- rep(NA, length(strLoopsGr2))  # Initialize motif vector
anchor2_ctcf_motif[queryHits(anchor2_with_ctcf)] <- ctcfGr[subjectHits(anchor2_with_ctcf)]$motif

# Add motif information to strLoops
strLoops <- strLoops %>%
  mutate(
    anchor1_motif = anchor1_ctcf_motif,
    anchor2_motif = anchor2_ctcf_motif
  )

# Filter for convergent CTCF interactions (forward in anchor1, reverse in anchor2)
convergent_loops <- strLoops %>%
  filter(anchor1_motif %in% c("fwd", "fwdrev") &
           (anchor2_motif %in% c("rev", "fwdrev")))
```

#### Processing
```{r}
# Extract loop boundary from left to right
temp <- convergent_loops %>% dplyr::select(c(1, 2, 6))
colnames(temp) <- c("chrom", "start", "end")
strLoopsGr <- makeGRangesFromDataFrame(temp)

# IMPORT TSS
flankSize <- 10
gene.tb <- fread(here(refDir, "mm10", "mm10_GRCm38.p6_gene_sorted.bed")) %>%
  dplyr::mutate(TSS = ifelse(V4 == "+", V2, V3),
                TSSstart = TSS - flankSize,
                TSSend = TSS + flankSize) %>%
  dplyr::select(V1, TSSstart, TSSend, V6, V5)
colnames(gene.tb) <- c("chrom", "start", "end", "ensembl", "gene")
geneGr <- makeGRangesFromDataFrame(gene.tb, keep.extra.columns = TRUE)


# Find overlaps between loops and TSS
overlaps <- findOverlaps(geneGr, strLoopsGr)

# Annotate overlaps
tss_with_loops <- geneGr[queryHits(overlaps)]
loops_with_tss <- strLoopsGr[subjectHits(overlaps)]

# Combine into a data frame for processing
loop_data <- data.frame(gene = tss_with_loops$ensembl,
                        loop_chr = seqnames(loops_with_tss),
                        loop_start = start(loops_with_tss),
                        loop_end = end(loops_with_tss),
                        loop_width = width(loops_with_tss))

# Identify the largest loop for each gene
largest_loops <- loop_data[order(loop_data$gene, -loop_data$loop_width), ]
largest_loops <- largest_loops[!duplicated(largest_loops$gene), ]
largest_loops <- largest_loops %>% dplyr::mutate(loopID = paste(loop_chr, loop_start, loop_end, sep = "_"))
largest_loops <- largest_loops[order(largest_loops$loop_chr, largest_loops$loop_start),]
rownames(largest_loops) <- NULL

# Checking smaller loops
allLoopsGr <- makeGRangesFromDataFrame(allLoops, 
                                       seqnames.field = "chrom1", 
                                       start.field = "start1", 
                                       end.field = "end2")
largestLoopsGr <- makeGRangesFromDataFrame(largest_loops, 
                                           seqnames.field = "loop_chr", 
                                           start.field = "loop_start", 
                                           end.field = "loop_end")
largestLoopsGr <- sort(unique(largestLoopsGr))
complete_overlaps <- subsetByOverlaps(allLoopsGr, largestLoopsGr, type = "within")
overlaps <- findOverlaps(complete_overlaps, largestLoopsGr)


completeOverlapDf <- as.data.frame(complete_overlaps) %>% dplyr::mutate(loopIDAll = paste(seqnames, start, end, sep = "_"))
largestLoopsDf <- as.data.frame(largestLoopsGr) %>% dplyr::mutate(loopID = paste(seqnames, start, end, sep = "_"))

loopIDpairs <- data.frame(loopID = completeOverlapDf$loopIDAll[queryHits(overlaps)],
                          loopID2 = largestLoopsDf$loopID[subjectHits(overlaps)])

##################################
name <- "chromosight"
geneAnnoData <- loadLoopAnnoData(here(commonLoopDir, paste0("loopScore_", name, "_p-n_complex_ensemblList.tsv"))) %>% unnest(gene) %>%
  dplyr::mutate(loopID = paste(chrom1, start1, end2, sep = "_"))

temp <- geneAnnoData %>% dplyr::left_join(loopIDpairs, by = c("loopID"))
temp2 <- temp %>% dplyr::select(gene, loopID2) %>%
  group_by(gene) %>%              # Group data by the 'gene' column
  summarize(geneHasSS = !all(is.na(loopID2)))  # Check if all loopID2 values for each gene are NA
temp <- temp %>% left_join(temp2, by = c("gene"))
data.all <- temp %>% dplyr::mutate(group = ifelse(!geneHasSS, "No",
                                                  ifelse(is.na(loopID2), "Outside", "Within")),
                                   size = start2 - start1)

##################################
outfile <- data.all %>% dplyr::select(loopID2, gene) %>% dplyr::filter(!is.na(loopID2)) %>%
  separate(loopID2, into = c("chrom", "start", "end"), sep = "_", convert = TRUE)
fwrite(outfile, here(commonLoopDir, "convergent_ss_domain.bed"), sep = "\t", col.names = FALSE)
```

#### [FIG] Plotting
```{r}
data <- data.all %>% dplyr::filter(Anno %in% c("P-SE", "P-E", "E-SP", "SP-SE"))

temp <- data %>% dplyr::select(chrom1, start1, end1, chrom2, start2, end2, group) %>% distinct()

temp1 <- temp %>% dplyr::filter(group == "No") %>% dplyr::select(-group)
fwrite(temp1, here(commonLoopDir, "insulated_domain_ss_no.bedpe"), sep = "\t", col.names = FALSE)
temp2 <- temp %>% dplyr::filter(group == "Within") %>% dplyr::select(-group)
fwrite(temp2, here(commonLoopDir, "insulated_domain_ss_within.bedpe"), sep = "\t", col.names = FALSE)
temp3 <- temp %>% dplyr::filter(group == "Outside") %>% dplyr::select(-group)
fwrite(temp3, here(commonLoopDir, "insulated_domain_ss_outside.bedpe"), sep = "\t", col.names = FALSE)

temp <- data %>% dplyr::select(id, group, size) %>% distinct()


getPvalWilcox <- function(data, group1, group2){
  distance1 <- (data %>% dplyr::filter(group ==group1) )$size
  distance2 <- (data %>% dplyr::filter(group ==group2) )$size
  wil <- wilcox.test(distance1, distance2)
  return(wil$p.value)
}

convPvalue(getPvalWilcox(temp, "No", "Within"))
convPvalue(getPvalWilcox(temp, "No", "Outside"))
convPvalue(getPvalWilcox(temp, "Outside", "Within"))


p <- ggplot(temp, aes(x = group, y = size, fill = group)) + 
  geom_violin(linewidth = lineMedium* mmToLineUnit, lineend = "square", alpha = .4, , show.legend = FALSE) + 
  geom_boxplot(width = 0.3, color = "black",
               linewidth = lineMedium * mmToLineUnit, lineend = "square",
               outlier.shape = NA,  alpha = 0.6, show.legend = FALSE) + 
  theme_classic() + scale_y_continuous(labels = label_kb_mb) +
  scale_fill_manual(values = c("#777777", "#F28E2C", "#F28E2C")) +
  stat_summary(
    aes(group = group), fun = mean,
    geom = "point", shape = 21, size = 0.5,
    fill = "black", color = "black", position = position_dodge(.3)
  ) +
  theme(
    axis.title.x = element_blank(),
    axis.title.y = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),    
    axis.text.x = element_text(
      size = fontSizeM,
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) +labs(y = "Loop size")

fileName <- paste0("insulationBoundary_size_convSS_PE_EpiLC")
width <-33*mmToInch
height <- 35*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()


###################################


pseudo_ooe <- 0.01

data <- data %>% dplyr::filter(!is.na(ooe_EpiDMSO), !is.na(ooe_EpidTAG)) %>%
  dplyr::mutate(
    ooe_EpiDMSO = if_else(ooe_EpiDMSO == 0, pseudo_ooe, ooe_EpiDMSO),
    ooe_EpidTAG = if_else(ooe_EpidTAG == 0, pseudo_ooe, ooe_EpidTAG),
    fc_EpidTAG    = log2(ooe_EpidTAG / ooe_EpiDMSO),
    diff_EpidTAG = score_EpidTAG - score_EpiDMSO
  )

getPvalWilcox <- function(data, group1, group2){
  distance1 <- (data %>% dplyr::filter(group ==group1) )$diff_EpidTAG
  distance2 <- (data %>% dplyr::filter(group ==group2) )$diff_EpidTAG
  wil <- wilcox.test(distance1, distance2)
  return(wil$p.value)
}

temp <- data %>% dplyr::select(group, id, diff_EpidTAG) %>% distinct()

convPvalue(getPvalWilcox(temp, "No", "Within"))
convPvalue(getPvalWilcox(temp, "No", "Outside"))
convPvalue(getPvalWilcox(temp, "Outside", "Within"))

p <- ggplot(data, aes(x = group, y = diff_EpidTAG, fill = group)) + 
  geom_violin(linewidth = lineMedium * mmToLineUnit, lineend = "square", alpha = .4, , show.legend = FALSE) + 
  geom_boxplot(width = 0.3, color = "black",
               linewidth = lineMedium * mmToLineUnit, lineend = "square",
               outlier.shape = NA,  alpha = 0.6, show.legend = FALSE) + 
  theme_classic() +   
  scale_fill_manual(values = c("#777777", "#F28E2C", "#F28E2C")) +
  stat_summary(
    aes(group = group), fun = mean,
    geom = "point", shape = 21, size = 0.5,
    fill = "black", color = "black", position = position_dodge(.3)
  ) +
  theme(
    axis.title.x = element_blank(),
    axis.title.y = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),    
    axis.text.x = element_text(
      size = fontSizeM,
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) +labs(y = "Δ loop score") +    
  geom_hline(yintercept = 0,
             alpha = 1,
             color = "black",
             size = lineThick*mmToLineUnit,
             lineend = "square") + 
  geom_hline(yintercept = -0.2,
             alpha = 0.5, 
             color = "black",
             size = lineThick*mmToLineUnit,
             lineend = "square", linetype = "dashed") +
  coord_cartesian(ylim = c(-0.7, 0.5))

fileName <- paste0("insulationBoundary_delta_convSS_pe_EpiLC")
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()  


###################### 
convPvalue(getPvalWilcox(temp, "No", "Within"))
convPvalue(getPvalWilcox(temp, "No", "Outside"))
convPvalue(getPvalWilcox(temp, "Outside", "Within"))

temp <- data %>% dplyr::select(group, id, fc_EpidTAG) %>% distinct()

p <- ggplot(temp, aes(x = group, y = fc_EpidTAG, fill = group)) + 
  geom_violin(linewidth = lineMedium* mmToLineUnit, lineend = "square", alpha = .4, , show.legend = FALSE) + 
  geom_boxplot(width = 0.3, color = "black",
               linewidth = lineMedium * mmToLineUnit, lineend = "square",
               outlier.shape = NA,  alpha = 0.6, show.legend = FALSE) + 
  theme_classic() + 
  scale_fill_manual(values = c("#777777", "#F28E2C", "#F28E2C")) +
  
  stat_summary(
    aes(group = group), fun = mean,
    geom = "point", shape = 21, size = 0.5,
    fill = "black", color = "black", position = position_dodge(.3)
  ) +
  theme(
    axis.title.x = element_blank(),
    axis.title.y = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),    
    axis.text.x = element_text(
      size = fontSizeM,
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) +labs(y = "log2(fc of obs/exp)") +    
  geom_hline(yintercept = 0,
             alpha = 1,
             color = "black",
             size = lineThick*mmToLineUnit,
             lineend = "square") + 
  coord_cartesian(ylim = c(-2.5, 2))
width <-30*mmToInch
height <- 35*mmToInch
fileName <- paste0("insulationBoundary_obsexp_convSS_pe_EpiLC")
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()  





#####################

data <- data.all
gene.withinBoundary <- unique((data %>% dplyr::filter(group == "Within"))$gene)
gene.outsideBoundary <- unique((data %>% dplyr::filter(group == "Outside"))$gene)
gene.noBoundary <- unique((data %>% dplyr::filter(group == "No"))$gene)

gene.wBoundary <- unique(c(gene.withinBoundary, gene.outsideBoundary))

dataDir <- here("..", "data")
gene.group1 <- fread(here(dataDir, "loop_analysis",
                          "geneList_chromosight_EpidTAG_vs_EpiDMSO_p-n_complex_Group1.tsv"))$gene
gene.group2 <- fread(here(dataDir, "loop_analysis",
                          "geneList_chromosight_EpidTAG_vs_EpiDMSO_p-n_complex_Group2.tsv"))$gene


print("perc of gene with loop within Boundary")
sum(gene.group1 %in% gene.withinBoundary)/length(gene.group1)*100
sum(gene.group2 %in% gene.withinBoundary)/length(gene.group2)*100

print("perc of gene with loop outside Boundary")
sum(gene.group1 %in% gene.outsideBoundary)/length(gene.group1)*100
sum(gene.group2 %in% gene.outsideBoundary)/length(gene.group2)*100

print("perc of gene with loop with no Boundary")
sum(gene.group1 %in% gene.noBoundary)/length(gene.group1)*100
sum(gene.group2 %in% gene.noBoundary)/length(gene.group2)*100


diff.RNA <- fread(here(dataDir, "RNA_diff", "diff_G1.dTAG_G1.Epi.dTAG_vs_G1.Epi.DMSO.tsv")) %>%
  dplyr::select(ensembl_gene_id, log2FoldChange, shrinked_log2FC, padj, external_gene_name) %>%
  dplyr::mutate(group = case_when(
    ensembl_gene_id %in% gene.wBoundary ~ "withinBoundary",
    ensembl_gene_id %in% gene.noBoundary ~ "noBoundary",
    TRUE ~ NA
  )) %>%
  dplyr::filter(!is.na(group)) %>%
  dplyr::mutate(absLog2FC = abs(log2FoldChange))
ks_result <- ks.test(
  diff.RNA %>% dplyr::filter(group == "withinBoundary") %>% pull(absLog2FC),
  diff.RNA %>% dplyr::filter(group == "noBoundary") %>% pull(absLog2FC)
)

# ggplot(diff.RNA, aes(x = group, y = log2FoldChange)) + geom_violin() + geom_boxplot(width = 0.1)

# getPvalWilcox <- function(data, group1, group2){
#   distance1 <- (data %>% dplyr::filter(group ==group1) )$absLog2FC
#   distance2 <- (data %>% dplyr::filter(group ==group2) )$absLog2FC
#   wil <- wilcox.test(distance1, distance2)
#   return(wil$p.value)
# }
# convPvalue(getPvalWilcox(diff.RNA, "withBoundary", "noBoundary"))
# 

# Create the CDF plot
p <- ggplot(diff.RNA, aes(x = absLog2FC, color = group)) +
  scale_color_manual(values = (c("#777777", "#F28E2C"))) +
  
  stat_ecdf(size = 0.4, linewidth = lineMedium * mmToLineUnit, lineend = "square"  ) + # Use stat_ecdf to plot the empirical CDF
  labs(
    x = "Abs. log2(fold change)",
    y = "Cumulative Probability"
  ) + coord_cartesian(xlim = c(0, 1)) +
  theme_classic() + # Clean theme
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.position = "none",
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  )

fileName <- paste0("log2FC_cdf_insulationBoundary_convSS_pe_EpiLC")
width <- 33*mmToInch
height <-33*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()
```

# [6] Visualization of heatmaps
https://bioconductor.org/books/devel/OHCA/pages/visualization.html
## Loops
```{r}
# Drawing parameters
hicDir <- here("../data/hic_pooled")
windowSize <- 1*1e6

# Importing loops
commonLoopDir <- here("../data/loop_analysis")
loop_chromo <- as_tibble(fread(here(commonLoopDir, "loopScore_chromosight.tsv")))

################################################################################
#Filtering loops to plot
comparison <- "ESC_vs_EpiLC"
note <- "strongInBoth_diff"

loop_chromo_filtered <- loop_chromo %>%
  dplyr::mutate(size = start2 - start1,
                score_avg_DMSO_EpiDMSO = (score_DMSO + score_EpiDMSO)/2,
                score_diff = score_EpiDMSO - score_DMSO,
                ooe_diff = ooe_EpiDMSO - ooe_DMSO) %>%
  dplyr::filter(score_DMSO > 0.5,
                score_EpiDMSO > 0.5,
                score_EpiDMSO > score_DMSO,
                ooe_EpiDMSO > ooe_DMSO,
                size < 500*1000,
                size > 100*1000) %>%
  dplyr::arrange(desc(ooe_diff))
temp <- loop_chromo_filtered
colnames(temp) <- c("V1", "V2", "V3", "V4", "V5", "V6")
loops <- importBedpe(temp)

# Visualization
# for(i in seq(1, nrow(temp))){
for(i in seq(1, 3)){
  ## Loading hic and plotting
  cf.G1DMSO <- HicFile(path = here(hicDir, "G1DMSO_pooled_allRes.hic"))
  cf.G1dTAG <- HicFile(path = here(hicDir, "G1dTAG_pooled_allRes.hic"))
  cf.EpiG1DMSO <- HicFile(path = here(hicDir, "EpiG1DMSO_pooled_allRes.hic"))
  cf.EpiG1dTAG <- HicFile(path = here(hicDir, "EpiG1dTAG_pooled_allRes.hic"))

  width <- panelSize(40/5)*mmToInch
  height <- panelSize(10/5)*mmToInch

  chr <- as_tibble(loops[i])$seqnames1
  center <- (as_tibble(loops[i])$start1 + as_tibble(loops[i])$end2)/2
  start <- floor(center - 0.5*windowSize)
  end <- floor(center + 0.5*windowSize)
  # Resolution 5kb
  for(resolution in c(5, 10)){
    res <- resolution*1000
    zmax <- 2.5

    hic1 <- import(cf.G1DMSO, focus = paste0(chr, ":", start, "-", end), resolution = res)
    hic2 <- import(cf.G1dTAG, focus = paste0(chr, ":", start, "-", end), resolution = res)
    hic3 <- import(cf.EpiG1DMSO, focus = paste0(chr, ":", start, "-", end), resolution = res)
    hic4 <- import(cf.EpiG1dTAG, focus = paste0(chr, ":", start, "-", end), resolution = res)
    p1 <- plotMatrix(hic1, dpi = 1000, limits = c(0, zmax), loop = loops)
    p2 <- plotMatrix(hic2, dpi = 1000, limits = c(0, zmax), loop = loops)
    p3 <- plotMatrix(hic3, dpi = 1000, limits = c(0, zmax), loop = loops)
    p4 <- plotMatrix(hic4, dpi = 1000, limits = c(0, zmax), loop = loops)

    fileName <- here(figDir, sprintf("visuzliation_%s_%s_%05d_%dkb", comparison, note, i, resolution))
    png(paste0(fileName, ".png"), res = 600, units = "in", width = width, height = height)
    print(cowplot::plot_grid(p1, p2, p3, p4, nrow = 1, align = "h"))
    dev.off()
  }
}


################################################################################
# Filtering loops to plot
comparison <- "ESC_vs_EpiLC"
note <- "strongInBoth_reg"

loop_chromo_filtered <- loop_chromo %>%
  dplyr::mutate(size = start2 - start1,
                score_avg_DMSO_EpiDMSO = (score_DMSO + score_EpiDMSO)/2) %>%
  dplyr::filter(score_DMSO > 0.5,
                score_EpiDMSO > 0.5,
                abs(score_DMSO - score_EpiDMSO) < 0.2,
                size < 500*1000,
                size > 100*1000) %>%
  dplyr::arrange(desc(score_avg_DMSO_EpiDMSO))
temp <- loop_chromo_filtered %>% dplyr::filter(Anno %in% c("P-P", "P-E", "E-E"))
colnames(temp) <- c("V1", "V2", "V3", "V4", "V5", "V6")
loops <- importBedpe(temp)

# Visualization
for(i in seq(1, nrow(temp))){
  ## Loading hic and plotting
  cf.G1DMSO <- HicFile(path = here(hicDir, "G1DMSO_pooled_allRes.hic"))
  cf.G1dTAG <- HicFile(path = here(hicDir, "G1dTAG_pooled_allRes.hic"))
  cf.EpiG1DMSO <- HicFile(path = here(hicDir, "EpiG1DMSO_pooled_allRes.hic"))
  cf.EpiG1dTAG <- HicFile(path = here(hicDir, "EpiG1dTAG_pooled_allRes.hic"))
  
  width <- panelSize(40/5)*mmToInch
  height <- panelSize(10/5)*mmToInch
  
  chr <- as_tibble(loops[i])$seqnames1
  center <- (as_tibble(loops[i])$start1 + as_tibble(loops[i])$end2)/2
  start <- floor(center - 0.5*windowSize)
  end <- floor(center + 0.5*windowSize)
  # Resolution 5kb
  for(resolution in c(5, 10)){
    res <- resolution*1000
    zmax <- 2.5
    
    hic1 <- import(cf.G1DMSO, focus = paste0(chr, ":", start, "-", end), resolution = res)
    hic2 <- import(cf.G1dTAG, focus = paste0(chr, ":", start, "-", end), resolution = res)
    hic3 <- import(cf.EpiG1DMSO, focus = paste0(chr, ":", start, "-", end), resolution = res)
    hic4 <- import(cf.EpiG1dTAG, focus = paste0(chr, ":", start, "-", end), resolution = res)
    p1 <- plotMatrix(hic1, dpi = 1000, limits = c(0, zmax), loop = loops)
    p2 <- plotMatrix(hic2, dpi = 1000, limits = c(0, zmax), loop = loops)
    p3 <- plotMatrix(hic3, dpi = 1000, limits = c(0, zmax), loop = loops)
    p4 <- plotMatrix(hic4, dpi = 1000, limits = c(0, zmax), loop = loops)
    
    fileName <- here(figDir, sprintf("visuzliation_%s_%s_%05d_%dkb", comparison, note, i, resolution))
    png(paste0(fileName, ".png"), res = 600, units = "in", width = width, height = height)
    print(cowplot::plot_grid(p1, p2, p3, p4, nrow = 1, align = "h"))
    dev.off()
  }
}

################################################################################
# Filtering loops to plot
comparison <- "ESC_vs_EpiLC"
note <- "strongInESC_reg"

loop_chromo_filtered <- loop_chromo %>%
  dplyr::mutate(size = start2 - start1,
                score_avg_DMSO_EpiDMSO = (score_DMSO + score_EpiDMSO)/2) %>%
  dplyr::filter(score_DMSO > 0.5,
                score_EpiDMSO > 0.5,
                (score_DMSO - score_EpiDMSO) > 0.2,
                size < 500*1000,
                size > 100*1000) %>%
  dplyr::arrange(desc(score_avg_DMSO_EpiDMSO))
temp <- loop_chromo_filtered %>% dplyr::filter(Anno %in% c("P-P", "P-E", "E-E"))
colnames(temp) <- c("V1", "V2", "V3", "V4", "V5", "V6")
loops <- importBedpe(temp)

# Visualization
for(i in seq(1, nrow(temp))){
  ## Loading hic and plotting
  cf.G1DMSO <- HicFile(path = here(hicDir, "G1DMSO_pooled_allRes.hic"))
  cf.G1dTAG <- HicFile(path = here(hicDir, "G1dTAG_pooled_allRes.hic"))
  cf.EpiG1DMSO <- HicFile(path = here(hicDir, "EpiG1DMSO_pooled_allRes.hic"))
  cf.EpiG1dTAG <- HicFile(path = here(hicDir, "EpiG1dTAG_pooled_allRes.hic"))
  
  width <- panelSize(40/5)*mmToInch
  height <- panelSize(10/5)*mmToInch
  
  chr <- as_tibble(loops[i])$seqnames1
  center <- (as_tibble(loops[i])$start1 + as_tibble(loops[i])$end2)/2
  start <- floor(center - 0.5*windowSize)
  end <- floor(center + 0.5*windowSize)
  # Resolution 5kb
  for(resolution in c(5, 10)){
    res <- resolution*1000
    zmax <- 2.5
    
    hic1 <- import(cf.G1DMSO, focus = paste0(chr, ":", start, "-", end), resolution = res)
    hic2 <- import(cf.G1dTAG, focus = paste0(chr, ":", start, "-", end), resolution = res)
    hic3 <- import(cf.EpiG1DMSO, focus = paste0(chr, ":", start, "-", end), resolution = res)
    hic4 <- import(cf.EpiG1dTAG, focus = paste0(chr, ":", start, "-", end), resolution = res)
    p1 <- plotMatrix(hic1, dpi = 1000, limits = c(0, zmax), loop = loops)
    p2 <- plotMatrix(hic2, dpi = 1000, limits = c(0, zmax), loop = loops)
    p3 <- plotMatrix(hic3, dpi = 1000, limits = c(0, zmax), loop = loops)
    p4 <- plotMatrix(hic4, dpi = 1000, limits = c(0, zmax), loop = loops)
    
    fileName <- here(figDir, sprintf("visuzliation_%s_%s_%05d_%dkb", comparison, note, i, resolution))
    png(paste0(fileName, ".png"), res = 600, units = "in", width = width, height = height)
    print(cowplot::plot_grid(p1, p2, p3, p4, nrow = 1, align = "h"))
    dev.off()
  }
}

################################################################################
# Filtering loops to plot
comparison <- "ESC_vs_EpiLC"
note <- "strongInESC_str"

loop_chromo_filtered <- loop_chromo %>%
  dplyr::mutate(size = start2 - start1,
                score_avg_DMSO_EpiDMSO = (score_DMSO + score_EpiDMSO)/2) %>%
  dplyr::filter(score_DMSO > 0.5,
                score_EpiDMSO > 0.5,
                (score_DMSO - score_EpiDMSO) > 0.2,
                size < 500*1000,
                size > 100*1000) %>%
  dplyr::arrange(desc(score_avg_DMSO_EpiDMSO))
temp <- loop_chromo_filtered %>% dplyr::filter(Anno %in% c("S-S", "S-X"))
colnames(temp) <- c("V1", "V2", "V3", "V4", "V5", "V6")
loops <- importBedpe(temp)

# Visualization
for(i in seq(1, nrow(temp))){
  ## Loading hic and plotting
  cf.G1DMSO <- HicFile(path = here(hicDir, "G1DMSO_pooled_allRes.hic"))
  cf.G1dTAG <- HicFile(path = here(hicDir, "G1dTAG_pooled_allRes.hic"))
  cf.EpiG1DMSO <- HicFile(path = here(hicDir, "EpiG1DMSO_pooled_allRes.hic"))
  cf.EpiG1dTAG <- HicFile(path = here(hicDir, "EpiG1dTAG_pooled_allRes.hic"))
  
  width <- panelSize(40/5)*mmToInch
  height <- panelSize(10/5)*mmToInch
  
  chr <- as_tibble(loops[i])$seqnames1
  center <- (as_tibble(loops[i])$start1 + as_tibble(loops[i])$end2)/2
  start <- floor(center - 0.5*windowSize)
  end <- floor(center + 0.5*windowSize)
  # Resolution 5kb
  for(resolution in c(5, 10)){
    res <- resolution*1000
    zmax <- 2.5
    
    hic1 <- import(cf.G1DMSO, focus = paste0(chr, ":", start, "-", end), resolution = res)
    hic2 <- import(cf.G1dTAG, focus = paste0(chr, ":", start, "-", end), resolution = res)
    hic3 <- import(cf.EpiG1DMSO, focus = paste0(chr, ":", start, "-", end), resolution = res)
    hic4 <- import(cf.EpiG1dTAG, focus = paste0(chr, ":", start, "-", end), resolution = res)
    p1 <- plotMatrix(hic1, dpi = 1000, limits = c(0, zmax), loop = loops)
    p2 <- plotMatrix(hic2, dpi = 1000, limits = c(0, zmax), loop = loops)
    p3 <- plotMatrix(hic3, dpi = 1000, limits = c(0, zmax), loop = loops)
    p4 <- plotMatrix(hic4, dpi = 1000, limits = c(0, zmax), loop = loops)
    
    fileName <- here(figDir, sprintf("visuzliation_%s_%s_%05d_%dkb", comparison, note, i, resolution))
    png(paste0(fileName, ".png"), res = 600, units = "in", width = width, height = height)
    print(cowplot::plot_grid(p1, p2, p3, p4, nrow = 1, align = "h"))
    dev.off()
  }
}


################################################################################
# Filtering loops to plot
comparison <- "ESC_vs_EpiLC"
note <- "strongInESC2_str"

loop_chromo_filtered <- loop_chromo %>%
  dplyr::mutate(size = start2 - start1,
                score_avg_DMSO_EpiDMSO = (score_DMSO + score_EpiDMSO)/2) %>%
  dplyr::filter(score_DMSO > 0.5,
                score_EpiDMSO < 0.3,
                size < 500*1000,
                size > 100*1000) %>%
  dplyr::arrange(desc(score_avg_DMSO_EpiDMSO))
temp <- loop_chromo_filtered %>% dplyr::filter(Anno %in% c("S-S", "S-X"))
colnames(temp) <- c("V1", "V2", "V3", "V4", "V5", "V6")
loops <- importBedpe(temp)

# Visualization
for(i in seq(1, nrow(temp))){
  ## Loading hic and plotting
  cf.G1DMSO <- HicFile(path = here(hicDir, "G1DMSO_pooled_allRes.hic"))
  cf.G1dTAG <- HicFile(path = here(hicDir, "G1dTAG_pooled_allRes.hic"))
  cf.EpiG1DMSO <- HicFile(path = here(hicDir, "EpiG1DMSO_pooled_allRes.hic"))
  cf.EpiG1dTAG <- HicFile(path = here(hicDir, "EpiG1dTAG_pooled_allRes.hic"))
  
  width <- panelSize(40/5)*mmToInch
  height <- panelSize(10/5)*mmToInch
  
  chr <- as_tibble(loops[i])$seqnames1
  center <- (as_tibble(loops[i])$start1 + as_tibble(loops[i])$end2)/2
  start <- floor(center - 0.5*windowSize)
  end <- floor(center + 0.5*windowSize)
  # Resolution 5kb
  for(resolution in c(5, 10)){
    res <- resolution*1000
    zmax <- 2.5
    
    hic1 <- import(cf.G1DMSO, focus = paste0(chr, ":", start, "-", end), resolution = res)
    hic2 <- import(cf.G1dTAG, focus = paste0(chr, ":", start, "-", end), resolution = res)
    hic3 <- import(cf.EpiG1DMSO, focus = paste0(chr, ":", start, "-", end), resolution = res)
    hic4 <- import(cf.EpiG1dTAG, focus = paste0(chr, ":", start, "-", end), resolution = res)
    p1 <- plotMatrix(hic1, dpi = 1000, limits = c(0, zmax), loop = loops)
    p2 <- plotMatrix(hic2, dpi = 1000, limits = c(0, zmax), loop = loops)
    p3 <- plotMatrix(hic3, dpi = 1000, limits = c(0, zmax), loop = loops)
    p4 <- plotMatrix(hic4, dpi = 1000, limits = c(0, zmax), loop = loops)
    
    fileName <- here(figDir, sprintf("visuzliation_%s_%s_%05d_%dkb", comparison, note, i, resolution))
    png(paste0(fileName, ".png"), res = 600, units = "in", width = width, height = height)
    print(cowplot::plot_grid(p1, p2, p3, p4, nrow = 1, align = "h"))
    dev.off()
  }
}

```
## Microcompartments
```{r}
# Drawing parameters
hicDir <- here("../data/hic_pooled")
windowSize <- 1*1e6

# Importing loops
microcompartment <- importBedpe(fread(here("..", "reference", "microcompartment", "MicrocompartmentLoops_PlusMin1kb.bedpe")))

comparison <- "microcompartment"
note <- "Gopel"
# Visualization
# for(i in seq(1, nrow(temp))){
for(i in seq(1, 3)){
  ## Loading hic and plotting
  cf.G1DMSO <- HicFile(path = here(hicDir, "G1DMSO_pooled_allRes.hic"))
  cf.G1dTAG <- HicFile(path = here(hicDir, "G1dTAG_pooled_allRes.hic"))

  width <- panelSize(20/5)*mmToInch
  height <- panelSize(10/5)*mmToInch

  chr <- as_tibble(microcompartment[i])$seqnames1
  center <- (as_tibble(microcompartment[i])$start1 + as_tibble(microcompartment[i])$end2)/2
  start <- floor(center - 0.5*windowSize)
  end <- floor(center + 0.5*windowSize)
  # Resolution 5kb
  for(resolution in c(5, 10)){
    res <- resolution*1000
    zmax <- 2.5

    hic1 <- import(cf.G1DMSO, focus = paste0(chr, ":", start, "-", end), resolution = res)
    hic2 <- import(cf.G1dTAG, focus = paste0(chr, ":", start, "-", end), resolution = res)
    p1 <- plotMatrix(hic1, dpi = 1000, limits = c(0, zmax), loop = microcompartment)
    p2 <- plotMatrix(hic2, dpi = 1000, limits = c(0, zmax), loop = microcompartment)

    fileName <- here(figDir, sprintf("visuzliation_%s_%s_%05d_%dkb", comparison, note, i, resolution))
    png(paste0(fileName, ".png"), res = 600, units = "in", width = width, height = height)
    print(cowplot::plot_grid(p1, p2, nrow = 1, align = "h"))
    dev.off()
  }
}
```


# [7] EpiLC Micro-C & ChIP-seq analysis
## Gene-centric approach
### Linking differential ChIP-seq to gene
```{r}
# IMPORTING PEAKS
chipDir <- here(refDir, "ChIP")

peak.H3K4me3.EpiLC <- makeGRangesFromDataFrame(fread(
  here(chipDir, "differential", "H3K4me3_diff_specific_EpiLC.bed"),
  select = 1:3,
  col.names = c("chr", "start", "end")
))

peak.H3K27ac.EpiLC <- makeGRangesFromDataFrame(fread(
  here(chipDir, "differential", "H3K27ac_diff_specific_EpiLC.bed"),
  select = 1:3,
  col.names = c("chr", "start", "end")
))
```

### H3K4me3 TSS genes
```{r}
# Linking H3K4me3 with TSS
flankSize <- 2500
gene.tb <- fread(here(refDir, "mm10", "gencode.vM25.mm10.GRCm38.p6_gene_sorted.bed")) %>%
  dplyr::mutate(TSS = ifelse(V4 == "+", V2, V3),
                TSSstart = TSS - flankSize,
                TSSend = TSS + flankSize)
gene.gr <- makeGRangesFromDataFrame(gene.tb %>%
                                      dplyr::rename(
                                        seqnames = V1,
                                        start = TSSstart,
                                        end = TSSend
                                      ) %>%
                                      dplyr::select(seqnames, start, end))

overlaps <- findOverlaps(gene.gr, peak.H3K4me3.EpiLC)
gene.filtered.tb <- gene.tb[queryHits(overlaps)]

fwrite(gene.filtered.tb, here(dataDir, "gene_H3K4me3_diff_specific_EpiLC.tsv"), sep = "\t", col.names = TRUE)

geneList.H3K4me3 <- gene.filtered.tb$V5
```

### H3K27ac genes
#### Anchor with P
```{r}
# Linking H3K27ac with TSS
loop.tb <- fread(here(dataDir, "loopScore_chromosight.tsv"))
loop.gr <- loop.tb %>% dplyr::select(seq(1, 6))
colnames(loop.gr) <- c("V1", "V2", "V3", "V4", "V5", "V6")
loop.gr <- importBedpe(loop.gr)

index1 <- queryHits(findOverlaps(first(loop.gr), peak.H3K27ac.EpiLC))
index2 <- queryHits(findOverlaps(second(loop.gr), peak.H3K27ac.EpiLC))

loop.filtered1.tb <- loop.tb[index1] %>% dplyr::filter(str_detect(A1, "E"))
loop.filtered2.tb <- loop.tb[index2] %>% dplyr::filter(str_detect(A2, "E"))

loop.filtered.tb <- distinct(bind_rows(loop.filtered1.tb, loop.filtered2.tb))


loop.filtered.hasP.tb <- loop.filtered.tb %>% dplyr::filter(str_detect(Anno, "P"))
loop.filtered.noP.tb <- loop.filtered.tb %>% dplyr::filter(!str_detect(Anno, "P"))


# annotating hasP
geneAnnoData <- loadLoopAnnoData(here(commonLoopDir, paste0("loopScore_", "chromosight", "_p-n_complex_ensemblList.tsv")))
temp <- (geneAnnoData %>% dplyr::filter(id %in% loop.filtered.hasP.tb$id))
geneList.hasP <- unlist(temp$gene)
```
#### Anchor without P
##### Strategy 1, assignign nonP reg loops to nearest expressed genes regardless of TAD domain.
```{r}
# Import TAD
# tad <- fread(here(dataDir, "../TAD", "TAD_25kb_125kb_otsu_G1DMSO_pooled.bed"))
# colnames(tad) <- c("chr", "start", "end")
# tad <- tad %>% dplyr::mutate(tadID = paste(chr, start, end, sep = "_"))
# tad.gr <- makeGRangesFromDataFrame(tad, 
#                                    seqnames.field = "chr",
#                                    start.field = "start",
#                                    end.field = "end")
# names(tad.gr) <- tad$tadID

# Import gene & filter for expressed genes
diffDir <- here("../data/RNA_diff")
diff.RNA.ESC <- fread(here(diffDir, "diff_G1.dTAG_G1.2i.dTAG_vs_G1.2i.DMSO.tsv"))
diff.RNA.EpiLC <- fread(here(diffDir, "diff_G1.dTAG_G1.Epi.dTAG_vs_G1.Epi.DMSO.tsv"))
geneList <- unique(c(diff.RNA.ESC$ensembl_gene_id, diff.RNA.EpiLC$ensembl_gene_id))
gene.expressed.tb <- gene.tb %>% dplyr::filter(V5 %in% geneList)
gene.expressed.gr <- makeGRangesFromDataFrame(gene.expressed.tb, 
                                   seqnames.field = "V1",
                                   start.field = "TSSstart",
                                   end.field = "TSSend",
                                   keep.extra.columns = TRUE)

# Loops
anchor1.gr <- makeGRangesFromDataFrame(loop.filtered.noP.tb,
                                       seqnames.field = "chrom1",
                                       start.field = "start1",
                                       end.field = "end1",
                                       keep.extra.columns = TRUE)

anchor2.gr <- makeGRangesFromDataFrame(loop.filtered.noP.tb,
                                       seqnames.field = "chrom2",
                                       start.field = "start2",
                                       end.field = "end2",
                                       keep.extra.columns = TRUE)

# Step 4: Find the Nearest Gene for Each Loop Anchor
# ------------------------------------------------
# The distanceToNearest function efficiently finds the closest feature in another GRanges object.
# It automatically respects chromosome boundaries (i.e., won't link chr1 to chr2).

# Find the nearest gene to the first anchor of each loop
dist1 <- distanceToNearest(anchor1.gr, gene.expressed.gr)
# Find the nearest gene to the second anchor of each loop
dist2 <- distanceToNearest(anchor2.gr, gene.expressed.gr)

# Step 5: Determine the Overall Nearest Gene for the Loop
# ------------------------------------------------
# For each loop, we compare the distance from anchor1 vs. anchor2 and pick the smaller one.

# Extract distances and gene indices from the results
dist1_vals <- mcols(dist1)$distance
dist2_vals <- mcols(dist2)$distance

# Create a results data frame
nearest_gene_results <- tibble(
  loop_id = anchor1.gr$id,
  # Choose the distance that is smaller
  distance = pmin(dist1_vals, dist2_vals, na.rm = TRUE),
  # Based on which distance was smaller, get the corresponding gene ID
  nearest_gene = ifelse(
    dist1_vals <= dist2_vals,
    gene.expressed.gr$V5[subjectHits(dist1)],
    gene.expressed.gr$V5[subjectHits(dist2)]
  )
) %>%
  dplyr::filter(distance < 500000)

# Step 6: Combine and Display Final Results
# ------------------------------------------------
# Join the nearest gene results back to the original loop data frame.

final_annotated_loops <- loop.filtered.noP.tb %>%
  left_join(nearest_gene_results, by = c("id" = "loop_id"))

geneList.noP.strat1 <- (final_annotated_loops %>% dplyr::filter(!is.na(nearest_gene)))$nearest_gene

# geneList <- unique(c(geneList1, geneList2))
# 
# out <- tibble(ensembl = geneList)
# fwrite(out, here(dataDir, "gene_H3K27ac_diff_specific_EpiLC.tsv"), sep = "\t")
```

##### Strategy 2, Find connected network and use the genes annotated to promoter within the network
```{r}
library(igraph)

# To be able to construct a graph, each anchor need to have same resolution.
# Expand the size to 25 kb

dataToGraph <- loop.tb %>% dplyr::mutate(binSize = end1 - start1,
                adj_start1 = ifelse(binSize == 25000, start1,
                                    floor(start1/25000)*25000),
                adj_end1 = adj_start1 + 25000,
                adj_start2 = ifelse(binSize == 25000, start2,
                                    floor(start2/25000)*25000),
                adj_end2 = adj_start2 + 25000,
                anchor1 = paste(chrom1, adj_start1, adj_end1, sep = "_"),
                anchor2 = paste(chrom2, adj_start2, adj_end2, sep = "_")) %>%
  dplyr::select(c(1, 2, 3, 4, 5, 6, 7, 47, 48))


graph <- graph_from_data_frame(dataToGraph %>% dplyr::select(anchor1, anchor2))
components <- components(graph)
dataToGraph$hub <- components$membership[dataToGraph$anchor1]

hubInfo <- dataToGraph %>% dplyr::select(c(id, hub))

hubAnno <- geneAnnoData %>% dplyr::select(id, gene) %>% left_join(hubInfo, by = c("id")) %>%
  group_by(hub) %>%
  summarise(gene = list(unlist(gene)))


geneList.noP.strat2 <- unlist((loop.filtered.noP.tb %>% left_join(hubInfo, by = c("id")) %>% left_join(hubAnno, by = c("hub")))$gene)

```

#### Merge and export gene list
```{r}
venn_list <- list(
  "K27ac Has Promoter" = geneList.hasP,
  "K27ac No Promoter (Strat 1)" = geneList.noP.strat1,
  "K27ac No Promoter (Strat 2)" = geneList.noP.strat2,
  "H3K4me3" = geneList.H3K4me3
)

venn.diagram(
  x = venn_list,
  
  # Output file
  filename = 'venn_diagram.png',
  output = TRUE,
  
  # Circles
  lwd = 2,                             # Border thickness
  lty = 'blank',                       # Border line type (blank for solid fill)
  fill = c("#440154ff", "#31688eff", "#35b779ff", "#fde725ff"), # Colors for circles
  
  # Numbers
  cex = 1.2,                           # Size of the numbers in intersections
  fontface = "bold",                   # Font style for numbers
  fontfamily = "sans",                 # Font family for numbers
  
  # Category Names (Labels)
  cat.cex = 1.2,                         # Size of the category names
  cat.fontface = "bold",                 # Font style for category names
  cat.fontfamily = "sans",               # Font family for category names
  cat.default.pos = "outer",             # Position of the labels
  cat.dist = c(0.055, 0.055, 0.055, 0.055), # Distance of labels from circles
  
  # Title
  main = "Gene List Overlap",
  main.cex = 1.5,
  main.fontface = "bold",
  main.fontfamily = "sans"
)

```

```{r}
fwrite(gene.tb %>% dplyr::filter(V5 %in% c(geneList.hasP, geneList.noP.strat1)),
       here(dataDir, "gene_H3K27ac_diff_specific_EpiLC_strategy1.tsv"), sep = "\t")

fwrite(gene.tb %>% dplyr::filter(V5 %in% c(geneList.hasP, geneList.noP.strat2)),
       here(dataDir, "gene_H3K27ac_diff_specific_EpiLC_strategy2.tsv"), sep = "\t")

fwrite(gene.tb %>% dplyr::filter(V5 %in% c(geneList.H3K4me3, geneList.hasP, geneList.noP.strat2)),
       here(dataDir, "gene_both_diff_specific_EpiLC_strategy2.tsv"), sep = "\t")


fwrite(gene.tb %>% dplyr::filter(V5 %in% c(geneList.H3K4me3, geneList.hasP, geneList.noP.strat2)),
       here(dataDir, "gene_both_diff_specific_EpiLC_strategy2.tsv"), sep = "\t")

```
#### Statistical analysis
```{r}
alpha <- 0.05
fcCutoff <- 0.5
diffCutoff <- 0.2
commonLoopDir <- here("../data/loop_analysis")
diffDir <- here("../data/RNA_diff")
name <- "chromosight"

diff.RNA <- fread(here(diffDir, "diff_G1.dTAG_G1.Epi.dTAG_vs_G1.Epi.DMSO.tsv")) %>%
  dplyr::select(ensembl_gene_id, log2FoldChange, shrinked_log2FC, padj, external_gene_name)

runFisher <- function(diff.RNA, geneList){
  # geneList <- unique(c(geneList.H3K4me3))
  
  n.total <- nrow(diff.RNA)
  n.epi <- length(geneList)
  n.nonEpi <- n.total - n.epi  
  
  diff.filtered <- diff.RNA %>% dplyr::filter(padj < alpha)
  n.total.sig.p <- nrow(diff.filtered)
  n.epi.sig.p <- nrow(diff.filtered %>% dplyr::filter(ensembl_gene_id %in% geneList))
  n.nonEpi.sig.p <- n.total.sig.p - n.epi.sig.p
  
  diff.filtered <- diff.RNA %>% dplyr::filter(padj < alpha, shrinked_log2FC < -fcCutoff)
  n.total.sig.pfc <- nrow(diff.filtered)
  n.epi.sig.pfc <- nrow(diff.filtered %>% dplyr::filter(ensembl_gene_id %in% geneList))
  n.nonEpi.sig.pfc <- n.total.sig.pfc - n.epi.sig.pfc
  
  
  contingency_table.p <- matrix(c(n.epi.sig.p, n.epi - n.epi.sig.p,
                                  n.nonEpi.sig.p, n.nonEpi - n.nonEpi.sig.p), nrow = 2, byrow = TRUE)
  colnames(contingency_table.p) <- c("Significant", "Not significant")
  rownames(contingency_table.p) <- c("Epi genes", "Non Epi genes")
  fisher_test_result.p <- fisher.test(contingency_table.p)
  
  
  contingency_table.pfc <- matrix(c(n.epi.sig.pfc, n.epi - n.epi.sig.pfc,
                                    n.nonEpi.sig.pfc, n.nonEpi - n.nonEpi.sig.pfc), nrow = 2, byrow = TRUE)
  colnames(contingency_table.pfc) <- c("Significant", "Not significant")
  rownames(contingency_table.pfc) <- c("Epi genes", "Non Epi genes")
  fisher_test_result.pfc <- fisher.test(contingency_table.pfc)
  
  print(contingency_table.p)
  print(fisher_test_result.p)
  print(contingency_table.pfc)
  print(fisher_test_result.pfc)
  
}

runFisher(diff.RNA, unique(c(geneList.H3K4me3)))
runFisher(diff.RNA, unique(c(geneList.H3K4me3, geneList.hasP, geneList.noP.strat2)))
runFisher(diff.RNA, unique(c(geneList.H3K4me3)))


################################################################################


ggplot(diff.RNA %>% dplyr::filter(padj < alpha) %>% mutate(flag = if_else(ensembl_gene_id %in% unique(c(geneList.H3K4me3, geneList.hasP, geneList.noP.strat2)), "Epi", "noEpi")), aes(x = flag, y = shrinked_log2FC)) +
  geom_violin()





```
### H3K27ac gene with different strategy
```{r}
# 1. Link peaks to genes by distance (within 10, 50, 100 or 250kb from TSS) or by looping (present in any direct loop from their promoters), 2. Focus on the cosehin sensitive DEGs (from acute or prolonged teratment (either nipbl or rad21) and calculate the percentages that have at least one new H3K27ac peak.

for(flankSize in c(10000, 50000, 100000, 250000, 500000)){
  
# flankSize <- 10000
# Link peaks to gene by distance from TSS
gene.tb <- fread(here(refDir, "mm10", "gencode.vM25.mm10.GRCm38.p6_gene_sorted.bed")) %>%
  dplyr::mutate(TSS = ifelse(V4 == "+", V2, V3),
                TSSstart = TSS - flankSize,
                TSSend = TSS + flankSize)
gene.gr <- makeGRangesFromDataFrame(gene.tb %>%
                                      dplyr::rename(
                                        seqnames = V1,
                                        start = TSSstart,
                                        end = TSSend
                                      ) %>%
                                      dplyr::select(seqnames, start, end))

overlaps <- findOverlaps(gene.gr, peak.H3K27ac.EpiLC)
geneList <- (gene.tb[queryHits(overlaps)])$V5

# Check how many of the genes are cohesin sensitive
diff.tb <- fread(here(extDir, "data", "diff_G1.dTAG_G1.Epi.dTAG_vs_G1.Epi.DMSO.tsv")) %>%
  dplyr::filter(padj < 0.05, shrinked_log2FC < 0)
n1 <- nrow(diff.tb)
n2 <- nrow(diff.tb %>% dplyr::filter(ensembl_gene_id %in% geneList))
diff.tb <- fread(here(extDir, "data", "diff_G1.dTAG_G1.Epi.dTAG_vs_G1.Epi.DMSO.tsv")) %>%
  dplyr::filter(padj < 0.05, shrinked_log2FC < -0.5)
n3 <- nrow(diff.tb)
n4 <- nrow(diff.tb %>% dplyr::filter(ensembl_gene_id %in% geneList))
print(paste0("Flank size: ", flankSize/1000, "kb"))
print(paste0("EpiLC down DEG (", n2, "/", n1, ", ", round(n2/n1*100, digits = 1), "%), strong down deg (", n4, "/", n3, ", ", round(n2/n1*100, digits = 1), "%)"))

diff.tb <- fread(here(extDir, "data", "diff_G1.dTAG_G1.2i.dTAG_vs_G1.2i.DMSO.tsv")) %>%
  dplyr::filter(padj < 0.05, shrinked_log2FC < 0)
n1 <- nrow(diff.tb)
n2 <- nrow(diff.tb %>% dplyr::filter(ensembl_gene_id %in% geneList))
diff.tb <- fread(here(extDir, "data", "diff_G1.dTAG_G1.2i.dTAG_vs_G1.2i.DMSO.tsv")) %>%
  dplyr::filter(padj < 0.05, shrinked_log2FC < -0.5)
n3 <- nrow(diff.tb)
n4 <- nrow(diff.tb %>% dplyr::filter(ensembl_gene_id %in% geneList))
print(paste0("ESC down DEG (", n2, "/", n1, ", ", round(n2/n1*100, digits = 1), "%), strong down deg (", n4, "/", n3, ", ", round(n2/n1*100, digits = 1), "%)"))

# Long term perturbed
geneListPerturbed <- c((fread(here("..", "RNA_241223_H202SC24124443_RAD21_EpiRecovery", "pub_EpiUP_DEGfiltered_kmeans_group2.tsv" )) %>%
                 dplyr::filter(group == "K1"))$gene,
              (fread(here("..", "RNA_241223_H202SC24124443_RAD21_EpiRecovery", "pub_EpiDOWN_DEGfiltered_kmeans_group2.tsv" )) %>%
                  dplyr::filter(group == "K1"))$gene)
n1 <- length(geneListPerturbed)      
n2 <- length(geneListPerturbed[geneListPerturbed %in% geneList])
print(paste0("Long term DOWN DEG (", n2, "/", n1, ", ", round(n2/n1*100, digits = 1), "%)"))
}
```

## Loop-centric approach
### APA on EpiLC only called loops
```{r}
# diffCutoff <- 0.2
# data <- fread(here(commonLoopDir, "loopScore_chromosight_extra.tsv"))
# 
# temp <- data %>% dplyr::filter((!DMSOcalledProcessed) & (EpiDMSOcalledProcessed)) %>%
#   dplyr::mutate(density = get_density(score_DMSO, score_EpiDMSO, n = 200))
# 
#   density_limits <- range(temp$density, na.rm = TRUE)
#   
#   bedpeUP <- temp %>% dplyr::select(chrom1, start1, end1, chrom2, start2, end2)
#   fwrite(bedpeUP, here(loopDir, "temp_EpiLCDMSO_onlyCalled.bedpe"), sep = "\t", col.names = FALSE)
#   
#   # ------ Scatterplot ------ 
#   p1 <- ggplot(temp, aes(x = score_DMSO, y = score_EpiDMSO)) +
#     # Layer 1: Plot "NO" points with a grey scale
#     geom_point(data = . %>% arrange(density),
#                aes(color = density), size = 0.1) +
#     # Add annotations and theme elements
#     xlim(-0.5, 1) + ylim(-0.5, 1) + coord_fixed() +
#     geom_abline(slope = 1, intercept = 0, col = "grey50", linetype = "dashed",
#                 color = "black",
#                 size = lineThick*mmToLineUnit,
#                 lineend = "square") +
#     geom_abline(slope = 1, intercept = -diffCutoff, col = "grey", linetype = "dotted",
#                 color = "black",
#                 size = lineThick*mmToLineUnit,
#                 lineend = "square") +
#     geom_abline(slope = 1, intercept = diffCutoff, col = "grey", linetype = "dotted",
#                 color = "black",
#                 size = lineThick*mmToLineUnit,
#                 lineend = "square") +
#     geom_hline(yintercept = 0, alpha = 0.5, color = "grey",
#                color = "black",
#                size = lineThick*mmToLineUnit,
#                lineend = "square") +
#     geom_vline(xintercept = 0, alpha = 0.5, color = "grey",
#                color = "black",
#                size = lineThick*mmToLineUnit,
#                lineend = "square") +
#     theme_classic() +
#     theme(
#       legend.position = "none",
#       plot.title = element_text(
#         hjust = 0.5,
#         size = fontSizeS,
#         family = fontType
#       ),
#       axis.title = element_text(
#         size = fontSizeS,
#         family = fontType,
#         color = "#000000"
#       ),
#       axis.text = element_text(
#         size = fontSizeS,
#         family = fontType,
#         color = "#000000"
#       ),
#       axis.line = element_line(
#         color = "#000000",
#         size = lineThick*mmToLineUnit,
#         lineend = "square"
#       ),
#       axis.ticks = element_line(
#         color = "#000000",
#         size = lineThick*mmToLineUnit,
#         lineend = "square"
#       ),
#       panel.background = element_rect(fill = "transparent"),
#     )
#   width <- panelSize(1.5)*mmToInch
#   height <- panelSize(1.5)*mmToInch  
#   
#   name = "EpiLCDMSO_Onlycalled"
#   fileName <- paste0("scatterplot_", name)
#   
#   png(here(figDir, paste0(fileName, ".png")), 
#       width = width, height = height, units = "in", res = 600)
#   print(p1)
#   dev.off()

```
### Filtering loops associated with differential chip-seq peaks (both H3K4me3 and H3K27ac)
```{r}

# --- 1. PROGRAMMATICALLY IMPORT PEAKS ---
chipDir <- here("..", "reference", "ChIP")

# Create a list of peak files and generate clean names
peak_files <- list.files(here(chipDir, "differential"),
                         pattern = "specific.*\\.bed$", full.names = TRUE)
peak_names <- basename(peak_files) %>%
  gsub("_diff_specific_", ".", .) %>%
  gsub("\\.bed$", "", .)

# Read all peak files into a named list of GRanges objects
peak_list <- lapply(peak_files, function(f) {
  makeGRangesFromDataFrame(fread(f, select = 1:3, col.names = c("chr", "start", "end")))
}) %>% setNames(peak_names)

# Combine peaks for analysis
peak.Epi <- c(peak_list$H3K27ac.EpiLC, peak_list$H3K4me3.EpiLC)
peak.ESC <- c(peak_list$H3K27ac.ESC, peak_list$H3K4me3.ESC)




# --- 2. IMPORT AND FILTER LOOPS ---
data <- fread(here(commonLoopDir, "loopScore_chromosight_extra.tsv"))

loop <- data %>% dplyr::select(seq(1, 6))
colnames(loop) <- c("V1", "V2", "V3", "V4", "V5", "V6")
loop <- importBedpe(loop)


# Helper function to find loops overlapping with peaks
find_overlapping_loops <- function(loops, peaks) {
  idx <- unique(c(
    queryHits(findOverlaps(first(loops), peaks)),
    queryHits(findOverlaps(second(loops), peaks))
  ))
  loops[idx]
}

loop.Epi <- find_overlapping_loops(loop, peak.Epi)
loop.ESC <- find_overlapping_loops(loop, peak.ESC)

fwrite(as_tibble(loop.Epi) %>% dplyr::select(c(1, 2, 3, 6, 7, 8)), here(dataDir, "loop_ChIP_diff_EpiLCspecific.bedpe"), sep = "\t", col.names = FALSE)
fwrite(as_tibble(loop.ESC) %>% dplyr::select(c(1, 2, 3, 6, 7, 8)), here(dataDir, "loop_ChIP_diff_ESCspecific.bedpe"), sep = "\t", col.names = FALSE)

# Generate IDs for joining
temp.Epi <- as_tibble(loop.Epi) %>%
  mutate(id = paste(seqnames1, start1-1, end1, seqnames2, start2-1, end2, sep = "_"))
loopIDList.Epi <- temp.Epi$id

temp.ESC <- as_tibble(loop.ESC) %>%
  mutate(id = paste(seqnames1, start1-1, end1, seqnames2, start2-1, end2, sep = "_"))
loopIDList.ESC <- temp.ESC$id

# --- 3. REUSABLE PLOTTING FUNCTION & THEME ---
# Define a common theme to reduce code duplication
common_theme <- theme_classic() +
  theme(
    legend.position = "none",
    axis.title = element_text(size = fontSizeS, family = fontType, color = "black"),
    axis.text = element_text(size = fontSizeS, family = fontType, color = "black"),
    axis.line = element_line(color = "black", size = lineThick * mmToLineUnit, lineend = "square"),
    axis.ticks = element_line(color = "black", size = lineThick * mmToLineUnit, lineend = "square"),
    panel.background = element_rect(fill = "transparent")
  )

# Function to generate and save scatterplots
generate_scatterplot <- function(df, x_var, y_var, name, limits = c(-0.5, 1), diff_cutoff = 0.2) {
  plot_data <- df %>%
    mutate(density = get_density(.data[[x_var]], .data[[y_var]], n = 200))

  p <- ggplot(plot_data, aes(x = .data[[x_var]], y = .data[[y_var]])) +
    geom_point(aes(color = density), size = 0.1, data = . %>% arrange(density)) +
    coord_fixed() +
    scale_x_continuous(limits = limits) +
    scale_y_continuous(limits = limits) +
    geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black", size = lineThick*mmToLineUnit) +
    geom_abline(slope = 1, intercept = c(-diff_cutoff, diff_cutoff), linetype = "dotted", color = "black", size = lineThick*mmToLineUnit) +
    geom_hline(yintercept = 0, alpha = 0.5, color = "black", size = lineThick*mmToLineUnit) +
    geom_vline(xintercept = 0, alpha = 0.5, color = "black", size = lineThick*mmToLineUnit) +
    common_theme

  # Save the plot
  ggsave(
    filename = here(figDir, paste0("scatterplot_", name, ".png")),
    plot = p,
    width = panelSize(1.5) * mmToInch,
    height = panelSize(1.5) * mmToInch,
    dpi = 600
  )
}

# --- 4. GENERATE PLOTS ---

# Generate the first three scatterplots
data_to_plot <- data %>% dplyr::filter(id %in% loopIDList.Epi)
generate_scatterplot(data_to_plot, "score_DMSO",    "score_EpiDMSO", "ChIP_diff_EpiLCspecific_EpiLCvsESC")
generate_scatterplot(data_to_plot, "score_DMSO",    "score_dTAG",    "ChIP_diff_EpiLCspecific_ESCdTAGvsESCDMSO")
generate_scatterplot(data_to_plot, "score_EpiDMSO", "score_EpidTAG", "ChIP_diff_EpiLCspecific_EpiLCdTAGvsEpiLCDMSO")
data_to_plot <- data %>% dplyr::filter(id %in% loopIDList.ESC)
generate_scatterplot(data_to_plot, "score_DMSO",    "score_EpiDMSO", "ChIP_diff_ESCspecific_EpiLCvsESC")
generate_scatterplot(data_to_plot, "score_DMSO",    "score_dTAG",    "ChIP_diff_ESCspecific_ESCdTAGvsESCDMSO")
generate_scatterplot(data_to_plot, "score_EpiDMSO", "score_EpidTAG", "ChIP_diff_ESCspecific_EpiLCdTAGvsEpiLCDMSO")


# Prepare data for O/E analysis
pseudo_ooe <- 0.01
## EpiLC
data_to_plot <- data %>% dplyr::filter(id %in% loopIDList.Epi)
plot_df_ooe <- data_to_plot %>%
  dplyr::select(id, ooe_DMSO, ooe_dTAG, ooe_EpiDMSO, ooe_EpidTAG) %>%
  na.omit() %>%
  mutate(
    across(starts_with("ooe_"), ~ if_else(. == 0, pseudo_ooe, .)),
    fc_ESC = log2(ooe_dTAG / ooe_DMSO),
    fc_Epi = log2(ooe_EpidTAG / ooe_EpiDMSO)
  )

# --- Violin Plot ---
plot_df_ooe_long <- plot_df_ooe %>%
  select(fc_ESC, fc_Epi) %>%
  pivot_longer(everything(), names_to = "condition", values_to = "value") %>%
  mutate(condition = factor(condition, levels = c("fc_ESC", "fc_Epi")))

p_violin <- ggplot(plot_df_ooe_long, aes(x = condition, y = value, fill = condition)) +
  geom_violin(linewidth = lineThick * mmToLineUnit, alpha = .4, show.legend = FALSE) +
  geom_boxplot(width = 0.3, outlier.shape = NA, alpha = 0.6, show.legend = FALSE, linewidth = lineThick * mmToLineUnit) +
  stat_summary(fun = mean, geom = "point", shape = 21, size = 0.5, fill = "black") +
  geom_hline(yintercept = c(-0.5, 0.5), linetype = "dashed", alpha = 0.5, color = "black", size = lineThick * mmToLineUnit) +
  geom_hline(yintercept = 0, color = "black", size = lineThick * mmToLineUnit) +
  coord_cartesian(ylim = c(-3, 3)) +
  scale_fill_manual(values = c(strong_red, strong_blue)) +
  labs(y = "log2 fc of obs/exp") +
  common_theme +
  theme(
    axis.title.x = element_blank(),
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = fontSizeM),
    axis.title.y = element_text(size = fontSizeM)
  )

svglite(here(figDir, "violin_ChIP_diff_EpiLCspecific_EpiLCvsESC.svg"), height = panelSize(1.5)*mmToInch, width = panelSize(1.5)*mmToInch)
print(p_violin)
dev.off()

# --- Final Scatterplot ---
generate_scatterplot(
  df = plot_df_ooe,
  x_var = "fc_ESC",
  y_var = "fc_Epi",
  name = "ChIP_diff_EpiLCspecific_EpiLCvsESC",
  limits = NULL, # Use default data limits
  diff_cutoff = 0.5
)


## ESC
data_to_plot <- data %>% dplyr::filter(id %in% loopIDList.ESC)
plot_df_ooe <- data_to_plot %>%
  dplyr::select(id, ooe_DMSO, ooe_dTAG, ooe_EpiDMSO, ooe_EpidTAG) %>%
  na.omit() %>%
  mutate(
    across(starts_with("ooe_"), ~ if_else(. == 0, pseudo_ooe, .)),
    fc_ESC = log2(ooe_dTAG / ooe_DMSO),
    fc_Epi = log2(ooe_EpidTAG / ooe_EpiDMSO)
  )

# --- Violin Plot ---
plot_df_ooe_long <- plot_df_ooe %>%
  select(fc_ESC, fc_Epi) %>%
  pivot_longer(everything(), names_to = "condition", values_to = "value") %>%
  mutate(condition = factor(condition, levels = c("fc_ESC", "fc_Epi")))

p_violin <- ggplot(plot_df_ooe_long, aes(x = condition, y = value, fill = condition)) +
  geom_violin(linewidth = lineThick * mmToLineUnit, alpha = .4, show.legend = FALSE) +
  geom_boxplot(width = 0.3, outlier.shape = NA, alpha = 0.6, show.legend = FALSE, linewidth = lineThick * mmToLineUnit) +
  stat_summary(fun = mean, geom = "point", shape = 21, size = 0.5, fill = "black") +
  geom_hline(yintercept = c(-0.5, 0.5), linetype = "dashed", alpha = 0.5, color = "black", size = lineThick * mmToLineUnit) +
  geom_hline(yintercept = 0, color = "black", size = lineThick * mmToLineUnit) +
  coord_cartesian(ylim = c(-3, 3)) +
  scale_fill_manual(values = c(strong_red, strong_blue)) +
  labs(y = "log2 fc of obs/exp") +
  common_theme +
  theme(
    axis.title.x = element_blank(),
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = fontSizeM),
    axis.title.y = element_text(size = fontSizeM)
  )

svglite(here(figDir, "violin_ChIP_diff_ESCspecific_EpiLCvsESC.svg"), height = panelSize(1.5)*mmToInch, width = panelSize(1.5)*mmToInch)
print(p_violin)
dev.off()

# --- Final Scatterplot ---
generate_scatterplot(
  df = plot_df_ooe,
  x_var = "fc_ESC",
  y_var = "fc_Epi",
  name = "ChIP_diff_ESCspecific_EpiLCvsESC",
  limits = NULL, # Use default data limits
  diff_cutoff = 0.5
)



```
### Union loops, checking differential perturbation
#### OOE fc
```{r}
pseudo_ooe <- 0.01
data_to_plot <- data
plot_df_ooe <- data_to_plot %>%
  dplyr::select(id, ooe_DMSO, ooe_dTAG, ooe_EpiDMSO, ooe_EpidTAG) %>%
  na.omit() %>%
  mutate(
    across(starts_with("ooe_"), ~ if_else(. == 0, pseudo_ooe, .)),
    fc_ESC = log2(ooe_dTAG / ooe_DMSO),
    fc_Epi = log2(ooe_EpidTAG / ooe_EpiDMSO)
  )

generate_scatterplot(
  df = plot_df_ooe,
  x_var = "fc_ESC",
  y_var = "fc_Epi",
  name = "allLoops_ooeFC_EpiLCvsESC",
  limits = NULL, # Use default data limits
  diff_cutoff = 0.5
)

# Filtering out loops of interest
library(dbscan)

# Select only the columns for clustering
cluster_data <- plot_df_ooe %>% 
  select(fc_ESC, fc_Epi)

# Run DBSCAN. You may need to adjust 'eps' based on your data's scale.
# 'eps = 1.5' is a good starting point for data on this scale.
for(eps in c(0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 1.1, 1.2)){
  set.seed(123) # for reproducibility
  db_clusters <- dbscan(cluster_data, eps = eps, minPts = 10)
  
  # Add the cluster assignments back to your original data
  # Cluster 0 represents noise/outliers
  plot_df_ooe$cluster <- db_clusters$cluster
  plot_df_ooe$cluster <- as.factor(plot_df_ooe$cluster)
  
  plot_data <- plot_df_ooe
  diff_cutoff <- 0.5
  p <- ggplot(plot_data, aes(x = fc_ESC, fc_Epi, colour = cluster)) +
    geom_point(size = 0.5, alpha = 0.7) + # Simplified geom_point
    coord_fixed() +
    scale_y_continuous(limits = NULL) +
    geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black", size = lineThick*mmToLineUnit) +
    geom_abline(slope = 1, intercept = c(-diff_cutoff, diff_cutoff), linetype = "dotted", color = "black", size = lineThick*mmToLineUnit) +
    geom_hline(yintercept = 0, alpha = 0.5, color = "black", size = lineThick*mmToLineUnit) +
    geom_vline(xintercept = 0, alpha = 0.5, color = "black", size = lineThick*mmToLineUnit) +
    common_theme + scale_color_brewer(palette = "Set1")
  
  # Save the plot
  name <- paste0("allLoops_ooeFC_EpiLCvsESC_clusters_", eps)
  ggsave(
    filename = here(figDir, paste0("scatterplot_", name, ".png")),
    plot = p,
    width = panelSize(1.5) * mmToInch,
    height = panelSize(1.5) * mmToInch,
    dpi = 600
  )
}



eps <- 0.5

set.seed(123) # for reproducibility
db_clusters <- dbscan(cluster_data, eps = eps, minPts = 10)

# Add the cluster assignments back to your original data
# Cluster 0 represents noise/outliers
plot_df_ooe$cluster <- db_clusters$cluster
plot_df_ooe$cluster <- as.factor(plot_df_ooe$cluster)

for(i in seq(1, 6)){
plot_data <- plot_df_ooe %>% dplyr::filter(cluster == i)
diff_cutoff <- 0.5
p <- ggplot(plot_data, aes(x = fc_ESC, fc_Epi, colour = cluster)) +
  geom_point(size = 0.5, alpha = 0.7) + # Simplified geom_point
  coord_fixed() +
  scale_y_continuous(limits = c(-11, 10)) +
    scale_x_continuous(limits = c(-11, 10)) +

  geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black", size = lineThick*mmToLineUnit) +
  geom_abline(slope = 1, intercept = c(-diff_cutoff, diff_cutoff), linetype = "dotted", color = "black", size = lineThick*mmToLineUnit) +
  geom_hline(yintercept = 0, alpha = 0.5, color = "black", size = lineThick*mmToLineUnit) +
  geom_vline(xintercept = 0, alpha = 0.5, color = "black", size = lineThick*mmToLineUnit) +
  common_theme + scale_color_brewer(palette = "Set1")
  name <- paste0("allLoops_ooeFC_EpiLCvsESC_clusters_", eps, "_", i)
  ggsave(
    filename = here(figDir, paste0("scatterplot_", name, ".png")),
    plot = p,
    width = panelSize(1.5) * mmToInch,
    height = panelSize(1.5) * mmToInch,
    dpi = 600
  )}

# Saving specific cluster loops
temp <- plot_df_ooe %>% dplyr::filter(cluster == 3)
fwrite((data %>% dplyr::filter(id %in% temp$id)) %>% dplyr::select(seq(1, 6)), here(dataDir, "loop_fcCluster_EpiLCspecificPerturb.bedpe"), sep = "\t", col.names = FALSE)

temp <- plot_df_ooe %>% dplyr::filter(cluster == 4)
fwrite((data %>% dplyr::filter(id %in% temp$id)) %>% dplyr::select(seq(1, 6)), here(dataDir, "loop_fcCluster_ESCspecificPerturb.bedpe"), sep = "\t", col.names = FALSE)
```
##### Checking characteristics of specific perturb loops
```{r}

loop.EpiLCSpecificPerturb <- fread(here(dataDir, "loop_fcCluster_EpiLCspecificPerturb.bedpe"))
colnames(loop.EpiLCSpecificPerturb) <- c("V1", "V2", "V3", "V4", "V5", "V6")
loop.EpiLCSpecificPerturb <- importBedpe(loop.EpiLCSpecificPerturb)


loop.ESCSpecificPerturb <- fread(here(dataDir, "loop_fcCluster_ESCspecificPerturb.bedpe"))
colnames(loop.ESCSpecificPerturb) <- c("V1", "V2", "V3", "V4", "V5", "V6")
loop.ESCSpecificPerturb <- importBedpe(loop.ESCSpecificPerturb)


# Helper function to find loops overlapping with peaks
find_overlapping_loops <- function(loops, peaks) {
  idx <- unique(c(
    queryHits(findOverlaps(first(loops), peaks)),
    queryHits(findOverlaps(second(loops), peaks))
  ))
  loops[idx]
}


find_overlapping_loops(loop.EpiLCSpecificPerturb, peak.Epi)
find_overlapping_loops(loop.EpiLCSpecificPerturb, peak.ESC)

find_overlapping_loops(loop.ESCSpecificPerturb, peak.Epi)
find_overlapping_loops(loop.ESCSpecificPerturb, peak.ESC)


### classification
make_table <- function(loop, name){
  n <- nrow(loop)
  temp <- loop %>% dplyr::select(id, Anno) %>%
    dplyr::mutate(AnnoGroup = case_when(
      
      
      Anno %in% c("P-P", "P-E", "E-E") ~ "Reg (P-P, P-E, E-E)",
      
      Anno %in% c("P-S", "E-S") ~ "Reg/Str mixed",
      Anno %in% c("P-SP", "P-SE", "E-SP", "E-SE") ~ "Reg/Str mixed",
      Anno %in% c("SP-SP", "SP-SE", "SE-SE") ~ "Reg/Str mixed",
      Anno %in% c("SP-S", "SE-S") ~ "Reg/Str mixed",
      
      Anno %in% c("S-S", "S-X") ~ "Str (S-S, S-X)",
      Anno %in% c("P-X", "E-X") ~ "Reg-Other",
      Anno %in% c("SP-X", "SE-X") ~ "StrReg-Other",
      Anno %in% c("X-X") ~ "Other",
      TRUE ~ NA
    ))
  
  
  temp$AnnoGroup <- factor(temp$AnnoGroup, levels = c("Reg (P-P, P-E, E-E)",
                                                      "Reg/Str mixed",
                                                      "Str (S-S, S-X)",
                                                      "Reg-Other",
                                                      "StrReg-Other",
                                                      "Other"))
  
  
  pie_df <- temp %>%
    as_tibble() %>%                          # if temp was a list, this will error early
    group_by(AnnoGroup) %>%
    summarise(n = n(), .groups = "drop")
  colnames(pie_df) <- c("AnnoGroup", name)
  return(pie_df)
}

loop.EpiLCSpecificPerturb <- fread(here(dataDir, "loop_fcCluster_EpiLCspecificPerturb.bedpe")) %>%
  dplyr::mutate(id = paste(V1, V2, V3, V4, V5, V6, sep = "_"))
loop.ESCSpecificPerturb <- fread(here(dataDir, "loop_fcCluster_ESCspecificPerturb.bedpe")) %>%
  dplyr::mutate(id = paste(V1, V2, V3, V4, V5, V6, sep = "_"))

loop <- data %>% dplyr::filter(id %in% loop.EpiLCSpecificPerturb$id)
temp1 <- make_table(loop, "EpiLCspecificPerturb")
loop <- data %>% dplyr::filter(id %in% loop.ESCSpecificPerturb$id)
temp2 <- make_table(loop, "ESCspecificPerturb")


temp <- temp1 %>% dplyr::full_join(temp2, by = "AnnoGroup")


data_long <- temp %>% pivot_longer(cols = c("EpiLCspecificPerturb", "ESCspecificPerturb"))

data_long$name <- factor(data_long$name, levels = c("EpiLCspecificPerturb", "ESCspecificPerturb"))

p <- ggplot(data_long, aes(x = name, y = value, fill = AnnoGroup)) +
  geom_bar(position="fill", stat="identity") +
  labs(
    x = NULL,         # Remove x-axis title
    y = "Ratio"       # Change y-axis title
  ) +
  theme(
    axis.title = element_text(
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45, hjust = 1,          # Rotate x labels
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text.y = element_text(
      size = fontSizeS, family = fontType, color = "#000000"
    ),
    axis.line = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) + 
guides(
    fill = guide_legend(
      keywidth = 0.2,  # Adjust the width of the legend keys
      keyheight = 0.2  # Adjust the height of the legend keys
    )
  )

```
#### Plotting scores (Taken from async vs G1)
```{r}
# --- 4. GENERATE PLOTS ---
data <- fread(here(commonLoopDir, "loopScore_chromosight.tsv"))
    # dplyr::filter(if_any(c(score_DMSO, score_dTAG), ~ . > 0.3)) %>%
    # dplyr::filter(if_any(c(score_UT, score_AID), ~ . > 0.3))

loopName <- "EpiLC"

# Checking the distribution of scores
dataToPlot <- data %>% dplyr::select(score_DMSO, score_dTAG, score_EpiDMSO, score_EpidTAG) %>%
  pivot_longer(everything(), names_to = "condition", values_to = "score" )
dataToPlot$condition <- factor(dataToPlot$condition, levels = c("score_DMSO", "score_dTAG", "score_EpiDMSO", "score_EpidTAG"))

ggplot(dataToPlot, aes(x = condition, y = score, fill = condition)) +
  geom_violin() +
  scale_y_continuous(limits = c(-0.5, 1)) +
  geom_boxplot(width = 0.4) +
  ggtitle("Chromo EpiLC") +
  theme_classic() +
  geom_hline(yintercept = c(0)) +
  geom_hline(yintercept = c(0.3), linetype = "dashed")
  
generate_scatterplot(data, "score_DMSO",    "score_EpiDMSO", paste0(loopName, "_EpiG1DMSOvsG1DMSO"))
generate_scatterplot(data, "score_EpiDMSO",    "score_EpidTAG", paste0(loopName, "_EpiG1dTAGvsEpiG1DMSO"))
generate_scatterplot(data, "score_DMSO",    "score_dTAG", paste0(loopName, "_G1dTAGvsG1DMSO"))



```


#### Delta score fc
```{r}
data_to_plot <- fread(here(commonLoopDir, "loopScore_chromosight_extra.tsv"))

plot_df_delta <- data_to_plot %>%
  dplyr::mutate(
    delta_ESC = score_dTAG - score_DMSO,
    delta_EpiLC = score_EpidTAG - score_EpiDMSO
  )

common_theme <- theme_classic() +
  theme(
    legend.position = "none",
    axis.title = element_text(size = fontSizeS, family = fontType, color = "black"),
    axis.text = element_text(size = fontSizeS, family = fontType, color = "black"),
    axis.line = element_line(color = "black", size = lineThick * mmToLineUnit, lineend = "square"),
    axis.ticks = element_line(color = "black", size = lineThick * mmToLineUnit, lineend = "square"),
    panel.background = element_rect(fill = "transparent")
  )

# Function to generate and save scatterplots
generate_scatterplot2 <- function(df, x_var, y_var, name, limits = c(-0.5, 1), diff_cutoff = 0.2) {
  plot_data <- df %>%
    mutate(density = get_density(.data[[x_var]], .data[[y_var]], n = 200))

  p <- ggplot(plot_data, aes(x = .data[[x_var]], y = .data[[y_var]])) +
    geom_point(aes(color = density), size = 0.1, data = . %>% arrange(density)) +
    coord_fixed() +
    scale_x_continuous(limits = limits) +
    scale_y_continuous(limits = limits) +
    geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black", size = lineThick*mmToLineUnit) +
    geom_abline(slope = 1, intercept = c(-diff_cutoff, diff_cutoff), linetype = "dotted", color = "black", size = lineThick*mmToLineUnit) +
    geom_hline(yintercept = c(-diff_cutoff, diff_cutoff), linetype = "dotted", color = "black", size = lineThick*mmToLineUnit) +
    geom_vline(xintercept = c(-diff_cutoff, diff_cutoff), linetype = "dotted", color = "black", size = lineThick*mmToLineUnit) +
    geom_hline(yintercept = 0, alpha = 0.5, color = "black", size = lineThick*mmToLineUnit) +
    geom_vline(xintercept = 0, alpha = 0.5, color = "black", size = lineThick*mmToLineUnit) +
    common_theme

  # Save the plot
  ggsave(
    filename = here(figDir, paste0("scatterplot_", name, ".png")),
    plot = p,
    width = panelSize(1.5) * mmToInch,
    height = panelSize(1.5) * mmToInch,
    dpi = 600
  )
}
generate_scatterplot2(
  df = plot_df_delta,
  x_var = "delta_ESC",
  y_var = "delta_EpiLC",
  name = "allLoops_deltaScore_EpiLCvsESC",
  limits = c(-0.75, 0.75), # Use default data limits
  diff_cutoff = 0.2
)

temp <- plot_df_delta %>% dplyr::filter(delta_EpiLC < -0.2, 
                                abs(delta_ESC) < 0.2,
                                delta_EpiLC - delta_ESC < -0.2)
fwrite((temp %>% dplyr::select(seq(1, 6))), here(dataDir, "loop_deltaScore_EpiLCspecificPerturb.bedpe"), sep = "\t", col.names = FALSE)

generate_scatterplot2(
  df = temp,
  x_var = "delta_ESC",
  y_var = "delta_EpiLC",
  name = "allLoops_deltaScore_EpiLCvsESC_EpiLCspecific",
  limits = c(-0.75, 0.75), # Use default data limits
  diff_cutoff = 0.2
)

temp <- plot_df_delta %>% dplyr::filter(delta_ESC < -0.2, 
                                abs(delta_EpiLC) < 0.2,
                                 delta_ESC - delta_EpiLC < -0.2)
fwrite((temp %>% dplyr::select(seq(1, 6))), here(dataDir, "loop_deltaScore_ESCspecificPerturb.bedpe"), sep = "\t", col.names = FALSE)

generate_scatterplot2(
  df = temp,
  x_var = "delta_ESC",
  y_var = "delta_EpiLC",
  name = "allLoops_deltaScore_EpiLCvsESC_ESCspecific",
  limits = c(-0.75, 0.75), # Use default data limits
  diff_cutoff = 0.2
)


```



```{r}
make_table <- function(loop, name){
  n <- nrow(loop)
  temp <- loop %>% dplyr::select(id, Anno) %>%
    dplyr::mutate(AnnoGroup = case_when(
      
      
      Anno %in% c("P-P", "P-E", "E-E") ~ "Reg (P-P, P-E, E-E)",
      
      Anno %in% c("P-S", "E-S") ~ "Reg/Str mixed",
      Anno %in% c("P-SP", "P-SE", "E-SP", "E-SE") ~ "Reg/Str mixed",
      Anno %in% c("SP-SP", "SP-SE", "SE-SE") ~ "Reg/Str mixed",
      Anno %in% c("SP-S", "SE-S") ~ "Reg/Str mixed",
      
      Anno %in% c("S-S", "S-X") ~ "Str (S-S, S-X)",
      Anno %in% c("P-X", "E-X") ~ "Reg-Other",
      Anno %in% c("SP-X", "SE-X") ~ "StrReg-Other",
      Anno %in% c("X-X") ~ "Other",
      TRUE ~ NA
    ))
  
  
  temp$AnnoGroup <- factor(temp$AnnoGroup, levels = c("Reg (P-P, P-E, E-E)",
                                                      "Reg/Str mixed",
                                                      "Str (S-S, S-X)",
                                                      "Reg-Other",
                                                      "StrReg-Other",
                                                      "Other"))
  
  
  pie_df <- temp %>%
    as_tibble() %>%                          # if temp was a list, this will error early
    group_by(AnnoGroup) %>%
    summarise(n = n(), .groups = "drop")
  colnames(pie_df) <- c("AnnoGroup", name)
  return(pie_df)
}

temp <- as_tibble(loop.ESC) %>% dplyr::mutate(id = paste(seqnames1, start1-1, end1, seqnames2, start2-1, end2, sep = "_"))
data_to_plot <- data %>% dplyr::filter(id %in% temp$id)
data.esc <- make_table(data_to_plot, "ESC")

temp <- as_tibble(loop.Epi) %>% dplyr::mutate(id = paste(seqnames1, start1-1, end1, seqnames2, start2-1, end2, sep = "_"))
data_to_plot <- data %>% dplyr::filter(id %in% temp$id)
data.epi <- make_table(data_to_plot, "Epi")

data <- data.esc %>%
  dplyr::full_join(data.epi, by = c("AnnoGroup"))

data_long <- data %>% pivot_longer(cols = c("ESC", "Epi"))

data_long$name <- factor(data_long$name, levels = c("ESC", "Epi"))

p <- ggplot(data_long, aes(x = name, y = value, fill = AnnoGroup)) +
  geom_bar(position="fill", stat="identity") +
  labs(
    x = NULL,         # Remove x-axis title
    y = "Ratio"       # Change y-axis title
  ) +
  theme(
    axis.title = element_text(
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45, hjust = 1,          # Rotate x labels
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text.y = element_text(
      size = fontSizeS, family = fontType, color = "#000000"
    ),
    axis.line = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) + 
guides(
    fill = guide_legend(
      keywidth = 0.2,  # Adjust the width of the legend keys
      keyheight = 0.2  # Adjust the height of the legend keys
    )
  )

fileName <- paste0("Barplot_loopAnno_comparison")
width <- panelSize(2.2)*mmToInch
height <- panelSize(1.75)*mmToInch
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()
```
##### Checking characteristics of specific perturb loops
```{r}

loop.EpiLCSpecificPerturb <- fread(here(dataDir, "loop_deltaScore_EpiLCspecificPerturb.bedpe"))
colnames(loop.EpiLCSpecificPerturb) <- c("V1", "V2", "V3", "V4", "V5", "V6")
loop.EpiLCSpecificPerturb <- importBedpe(loop.EpiLCSpecificPerturb)


loop.ESCSpecificPerturb <- fread(here(dataDir, "loop_deltaScore_ESCspecificPerturb.bedpe"))
colnames(loop.ESCSpecificPerturb) <- c("V1", "V2", "V3", "V4", "V5", "V6")
loop.ESCSpecificPerturb <- importBedpe(loop.ESCSpecificPerturb)


# Helper function to find loops overlapping with peaks
find_overlapping_loops <- function(loops, peaks) {
  idx <- unique(c(
    queryHits(findOverlaps(first(loops), peaks)),
    queryHits(findOverlaps(second(loops), peaks))
  ))
  loops[idx]
}


length(find_overlapping_loops(loop.EpiLCSpecificPerturb, peak.Epi))/length(loop.EpiLCSpecificPerturb)
find_overlapping_loops(loop.EpiLCSpecificPerturb, peak.ESC)

find_overlapping_loops(loop.ESCSpecificPerturb, peak.Epi)
find_overlapping_loops(loop.ESCSpecificPerturb, peak.ESC)


### classification
make_table <- function(loop, name){
  n <- nrow(loop)
  temp <- loop %>% dplyr::select(id, Anno) %>%
    dplyr::mutate(AnnoGroup = case_when(
      
      
      Anno %in% c("P-P", "P-E", "E-E") ~ "Reg (P-P, P-E, E-E)",
      
      Anno %in% c("P-S", "E-S") ~ "Reg/Str mixed",
      Anno %in% c("P-SP", "P-SE", "E-SP", "E-SE") ~ "Reg/Str mixed",
      Anno %in% c("SP-SP", "SP-SE", "SE-SE") ~ "Reg/Str mixed",
      Anno %in% c("SP-S", "SE-S") ~ "Reg/Str mixed",
      
      Anno %in% c("S-S", "S-X") ~ "Str (S-S, S-X)",
      Anno %in% c("P-X", "E-X") ~ "Reg-Other",
      Anno %in% c("SP-X", "SE-X") ~ "StrReg-Other",
      Anno %in% c("X-X") ~ "Other",
      TRUE ~ NA
    ))
  
  
  temp$AnnoGroup <- factor(temp$AnnoGroup, levels = c("Reg (P-P, P-E, E-E)",
                                                      "Reg/Str mixed",
                                                      "Str (S-S, S-X)",
                                                      "Reg-Other",
                                                      "StrReg-Other",
                                                      "Other"))
  
  
  pie_df <- temp %>%
    as_tibble() %>%                          # if temp was a list, this will error early
    group_by(AnnoGroup) %>%
    summarise(n = n(), .groups = "drop")
  colnames(pie_df) <- c("AnnoGroup", name)
  return(pie_df)
}

loop.EpiLCSpecificPerturb <- fread(here(dataDir, "loop_deltaScore_EpiLCspecificPerturb.bedpe")) %>%
  dplyr::mutate(id = paste(V1, V2, V3, V4, V5, V6, sep = "_")) 
loop.ESCSpecificPerturb <- fread(here(dataDir, "loop_deltaScore_ESCspecificPerturb.bedpe")) %>%
  dplyr::mutate(id = paste(V1, V2, V3, V4, V5, V6, sep = "_"))

loop <- data %>% dplyr::filter(id %in% loop.EpiLCSpecificPerturb$id)
temp1 <- make_table(loop, "EpiLCspecificPerturb")
loop <- data %>% dplyr::filter(id %in% loop.ESCSpecificPerturb$id)
temp2 <- make_table(loop, "ESCspecificPerturb")


temp <- temp1 %>% dplyr::full_join(temp2, by = "AnnoGroup")


data_long <- temp %>% pivot_longer(cols = c("EpiLCspecificPerturb", "ESCspecificPerturb"))

data_long$name <- factor(data_long$name, levels = c("EpiLCspecificPerturb", "ESCspecificPerturb"))

p <- ggplot(data_long, aes(x = name, y = value, fill = AnnoGroup)) +
  geom_bar(position="fill", stat="identity") +
  labs(
    x = NULL,         # Remove x-axis title
    y = "Ratio"       # Change y-axis title
  ) +
  theme(
    axis.title = element_text(
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45, hjust = 1,          # Rotate x labels
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text.y = element_text(
      size = fontSizeS, family = fontType, color = "#000000"
    ),
    axis.line = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) + 
guides(
    fill = guide_legend(
      keywidth = 0.2,  # Adjust the width of the legend keys
      keyheight = 0.2  # Adjust the height of the legend keys
    )
  )

```
##### Checking related genes
```{r}

# IMPORTING GENE LIST OF INTERST
extDir <- here("..", "RNA_231102_14065-14458-15127-H202SC24044572")
alpha <- 0.05
foldchange <- 0.5

# Create a mapping of ensembl gene ID to external gene name for visualization
geneIDNamePair <- bind_rows(fread(here(extDir, "data", "diff_G1.dTAG_G1.Epi.DMSO_vs_G1.2i.DMSO.tsv")),
                            fread(here(extDir, "data", "diff_G1.dTAG_G1.2i.dTAG_vs_G1.2i.DMSO.tsv")),
                            fread(here(extDir, "data", "diff_G1.dTAG_G1.Epi.dTAG_vs_G1.Epi.DMSO.tsv"))) %>%
  dplyr::select(ensembl_gene_id, external_gene_name) %>%
  distinct()


loop.EpiLCSpecificPerturb <- fread(here(dataDir, "loop_deltaScore_EpiLCspecificPerturb.bedpe")) %>%
  dplyr::mutate(id = paste(V1, V2, V3, V4, V5, V6, sep = "_"))
loop.ESCSpecificPerturb <- fread(here(dataDir, "loop_deltaScore_ESCspecificPerturb.bedpe")) %>%
  dplyr::mutate(id = paste(V1, V2, V3, V4, V5, V6, sep = "_"))

# EpiLC
# annotating hasP
geneAnnoData <- loadLoopAnnoData(here(commonLoopDir, paste0("loopScore_", "chromosight", "_p-n_complex_ensemblList.tsv")))
temp <- (geneAnnoData %>% dplyr::filter(id %in% loop.EpiLCSpecificPerturb$id))
geneList.hasP.EpiLC <- unlist(temp$gene)


# ESC
temp <- (geneAnnoData %>% dplyr::filter(id %in% loop.ESCSpecificPerturb$id))
geneList.hasP.ESC <- unlist(temp$gene)


gene_background <- geneIDNamePair$ensembl_gene_id
GO.1 <- enrichGO(gene = geneList.hasP.ESC, OrgDb = org.Mm.eg.db,
                  keyType = "ENSEMBL", ont = "BP", readable = TRUE, universe = gene_background)
GO.2 <- enrichGO(gene = geneList.hasP.EpiLC, OrgDb = org.Mm.eg.db,
                  keyType = "ENSEMBL", ont = "BP", readable = TRUE, universe = gene_background)
GO.1.df <- as.data.frame(GO.1)
GO.2.df <- as.data.frame(GO.2)

```
##### Checking how much of them overlap with DEGs
```{r}
loop.EpiLCSpecificPerturb <- fread(here(dataDir, "loop_deltaScore_EpiLCspecificPerturb.bedpe")) %>%
  dplyr::mutate(id = paste(V1, V2, V3, V4, V5, V6, sep = "_"))
loop.ESCSpecificPerturb <- fread(here(dataDir, "loop_deltaScore_ESCspecificPerturb.bedpe")) %>%
  dplyr::mutate(id = paste(V1, V2, V3, V4, V5, V6, sep = "_"))

geneAnnoData <- loadLoopAnnoData(here(commonLoopDir, paste0("loopScore_", "chromosight", "_p-n_complex_ensemblList.tsv")))
temp <- (geneAnnoData %>% dplyr::filter(id %in% loop.EpiLCSpecificPerturb$id))
geneList.hasP.EpiLC <- unlist(temp$gene)

extDir <- here("..", "RNA_231102_14065-14458-15127-H202SC24044572")
diff.tb <- fread(here(extDir, "data", "diff_G1.dTAG_G1.Epi.dTAG_vs_G1.Epi.DMSO.tsv")) %>%
  dplyr::filter(padj < 0.05)

nrow(diff.tb)
diff.tb %>% dplyr::filter(ensembl_gene_id %in% geneList.hasP.EpiLC)


### Checking how many of the down DEGs from EpiLC has any perturbed loops from EpiLC
data_to_plot <- fread(here(commonLoopDir, "loopScore_chromosight_extra.tsv"))
plot_df_delta <- data_to_plot %>%
  dplyr::mutate(
    delta_ESC = score_dTAG - score_DMSO,
    delta_EpiLC = score_EpidTAG - score_EpiDMSO
  )
geneList <- unlist((geneAnnoData %>% dplyr::filter(id %in% (plot_df_delta %>% dplyr::filter(delta_EpiLC < -0.2))$id))$gene)
diff.tb <- fread(here(extDir, "data", "diff_G1.dTAG_G1.Epi.dTAG_vs_G1.Epi.DMSO.tsv")) %>%
  dplyr::filter(padj < 0.05, shrinked_log2FC < 0)
nrow(diff.tb)
nrow(diff.tb %>% dplyr::filter(ensembl_gene_id %in% geneList))
diff.tb <- fread(here(extDir, "data", "diff_G1.dTAG_G1.Epi.dTAG_vs_G1.Epi.DMSO.tsv")) %>%
  dplyr::filter(padj < 0.05, shrinked_log2FC < -0.5)
nrow(diff.tb)
nrow(diff.tb %>% dplyr::filter(ensembl_gene_id %in% geneList))


### Checking how many of the down DEG from ESC has any perturbed loops from ESC
geneList <- unlist((geneAnnoData %>% dplyr::filter(id %in% (plot_df_delta %>% dplyr::filter(delta_ESC < -0.2))$id))$gene)
diff.tb <- fread(here(extDir, "data", "diff_G1.dTAG_G1.2i.dTAG_vs_G1.2i.DMSO.tsv")) %>%
  dplyr::filter(padj < 0.05, shrinked_log2FC < 0)
nrow(diff.tb)
nrow(diff.tb %>% dplyr::filter(ensembl_gene_id %in% geneList))
diff.tb <- fread(here(extDir, "data", "diff_G1.dTAG_G1.2i.dTAG_vs_G1.2i.DMSO.tsv")) %>%
  dplyr::filter(padj < 0.05, shrinked_log2FC < -0.5)
nrow(diff.tb)
nrow(diff.tb %>% dplyr::filter(ensembl_gene_id %in% geneList))


### Checking how many of the down DEG from ESC or EpiLC has any perturbed loops from ESC
geneList <- unlist((geneAnnoData %>% dplyr::filter(id %in% (plot_df_delta %>% dplyr::filter((delta_EpiLC < -0.2) | (delta_ESC < -0.2)))$id))$gene)
diff.tb <- fread(here(extDir, "data", "diff_G1.dTAG_G1.Epi.dTAG_vs_G1.Epi.DMSO.tsv")) %>%
  dplyr::filter(padj < 0.05, shrinked_log2FC < 0)
nrow(diff.tb)
nrow(diff.tb %>% dplyr::filter(ensembl_gene_id %in% geneList))
diff.tb <- fread(here(extDir, "data", "diff_G1.dTAG_G1.Epi.dTAG_vs_G1.Epi.DMSO.tsv")) %>%
  dplyr::filter(padj < 0.05, shrinked_log2FC < -0.5)
nrow(diff.tb)
nrow(diff.tb %>% dplyr::filter(ensembl_gene_id %in% geneList))
diff.tb <- fread(here(extDir, "data", "diff_G1.dTAG_G1.2i.dTAG_vs_G1.2i.DMSO.tsv")) %>%
  dplyr::filter(padj < 0.05, shrinked_log2FC < 0)
nrow(diff.tb)
nrow(diff.tb %>% dplyr::filter(ensembl_gene_id %in% geneList))
diff.tb <- fread(here(extDir, "data", "diff_G1.dTAG_G1.2i.dTAG_vs_G1.2i.DMSO.tsv")) %>%
  dplyr::filter(padj < 0.05, shrinked_log2FC < -0.5)
nrow(diff.tb)
nrow(diff.tb %>% dplyr::filter(ensembl_gene_id %in% geneList))

# Long term perturbed
geneListPerturbed <- c((fread(here("..", "RNA_241223_H202SC24124443_RAD21_EpiRecovery", "pub_EpiUP_DEGfiltered_kmeans_group2.tsv" )) %>%
                 dplyr::filter(group == "K1"))$gene,
              (fread(here("..", "RNA_241223_H202SC24124443_RAD21_EpiRecovery", "pub_EpiDOWN_DEGfiltered_kmeans_group2.tsv" )) %>%
                  dplyr::filter(group == "K1"))$gene)
length(geneListPerturbed)      
length(geneListPerturbed[geneListPerturbed %in% geneList])

```

# [8] Async vs G1
## Common functions
```{r}
# Define a common theme to reduce code duplication
common_theme <- theme_classic() +
  theme(
    legend.position = "none",
    axis.title = element_text(size = fontSizeS, family = fontType, color = "black"),
    axis.text = element_text(size = fontSizeS, family = fontType, color = "black"),
    axis.line = element_line(color = "black", size = lineThick * mmToLineUnit, lineend = "square"),
    axis.ticks = element_line(color = "black", size = lineThick * mmToLineUnit, lineend = "square"),
    panel.background = element_rect(fill = "transparent")
  )

# Function to generate and save scatterplots
generate_scatterplot <- function(df, x_var, y_var, name, limits = c(-0.5, 1), diff_cutoff = 0.2) {
  plot_data <- df %>%
    mutate(density = get_density(.data[[x_var]], .data[[y_var]], n = 200))

  p <- ggplot(plot_data, aes(x = .data[[x_var]], y = .data[[y_var]])) +
    geom_point(aes(color = density), size = 0.1, data = . %>% arrange(density)) +
    coord_fixed() +
    scale_x_continuous(limits = limits) +
    scale_y_continuous(limits = limits) +
    geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black", size = lineThick*mmToLineUnit) +
    geom_abline(slope = 1, intercept = c(-diff_cutoff, diff_cutoff), linetype = "dotted", color = "black", size = lineThick*mmToLineUnit) +
    geom_hline(yintercept = 0, alpha = 0.5, color = "black", size = lineThick*mmToLineUnit) +
    geom_vline(xintercept = 0, alpha = 0.5, color = "black", size = lineThick*mmToLineUnit) +
    common_theme

  # Save the plot
  ggsave(
    filename = here(figDir, paste0("scatterplot_", name, ".png")),
    plot = p,
    width = panelSize(1.5) * mmToInch,
    height = panelSize(1.5) * mmToInch,
    dpi = 600
  )
}

make_table <- function(loop, name){
  n <- nrow(loop)
  temp <- loop %>% dplyr::select(id, Anno) %>%
    dplyr::mutate(AnnoGroup = case_when(
      
      
      Anno %in% c("P-P", "P-E", "E-E") ~ "Reg (P-P, P-E, E-E)",
      
      Anno %in% c("P-S", "E-S") ~ "Reg/Str mixed",
      Anno %in% c("P-SP", "P-SE", "E-SP", "E-SE") ~ "Reg/Str mixed",
      Anno %in% c("SP-SP", "SP-SE", "SE-SE") ~ "Reg/Str mixed",
      Anno %in% c("SP-S", "SE-S") ~ "Reg/Str mixed",
      
      Anno %in% c("S-S", "S-X") ~ "Str (S-S, S-X)",
      Anno %in% c("P-X", "E-X") ~ "Reg-Other",
      Anno %in% c("SP-X", "SE-X") ~ "StrReg-Other",
      Anno %in% c("X-X") ~ "Other",
      TRUE ~ NA
    ))
  
  
  temp$AnnoGroup <- factor(temp$AnnoGroup, levels = c("Reg (P-P, P-E, E-E)",
                                                      "Reg/Str mixed",
                                                      "Str (S-S, S-X)",
                                                      "Reg-Other",
                                                      "StrReg-Other",
                                                      "Other"))
  
  
  pie_df <- temp %>%
    as_tibble() %>%                          # if temp was a list, this will error early
    group_by(AnnoGroup) %>%
    summarise(n = n(), .groups = "drop")
  colnames(pie_df) <- c("AnnoGroup", name)
  return(pie_df)
}

make_table2 <- function(loop, name){
  n <- nrow(loop)
  temp <- loop %>% dplyr::select(id, Anno) %>%
    dplyr::mutate(AnnoGroup = case_when(
      Anno %in% c("P-P") ~ "P-P",
      Anno %in% c("P-E") ~ "P-E",
      Anno %in% c("E-E") ~ "E-E",
      TRUE ~ NA
    ))
  
  
  temp$AnnoGroup <- factor(temp$AnnoGroup, levels = c("P-P",
                                                      "P-E",
                                                      "E-E"))
  
  
  pie_df <- temp %>%
    as_tibble() %>%                          # if temp was a list, this will error early
    group_by(AnnoGroup) %>%
    summarise(n = n(), .groups = "drop")
  colnames(pie_df) <- c("AnnoGroup", name)
  return(pie_df)
}

```

## Chromosight loops
```{r}
# --- 4. GENERATE PLOTS ---
data <- fread(here(commonLoopDir, "loopScore_chromosight.tsv")) %>%
  # dplyr::filter(if_any(c(score_DMSO, score_dTAG), ~ . > 0.3)) %>%
  # dplyr::filter(if_any(c(score_UT, score_AID), ~ . > 0.3)) %>%
  dplyr::filter(Anno %in% c("P-P", "P-E", "E-E"))
  # dplyr::filter(Anno %in% c("S-S", "S-X"))

# loopName <- "strongLoopsBothStr"
loopName <- "allLoopsReg"
# Checking the distribution of scores
# dataToPlot <- data %>% dplyr::select(score_DMSO, score_UT, score_dTAG, score_AID) %>%
#   pivot_longer(everything(), names_to = "condition", values_to = "score" )
# dataToPlot$condition <- factor(dataToPlot$condition, levels = c("score_DMSO", "score_dTAG", "score_UT", "score_AID"))
# 
# ggplot(dataToPlot, aes(x = condition, y = score, fill = condition)) +
#   geom_violin() +
#   scale_y_continuous(limits = c(-0.5, 1)) +
#   geom_boxplot(width = 0.4) +
#   ggtitle(paste0("Chromosight ", loopName)) +
#   theme_classic() +
#   geom_hline(yintercept = c(0)) +
#   geom_hline(yintercept = c(0.3), linetype = "dashed")

# Differential
# dataToPlot2 <- data %>% dplyr::mutate(diffScore_G1 = score_dTAG - score_DMSO,
#                                       diffScore_Async = score_AID - score_UT) %>%
#   dplyr::select(diffScore_G1, diffScore_Async) %>%
#   pivot_longer(everything(), names_to = "condition", values_to = "score")
# dataToPlot2$condition <- factor(dataToPlot2$condition, levels = c("diffScore_G1", "diffScore_Async"))
# ggplot(dataToPlot2, aes(x = condition, y = score, fill = condition)) +
#   geom_violin() +
#   scale_y_continuous(limits = c(-1, 1)) +
#   geom_boxplot(width = 0.4) +
#   ggtitle(paste0("Chromosight ", loopName)) +
#   theme_classic() +
#   geom_hline(yintercept = c(0)) +
#   geom_hline(yintercept = c(0.3), linetype = "dashed")
#   
# generate_scatterplot(data, "score_DMSO",    "score_UT", paste0(loopName, "_AsyncUTvsG1DMSO"))
# generate_scatterplot(data, "score_UT",    "score_AID", paste0(loopName, "_AsyncAIDvsAsyncUT"))
# generate_scatterplot(data, "score_DMSO",    "score_dTAG", paste0(loopName, "_G1dTAGvsG1DMSO"))
# 
# ## Plot diff score
plot_df_delta <- data %>%
  dplyr::mutate(
    delta_G1 = score_dTAG - score_DMSO,
    delta_Async = score_AID - score_UT
  )
# 
# 
# # Function to generate and save scatterplots
# generate_scatterplot2 <- function(df, x_var, y_var, name, limits = c(-0.5, 1), diff_cutoff = 0.2) {
#   plot_data <- df %>%
#     mutate(density = get_density(.data[[x_var]], .data[[y_var]], n = 200))
# 
#   p <- ggplot(plot_data, aes(x = .data[[x_var]], y = .data[[y_var]])) +
#     geom_point(aes(color = density), size = 0.1, data = . %>% arrange(density)) +
#     coord_fixed() +
#     scale_x_continuous(limits = limits) +
#     scale_y_continuous(limits = limits) +
#     geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black", size = lineThick*mmToLineUnit) +
#     geom_abline(slope = 1, intercept = c(-diff_cutoff, diff_cutoff), linetype = "dotted", color = "black", size = lineThick*mmToLineUnit) +
#     geom_hline(yintercept = c(-diff_cutoff, diff_cutoff), linetype = "dotted", color = "black", size = lineThick*mmToLineUnit) +
#     geom_vline(xintercept = c(-diff_cutoff, diff_cutoff), linetype = "dotted", color = "black", size = lineThick*mmToLineUnit) +
#     geom_hline(yintercept = 0, alpha = 0.5, color = "black", size = lineThick*mmToLineUnit) +
#     geom_vline(xintercept = 0, alpha = 0.5, color = "black", size = lineThick*mmToLineUnit) +
#     common_theme
# 
#   # Save the plot
#   ggsave(
#     filename = here(figDir, paste0("scatterplot_", name, ".png")),
#     plot = p,
#     width = panelSize(1.5) * mmToInch,
#     height = panelSize(1.5) * mmToInch,
#     dpi = 600
#   )
# }
# generate_scatterplot2(
#   df = plot_df_delta,
#   x_var = "delta_G1",
#   y_var = "delta_Async",
#   name = paste0(loopName, "_deltaScore_AsyncvsG1"),
#   limits = c(-0.75, 0.75), # Use default data limits
#   diff_cutoff = 0.2
# )


temp <- plot_df_delta %>% dplyr::filter(delta_G1 < -0.2, 
                                        abs(delta_Async) < 0.2,
                                        delta_G1 - delta_Async < -0.2)
# print(nrow(temp))
# temp1 <- make_table(temp, "G1SpecificPerturb")
# temp3 <- make_table2(temp, "G1SpecificPerturb") %>%
#   dplyr::filter(!is.na(AnnoGroup))
# 
# generate_scatterplot2(
#   df = temp,
#   x_var = "delta_G1",
#   y_var = "delta_Async",
#   name = paste0(loopName, "_deltaScore_AsyncvsG1_G1specific"),
#   limits = c(-0.75, 0.75), # Use default data limits
#   diff_cutoff = 0.2
# )

temp <- plot_df_delta %>% dplyr::filter(delta_Async < -0.2, 
                                        abs(delta_G1) < 0.2,
                                        delta_Async - delta_G1 < -0.2)
print(nrow(temp))
# temp2 <- make_table(temp, "AsyncSpecificPerturb")
# temp4 <- make_table2(temp, "AsyncSpecificPerturb") %>%
#   dplyr::filter(!is.na(AnnoGroup))
# 
# generate_scatterplot2(
#   df = temp,
#   x_var = "delta_G1",
#   y_var = "delta_Async",
#   name = paste0(loopName, "_deltaScore_AsyncvsG1_Asyncspecific"),
#   limits = c(-0.75, 0.75), # Use default data limits
#   diff_cutoff = 0.2
# )
```


```{r}
temp <- temp1 %>% dplyr::full_join(temp2, by = "AnnoGroup")


data_long <- temp %>% pivot_longer(cols = c("G1SpecificPerturb", "AsyncSpecificPerturb"))

data_long$name <- factor(data_long$name, levels = c("G1SpecificPerturb", "AsyncSpecificPerturb"))

p <- ggplot(data_long, aes(x = name, y = value, fill = AnnoGroup)) +
  geom_bar(position="fill", stat="identity") +
  labs(
    x = NULL,         # Remove x-axis title
    y = "Ratio"       # Change y-axis title
  ) +
  theme(
    axis.title = element_text(
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45, hjust = 1,          # Rotate x labels
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text.y = element_text(
      size = fontSizeS, family = fontType, color = "#000000"
    ),
    axis.line = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) + 
  guides(
    fill = guide_legend(
      keywidth = 0.2,  # Adjust the width of the legend keys
      keyheight = 0.2  # Adjust the height of the legend keys
    )
  )

# Save the plot
ggsave(
  filename = here(figDir, paste0("loopClassify_", loopName, ".png")),
  plot = p,
  width = panelSize(2) * mmToInch,
  height = panelSize(2.5) * mmToInch,
  dpi = 600
)




temp <- temp3 %>% dplyr::full_join(temp4, by = "AnnoGroup")


data_long <- temp %>% pivot_longer(cols = c("G1SpecificPerturb", "AsyncSpecificPerturb"))

data_long$name <- factor(data_long$name, levels = c("G1SpecificPerturb", "AsyncSpecificPerturb"))

p <- ggplot(data_long, aes(x = name, y = value, fill = AnnoGroup)) +
  geom_bar(position="fill", stat="identity") +
  labs(
    x = NULL,         # Remove x-axis title
    y = "Ratio"       # Change y-axis title
  ) +
  theme(
    axis.title = element_text(
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45, hjust = 1,          # Rotate x labels
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text.y = element_text(
      size = fontSizeS, family = fontType, color = "#000000"
    ),
    axis.line = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) + 
  guides(
    fill = guide_legend(
      keywidth = 0.2,  # Adjust the width of the legend keys
      keyheight = 0.2  # Adjust the height of the legend keys
    )
  )

# Save the plot
ggsave(
  filename = here(figDir, paste0("loopClassify_", loopName, "_Reg.png")),
  plot = p,
  width = panelSize(2) * mmToInch,
  height = panelSize(2.5) * mmToInch,
  dpi = 600
)

```
###Bookmarking
```{r}
background.gr <- extractAnchor(data)

temp <- plot_df_delta %>% dplyr::filter(delta_G1 < -0.2, 
                                        abs(delta_Async) < 0.2,
                                        delta_G1 - delta_Async < -0.2)

G1specific.gr <- extractAnchor(temp)

temp <- plot_df_delta %>% dplyr::filter(
                                        abs(delta_G1) < 0.2)

G1noChange.gr <- extractAnchor(temp)

temp <- plot_df_delta %>% dplyr::filter(delta_Async < -0.2, 
                                        abs(delta_G1) < 0.2,
                                        delta_Async - delta_G1 < -0.2)
Asyncspecific.gr <- extractAnchor(temp)

####
targetName <- "H3K27ac_effie"
# RETAINED loop
retained <- fread(here(refDir, "TF_bookmarking", targetName, "am_a.bed")) %>% dplyr::select(c(1, 2, 3))
colnames(retained) <- c("chr", "start", "end")
retained.gr <- makeGRangesFromDataFrame(retained)
# LOST loop
lost <- fread(here(refDir, "TF_bookmarking", targetName, "oa.bed")) %>% dplyr::select(c(1, 2, 3))
colnames(lost) <- c("chr", "start", "end")
lost.gr <- makeGRangesFromDataFrame(lost)




# SEEDING
interestName <- "ChromoAllReg"
temp <- runFisherExact(G1specific.gr, background.gr, retained.gr)
result.tb <- tibble(interest = paste0(interestName, "_G1"),
                    target = paste0(targetName, "_retained"),
                    pvalue = temp$p.value,
                    oddsRatio = temp$estimate)
# ADDING ROWS
temp <- runFisherExact(Asyncspecific.gr, background.gr, retained.gr)
result.tb <- result.tb %>% 
  add_row(interest = paste0(interestName, "_Async"),
          target = paste0(targetName, "_retained"),
          pvalue = temp$p.value,
          oddsRatio = temp$estimate)

# ADDING ROWS
temp <- runFisherExact(G1specific.gr, background.gr, lost.gr)
result.tb <- result.tb %>% 
add_row(interest = paste0(interestName, "_G1"),
          target = paste0(targetName, "_lost"),
          pvalue = temp$p.value,
          oddsRatio = temp$estimate)

# ADDING ROWS
temp <- runFisherExact(Asyncspecific.gr, background.gr, lost.gr)
result.tb <- result.tb %>% 
  add_row(interest = paste0(interestName, "_Async"),
          target = paste0(targetName, "_lost"),
          pvalue = temp$p.value,
          oddsRatio = temp$estimate)


result.tb <- result.tb %>%
  mutate(
         target = recode(target,
                         "H3K27ac_effie_retained" = "Retained",
                         "H3K27ac_effie_lost" = "Lost"))

# result.tb$interest <- factor(result.tb$interest, levels = c("UP/NO", "DOWN"))

p <- ggplot(result.tb, aes(x = interest, y = target, size = -log10(pvalue), fill = oddsRatio)) +
  geom_point(shape = 21,        # Ensures a point with an outline
             stroke = 0.5*ptToMM      # Line width for the border
  ) + theme_bw() + 
  scale_size_continuous(range = c(1, 3)) +  # Set min and max point sizes here
  scale_fill_gradientn(colors = c("#4852A0", "white", "#CB333A"),  # Define gradient colors
                       values = scales::rescale(c(0.5, 1, 1.5)), limits = c(0.5, 1.5), 
                       #low = "white", high = "#CB333A",
                       #                  limits = c(1, 3),
                       oob = scales::squish, # Define gradient colors
                       guide = guide_colorbar(
                         barwidth = 1.5/5.08,  # Adjust width of the color bar
                         barheight = 15/5.08   # Adjust height of the color bar
                       )
  ) +
  labs(x = NULL, y = NULL)  +
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  )





# SEEDING
interestName <- "ChromoAllReg"
temp <- runFisherExact(G1specific.gr, background.gr, retained.gr)
result.tb <- tibble(interest = paste0(interestName, "_G1"),
                    target = paste0(targetName, "_retained"),
                    pvalue = temp$p.value,
                    oddsRatio = temp$estimate)
# ADDING ROWS
temp <- runFisherExact(G1noChange.gr, background.gr, retained.gr)
result.tb <- result.tb %>% 
  add_row(interest = paste0(interestName, "_G1nochange"),
          target = paste0(targetName, "_retained"),
          pvalue = temp$p.value,
          oddsRatio = temp$estimate)

# ADDING ROWS
temp <- runFisherExact(G1specific.gr, background.gr, lost.gr)
result.tb <- result.tb %>% 
add_row(interest = paste0(interestName, "_G1"),
          target = paste0(targetName, "_lost"),
          pvalue = temp$p.value,
          oddsRatio = temp$estimate)

# ADDING ROWS
temp <- runFisherExact(G1noChange.gr, background.gr, lost.gr)
result.tb <- result.tb %>% 
  add_row(interest = paste0(interestName, "_G1nochange"),
          target = paste0(targetName, "_lost"),
          pvalue = temp$p.value,
          oddsRatio = temp$estimate)


result.tb <- result.tb %>%
  mutate(
         target = recode(target,
                         "H3K27ac_effie_retained" = "Retained",
                         "H3K27ac_effie_lost" = "Lost"))

# result.tb$interest <- factor(result.tb$interest, levels = c("UP/NO", "DOWN"))

p <- ggplot(result.tb, aes(x = interest, y = target, size = -log10(pvalue), fill = oddsRatio)) +
  geom_point(shape = 21,        # Ensures a point with an outline
             stroke = 0.5*ptToMM      # Line width for the border
  ) + theme_bw() + 
  scale_size_continuous(range = c(1, 3)) +  # Set min and max point sizes here
  scale_fill_gradientn(colors = c("#4852A0", "white", "#CB333A"),  # Define gradient colors
                       values = scales::rescale(c(0.5, 1, 1.5)), limits = c(0.5, 1.5), 
                       #low = "white", high = "#CB333A",
                       #                  limits = c(1, 3),
                       oob = scales::squish, # Define gradient colors
                       guide = guide_colorbar(
                         barwidth = 1.5/5.08,  # Adjust width of the color bar
                         barheight = 15/5.08   # Adjust height of the color bar
                       )
  ) +
  labs(x = NULL, y = NULL)  +
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  )
```

### Level
```{r}

plotSumScoresBinary <- function(track, peak, name, note = "", anchor.upno, anchor.down, loopName1, loopName2){
  peakTrack <- track[unique(queryHits(findOverlaps(track, peak)))]
  b <- getSumScores(peakTrack, anchor.upno)
  c <- getSumScores(peakTrack, anchor.down)
  b.tb <- tibble(loop = loopName1,
                 sumScore = b)
  c.tb <- tibble(loop = loopName2,
                 sumScore = c)
  data <- bind_rows(b.tb, c.tb) %>% drop_na()
  data$loop <- factor(data$loop, levels = c(loopName1, loopName2))
  
  p12 <- getPvalWilcox(data, loopName1, loopName2)
  p <-  ggplot(data, aes(x = name, fill = loop, y = sumScore)) + 
    labs(x = NULL, y = paste0("Average ChIP peak score at anchor)")) +
    introdataviz::geom_split_violin(linewidth = lineThick * mmToLineUnit, lineend = "square",
                                    alpha = .4) +
    geom_boxplot(width = 0.3, color = "black",
                 linewidth = lineThick * mmToLineUnit, lineend = "square",
                 outlier.shape = NA,  alpha = 0.6, show.legend = FALSE) + theme_classic() +
    scale_fill_manual(
      name = "Reg loop type",
      values = c("grey30", strong_blue)
    ) +
    stat_summary(
      aes(group = loop), fun = mean,
      geom = "point", shape = 21, size = 0.5,
      fill = "black", color = "black", position = position_dodge(.3)
    ) + theme(
      axis.title = element_text(
        size = fontSizeM, family = fontType, color = "#000000"
      ),
      axis.text = element_text(
        size = fontSizeM, family = fontType, color = "#000000"
      ),
      axis.text.x = element_text(
        angle = 0,
      ),
      axis.text.y = element_text(
        size = fontSizeS,
      ),
      axis.line = element_line(
        color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
      ),
      axis.ticks = element_line(
        color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
      ),
      panel.background = element_rect(fill = "transparent"),
      legend.text = element_text(family = fontType, size = fontSizeS),
      legend.title = element_text(family = fontType, size = fontSizeS)
    ) + 
    coord_cartesian(ylim = c(quantile(data$sumScore, 0.0), quantile(data$sumScore, 0.95))) + 
    annotate(
      "text", x = 1, y = quantile(data$sumScore, 0.5),
      label = paste0("p12: ", convPvalue(p12)),
      color = "black", hjust = 0, size = 1
    ) +   guides(
      fill = guide_legend(
        keywidth = 0.2,  # Adjust the width of the legend keys
        keyheight = 0.2  # Adjust the height of the legend keys
      )
    )
  
  fileName <- paste0("ChIP_peak_avgPeakScore_", name, note)
  width <- panelSize(1.5)*mmToInch
  height <- panelSize(1.2)*mmToInch
  png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
  print(p)
  dev.off()
  svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
  print(p)
  dev.off()
}


track <- import(here(refDir, "ChIP", "H3K27ac", "H3K27ac_ESC_avg_depthNorm.bw"), format = "BigWig")
peak <- importPeak(here(refDir, "ChIP", "H3K27ac", "H3K27ac_ESC_unionPeaks.bed"))
plotSumScoresBinary(track, peak, "H3K27ac", "_G1vsAsync", G1specific.gr, Asyncspecific.gr, "G1specific", "Asyncspecific")
plotSumScoresBinary(track, peak, "H3K27ac", "_G1_specificvsnochange", G1specific.gr, G1noChange.gr, "G1specific", "G1nochange")


```


### Characteristics
```{r}
data <- fread(here(commonLoopDir, "loopScore_chromosight.tsv"))

loop.G1SpecificPerturb <- fread(here(dataDir, "loop_deltaScore_G1specificPerturb.bedpe")) %>%
  dplyr::mutate(id = paste(V1, V2, V3, V4, V5, V6, sep = "_")) 
loop.AsyncSpecificPerturb <- fread(here(dataDir, "loop_deltaScore_AsyncspecificPerturb.bedpe")) %>%
  dplyr::mutate(id = paste(V1, V2, V3, V4, V5, V6, sep = "_"))

temp <- data %>% dplyr::mutate(
  group = case_when(
    id %in% loop.G1SpecificPerturb$id ~ "G1SpecificPerturb",
    id %in% loop.AsyncSpecificPerturb$id ~ "AsyncSpecificPerturb",
    TRUE ~ NA
  ),
  size = start2 - start1
) %>%
  dplyr::filter(!is.na(group))

ggplot(temp, aes(x = group, y = size, fill = group)) +
  geom_violin() +
  geom_boxplot(width = 0.25) +
  theme_classic()

################
temp.reg <- temp %>% dplyr::filter(Anno %in% c("P-P", "P-E", "E-E"))


###################
# Prepare data for O/E analysis
pseudo_ooe <- 0.01
data_to_plot <- temp.reg
plot_df_ooe <- data_to_plot %>%
  dplyr::select(id, ooe_DMSO, ooe_dTAG, ooe_UT, ooe_AID, group) %>%
  na.omit() %>%
  mutate(
    across(starts_with("ooe_"), ~ if_else(. == 0, pseudo_ooe, .)),
    fc_G1 = log2(ooe_dTAG / ooe_DMSO),
    fc_Async = log2(ooe_AID / ooe_UT)
  )

# --- Violin Plot ---
plot_df_ooe_long <- plot_df_ooe %>%
  select(fc_G1, fc_Async, group) %>%
  pivot_longer(cols = -group, names_to = "condition", values_to = "value") %>%
  mutate(condition = factor(condition, levels = c("fc_G1", "fc_Async")))

p_violin <- ggplot(plot_df_ooe_long, aes(x = condition, y = value, fill = condition)) +
  facet_wrap(~group) +
  geom_violin(linewidth = lineThick * mmToLineUnit, alpha = .4, show.legend = FALSE) +
  geom_boxplot(width = 0.3, outlier.shape = NA, alpha = 0.6, show.legend = FALSE, linewidth = lineThick * mmToLineUnit) +
  stat_summary(fun = mean, geom = "point", shape = 21, size = 0.5, fill = "black") +
  geom_hline(yintercept = c(-0.5, 0.5), linetype = "dashed", alpha = 0.5, color = "black", size = lineThick * mmToLineUnit) +
  geom_hline(yintercept = 0, color = "black", size = lineThick * mmToLineUnit) +
  coord_cartesian(ylim = c(-3, 3)) +
  scale_fill_manual(values = c(strong_red, strong_blue)) +
  labs(y = "log2 fc of obs/exp") +
  common_theme +
  theme(
    axis.title.x = element_blank(),
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = fontSizeM),
    axis.title.y = element_text(size = fontSizeM)
  )

```
### LOLA
##### ESC
```{r}
#-------------------------------------------------------------------------------
# LOADING LOLA DB
library("simpleCache")
library("LOLA")
lolaDB = loadRegionDB("/Volumes/UKJIN_SSD/reference/LOLACore_cached/mm10")

atac <- fread(here(refDir, "mm10", "GSM3106257_ATAC_ESC_1.bed")) %>% dplyr::select(V1, V2, V3)
colnames(atac) <- c("chr", "start", "end")
atac.gr <- makeGRangesFromDataFrame(atac)
```
###### All loops
```{r}
#-------------------------------------------------------------------------------
# LOADING LOOPS
### Importing differential regulatory loops & extract anchor
commonLoopDir <- here("../data/loop_analysis")
loop.all <- fread(here(commonLoopDir, "loopScore_chromosight.tsv"))

anchor.all <- (extractAnchor(loop.all))
overlaps <- findOverlaps(anchor.all, atac.gr)
anchor.all <- pintersect(anchor.all[queryHits(overlaps)], atac.gr[subjectHits(overlaps)])

loop.G1 <- fread(here(commonLoopDir, "loop_deltaScore_G1specificPerturb.bedpe"))
anchor.G1 <- (extractAnchor(loop.G1))
overlaps <- findOverlaps(anchor.G1, atac.gr)
anchor.G1 <- pintersect(anchor.G1[queryHits(overlaps)], atac.gr[subjectHits(overlaps)])

loop.Async <- fread(here(commonLoopDir, "loop_deltaScore_AsyncspecificPerturb.bedpe"))
anchor.Async <- (extractAnchor(loop.Async))
overlaps <- findOverlaps(anchor.Async, atac.gr)
anchor.Async <- pintersect(anchor.Async[queryHits(overlaps)], atac.gr[subjectHits(overlaps)])

#-------------------------------------------------------------------------------
# RUNNING LOLA
# G1
lolaDir <- here(commonLoopDir, "LOLA")
dir.create(lolaDir, showWarnings = FALSE, recursive = TRUE)


name <- "chromosight_score_G1vsAsync_G1Specific"
result <- runLOLA(anchor.G1, anchor.all, lolaDB)
tb <- as_tibble(result)
fwrite(tb, here(lolaDir, paste0("LOLA_", name, "_ATAC.tsv")), sep = "\t")

# Async
name <- "chromosight_score_G1vsAsync_AsyncSpecific"
result <- runLOLA(anchor.Async, anchor.all, lolaDB)
tb <- as_tibble(result)
fwrite(tb, here(lolaDir, paste0("LOLA_", name, "_ATAC.tsv")), sep = "\t")



### Visualizing p-value and OR
alpha <- 0.05
name <- "chromosight_score_G1vsAsync_G1Specific"
tb.G1 <- fread(here(lolaDir, paste0("LOLA_", name, "_ATAC.tsv"))) %>%
  dplyr::mutate(target = toupper(antibody),
                qValueLog = -log2(qValue),
                group = "UP") %>%
  filter(str_to_lower(cellType) == "embryonic stem cell") %>%
  dplyr::filter(qValue < alpha) %>% dplyr::group_by(target) %>%
  slice_min(meanRnk, with_ties = FALSE)
name <- "chromosight_score_G1vsAsync_AsyncSpecific"
tb.Async <- fread(here(lolaDir, paste0("LOLA_", name, "_ATAC.tsv"))) %>%
  dplyr::mutate(target = toupper(antibody),
                qValueLog = -log2(qValue),
                group = "NO") %>%
  filter(str_to_lower(cellType) == "embryonic stem cell") %>%
  dplyr::filter(qValue < alpha) %>% dplyr::group_by(target) %>%
  slice_min(meanRnk, with_ties = FALSE)


```
## Hansen loops
```{r}
loopName <- "Hansen_strongLoopsG1"

# --- 4. GENERATE PLOTS ---
data <- fread(here(commonLoopDir, "loopScore_Hansen.tsv")) %>%
 drop_na(score_DMSO, score_dTAG, score_A485, score_EpiDMSO, score_EpidTAG, score_UT, score_AID) %>%
  dplyr::mutate(res = end1 - start1) %>%
    # dplyr::filter(if_any(c(score_DMSO, score_dTAG, score_A485, score_EpiDMSO, score_EpidTAG, score_UT, score_AID), ~ . > 0.3))
    dplyr::filter(if_any(c(score_DMSO, score_dTAG), ~ . > 0.3)) %>%
    dplyr::filter(if_any(c(score_UT, score_AID), ~ . > 0.3))


# Checking the distribution of scores
dataToPlot <- data %>% dplyr::select(score_DMSO, score_UT, score_dTAG, score_AID) %>%
  pivot_longer(everything(), names_to = "condition", values_to = "score" )
dataToPlot$condition <- factor(dataToPlot$condition, levels = c("score_DMSO", "score_dTAG", "score_UT", "score_AID"))
ggplot(dataToPlot, aes(x = condition, y = score, fill = condition)) +
  geom_violin() +
  scale_y_continuous(limits = c(-0.5, 1)) +
  geom_boxplot(width = 0.4) +
  ggtitle("Hansen >0.3 G1 loops") +
  theme_classic() +
  geom_hline(yintercept = c(0)) +
  geom_hline(yintercept = c(0.3), linetype = "dashed")


generate_scatterplot(data, "score_DMSO",    "score_UT", paste0(loopName, "_AsyncUTvsG1DMSO"))
generate_scatterplot(data, "score_UT",    "score_AID", paste0(loopName, "_AsyncAIDvsAsyncUT"))
generate_scatterplot(data, "score_DMSO",    "score_dTAG", paste0(loopName, "_G1dTAGvsG1DMSO"))

## Plot diff score
plot_df_delta <- data %>%
  dplyr::mutate(
    delta_G1 = score_dTAG - score_DMSO,
    delta_Async = score_AID - score_UT
  )


# Function to generate and save scatterplots
generate_scatterplot2 <- function(df, x_var, y_var, name, limits = c(-0.5, 1), diff_cutoff = 0.2) {
  plot_data <- df %>%
    mutate(density = get_density(.data[[x_var]], .data[[y_var]], n = 200))

  p <- ggplot(plot_data, aes(x = .data[[x_var]], y = .data[[y_var]])) +
    geom_point(aes(color = density), size = 0.1, data = . %>% arrange(density)) +
    coord_fixed() +
    scale_x_continuous(limits = limits) +
    scale_y_continuous(limits = limits) +
    geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black", size = lineThick*mmToLineUnit) +
    geom_abline(slope = 1, intercept = c(-diff_cutoff, diff_cutoff), linetype = "dotted", color = "black", size = lineThick*mmToLineUnit) +
    geom_hline(yintercept = c(-diff_cutoff, diff_cutoff), linetype = "dotted", color = "black", size = lineThick*mmToLineUnit) +
    geom_vline(xintercept = c(-diff_cutoff, diff_cutoff), linetype = "dotted", color = "black", size = lineThick*mmToLineUnit) +
    geom_hline(yintercept = 0, alpha = 0.5, color = "black", size = lineThick*mmToLineUnit) +
    geom_vline(xintercept = 0, alpha = 0.5, color = "black", size = lineThick*mmToLineUnit) +
    common_theme

  # Save the plot
  ggsave(
    filename = here(figDir, paste0("scatterplot_", name, ".png")),
    plot = p,
    width = panelSize(1.5) * mmToInch,
    height = panelSize(1.5) * mmToInch,
    dpi = 600
  )
}
generate_scatterplot2(
  df = plot_df_delta,
  x_var = "delta_G1",
  y_var = "delta_Async",
  name = paste0(loopName, "_deltaScore_AsyncvsG1"),
  limits = c(-0.75, 0.75), # Use default data limits
  diff_cutoff = 0.2
)


temp <- plot_df_delta %>% dplyr::filter(delta_G1 < -0.2, 
                                        abs(delta_Async) < 0.2,
                                        delta_G1 - delta_Async < -0.2)
print(nrow(temp))
generate_scatterplot2(
  df = temp,
  x_var = "delta_G1",
  y_var = "delta_Async",
  name = paste0(loopName, "_deltaScore_AsyncvsG1_G1specific"),
  limits = c(-0.75, 0.75), # Use default data limits
  diff_cutoff = 0.2
)

temp <- plot_df_delta %>% dplyr::filter(delta_Async < -0.2, 
                                        abs(delta_G1) < 0.2,
                                        delta_Async - delta_G1 < -0.2)
print(nrow(temp))
generate_scatterplot2(
  df = temp,
  x_var = "delta_G1",
  y_var = "delta_Async",
  name = paste0(loopName, "_deltaScore_AsyncvsG1_Asyncspecific"),
  limits = c(-0.75, 0.75), # Use default data limits
  diff_cutoff = 0.2
)

```
## Tjian loops
```{r}
# --- 4. GENERATE PLOTS ---
data <- fread(here(commonLoopDir, "loopScore_TjianChromo.tsv")) %>%
 drop_na(score_DMSO, score_dTAG, score_A485, score_EpiDMSO, score_EpidTAG, score_UT, score_AID) %>%
  dplyr::mutate(res = end1 - start1) %>%
  dplyr::filter(res >= 1000) %>%
    dplyr::filter(if_any(c(score_DMSO, score_dTAG, score_A485, score_EpiDMSO, score_EpidTAG, score_UT, score_AID), ~ . > 0.3)) %>%
    # dplyr::filter(if_any(c(score_DMSO, score_dTAG), ~ . > 0.3)) %>%
    # dplyr::filter(if_any(c(score_UT, score_AID), ~ . > 0.3)) %>%
    dplyr::filter(Anno %in% c("P-P", "P-E", "E-E"))
  # dplyr::filter(Anno %in% c("S-S", "S-X"))


# loopName <- "TjianChromo_strongAnyEE"
# loopName <- "TjianChromo_strongLoopsG1Async"

# Checking the distribution of scores
# dataToPlot <- data %>% dplyr::select(score_DMSO, score_UT, score_dTAG, score_AID) %>%
#   pivot_longer(everything(), names_to = "condition", values_to = "score" )
# dataToPlot$condition <- factor(dataToPlot$condition, levels = c("score_DMSO", "score_dTAG", "score_UT", "score_AID"))
# ggplot(dataToPlot, aes(x = condition, y = score, fill = condition)) +
#   geom_violin() +
#   scale_y_continuous(limits = c(-0.5, 1)) +
#   geom_boxplot(width = 0.4) +
#   ggtitle("TjianChromo over1kb strongG1Async loops") +
#   theme_classic() +
#   geom_hline(yintercept = c(0)) +
#   geom_hline(yintercept = c(0.3), linetype = "dashed")


# Differential
dataToPlot2 <- data %>% dplyr::mutate(diffScore_G1 = score_dTAG - score_DMSO,
                                      diffScore_Async = score_AID - score_UT) %>%
  dplyr::select(diffScore_G1, diffScore_Async) %>%
  pivot_longer(everything(), names_to = "condition", values_to = "score")
dataToPlot2$condition <- factor(dataToPlot2$condition, levels = c("diffScore_G1", "diffScore_Async"))
ggplot(dataToPlot2, aes(x = condition, y = score, fill = condition)) +
  geom_violin() +
  scale_y_continuous(limits = c(-1, 1)) +
  geom_boxplot(width = 0.4) +
  ggtitle(paste0("Chromosight ", loopName)) +
  theme_classic() +
  geom_hline(yintercept = c(0)) +
  geom_hline(yintercept = c(0.3), linetype = "dashed")  


generate_scatterplot(data, "score_DMSO",    "score_UT", paste0(loopName, "_AsyncUTvsG1DMSO"))
generate_scatterplot(data, "score_UT",    "score_AID", paste0(loopName, "_AsyncAIDvsAsyncUT"))
generate_scatterplot(data, "score_DMSO",    "score_dTAG", paste0(loopName, "_G1dTAGvsG1DMSO"))

## Plot diff score
plot_df_delta <- data %>%
  dplyr::mutate(
    delta_G1 = score_dTAG - score_DMSO,
    delta_Async = score_AID - score_UT
  )


# Function to generate and save scatterplots
generate_scatterplot2 <- function(df, x_var, y_var, name, limits = c(-0.5, 1), diff_cutoff = 0.2) {
  plot_data <- df %>%
    mutate(density = get_density(.data[[x_var]], .data[[y_var]], n = 200))

  p <- ggplot(plot_data, aes(x = .data[[x_var]], y = .data[[y_var]])) +
    geom_point(aes(color = density), size = 0.1, data = . %>% arrange(density)) +
    coord_fixed() +
    scale_x_continuous(limits = limits) +
    scale_y_continuous(limits = limits) +
    geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "black", size = lineThick*mmToLineUnit) +
    geom_abline(slope = 1, intercept = c(-diff_cutoff, diff_cutoff), linetype = "dotted", color = "black", size = lineThick*mmToLineUnit) +
    geom_hline(yintercept = c(-diff_cutoff, diff_cutoff), linetype = "dotted", color = "black", size = lineThick*mmToLineUnit) +
    geom_vline(xintercept = c(-diff_cutoff, diff_cutoff), linetype = "dotted", color = "black", size = lineThick*mmToLineUnit) +
    geom_hline(yintercept = 0, alpha = 0.5, color = "black", size = lineThick*mmToLineUnit) +
    geom_vline(xintercept = 0, alpha = 0.5, color = "black", size = lineThick*mmToLineUnit) +
    common_theme

  # Save the plot
  ggsave(
    filename = here(figDir, paste0("scatterplot_", name, ".png")),
    plot = p,
    width = panelSize(1.5) * mmToInch,
    height = panelSize(1.5) * mmToInch,
    dpi = 600
  )
}
generate_scatterplot2(
  df = plot_df_delta,
  x_var = "delta_G1",
  y_var = "delta_Async",
  name = paste0(loopName, "_deltaScore_AsyncvsG1"),
  limits = c(-0.75, 0.75), # Use default data limits
  diff_cutoff = 0.2
)


temp <- plot_df_delta %>% dplyr::filter(delta_G1 < -0.2, 
                                        abs(delta_Async) < 0.2,
                                        delta_G1 - delta_Async < -0.2)
print(nrow(temp))
temp1 <- make_table(temp, "G1SpecificPerturb")
temp3 <- make_table2(temp, "G1SpecificPerturb") %>%
  dplyr::filter(!is.na(AnnoGroup))
generate_scatterplot2(
  df = temp,
  x_var = "delta_G1",
  y_var = "delta_Async",
  name = paste0(loopName, "_deltaScore_AsyncvsG1_G1specific"),
  limits = c(-0.75, 0.75), # Use default data limits
  diff_cutoff = 0.2
)

temp <- plot_df_delta %>% dplyr::filter(delta_Async < -0.2, 
                                        abs(delta_G1) < 0.2,
                                        delta_Async - delta_G1 < -0.2)
print(nrow(temp))
temp2 <- make_table(temp, "AsyncSpecificPerturb")
temp4 <- make_table2(temp, "AsyncSpecificPerturb") %>%
  dplyr::filter(!is.na(AnnoGroup))

generate_scatterplot2(
  df = temp,
  x_var = "delta_G1",
  y_var = "delta_Async",
  name = paste0(loopName, "_deltaScore_AsyncvsG1_Asyncspecific"),
  limits = c(-0.75, 0.75), # Use default data limits
  diff_cutoff = 0.2
)

```

```{r}
temp <- temp1 %>% dplyr::full_join(temp2, by = "AnnoGroup")


data_long <- temp %>% pivot_longer(cols = c("G1SpecificPerturb", "AsyncSpecificPerturb"))

data_long$name <- factor(data_long$name, levels = c("G1SpecificPerturb", "AsyncSpecificPerturb"))

p <- ggplot(data_long, aes(x = name, y = value, fill = AnnoGroup)) +
  geom_bar(position="fill", stat="identity") +
  labs(
    x = NULL,         # Remove x-axis title
    y = "Ratio"       # Change y-axis title
  ) +
  theme(
    axis.title = element_text(
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45, hjust = 1,          # Rotate x labels
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text.y = element_text(
      size = fontSizeS, family = fontType, color = "#000000"
    ),
    axis.line = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) + 
  guides(
    fill = guide_legend(
      keywidth = 0.2,  # Adjust the width of the legend keys
      keyheight = 0.2  # Adjust the height of the legend keys
    )
  )

# Save the plot
ggsave(
  filename = here(figDir, paste0("loopClassify_", loopName, ".png")),
  plot = p,
  width = panelSize(2) * mmToInch,
  height = panelSize(2.5) * mmToInch,
  dpi = 600
)




temp <- temp3 %>% dplyr::full_join(temp4, by = "AnnoGroup")


data_long <- temp %>% pivot_longer(cols = c("G1SpecificPerturb", "AsyncSpecificPerturb"))

data_long$name <- factor(data_long$name, levels = c("G1SpecificPerturb", "AsyncSpecificPerturb"))

p <- ggplot(data_long, aes(x = name, y = value, fill = AnnoGroup)) +
  geom_bar(position="fill", stat="identity") +
  labs(
    x = NULL,         # Remove x-axis title
    y = "Ratio"       # Change y-axis title
  ) +
  theme(
    axis.title = element_text(
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45, hjust = 1,          # Rotate x labels
      size = fontSizeM, family = fontType, color = "#000000"
    ),
    axis.text.y = element_text(
      size = fontSizeS, family = fontType, color = "#000000"
    ),
    axis.line = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000", size = lineThick * mmToLineUnit, lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) + 
  guides(
    fill = guide_legend(
      keywidth = 0.2,  # Adjust the width of the legend keys
      keyheight = 0.2  # Adjust the height of the legend keys
    )
  )

# Save the plot
ggsave(
  filename = here(figDir, paste0("loopClassify_", loopName, "_Reg.png")),
  plot = p,
  width = panelSize(2) * mmToInch,
  height = panelSize(2.5) * mmToInch,
  dpi = 600
)

```
# [9] A485 dual
```{r}
# Import gene grouping
geneGroup.dual <- fread(here("..", "RNA_250508_H202SC25035068_G1RAD21A485dual", "geneList_synergy.tsv"))
```

#### H3K4me3 peak number per 10kb
```{r}
process_group <- function(group_name, gene_data, bins) {
  # Filter for the specific group
  TSS_group <- makeGRangesFromDataFrame(
    gene_data %>% filter(ensembl %in% get(group_name)), 
    keep.extra.columns = TRUE
  )
  
  # Process each TSS in the group
  all_results <- map_dfr(seq_len(length(TSS_group)), function(i) {
    # Find overlapping bins
    overlapping_bins <- findOverlaps(TSS_group[i], bins)
    bins_near_tss <- as_tibble(bins[subjectHits(overlapping_bins)])
    
    # Calculate distance bin
    TSScenter <- (as_tibble(TSS_group[i]) %>%
                    mutate(center = (start + end) / 2))$center
    centerBinStart <- floor(TSScenter / 10e3) * 10e3 + 1
    bins_near_tss <- bins_near_tss %>%
      mutate(distanceBin = abs((start - centerBinStart)) / 10e3)
    
    # Summarize results
    result <- bins_near_tss %>%
      group_by(distanceBin) %>%
      summarise(mean_peak_counts = mean(peak_counts, na.rm = TRUE), .groups = "drop") %>%
      mutate(gene = TSS_group[i]$ensembl)
    return(result)
  })
  
  # Calculate mean and SD for the group
  mean_data <- all_results %>%
    group_by(distanceBin) %>%
    summarise(mean_peak_counts = mean(mean_peak_counts, na.rm = TRUE))
  
  sd_data <- all_results %>%
    group_by(distanceBin) %>%
    summarise(sd_peak_counts = sd(mean_peak_counts, na.rm = TRUE))
  
  # Join and add group name
  summary_data <- left_join(mean_data, sd_data, by = "distanceBin") %>%
    mutate(group = group_name)
  
  return(summary_data)
}

##########################################################


## Import peak
peak <- importPeak(here(refDir, "ChIP", "H3K27ac", "H3K27ac_ESC_unionPeaks.bed"))

## Import 10kb bin
temp <- fread(here(refDir, "mm10", "mm10.bin.10kb.bed")) %>% dplyr::select(c(1, 2, 3))
colnames(temp) <- c("chr", "start", "end")
temp <- temp %>% dplyr::mutate(start = start+1)
bins.10kb <- makeGRangesFromDataFrame(temp)

## Count overlap
counts <- countOverlaps(bins.10kb, peak)
mcols(bins.10kb)$peak_counts <- counts

## Getting TSS
flankSize <- 1e6
gene.tb <- fread(here(refDir, "mm10", "mm10_GRCm38.p6_gene_sorted.bed")) %>%
  dplyr::mutate(TSS = ifelse(V4 == "+", V2, V3),
                TSSstart = TSS - flankSize,
                TSSend = TSS + flankSize) %>%
  dplyr::select(V1, TSSstart, TSSend, V6)
colnames(gene.tb) <- c("chr", "start", "end", "ensembl")


group.down_syn <- (geneGroup.dual %>% dplyr::filter(group == "DOWN_syn"))$ensembl
group.down_ant <- (geneGroup.dual %>% dplyr::filter(group == "DOWN_ant"))$ensembl
group.down_noInt <- (geneGroup.dual %>% dplyr::filter(group == "DOWN_noInt"))$ensembl


##########################################################
# List of groups to process
groups <- c("group.down_syn", "group.down_ant", "group.down_noInt")

# Process each group and combine results
summary_data <- map_dfr(groups, ~ process_group(.x, gene.tb, bins.10kb))

n1 <- length(group1)
n2 <- length(group2)
n3 <- length(group3)



# Plot the results
p <- ggplot(summary_data, aes(x = distanceBin, y = mean_peak_counts, fill = group, color = group, group = group)) + 
  geom_smooth(
    linewidth = lineMedium,
    method = "loess", 
    formula = y ~ x, 
    span   = 0.2,    # smaller = wigglier, larger = smoother
    se     = TRUE,   # show confidence band
    alpha  = 0.2, 
  ) +
  labs(
    x = "Distance bin (10kb)",
    y = "H3K27ac peak count"
  ) +
  theme_classic() +
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    # legend.text = element_text(family = fontType, size = fontSizeS),
    # legend.title = element_text(family = fontType, size = fontSizeS),
    # legend.position = "none"  # Moves legend inside the plot (x, y relative coordinates)
  ) +
  guides(
    fill = guide_legend(
      keywidth = 0.2,  # Adjust the width of the legend keys
      keyheight = 0.2  # Adjust the height of the legend keys
    )) +
  # scale_x_continuous(limits = c(0, 50))
  scale_color_manual(values = rev(c("red", "darkgreen", "blue" ))) +
  scale_fill_manual(values = rev(c("red", "darkgreen", "blue")))


summary_data2 <- summary_data %>% dplyr::filter(group != "group.down_noInt")


# Plot the results
p2 <- ggplot(summary_data2, aes(x = distanceBin, y = mean_peak_counts, fill = group, color = group, group = group)) + 
  geom_smooth(
    linewidth = lineMedium,
    method = "loess", 
    formula = y ~ x, 
    span   = 0.2,    # smaller = wigglier, larger = smoother
    se     = TRUE,   # show confidence band
    alpha  = 0.2, 
  ) +
  labs(
    x = "Distance bin (10kb)",
    y = "H3K27ac peak count"
  ) +
  theme_classic() +
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeS,
      family = fontType,
      color = "#000000"
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    # legend.text = element_text(family = fontType, size = fontSizeS),
    # legend.title = element_text(family = fontType, size = fontSizeS),
    # legend.position = "none"  # Moves legend inside the plot (x, y relative coordinates)
  ) +
  guides(
    fill = guide_legend(
      keywidth = 0.2,  # Adjust the width of the legend keys
      keyheight = 0.2  # Adjust the height of the legend keys
    )) +
  # scale_x_continuous(limits = c(0, 50))
  scale_color_manual(values = rev(c("red", "blue"))) +
  scale_fill_manual(values = rev(c("red", "blue")))


# 
# ##########################################################
# 
# fileName <- paste0("binPeakDensity_H3K4me3_binaryGroup")
# width <- 38*mmToInch
# height <- 35*mmToInch
# png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
# print(p)
# dev.off()
# svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
# print(p)
# dev.off()
```


### Loop analysis per gene group
```{r}
commonLoopDir <- here("../data/loop_analysis")
diffDir <- here("../data/RNA_diff")
name <- "chromosight"

geneAnnoData <- loadLoopAnnoData(here(commonLoopDir, paste0("loopScore_", name, "_p-n_complex_ensemblList.tsv"))) %>%
  dplyr::mutate(distance = start2 - start1,
                loopID = paste(chrom1, start1, start2, sep = "_"))
```

##### Loop number per gene: P-N
```{r}
dataDir <- here("../data")

group.down_syn <- (geneGroup.dual %>% dplyr::filter(group == "DOWN_syn"))$ensembl
group.down_ant <- (geneGroup.dual %>% dplyr::filter(group == "DOWN_ant"))$ensembl
group.down_noInt <- (geneGroup.dual %>% dplyr::filter(group == "DOWN_noInt"))$ensembl


# Counting number of loop per genes
tempSum <- geneAnnoData %>% dplyr::select(loopID, gene, Anno) %>% unnest(gene) %>% group_by(gene) %>%  summarize(
  loop = list(loopID),
  anno = list(Anno),
  count = n())

tempSum <- tempSum %>% dplyr::mutate(
  group = ifelse(gene %in% group.down_syn, "down_syn",
                 ifelse(gene %in% group.down_ant, "down_ant", 
                        ifelse(gene %in% group.down_noInt, "down_noInt", NA)))
) %>% dplyr::filter(!is.na(group))

# # 
# getPvalWilcox <- function(data, group1, group2){
#   distance1 <- (data %>% dplyr::filter(group ==group1) )$count
#   distance2 <- (data %>% dplyr::filter(group ==group2) )$count
#   wil <- wilcox.test(distance1, distance2)
#   return(wil$p.value)
# }


# pv12 <- round(getPvalWilcox(tempSum,"group1", "group2"), 5)


p <- ggplot(tempSum, aes(x = group, y = count, fill = group)) + 
  geom_boxplot(width = 0.5, outlier.shape = NA, color = "black", alpha = 0.6,
               linewidth = lineThick * mmToLineUnit, lineend = "square", show.legend = FALSE) + theme_classic() +       
  scale_y_continuous(breaks = seq(0, 10, by = 2)) + 
  labs(x = NULL, y = "# of P-N loops per gene") +
  coord_cartesian(ylim = c(0, 8)) +
  stat_summary(
    aes(group = group), fun = mean,
    geom = "point", shape = 21, size = 0.5,
    fill = "black", color = "black"
  )+
  # annotate("text", x = 1, y = 3, label = paste0("pv12: ", convPvalue(pv12)),
  #          color = "black", hjust = 0, size = 1) +
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) + scale_fill_manual(values = (c("blue", "darkgreen", "red")))

fileName <- paste0("count_barplot_diffGroup_EpidTAG_vs_EpiDMSO_binaryGroup")
width <- panelSize(0.8)*mmToInch
height <- panelSize(1.5)*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()
```

##### Counting loop types
```{r}
#######
temp2 <- tempSum %>% rowwise() %>% mutate(total = length(anno),
                                          num_pp = sum(anno == "P-P"),
                                          num_pe = sum(anno == "P-E"),
                                          num_px = sum(anno == "P-X"),
                                          num_ps = sum(anno %in% c("P-S", "P-SP", "P-SE")),
                                          num_sps = sum(anno %in% c("SP-SP", "SP-SE", "SP-S")),
                                          num_spn = sum(anno %in% c("E-SP", "SP-X")),
                                          num_str = num_ps + num_sps + num_spn
)


loopType <- temp2 %>% group_by(group) %>% summarise(num_pp = sum(num_pp),
                                                    num_pe = sum(num_pe),
                                                    # num_str = sum(num_str),
                                                    num_ps = sum(num_ps),
                                                    num_sps = sum(num_sps),
                                                    num_spn = sum(num_spn),
                                                    num_px = sum(num_px))


loopTypeLong <- loopType %>% pivot_longer(-group, names_to = "type", values_to = "count")

loopTypeLong$type <- factor(loopTypeLong$type, levels = c("num_pp", "num_pe", "num_px",
                                                          "num_ps", "num_spn", 
                                                          "num_sps" ))

# Plotting
p <- ggplot(loopTypeLong, aes(fill=type, y=count, x=group)) + 
  geom_bar(position="fill", stat="identity") + theme_classic() +
  scale_fill_manual(values = c(strong_red, strong_orange, strong_yellow, strong_blue, "grey30", "grey10")) +theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  )+   
  guides(
    fill = guide_legend(
      keywidth = 0.2,  # Adjust the width of the legend keys
      keyheight = 0.2  # Adjust the height of the legend keys
    ))

fileName <- paste0("count_barplot_diffGroup_EpidTAG_vs_EpiDMSO_binaryGroup_alltypes")
width <- panelSize(1.5)*mmToInch
height <- panelSize(1.5)*mmToInch
png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
print(p)
dev.off()
svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
print(p)
dev.off()
```

##### Loop number per gene: structural
```{r}
########
data <- temp2 %>% dplyr::select(group, num_str)
colnames(data) <- c("group", "count")
# pv12 <- round(getPvalWilcox(data,"group1", "group2"), 5)

p <- ggplot(data, aes(x = group, y = count, fill = group)) + 
  geom_boxplot(width = 0.5, outlier.shape = NA, color = "black", alpha = 0.6,
               linewidth = lineThick * mmToLineUnit, lineend = "square", show.legend = FALSE) + theme_classic() +       
  scale_y_continuous(breaks = seq(0, 10, by = 2)) + 
  labs(x = NULL, y = "# of PromStr loops per gene") +
  coord_cartesian(ylim = c(0, 4)) +
  stat_summary(
    aes(group = group), fun = mean,
    geom = "point", shape = 21, size = 0.5,
    fill = "black", color = "black"
  )+
  # annotate("text", x = 1, y = 3, label = paste0("pv12: ", convPvalue(pv12)),
  #          color = "black", hjust = 0, size = 1) +
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) + scale_fill_manual(values = (c("blue", "darkgreen", "red")))

# fileName <- paste0("count_barplot_diffGroup_EpidTAG_vs_EpiDMSO_str_binaryGroup")
# width <- panelSize(0.8)*mmToInch
# height <- panelSize(1.5)*mmToInch
# png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
# print(p)
# dev.off()
# svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
# print(p)
# dev.off()
```


##### Loop number per gene: P-P
```{r}
########
data <- temp2 %>% dplyr::select(group, num_pp)
colnames(data) <- c("group", "count")
# pv12 <- round(getPvalWilcox(data,"group1", "group2"), 5)

p <- ggplot(data, aes(x = group, y = count, fill = group)) + 
  geom_boxplot(width = 0.5, outlier.shape = NA, color = "black", alpha = 0.6,
               linewidth = lineThick * mmToLineUnit, lineend = "square", show.legend = FALSE) + theme_classic() +       
  scale_y_continuous(breaks = seq(0, 10, by = 2)) + 
  labs(x = NULL, y = "# of P-P loops per gene") +
  coord_cartesian(ylim = c(0, 4)) +
  stat_summary(
    aes(group = group), fun = mean,
    geom = "point", shape = 21, size = 0.5,
    fill = "black", color = "black"
  )+
  # annotate("text", x = 1, y = 3, label = paste0("pv12: ", convPvalue(pv12)),
  #          color = "black", hjust = 0, size = 1) +
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) + scale_fill_manual(values = (c("blue", "darkgreen", "red")))

# fileName <- paste0("count_barplot_diffGroup_EpidTAG_vs_EpiDMSO_pp_binaryGroup")
# width <- panelSize(0.8)*mmToInch
# height <- panelSize(1.5)*mmToInch
# png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
# print(p)
# dev.off()
# svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
# print(p)
# dev.off()
```

####### Loop number per gene: P-E
```{r}
########
data <- temp2 %>% dplyr::select(group, num_pe)
colnames(data) <- c("group", "count")
# pv12 <- round(getPvalWilcox(data,"group1", "group2"), 5)

p <- ggplot(data, aes(x = group, y = count, fill = group)) + 
  geom_boxplot(width = 0.5, outlier.shape = NA, color = "black", alpha = 0.6,
               linewidth = lineThick * mmToLineUnit, lineend = "square", show.legend = FALSE) + theme_classic() +       
  scale_y_continuous(breaks = seq(0, 10, by = 2)) + 
  labs(x = NULL, y = "# of P-E loops per gene") +
  coord_cartesian(ylim = c(0, 4)) +
  stat_summary(
    aes(group = group), fun = mean,
    geom = "point", shape = 21, size = 0.5,
    fill = "black", color = "black"
  )+
  # annotate("text", x = 1, y = 3, label = paste0("pv12: ", convPvalue(pv12)),
  #          color = "black", hjust = 0, size = 1) +
  theme(
    axis.title = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text = element_text(
      size = fontSizeM,
      family = fontType,
      color = "#000000"
    ),
    axis.text.x = element_text(
      angle = 45,      # Rotate x-axis labels 45 degrees
      hjust = 1,       # Adjust horizontal justification
      vjust = 1        # Adjust vertical justification
    ),
    axis.line = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    axis.ticks = element_line(
      color = "#000000",
      size = lineThick*mmToLineUnit,
      lineend = "square"
    ),
    panel.background = element_rect(fill = "transparent"),
    legend.text = element_text(family = fontType, size = fontSizeS),
    legend.title = element_text(family = fontType, size = fontSizeS)
  ) + scale_fill_manual(values = (c("blue", "darkgreen", "red")))

# fileName <- paste0("count_barplot_diffGroup_EpidTAG_vs_EpiDMSO_pe_binaryGroup")
# width <- panelSize(0.8)*mmToInch
# height <- panelSize(1.5)*mmToInch
# png(here(figDir, paste0(fileName, ".png")), res = 600, unit = "in", height = height, width = width)
# print(p)
# dev.off()
# svglite(here(figDir, paste0(fileName, ".svg")),  height = height, width = width)
# print(p)
# dev.off()
```
##### Finding Enhancer distance
```{r}
group.down_syn <- (geneGroup.dual %>% dplyr::filter(group == "DOWN_syn"))$ensembl
group.down_ant <- (geneGroup.dual %>% dplyr::filter(group == "DOWN_ant"))$ensembl
group.down_noInt <- (geneGroup.dual %>% dplyr::filter(group == "DOWN_noInt"))$ensembl


dataToPlot <- geneAnnoData %>% dplyr::mutate(size = start2 - start1) %>%
  dplyr::select(loopID, gene, Anno, size) %>% unnest(gene) %>% dplyr::filter(Anno %in% c("E-SP", "P-E", "P-SE", "SP-SE")) %>%
  group_by(gene) %>%
  summarize(
    loop = list(loopID),
    sizes = list(size),
    count = n(),
        min_size = min(size),
    max_size = max(size),
    avg_size = mean(size)
  ) 

dataToPlot <- dataToPlot %>% dplyr::mutate(
  group = ifelse(gene %in% group.down_syn, "down_syn",
                 ifelse(gene %in% group.down_ant, "down_ant", 
                        ifelse(gene %in% group.down_noInt, "down_noInt", NA)))
) %>% dplyr::filter(!is.na(group))


ggplot(dataToPlot, aes(x = group, y = min_size, fill = group)) + geom_violin() + geom_boxplot(width = 0.3) + coord_cartesian(ylim = c(0, 1000000)) + 
  theme_classic() + scale_fill_manual(values = (c(lighten("blue", amount = 0.3),
                                                  lighten("darkgreen", amount = 0.3),
                                                  lighten("red", amount = 0.3))))
ggplot(dataToPlot, aes(x = group, y = avg_size, fill = group)) + geom_violin() + geom_boxplot(width = 0.3) + coord_cartesian(ylim = c(0, 1000000)) + 
  theme_classic() + scale_fill_manual(values = (c(lighten("blue", amount = 0.3),
                                                  lighten("darkgreen", amount = 0.3),
                                                  lighten("red", amount = 0.3))))
ggplot(dataToPlot, aes(x = group, y = max_size, fill = group)) + geom_violin() + geom_boxplot(width = 0.3) + coord_cartesian(ylim = c(0, 1000000)) + 
  theme_classic() + scale_fill_manual(values = (c(lighten("blue", amount = 0.3),
                                                  lighten("darkgreen", amount = 0.3),
                                                  lighten("red", amount = 0.3))))
```

